<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    
    <meta name="author" content="HE Yunye">
    
    <link rel="shortcut icon" href="../img/favicon.ico">

    
    <title>Harmonization - GWASLab</title>
    

    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.12.0/css/all.css">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.12.0/css/v4-shims.css">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/npm/hack-font@3.3.0/build/web/hack.min.css">
    <link href='//rsms.me/inter/inter.css' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,700italic,400,300,600,700&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href="../css/bootstrap-custom.min.css" rel="stylesheet">
    <link href="../css/base.min.css" rel="stylesheet">
    <link href="../css/cinder.min.css" rel="stylesheet">

    
        
        <link rel="stylesheet" href="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/styles/github.min.css">
        
    

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js"></script>
            <script src="https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js"></script>
        <![endif]-->

    

     
</head>

<body>

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->

            
              <a class="navbar-brand" href="..">GWASLab</a>
            
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="..">Home</a>
                    </li>
                
                
                
                    <li >
                        <a href="../Tutorial/">Tutorial</a>
                    </li>
                
                
                
                    <li class="dropdown active">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Sumstats manipulation <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../SumstatsObject/">SumstatsObject</a>
</li>

                        
                            
<li >
    <a href="../StatusCode/">StatusCode</a>
</li>

                        
                            
<li >
    <a href="../Standardization/">Standardization&normalization</a>
</li>

                        
                            
<li >
    <a href="../QC%26Filtering/">QC&Filtering</a>
</li>

                        
                            
<li class="active">
    <a href="./">Harmonization</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Visualization <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../Visualization/">Manhattan&QQ</a>
</li>

                        
                            
<li >
    <a href="../EffectSize/">Scatter: effect size comparison</a>
</li>

                        
                            
<li >
    <a href="../AlleleFrequency/">Scatter: allele frequency comparison</a>
</li>

                        
                            
<li >
    <a href="../GeneticCorrelation/">Heatmap: genetic correlation</a>
</li>

                        
                            
<li >
    <a href="../Gallery/">Gallery</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Utilities <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../ExtractLead/">Extract Lead Variants</a>
</li>

                        
                            
<li >
    <a href="../Format/">Formatting</a>
</li>

                        
                            
<li >
    <a href="../LoadLDSC/">Load LDSC log</a>
</li>

                        
                            
<li >
    <a href="../HeritabilityConversion/">Heritability conversion</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li >
                        <a href="../Reference/">References</a>
                    </li>
                
                
                </ul>

            <ul class="nav navbar-nav navbar-right">
                    <li>
                        <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                            <i class="fas fa-search"></i> Search
                        </a>
                    </li>
                    <li >
                        <a rel="prev" href="../QC%26Filtering/">
                            <i class="fas fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="next" href="../Visualization/">
                            Next <i class="fas fa-arrow-right"></i>
                        </a>
                    </li>
                    <li>
                        <a href="https://github.com/Cloufield/xmind/"><i class="fab fa-github"></i> GitHub</a>
                    </li>
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
        <li class="first-level active"><a href="#harmonization">Harmonization</a></li>
            <li class="second-level"><a href="#methods-summary">Methods summary</a></li>
                
            <li class="second-level"><a href="#align-nea-with-ref-in-reference-genome">Align NEA with REF in reference genome</a></li>
                
            <li class="second-level"><a href="#assign-chr-and-pos-according-to-rsid-and-reference-data">Assign CHR and POS according to rsID and reference data</a></li>
                
            <li class="second-level"><a href="#assign-rsid-according-to-chr-pos-refalt">Assign rsID according to CHR, POS, REF/ALT</a></li>
                
            <li class="second-level"><a href="#check-panlidromic-snps-or-undistingushable-indels">Check panlidromic SNPs or undistingushable Indels</a></li>
                
            <li class="second-level"><a href="#check-difference-in-allele-frequency">Check difference in allele frequency</a></li>
                
            <li class="second-level"><a href="#flipping-based-on-status-code">Flipping based on status code</a></li>
                
            <li class="second-level"><a href="#liftover">Liftover</a></li>
                
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<h1 id="harmonization">Harmonization</h1>
<h2 id="methods-summary">Methods summary</h2>
<table>
<thead>
<tr>
<th>Sumstats Methods</th>
<th>Options</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>.check_ref()</code></td>
<td>ref_path</td>
<td>check alignment with a reference sequence</td>
</tr>
<tr>
<td><code>.rsid_to_chrpos()</code></td>
<td>path, n_cores</td>
<td>use rsid to fill CHR and POS</td>
</tr>
<tr>
<td><code>.rsid_to_chrpos2()</code></td>
<td>path</td>
<td>use rsid to fill CHR and POS (muilti-thread, need hd5 file)</td>
</tr>
<tr>
<td><code>.assign_rsid()</code></td>
<td>path</td>
<td>annotate rsid using a reference vcf file</td>
</tr>
<tr>
<td><code>.infer_strand()</code></td>
<td>ref_infer="" , ref_alt_freq=None,  maf_threshold=0.43</td>
<td>infer the strand of a variant using reference vcf file with EAF in INFO</td>
</tr>
<tr>
<td><code>.check_daf()</code></td>
<td>ref_infer="" , ref_alt_freq=None,</td>
<td>calculate difference in allele frequencies</td>
</tr>
<tr>
<td><code>.flip_allele_stats()</code></td>
<td></td>
<td>After alignment and inferring, flip the alleles to harmonise the variants.</td>
</tr>
<tr>
<td><code>.liftover()</code></td>
<td>n_cores=1,from_build="19", to_build="38"</td>
<td>perform liftover</td>
</tr>
<tr>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<h2 id="align-nea-with-ref-in-reference-genome">Align NEA with REF in reference genome</h2>
<pre><code class="language-python">mysumstats.check_ref(ref_path=&quot;ref_genome.fa&quot;)
</code></pre>
<p>(! Only changing the status code) </p>
<p>Check if NEA is aligned with the reference sequence. After checking, the tracking status code will be changed accordingly.   </p>
<h2 id="assign-chr-and-pos-according-to-rsid-and-reference-data">Assign CHR and POS according to rsID and reference data</h2>
<pre><code class="language-python">mysumstats.rsid_to_chrpos()     #single thread
mysumstats.rsid_to_chrpos2()    #multithread
</code></pre>
<p>(to be tested)</p>
<h2 id="assign-rsid-according-to-chr-pos-refalt">Assign rsID according to CHR, POS, REF/ALT</h2>
<pre><code class="language-python">mysumstats.assign_rsid(path=&quot;reference.vcf.gz&quot;)
</code></pre>
<p>Annotated variants with rsID using a reference vcf file (tabix indexd).</p>
<p>Gwaslab will first extract all variants in reference file with matching CHR and POS. And then comapre EA/NEA in sumstats with REF/ALT in reference vcf. When matching, it will annotate the vairant in sumstats with the matching rsID in reference vcf.  </p>
<h2 id="check-panlidromic-snps-or-undistingushable-indels">Check panlidromic SNPs or undistingushable Indels</h2>
<pre><code class="language-python">mysumstats.infer_strand()
</code></pre>
<p>(! Only changing the status code)</p>
<ol>
<li>
<p>Infer the strand for palindromic SNPs (AT, or CG), the default threshlod is 0.43. </p>
</li>
<li>
<p>make sure specify the right allele frequency for you target ancestry in INFO field.</p>
</li>
<li>
<p>Checking the alignment status of  indels with the REF allele in reference vcf file.</p>
</li>
</ol>
<h2 id="check-difference-in-allele-frequency">Check difference in allele frequency</h2>
<pre><code>mysumstats.check_daf()
</code></pre>
<p>You may want to check the allele frequency discrepancy with a reference vcf. Just specify the path and the right allele frequency for you target ancestry in INFO field.</p>
<p>GWASlab will simply calculate DAF = AF-EAF - AF-ALT , and store the results in DAF column. DAF can then be used to plot or filter variants.</p>
<pre><code class="language-python">mysumstats.plot_daf()
</code></pre>
<h2 id="flipping-based-on-status-code">Flipping based on status code</h2>
<pre><code class="language-python">mysumstats.flip_allele_stats()
</code></pre>
<p>Flip allele-specific statistics to harmonise the variants based on the tracking status code</p>
<h2 id="liftover">Liftover</h2>
<pre><code class="language-python">mysumstats.liftover(n_cores=1,from_build=&quot;19&quot;, to_build=&quot;38&quot;)
</code></pre>
<p>Perform liftover for POS (based on liftover <a href="https://github.com/jeremymcrae/liftover">GitHub - jeremymcrae/liftover: liftover for python, made fast with cython</a>)</p></div>
        
        
    </div>

    
      <footer class="col-md-12 text-center">
          
          
            <hr>
            <p>
            <small>GWASLab is licensed under the MIT license</small><br>
            
            <small>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</small>
            </p>
          

          
          
      </footer>
    
    <script src="//ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="../js/bootstrap-3.0.3.min.js"></script>

    
    <script src="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/highlight.min.js"></script>
        
    <script>hljs.initHighlightingOnLoad();</script>
    

    <script>var base_url = ".."</script>
    
    <script src="../js/base.js"></script>
    <script src="../search/main.js"></script>

    <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="searchModalLabel" aria-hidden="true">
    <div class="modal-dialog modal-lg">
        <div class="modal-content">
            <div class="modal-header">
                <button type="button" class="close" data-dismiss="modal">
                    <span aria-hidden="true">&times;</span>
                    <span class="sr-only">Close</span>
                </button>
                <h4 class="modal-title" id="searchModalLabel">Search</h4>
            </div>
            <div class="modal-body">
                <p>
                    From here you can search these documents. Enter
                    your search terms below.
                </p>
                <form>
                    <div class="form-group">
                        <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query" title="Type search term here">
                    </div>
                </form>
                <div id="mkdocs-search-results"></div>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div><div class="modal" id="mkdocs_keyboard_modal" tabindex="-1" role="dialog" aria-labelledby="keyboardModalLabel" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
            </div>
            <div class="modal-body">
              <table class="table">
                <thead>
                  <tr>
                    <th style="width: 20%;">Keys</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td class="help shortcut"><kbd>?</kbd></td>
                    <td>Open this help</td>
                  </tr>
                  <tr>
                    <td class="next shortcut"><kbd>n</kbd></td>
                    <td>Next page</td>
                  </tr>
                  <tr>
                    <td class="prev shortcut"><kbd>p</kbd></td>
                    <td>Previous page</td>
                  </tr>
                  <tr>
                    <td class="search shortcut"><kbd>s</kbd></td>
                    <td>Search</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>
    </body>

</html>
