<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    
    <meta name="author" content="HE Yunye">
    
    <link rel="shortcut icon" href="../img/favicon.ico">

    
    <title>gwaslab.Sumstats methods - GWASLab</title>
    

    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.12.0/css/all.css">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.12.0/css/v4-shims.css">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/npm/hack-font@3.3.0/build/web/hack.min.css">
    <link href='//rsms.me/inter/inter.css' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,700italic,400,300,600,700&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href="../css/bootstrap-custom.min.css" rel="stylesheet">
    <link href="../css/base.min.css" rel="stylesheet">
    <link href="../css/cinder.min.css" rel="stylesheet">

    
        
        <link rel="stylesheet" href="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/styles/github.min.css">
        
    
    <link href="../stylesheets/extra.css" rel="stylesheet">

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js"></script>
            <script src="https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js"></script>
        <![endif]-->

    

     
</head>

<body>

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->

            
              <a class="navbar-brand" href="..">GWASLab</a>
            
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="..">Home</a>
                    </li>
                
                
                
                    <li >
                        <a href="../tutorial_3.3/">Tutorial for gwaslab</a>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Sumstats manipulation <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../SumstatsObject/">SumstatsObject</a>
</li>

                        
                            
<li >
    <a href="../StatusCode/">StatusCode</a>
</li>

                        
                            
<li >
    <a href="../Standardization/">Standardization&normalization</a>
</li>

                        
                            
<li >
    <a href="../QC%26Filtering/">QC&Filtering</a>
</li>

                        
                            
<li >
    <a href="../Harmonization/">Harmonization</a>
</li>

                        
                            
<li >
    <a href="../LiftOver/">Liftover</a>
</li>

                        
                            
<li >
    <a href="../Conversion/">Data conversion</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Visualization <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../Visualization/">Manhattan & QQ plot</a>
</li>

                        
                            
<li >
    <a href="../RegionalPlot/">Regional Plot</a>
</li>

                        
                            
<li >
    <a href="../BrisbanePlot/">Brisbane Plot</a>
</li>

                        
                            
<li >
    <a href="../MiamiPlot/">Miami Plot</a>
</li>

                        
                            
<li >
    <a href="../EffectSize/">Scatter: effect size comparison</a>
</li>

                        
                            
<li >
    <a href="../AlleleFrequency/">Scatter: allele frequency comparison</a>
</li>

                        
                            
<li >
    <a href="../GeneticCorrelation/">Heatmap: genetic correlation</a>
</li>

                        
                            
<li >
    <a href="../ForestPlot/">Forest Plot</a>
</li>

                        
                            
<li >
    <a href="../Gallery/">Gallery</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Utilities <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../ExtractLead/">Extract Lead Variants</a>
</li>

                        
                            
<li >
    <a href="../ExtractNovel/">Extract Novel Variants</a>
</li>

                        
                            
<li >
    <a href="../Format/">Format & Save</a>
</li>

                        
                            
<li >
    <a href="../LoadLDSC/">Load LDSC log</a>
</li>

                        
                            
<li >
    <a href="../HeritabilityConversion/">Convert Heritability</a>
</li>

                        
                            
<li >
    <a href="../InferBuild/">Infer Genome Build</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">References <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../Download/">Download</a>
</li>

                        
                            
<li >
    <a href="../Reference/">Reference data</a>
</li>

                        
                            
<li >
    <a href="../CommonData/">Common data</a>
</li>

                        
                            
<li >
    <a href="../UpdateLogs/">Update logs</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Examples <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Standardization and QC example</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../standardization_workflow/">Standardization Workflow</a>
</li>

        
            
<li >
    <a href="../quality_control_and_filtering/">QC &amp; Filtering</a>
</li>

        
    </ul>
  </li>

                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Harmonization example</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../harmonization_workflow/">Workflow</a>
</li>

        
            
<li >
    <a href="../harmonization_liftover/">Liftover</a>
</li>

        
    </ul>
  </li>

                        
                            
<li >
    <a href="../format_load_save/">Formatting and saving</a>
</li>

                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Utility example</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../utility_data_conversion/">Data conversion</a>
</li>

        
            
<li >
    <a href="../utility_get_novel/">Get novel variants</a>
</li>

        
    </ul>
  </li>

                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Visualization example</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../visualization_mqq/">Manhattan and Q-Q plot</a>
</li>

        
            
<li >
    <a href="../visualization_miami/">Miami plot</a>
</li>

        
            
<li >
    <a href="../visualization_regional/">Regional plot</a>
</li>

        
            
<li >
    <a href="../visualization_brisbane/">Brisbane plot</a>
</li>

        
    </ul>
  </li>

                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Reference management</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../Download_reference/">Download</a>
</li>

        
    </ul>
  </li>

                        
                        </ul>
                    </li>
                
                
                </ul>

            <ul class="nav navbar-nav navbar-right">
                    <li>
                        <a href="https://github.com/Cloufield/gwaslab/"><i class="fab fa-github"></i> GitHub</a>
                    </li>
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
        <li class="first-level active"><a href="#gwaslabsumstats-methods">gwaslab.Sumstats methods</a></li>
            <li class="second-level"><a href="#glsumstats">gl.Sumstats</a></li>
                
            <li class="second-level"><a href="#standardization">Standardization</a></li>
                
            <li class="second-level"><a href="#qc-and-filtering">QC and filtering</a></li>
                
            <li class="second-level"><a href="#harmonization">Harmonization</a></li>
                
        <li class="first-level "><a href="#standalone-functions">Standalone functions</a></li>
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<h1 id="gwaslabsumstats-methods">gwaslab.Sumstats methods</h1>
<h2 id="glsumstats">gl.Sumstats</h2>
<h2 id="standardization">Standardization</h2>
<table>
<thead>
<tr>
<th>Sumstats Methods</th>
<th>Options</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>.fix_id()</code></td>
<td><code>fixchrpos=False</code>, <br/><code>fixid=False</code>, <br/><code>fixsep=False</code>,<br/><code>overwrite=False</code>,<br/><code>forcefixid=False</code></td>
<td>check and  fix rsID or SNPID(chr:pos:ref:alt), or use snpid to fix CHR and POS</td>
</tr>
<tr>
<td><code>.fix_CHR()</code></td>
<td><code>remove=False</code></td>
<td>standardize chromsome notation</td>
</tr>
<tr>
<td><code>.fix_POS()</code></td>
<td><code>remove=False</code></td>
<td>standardize basepair posituion notation and filter out bad values</td>
</tr>
<tr>
<td><code>.fix_allele()</code></td>
<td><code>remove=False</code></td>
<td>standardize base notation to ATCG</td>
</tr>
<tr>
<td><code>.normalize_allele()</code></td>
<td><code>n_cores=1</code></td>
<td>normalize indels (only support ATA:AA -&gt; AT:A but not -:T)</td>
</tr>
<tr>
<td><code>.sort_coordinate()</code></td>
<td></td>
<td>sort the variant coordinates</td>
</tr>
<tr>
<td><code>.sort_column()</code></td>
<td></td>
<td>sort the column order to gwaslab default</td>
</tr>
<tr>
<td><code>.basic_check()</code></td>
<td></td>
<td>all in one function</td>
</tr>
</tbody>
</table>
<h2 id="qc-and-filtering">QC and filtering</h2>
<table>
<thead>
<tr>
<th>Sumstats Methods</th>
<th>Options</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>.check_sanity()</code></td>
<td><code>n=(0,float("Inf"))</code>, <br/><code>eaf=(0,1)</code>, <br/><code>mac=(5,float("Inf"))</code>, <br/><code>chisq=(0,float("Inf"))</code>, <br/><code>p=(5e-300,1)</code>, <br/><code>mlog10p=(0,float("Inf"))</code>, <br/><code>beta=(-10,10)</code>, <br/><code>z=(-37.5,37.5)</code>, <br/><code>se=(0,float("Inf"))</code>, <br/><code>OR=(-10,10)</code> , <br/><code>OR_95L=(0,float("Inf"))</code>, <br/><code>OR_95U=(0,float("Inf"))</code>, <br/><code>info=(0,float("Inf"))</code></td>
<td>sanity check for statistics including BETA, SE, Z, CHISQ, EAF, OR, N...</td>
</tr>
<tr>
<td><code>.remove_dup()</code></td>
<td><code>mode="md"</code>, <br/><code>keep='first'</code>, <br/><code>keep_col="P"</code>, <br/><code>remove=False</code></td>
<td>remove duplicated, multiallelic or NA variants</td>
</tr>
<tr>
<td><code>.filter_value()</code></td>
<td>expr</td>
<td>filter in variants base on expr</td>
</tr>
<tr>
<td><code>.filter_in()</code></td>
<td>lt, gt, eq, inplace</td>
<td>filter in variants base on given threshold</td>
</tr>
<tr>
<td><code>.filter_out()</code></td>
<td>lt, gt, eq, inplace</td>
<td>filter out variants base on given threshold</td>
</tr>
<tr>
<td><code>.filter_region_in()</code></td>
<td><code>path</code> , <br/> <code>inplace=True</code> , <br/><code>high_ld=False</code>, <br/> <code>build="19"</code></td>
<td>filter in variants in the specified region define by a bed file</td>
</tr>
<tr>
<td><code>.filter_region_out()</code></td>
<td><code>path</code> , <br/> <code>inplace=True</code> , <br/><code>high_ld=False</code>, <br/> <code>build="19"</code></td>
<td>filter out variants in the specified region define by a bed file</td>
</tr>
</tbody>
</table>
<h2 id="harmonization">Harmonization</h2>
<table>
<thead>
<tr>
<th>Sumstats Methods</th>
<th>Options</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>.check_ref()</code></td>
<td>ref_path</td>
<td>check alignment with a reference sequence</td>
</tr>
<tr>
<td><code>.rsid_to_chrpos()</code></td>
<td>path, n_cores</td>
<td>use rsid to fill CHR and POS</td>
</tr>
<tr>
<td><code>.rsid_to_chrpos2()</code></td>
<td>path</td>
<td>use rsid to fill CHR and POS (muilti-thread, need hd5 file)</td>
</tr>
<tr>
<td><code>.assign_rsid()</code></td>
<td>path</td>
<td>annotate rsid using a reference vcf file</td>
</tr>
<tr>
<td><code>.infer_strand()</code></td>
<td>ref_infer="" , ref_alt_freq=None,  maf_threshold=0.43</td>
<td>infer the strand of a variant using reference vcf file with EAF in INFO</td>
</tr>
<tr>
<td><code>.check_daf()</code></td>
<td>ref_infer="" , ref_alt_freq=None,</td>
<td>calculate difference in allele frequencies</td>
</tr>
<tr>
<td><code>.flip_allele_stats()</code></td>
<td></td>
<td>After alignment and inferring, flip the alleles to harmonise the variants.</td>
</tr>
<tr>
<td><code>.liftover()</code></td>
<td>n_cores=1,from_build="19", to_build="38"</td>
<td>perform liftover</td>
</tr>
</tbody>
</table>
<h1 id="standalone-functions">Standalone functions</h1>
<table>
<thead>
<tr>
<th>function</th>
<th>catagory</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table></div>
        
        
    </div>

    
      <footer class="col-md-12 text-center">
          
          
            <hr>
            <p>
            <small>GWASLab is licensed under the MIT license</small><br>
            
            <small>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</small>
            </p>
          

          
          
      </footer>
    
    <script src="//ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="../js/bootstrap-3.0.3.min.js"></script>

    
    <script src="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/highlight.min.js"></script>
        
    <script>hljs.initHighlightingOnLoad();</script>
    

    <script>var base_url = ".."</script>
    
    <script src="../js/base.js"></script>

    <div class="modal" id="mkdocs_keyboard_modal" tabindex="-1" role="dialog" aria-labelledby="keyboardModalLabel" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
            </div>
            <div class="modal-body">
              <table class="table">
                <thead>
                  <tr>
                    <th style="width: 20%;">Keys</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td class="help shortcut"><kbd>?</kbd></td>
                    <td>Open this help</td>
                  </tr>
                  <tr>
                    <td class="next shortcut"><kbd>n</kbd></td>
                    <td>Next page</td>
                  </tr>
                  <tr>
                    <td class="prev shortcut"><kbd>p</kbd></td>
                    <td>Previous page</td>
                  </tr>
                  <tr>
                    <td class="search shortcut"><kbd>s</kbd></td>
                    <td>Search</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>
    </body>

</html>
