<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    
    <meta name="author" content="HE Yunye">
    
    <link rel="shortcut icon" href="img/favicon.ico">

    
    <title>Home - GWASLab</title>
    

    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.12.0/css/all.css">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.12.0/css/v4-shims.css">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/npm/hack-font@3.3.0/build/web/hack.min.css">
    <link href='//rsms.me/inter/inter.css' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,700italic,400,300,600,700&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href="css/bootstrap-custom.min.css" rel="stylesheet">
    <link href="css/base.min.css" rel="stylesheet">
    <link href="css/cinder.min.css" rel="stylesheet">

    
        
        <link rel="stylesheet" href="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/styles/github.min.css">
        
    
    <link href="stylesheets/extra.css" rel="stylesheet">

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js"></script>
            <script src="https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js"></script>
        <![endif]-->

    

     
</head>

<body class="homepage" >

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->

            
              <a class="navbar-brand" href=".">GWASLab</a>
            
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li class="active">
                        <a href=".">Home</a>
                    </li>
                
                
                
                    <li >
                        <a href="tutorial_3.3/">Tutorial for gwaslab</a>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Sumstats manipulation <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="SumstatsObject/">SumstatsObject</a>
</li>

                        
                            
<li >
    <a href="StatusCode/">StatusCode</a>
</li>

                        
                            
<li >
    <a href="Standardization/">Standardization&normalization</a>
</li>

                        
                            
<li >
    <a href="QC%26Filtering/">QC&Filtering</a>
</li>

                        
                            
<li >
    <a href="Harmonization/">Harmonization</a>
</li>

                        
                            
<li >
    <a href="LiftOver/">Liftover</a>
</li>

                        
                            
<li >
    <a href="Conversion/">Data conversion</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Visualization <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="Visualization/">Manhattan & QQ plot</a>
</li>

                        
                            
<li >
    <a href="RegionalPlot/">Regional Plot</a>
</li>

                        
                            
<li >
    <a href="BrisbanePlot/">Brisbane Plot</a>
</li>

                        
                            
<li >
    <a href="MiamiPlot/">Miami Plot</a>
</li>

                        
                            
<li >
    <a href="EffectSize/">Scatter: effect size comparison</a>
</li>

                        
                            
<li >
    <a href="AlleleFrequency/">Scatter: allele frequency comparison</a>
</li>

                        
                            
<li >
    <a href="GeneticCorrelation/">Heatmap: genetic correlation</a>
</li>

                        
                            
<li >
    <a href="ForestPlot/">Forest Plot</a>
</li>

                        
                            
<li >
    <a href="Gallery/">Gallery</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Utilities <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="ExtractLead/">Extract Lead Variants</a>
</li>

                        
                            
<li >
    <a href="ExtractNovel/">Extract Novel Variants</a>
</li>

                        
                            
<li >
    <a href="Format/">Format & Save</a>
</li>

                        
                            
<li >
    <a href="LoadLDSC/">Load LDSC log</a>
</li>

                        
                            
<li >
    <a href="HeritabilityConversion/">Convert Heritability</a>
</li>

                        
                            
<li >
    <a href="InferBuild/">Infer Genome Build</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">References <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="Download/">Download</a>
</li>

                        
                            
<li >
    <a href="Reference/">Reference data</a>
</li>

                        
                            
<li >
    <a href="CommonData/">Common data</a>
</li>

                        
                            
<li >
    <a href="UpdateLogs/">Update logs</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Examples <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Standardization and QC example</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="standardization_workflow/">Standardization Workflow</a>
</li>

        
            
<li >
    <a href="quality_control_and_filtering/">QC &amp; Filtering</a>
</li>

        
    </ul>
  </li>

                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Harmonization example</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="harmonization_workflow/">Workflow</a>
</li>

        
            
<li >
    <a href="harmonization_liftover/">Liftover</a>
</li>

        
    </ul>
  </li>

                        
                            
<li >
    <a href="format_load_save/">Formatting and saving</a>
</li>

                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Utility example</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="utility_data_conversion/">Data conversion</a>
</li>

        
            
<li >
    <a href="utility_get_novel/">Get novel variants</a>
</li>

        
    </ul>
  </li>

                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Visualization example</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="visualization_mqq/">Manhattan and Q-Q plot</a>
</li>

        
            
<li >
    <a href="visualization_miami/">Miami plot</a>
</li>

        
            
<li >
    <a href="visualization_regional/">Regional plot</a>
</li>

        
            
<li >
    <a href="visualization_brisbane/">Brisbane plot</a>
</li>

        
    </ul>
  </li>

                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Reference management</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="Download_reference/">Download</a>
</li>

        
    </ul>
  </li>

                        
                        </ul>
                    </li>
                
                
                </ul>

            <ul class="nav navbar-nav navbar-right">
                    <li class="disabled">
                        <a rel="prev" >
                            <i class="fas fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="next" href="tutorial_3.3/">
                            Next <i class="fas fa-arrow-right"></i>
                        </a>
                    </li>
                    <li>
                        <a href="https://github.com/Cloufield/gwaslab/"><i class="fab fa-github"></i> GitHub</a>
                    </li>
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
        <li class="first-level active"><a href="#gwaslab">gwaslab</a></li>
            <li class="second-level"><a href="#install">Install</a></li>
                
            <li class="second-level"><a href="#functions">Functions</a></li>
                
                <li class="third-level"><a href="#loading-and-formatting">Loading and Formatting</a></li>
                <li class="third-level"><a href="#standardization-normalization">Standardization &amp; Normalization</a></li>
                <li class="third-level"><a href="#quality-control-value-conversion-filtering">Quality control, Value conversion &amp; Filtering</a></li>
                <li class="third-level"><a href="#harmonization">Harmonization</a></li>
                <li class="third-level"><a href="#visualization">Visualization</a></li>
                <li class="third-level"><a href="#visualization-examples">Visualization Examples</a></li>
                <li class="third-level"><a href="#other-utilities">Other Utilities</a></li>
            <li class="second-level"><a href="#requirements">Requirements:</a></li>
                
        <li class="first-level "><a href="#contacts">Contacts</a></li>
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<p><img width="600" alt="image" src="https://user-images.githubusercontent.com/40289485/197167760-5f761f5e-5856-4b27-a540-8b9cd90bdadb.png"></p>
<h1 id="gwaslab">gwaslab</h1>
<p><img alt="badge" src="https://img.shields.io/badge/release-v3.3.9-brightgreen" />
<img alt="badge_pip" src="https://img.shields.io/pypi/dm/gwaslab" />
<a href="https://hits.seeyoufarm.com"><img alt="Hits" src="https://hits.seeyoufarm.com/api/count/incr/badge.svg?url=https%3A%2F%2Fcloufield.github.io%2Fgwaslab%2F&amp;count_bg=%2379C83D&amp;title_bg=%23555555&amp;icon=&amp;icon_color=%23E7E7E7&amp;title=hits&amp;edge_flat=false" /></a></p>
<ul>
<li>A simple python toolkit for handling GWAS sumstats.</li>
<li>Each process is modularized and can be customized to your needs.</li>
<li>Sumstats-specific manipulations are designed as methods of a python object, <code>gwaslab.Sumstats</code>.</li>
</ul>
<p>Please check GWASLab document at <a href="https://cloufield.github.io/gwaslab/">https://cloufield.github.io/gwaslab/</a>
Note: gwaslab is being updated very frequently for now. I will release the first stable version soon (within this year)! Please stay tuned.</p>
<h2 id="install">Install</h2>
<pre><code>pip install gwaslab==3.3.9
</code></pre>
<pre><code class="language-python">import gwaslab as gl
# load plink2 output
mysumstats = gl.Sumstats(&quot;t2d_bbj.txt.gz&quot;, fmt=&quot;plink2&quot;)

# or you can specify the columns:
mysumstats = gl.Sumstats(&quot;t2d_bbj.txt.gz&quot;,
             snpid=&quot;SNP&quot;,
             chrom=&quot;CHR&quot;,
             pos=&quot;POS&quot;,
             ea=&quot;ALT&quot;,
             nea=&quot;REF&quot;,
             neaf=&quot;Frq&quot;,
             beta=&quot;BETA&quot;,
             se=&quot;SE&quot;,
             p=&quot;P&quot;,
             direction=&quot;Dir&quot;,
             n=&quot;N&quot;,
             build=&quot;19&quot;)

# manhattan and qq plot
mysumstats.plot_mqq()
...
</code></pre>
<h2 id="functions">Functions</h2>
<h3 id="loading-and-formatting">Loading and Formatting</h3>
<ul>
<li>Loading sumstats by simply specifying the software name or format name.</li>
<li>Converting GWAS sumstats to specific formats:</li>
<li>LDSC / MAGMA / METAL / MR-MEGA / FUMA / VCF / BED... </li>
<li><a href="https://github.com/Cloufield/formatbook">check available formats</a></li>
<li>Optional Filtering of variants in Hapmap3 SNPs / High-LD regions / HLA region </li>
</ul>
<h3 id="standardization-normalization">Standardization &amp; Normalization</h3>
<ul>
<li>Variant ID standardization</li>
<li>CHR and POS notation standardization</li>
<li>Variant POS and allele normalization</li>
<li>Genome build : Inference and Liftover </li>
</ul>
<h3 id="quality-control-value-conversion-filtering">Quality control, Value conversion &amp; Filtering</h3>
<ul>
<li>Statistics sanity check</li>
<li>Extreme value removal</li>
<li>Equivalent statistics conversion<ul>
<li>BETA/SE , OR/OR_95L/OR_95U</li>
<li>P, Z, CHISQ, MLOG10</li>
</ul>
</li>
<li>Customized value filtering</li>
</ul>
<h3 id="harmonization">Harmonization</h3>
<ul>
<li>rsID assignment based on CHR, POS, and REF/ALT</li>
<li>CHR POS assignment based on rsID using a reference text file</li>
<li>Palindromic SNPs and indels strand inference using a reference VCF</li>
<li>Check allele frequency discrepancy using a reference VCF</li>
<li>Reference allele alignment using a reference genome sequence FASTA file</li>
</ul>
<h3 id="visualization">Visualization</h3>
<ul>
<li>Mqq plot : Manhattan plot , QQ plot or MQQ plot (with a bunch of customizable features including auto-annotate nearest gene names)</li>
<li>Miami plot : Manhattan plot</li>
<li>Brisbane plot:  GWAS hits density plot</li>
<li>Regional plot : GWAS regional plot</li>
<li>Heatmap : ldsc-rg genetic correlation matrix</li>
<li>Scatter Plot : variant effect size comparison with sumstats</li>
<li>Scatter Plot : allele frequency comparison </li>
<li>Forest Plot : forest plots for meta-analysis of SNPs</li>
</ul>
<h3 id="visualization-examples">Visualization Examples</h3>
<p><img width="700" alt="image" src="https://user-images.githubusercontent.com/40289485/195526882-dff70593-752d-4672-901e-0b3cea3d8cda.png"><img width="600" alt="image" src="https://user-images.githubusercontent.com/40289485/197393168-e3e7076f-2801-4d66-9526-80778d44f3da.png"><img width="600" alt="image" src="https://user-images.githubusercontent.com/40289485/197463243-89352749-f882-418d-907d-27530fd4e922.png"><img width="600" alt="image" src="https://user-images.githubusercontent.com/40289485/197126045-b1c55adf-3391-4c3d-b2f6-eaeac7c26024.png"></p>
<p><img width="400" alt="image" src="https://user-images.githubusercontent.com/40289485/195526481-df060ad5-dc61-4e35-ab37-3ea45ed00618.png"></p>
<h3 id="other-utilities">Other Utilities</h3>
<ul>
<li>Read ldsc h2 or rg outputs directly as DataFrames (auto-parsing).</li>
<li>Extract lead variants given a sliding window size.</li>
<li>Extract novel loci given a list of known lead variants.</li>
<li>Logging : keep a complete record of manipulations from raw data to munged data.</li>
<li>Sumstats summary function: know your data better. </li>
</ul>
<h2 id="requirements">Requirements:</h2>
<ul>
<li>Python &gt;= 3.6</li>
<li>pySAM </li>
<li>pyensembl</li>
<li>scikit-allel</li>
<li>Biopython &gt;= 1.79</li>
<li>liftover &gt;= 1.1.13</li>
<li>pandas &gt;= 1.2.4</li>
<li>numpy &gt;= 1.21.2</li>
<li>matplotlib &gt;=3.5</li>
<li>seaborn &gt;=0.11.1</li>
<li>scipy &gt;=1.6.2</li>
<li>statsmodels &gt; =0.13</li>
<li>adjustText</li>
</ul>
<h1 id="contacts">Contacts</h1>
<ul>
<li>Github: <a href="https://github.com/Cloufield/gwaslab">https://github.com/Cloufield/gwaslab</a></li>
<li>Blog (in Chinese): <a href="https://gwaslab.com/">https://gwaslab.com/</a></li>
<li>Email: gwaslab@gmail.com</li>
<li>Stats: <a href="https://pypistats.org/packages/gwaslab">https://pypistats.org/packages/gwaslab</a></li>
</ul></div>
        
        
    </div>

    
      <footer class="col-md-12 text-center">
          
          
            <hr>
            <p>
            <small>GWASLab is licensed under the MIT license</small><br>
            
            <small>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</small>
            </p>
          

          
          
      </footer>
    
    <script src="//ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="js/bootstrap-3.0.3.min.js"></script>

    
    <script src="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/highlight.min.js"></script>
        
    <script>hljs.initHighlightingOnLoad();</script>
    

    <script>var base_url = "."</script>
    
    <script src="js/base.js"></script>

    <div class="modal" id="mkdocs_keyboard_modal" tabindex="-1" role="dialog" aria-labelledby="keyboardModalLabel" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
            </div>
            <div class="modal-body">
              <table class="table">
                <thead>
                  <tr>
                    <th style="width: 20%;">Keys</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td class="help shortcut"><kbd>?</kbd></td>
                    <td>Open this help</td>
                  </tr>
                  <tr>
                    <td class="next shortcut"><kbd>n</kbd></td>
                    <td>Next page</td>
                  </tr>
                  <tr>
                    <td class="prev shortcut"><kbd>p</kbd></td>
                    <td>Previous page</td>
                  </tr>
                  <tr>
                    <td class="search shortcut"><kbd>s</kbd></td>
                    <td>Search</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>
    </body>

</html>

<!--
MkDocs version : 1.3.0
Build Date UTC : 2022-11-05 09:50:49.615911+00:00
-->
