{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "deaf264d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-12-23 16:48:00--  https://koges.leelabsg.org/download/KoGES_BMI\n",
      "Resolving koges.leelabsg.org (koges.leelabsg.org)... 35.192.148.108\n",
      "Connecting to koges.leelabsg.org (koges.leelabsg.org)|35.192.148.108|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 158394405 (151M) [text/tab-separated-values]\n",
      "Saving to: ‘../../0_sample_data/koges_bmi.tsv.gz’\n",
      "\n",
      "../../0_sample_data 100%[===================>] 151.06M  18.8MB/s    in 20s     \n",
      "\n",
      "2024-12-23 16:48:21 (7.54 MB/s) - ‘../../0_sample_data/koges_bmi.tsv.gz’ saved [158394405/158394405]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#!wget -O \"../../0_sample_data/koges_bmi.tsv.gz\" https://koges.leelabsg.org/download/KoGES_BMI"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d88da0c",
   "metadata": {},
   "source": [
    "## load gwaslab and sample data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d3f8dcc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0,\"/home/yunye/work/gwaslab/src\")\n",
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c6adbd0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mysumstats = gl.Sumstats(\"../../0_sample_data/koges_bmi.tsv.gz\", \n",
    "                             chrom=\"chrom\",\n",
    "                             pos=\"pos\",\n",
    "                             ea=\"alt\",\n",
    "                             nea=\"ref\",\n",
    "                             eaf=\"af\",\n",
    "                             beta=\"beta\",\n",
    "                             se=\"sebeta\",\n",
    "                             p=\"pval\",nrows=1000000,\n",
    "                            verbose=False) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0164ab54",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024/12/24 12:35:09 Start to check SNPID/rsID...v3.5.4\n",
      "2024/12/24 12:35:09  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 75.34 MB\n",
      "2024/12/24 12:35:09 Finished checking SNPID/rsID.\n",
      "2024/12/24 12:35:10 Start to fix chromosome notation (CHR)...v3.5.4\n",
      "2024/12/24 12:35:10  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 75.34 MB\n",
      "2024/12/24 12:35:10  -Checking CHR data type...\n",
      "2024/12/24 12:35:10  -Variants with standardized chromosome notation: 1000000\n",
      "2024/12/24 12:35:10  -All CHR are already fixed...\n",
      "2024/12/24 12:35:11 Finished fixing chromosome notation (CHR).\n",
      "2024/12/24 12:35:11 Start to fix basepair positions (POS)...v3.5.4\n",
      "2024/12/24 12:35:11  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 76.29 MB\n",
      "2024/12/24 12:35:11  -Converting to Int64 data type ...\n",
      "2024/12/24 12:35:13  -Position bound:(0 , 250,000,000)\n",
      "2024/12/24 12:35:13  -Removed outliers: 0\n",
      "2024/12/24 12:35:13 Finished fixing basepair positions (POS).\n",
      "2024/12/24 12:35:13 Start to fix alleles (EA and NEA)...v3.5.4\n",
      "2024/12/24 12:35:13  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 84.88 MB\n",
      "2024/12/24 12:35:13  -Converted all bases to string datatype and UPPERCASE.\n",
      "2024/12/24 12:35:13  -Variants with bad EA  : 464\n",
      "2024/12/24 12:35:13  -Variants with bad NEA : 0\n",
      "2024/12/24 12:35:13  -Variants with NA for EA or NEA: 0\n",
      "2024/12/24 12:35:13  -Variants with same EA and NEA: 0\n",
      "2024/12/24 12:35:13  -A look at the non-ATCG EA: {'<CN0>', '<INS:ME:ALU>'} ...\n",
      "2024/12/24 12:35:13  -Detected 464 variants with alleles that contain bases other than A/C/T/G .\n",
      "2024/12/24 12:35:18 Finished fixing alleles (EA and NEA).\n",
      "2024/12/24 12:35:18 Start to perform sanity check for statistics...v3.5.4\n",
      "2024/12/24 12:35:18  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 85.07 MB\n",
      "2024/12/24 12:35:18  -Comparison tolerance for floats: 1e-07\n",
      "2024/12/24 12:35:18  -Checking if -1e-07 < EAF < 1.0000001 ...\n",
      "2024/12/24 12:35:18  -Removed 0 variants with bad/na EAF.\n",
      "2024/12/24 12:35:18  -Checking if -1e-07 < P < 1.0000001 ...\n",
      "2024/12/24 12:35:18  -Removed 0 variants with bad/na P.\n",
      "2024/12/24 12:35:18  -Checking if -100.0000001 < BETA < 100.0000001 ...\n",
      "2024/12/24 12:35:18  -Removed 0 variants with bad/na BETA.\n",
      "2024/12/24 12:35:18  -Checking if -1e-07 < SE < inf ...\n",
      "2024/12/24 12:35:18  -Removed 0 variants with bad/na SE.\n",
      "2024/12/24 12:35:18  -Checking STATUS and converting STATUS to categories....\n",
      "2024/12/24 12:35:19  -Removed 0 variants with bad statistics in total.\n",
      "2024/12/24 12:35:19  -Data types for each column:\n",
      "2024/12/24 12:35:19  -Column  : CHR   POS   EA       NEA      EAF     BETA    SE      P       STATUS  \n",
      "2024/12/24 12:35:19  -DType   : Int64 Int64 category category float32 float64 float64 float64 category\n",
      "2024/12/24 12:35:19  -Verified: T     T     T        T        T       T       T       T       T       \n",
      "2024/12/24 12:35:19 Finished sanity check for statistics.\n",
      "2024/12/24 12:35:19 Start to check data consistency across columns...v3.5.4\n",
      "2024/12/24 12:35:19  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 81.25 MB\n",
      "2024/12/24 12:35:19  -Tolerance: 0.001 (Relative) and 0.001 (Absolute)\n",
      "2024/12/24 12:35:19  -SNPID/rsID not available...SKipping\n",
      "2024/12/24 12:35:19 Finished checking data consistency across columns.\n",
      "2024/12/24 12:35:19 Start to normalize indels...v3.5.4\n",
      "2024/12/24 12:35:19  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 81.25 MB\n",
      "2024/12/24 12:35:20  -Not normalized allele IDs:[656] [1089] [1592] [2366] [3221] ... \n",
      "2024/12/24 12:35:20  -Not normalized allele:['TTGTGTG' 'TTGTG']['TTG' 'CTG']['TAC' 'TACAC']['CAGAGAGAG' 'CAGAGAG']['TGT' 'CGT']... \n",
      "2024/12/24 12:35:20  -Modified 3787 variants according to parsimony and left alignment principal.\n",
      "2024/12/24 12:35:20 Finished normalizing indels.\n",
      "2024/12/24 12:35:20 Start to sort the genome coordinates...v3.5.4\n",
      "2024/12/24 12:35:20  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 81.25 MB\n",
      "2024/12/24 12:35:20 Finished sorting coordinates.\n",
      "2024/12/24 12:35:20 Start to reorder the columns...v3.5.4\n",
      "2024/12/24 12:35:20  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 73.62 MB\n",
      "2024/12/24 12:35:20  -Reordering columns to    : CHR,POS,EA,NEA,EAF,BETA,SE,P,STATUS\n",
      "2024/12/24 12:35:20 Finished reordering the columns.\n"
     ]
    }
   ],
   "source": [
    "mysumstats.basic_check(verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c399f770",
   "metadata": {},
   "outputs": [],
   "source": [
    "mysumstats.fix_id(fixid=True,forcefixid=True,overwrite=True,verbose=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "aba4feb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>SNPID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>714646</td>\n",
       "      <td>T</td>\n",
       "      <td>TTTGAAAC</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.0093</td>\n",
       "      <td>0.0170</td>\n",
       "      <td>0.5900</td>\n",
       "      <td>9960399</td>\n",
       "      <td>1:714646:TTTGAAAC:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>725256</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.027</td>\n",
       "      <td>0.0076</td>\n",
       "      <td>0.0170</td>\n",
       "      <td>0.6500</td>\n",
       "      <td>9960099</td>\n",
       "      <td>1:725256:C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>726805</td>\n",
       "      <td>GA</td>\n",
       "      <td>G</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>0.0180</td>\n",
       "      <td>0.6600</td>\n",
       "      <td>9960399</td>\n",
       "      <td>1:726805:G:GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>726807</td>\n",
       "      <td>T</td>\n",
       "      <td>TTG</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>0.0180</td>\n",
       "      <td>0.6600</td>\n",
       "      <td>9960399</td>\n",
       "      <td>1:726807:TTG:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>726812</td>\n",
       "      <td>AG</td>\n",
       "      <td>A</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>0.0180</td>\n",
       "      <td>0.6600</td>\n",
       "      <td>9960399</td>\n",
       "      <td>1:726812:A:AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999995</th>\n",
       "      <td>2</td>\n",
       "      <td>142257738</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.020</td>\n",
       "      <td>-0.0030</td>\n",
       "      <td>0.0180</td>\n",
       "      <td>0.8700</td>\n",
       "      <td>9960099</td>\n",
       "      <td>2:142257738:G:A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999996</th>\n",
       "      <td>2</td>\n",
       "      <td>142257971</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.540</td>\n",
       "      <td>-0.0057</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.2700</td>\n",
       "      <td>9960099</td>\n",
       "      <td>2:142257971:T:G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999997</th>\n",
       "      <td>2</td>\n",
       "      <td>142258001</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.370</td>\n",
       "      <td>0.0140</td>\n",
       "      <td>0.0053</td>\n",
       "      <td>0.0068</td>\n",
       "      <td>9960099</td>\n",
       "      <td>2:142258001:G:A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999998</th>\n",
       "      <td>2</td>\n",
       "      <td>142258020</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.0093</td>\n",
       "      <td>0.0230</td>\n",
       "      <td>9960099</td>\n",
       "      <td>2:142258020:G:A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999999</th>\n",
       "      <td>2</td>\n",
       "      <td>142258089</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.530</td>\n",
       "      <td>-0.0110</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.0330</td>\n",
       "      <td>9960099</td>\n",
       "      <td>2:142258089:A:G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000000 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHR        POS  EA       NEA    EAF    BETA      SE       P   STATUS  \\\n",
       "0         1     714646   T  TTTGAAAC  0.026  0.0093  0.0170  0.5900  9960399   \n",
       "1         1     725256   T         C  0.027  0.0076  0.0170  0.6500  9960099   \n",
       "2         1     726805  GA         G  0.026  0.0078  0.0180  0.6600  9960399   \n",
       "3         1     726807   T       TTG  0.026  0.0078  0.0180  0.6600  9960399   \n",
       "4         1     726812  AG         A  0.026  0.0078  0.0180  0.6600  9960399   \n",
       "...     ...        ...  ..       ...    ...     ...     ...     ...      ...   \n",
       "999995    2  142257738   A         G  0.020 -0.0030  0.0180  0.8700  9960099   \n",
       "999996    2  142257971   G         T  0.540 -0.0057  0.0051  0.2700  9960099   \n",
       "999997    2  142258001   A         G  0.370  0.0140  0.0053  0.0068  9960099   \n",
       "999998    2  142258020   A         G  0.082  0.0210  0.0093  0.0230  9960099   \n",
       "999999    2  142258089   G         A  0.530 -0.0110  0.0051  0.0330  9960099   \n",
       "\n",
       "                      SNPID  \n",
       "0       1:714646:TTTGAAAC:T  \n",
       "1              1:725256:C:T  \n",
       "2             1:726805:G:GA  \n",
       "3            1:726807:TTG:T  \n",
       "4             1:726812:A:AG  \n",
       "...                     ...  \n",
       "999995      2:142257738:G:A  \n",
       "999996      2:142257971:T:G  \n",
       "999997      2:142258001:G:A  \n",
       "999998      2:142258020:G:A  \n",
       "999999      2:142258089:A:G  \n",
       "\n",
       "[1000000 rows x 10 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "771863c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "mysumstats2 = gl.Sumstats(\"../../0_sample_data/t2d_bbj.txt.gz\", \n",
    "                             snpid=\"SNP\",\n",
    "                             chrom=\"CHR\",\n",
    "                             pos=\"POS\",\n",
    "                             ea=\"REF\",\n",
    "                             nea=\"ALT\",\n",
    "                             eaf=\"Frq\",\n",
    "                             beta=\"BETA\",\n",
    "                             se=\"SE\",\n",
    "                             p=\"P\",\n",
    "                             nrows=1000000,\n",
    "                             verbose=False) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4f806e05",
   "metadata": {},
   "outputs": [],
   "source": [
    "mysumstats2.basic_check(verbose=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cb44314",
   "metadata": {},
   "source": [
    "## create a SumstatsPair object"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46128536",
   "metadata": {},
   "source": [
    "Allele and statistics in sumstats2 will be aligned using sumstats1 as reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "248584f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024/12/24 12:35:37 Start to create SumstatsPair object...\n",
      "2024/12/24 12:35:37  -Checking sumstats 1...\n",
      "2024/12/24 12:35:37  -Column  : CHR   POS   EA       NEA      EAF     BETA    SE      P       STATUS   SNPID \n",
      "2024/12/24 12:35:37  -DType   : Int64 Int64 category category float32 float64 float64 float64 category string\n",
      "2024/12/24 12:35:37  -Verified: T     T     T        T        T       T       T       T       T        T     \n",
      "2024/12/24 12:35:37  -Current Dataframe shape : 1000000 x 10 ; Memory usage: 81.25 MB\n",
      "2024/12/24 12:35:37  -Checking sumstats 2...\n",
      "2024/12/24 12:35:37  -Column  : SNPID  CHR   POS   EA       NEA      EAF     BETA    SE      P       STATUS  \n",
      "2024/12/24 12:35:37  -DType   : string Int64 Int64 category category float32 float64 float64 float64 category\n",
      "2024/12/24 12:35:37  -Verified: T      T     T     T        T        T       T       T       T       T       \n",
      "2024/12/24 12:35:37  -Current Dataframe shape : 1000000 x 10 ; Memory usage: 81.23 MB\n",
      "2024/12/24 12:35:37  -Variant Info columns: ['CHR', 'POS', 'EA', 'NEA', 'STATUS', 'SNPID']\n",
      "2024/12/24 12:35:37  -Variant statistics columns: ['EAF', 'BETA', 'SE', 'P']\n",
      "2024/12/24 12:35:37  -Sumstats1 other columns: []\n",
      "2024/12/24 12:35:37  -Sumstats2 other columns: []\n",
      "2024/12/24 12:35:37 Start to merge sumstats...\n",
      "2024/12/24 12:35:37  -Dropping old IDs:['SNPID']\n",
      "2024/12/24 12:35:37  -After merging by CHR and POS:613216\n",
      "2024/12/24 12:35:37  -Matching alleles and keeping only variants with same allele set: \n",
      "2024/12/24 12:35:37   -Perfect match: 330147\n",
      "2024/12/24 12:35:38   -Flipped match: 282001\n",
      "2024/12/24 12:35:38   -Unmatched : 1068\n",
      "2024/12/24 12:35:38  -Matched variants:612148\n",
      "2024/12/24 12:35:38  -Aligning alleles with reference: \n",
      "2024/12/24 12:35:38   -Perfect match: 330147\n",
      "2024/12/24 12:35:38   -Flipped match: 282001\n",
      "2024/12/24 12:35:38   -For perfect match: copy STATUS from reference...\n",
      "2024/12/24 12:35:38   -For Flipped match: convert STATUS xxxxx[456789]x to xxxxx3x...\n",
      "2024/12/24 12:35:39 Start to adjust statistics based on STATUS code...v3.5.4\n",
      "2024/12/24 12:35:39  -Current Dataframe shape : 612148 x 17 ; Memory usage: 106.02 MB\n",
      "2024/12/24 12:35:40 Start to flip allele-specific stats for SNPs with status xxxxx[35]x: ALT->EA , REF->NEA ...v3.5.4\n",
      "2024/12/24 12:35:40  -Flipping 282001 variants...\n",
      "2024/12/24 12:35:40  -Swapping column: NEA <=> EA...\n",
      "2024/12/24 12:35:40  -Flipping column: BETA = - BETA...\n",
      "2024/12/24 12:35:40  -Flipping column: EAF = 1 - EAF...\n",
      "2024/12/24 12:35:40  -Changed the status for flipped variants : xxxxx[35]x -> xxxxx[12]x\n",
      "2024/12/24 12:35:41 Finished adjusting statistics based on STATUS code.\n",
      "2024/12/24 12:35:41  -Renaming sumstats2 columns by adding suffix _2\n",
      "2024/12/24 12:35:41 Start to reorder the columns...\n",
      "2024/12/24 12:35:41  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,EAF_1,BETA_1,SE_1,P_1,STATUS_1,EAF_2,BETA_2,SE_2,P_2,STATUS_2\n",
      "2024/12/24 12:35:41 Finished sorting columns successfully!\n"
     ]
    }
   ],
   "source": [
    "pair = gl.SumstatsPair(mysumstats, mysumstats2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0638d2bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF_1</th>\n",
       "      <th>BETA_1</th>\n",
       "      <th>SE_1</th>\n",
       "      <th>P_1</th>\n",
       "      <th>STATUS_1</th>\n",
       "      <th>EAF_2</th>\n",
       "      <th>BETA_2</th>\n",
       "      <th>SE_2</th>\n",
       "      <th>P_2</th>\n",
       "      <th>STATUS_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:751343:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>751343</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.16</td>\n",
       "      <td>-0.00160</td>\n",
       "      <td>0.0074</td>\n",
       "      <td>0.82</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.1407</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:751488:G:GA</td>\n",
       "      <td>1</td>\n",
       "      <td>751488</td>\n",
       "      <td>GA</td>\n",
       "      <td>G</td>\n",
       "      <td>0.16</td>\n",
       "      <td>-0.00390</td>\n",
       "      <td>0.0077</td>\n",
       "      <td>0.61</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.1470</td>\n",
       "      <td>0.0187</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0.28300</td>\n",
       "      <td>9960319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:751756:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>751756</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.16</td>\n",
       "      <td>-0.00150</td>\n",
       "      <td>0.0074</td>\n",
       "      <td>0.84</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.1400</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>9960019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:752566:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>752566</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.00091</td>\n",
       "      <td>0.0073</td>\n",
       "      <td>0.90</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.8422</td>\n",
       "      <td>-0.0155</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.23500</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:752894:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>752894</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.00130</td>\n",
       "      <td>0.0074</td>\n",
       "      <td>0.86</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.8560</td>\n",
       "      <td>-0.0169</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.25250</td>\n",
       "      <td>9960019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613211</th>\n",
       "      <td>2:6347478:T:C</td>\n",
       "      <td>2</td>\n",
       "      <td>6347478</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.22</td>\n",
       "      <td>-0.00270</td>\n",
       "      <td>0.0063</td>\n",
       "      <td>0.67</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.2093</td>\n",
       "      <td>-0.0149</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>0.17790</td>\n",
       "      <td>9960019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613212</th>\n",
       "      <td>2:6347639:C:A</td>\n",
       "      <td>2</td>\n",
       "      <td>6347639</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.22</td>\n",
       "      <td>-0.00280</td>\n",
       "      <td>0.0063</td>\n",
       "      <td>0.66</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.2100</td>\n",
       "      <td>-0.0159</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>0.15100</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613213</th>\n",
       "      <td>2:6347694:G:C</td>\n",
       "      <td>2</td>\n",
       "      <td>6347694</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.22</td>\n",
       "      <td>-0.00270</td>\n",
       "      <td>0.0063</td>\n",
       "      <td>0.67</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.2093</td>\n",
       "      <td>-0.0151</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>0.17250</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613214</th>\n",
       "      <td>2:6348478:G:A</td>\n",
       "      <td>2</td>\n",
       "      <td>6348478</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.22</td>\n",
       "      <td>-0.00280</td>\n",
       "      <td>0.0063</td>\n",
       "      <td>0.65</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.2140</td>\n",
       "      <td>-0.0152</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>0.17160</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613215</th>\n",
       "      <td>2:6348754:A:C</td>\n",
       "      <td>2</td>\n",
       "      <td>6348754</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.14</td>\n",
       "      <td>-0.00350</td>\n",
       "      <td>0.0073</td>\n",
       "      <td>0.64</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.1572</td>\n",
       "      <td>-0.0297</td>\n",
       "      <td>0.0126</td>\n",
       "      <td>0.01846</td>\n",
       "      <td>9960019</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>612148 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                SNPID  CHR      POS  EA NEA  EAF_1   BETA_1    SE_1   P_1  \\\n",
       "0        1:751343:T:A    1   751343   A   T   0.16 -0.00160  0.0074  0.82   \n",
       "1       1:751488:G:GA    1   751488  GA   G   0.16 -0.00390  0.0077  0.61   \n",
       "2        1:751756:T:C    1   751756   C   T   0.16 -0.00150  0.0074  0.84   \n",
       "3        1:752566:G:A    1   752566   A   G   0.83  0.00091  0.0073  0.90   \n",
       "4        1:752894:T:C    1   752894   C   T   0.84  0.00130  0.0074  0.86   \n",
       "...               ...  ...      ...  ..  ..    ...      ...     ...   ...   \n",
       "613211  2:6347478:T:C    2  6347478   C   T   0.22 -0.00270  0.0063  0.67   \n",
       "613212  2:6347639:C:A    2  6347639   A   C   0.22 -0.00280  0.0063  0.66   \n",
       "613213  2:6347694:G:C    2  6347694   C   G   0.22 -0.00270  0.0063  0.67   \n",
       "613214  2:6348478:G:A    2  6348478   A   G   0.22 -0.00280  0.0063  0.65   \n",
       "613215  2:6348754:A:C    2  6348754   C   A   0.14 -0.00350  0.0073  0.64   \n",
       "\n",
       "       STATUS_1   EAF_2  BETA_2    SE_2      P_2 STATUS_2  \n",
       "0       9960099  0.1407  0.0172  0.0156  0.27050  9960099  \n",
       "1       9960399  0.1470  0.0187  0.0174  0.28300  9960319  \n",
       "2       9960099  0.1400  0.0172  0.0156  0.27050  9960019  \n",
       "3       9960099  0.8422 -0.0155  0.0131  0.23500  9960099  \n",
       "4       9960099  0.8560 -0.0169  0.0147  0.25250  9960019  \n",
       "...         ...     ...     ...     ...      ...      ...  \n",
       "613211  9960099  0.2093 -0.0149  0.0111  0.17790  9960019  \n",
       "613212  9960099  0.2100 -0.0159  0.0111  0.15100  9960099  \n",
       "613213  9960099  0.2093 -0.0151  0.0111  0.17250  9960099  \n",
       "613214  9960099  0.2140 -0.0152  0.0111  0.17160  9960099  \n",
       "613215  9960099  0.1572 -0.0297  0.0126  0.01846  9960019  \n",
       "\n",
       "[612148 rows x 15 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pair.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "337be427",
   "metadata": {},
   "source": [
    "### Run clumping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f8ae3443",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024/12/24 12:35:41 Start to perfrom clumping...v3.5.4\n",
      "2024/12/24 12:35:41  -Current Dataframe shape : 612148 x 15 ; Memory usage: 103.08 MB\n",
      "2024/12/24 12:35:41 Start to perform clumping...\n",
      "2024/12/24 12:35:41  -Clumping parameters for PLINK2:\n",
      "2024/12/24 12:35:41   -clump_p1 : 5e-08...\n",
      "2024/12/24 12:35:41   -clump_p2 : 5e-08...\n",
      "2024/12/24 12:35:41   -clump_kb : 250...\n",
      "2024/12/24 12:35:41   -clump_r2 : 0.01...\n",
      "2024/12/24 12:35:41  -Clumping will be performed using P_1\n",
      "2024/12/24 12:35:41  -Significant variants on CHR:  [1, 2]\n",
      "2024/12/24 12:35:41  -Processing VCF : /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz...\n",
      "2024/12/24 12:35:41  -PLINK version: PLINK v2.00a5.9LM AVX2 AMD (12 Dec 2023)\n",
      "2024/12/24 12:35:41   -Processing VCF for CHR 1...\n",
      "2024/12/24 12:35:41   -Plink bfile for CHR 1 exists: /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.1. Skipping...\n",
      "2024/12/24 12:35:41   -Processing VCF for CHR 2...\n",
      "2024/12/24 12:35:41   -Plink bfile for CHR 2 exists: /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.2. Skipping...\n",
      "2024/12/24 12:35:41  -Processing sumstats for CHR 1...\n",
      "2024/12/24 12:35:46  -Variants in reference file: 6471908...\n",
      "2024/12/24 12:35:46  -Variants in sumstats: 54...\n",
      "2024/12/24 12:35:46  -Variants available in both reference and sumstats: 54...\n",
      "2024/12/24 12:35:46  -Processing sumstats for CHR 2...\n",
      "2024/12/24 12:35:50  -Variants in reference file: 7085912...\n",
      "2024/12/24 12:35:50  -Variants in sumstats: 235...\n",
      "2024/12/24 12:35:50  -Variants available in both reference and sumstats: 235...\n",
      "2024/12/24 12:35:50  -Performing clumping for CHR 1...\n",
      "2024/12/24 12:35:50  -PLINK version: PLINK v2.00a5.9LM AVX2 AMD (12 Dec 2023)\n",
      "2024/12/24 12:35:51  -Saved results for CHR 1 to : clumping_plink2.1.clumps\n",
      "2024/12/24 12:35:51  -Performing clumping for CHR 2...\n",
      "2024/12/24 12:35:51  -PLINK version: PLINK v2.00a5.9LM AVX2 AMD (12 Dec 2023)\n",
      "2024/12/24 12:35:52  -Saved results for CHR 2 to : clumping_plink2.2.clumps\n",
      "2024/12/24 12:35:52 Finished clumping.\n",
      "2024/12/24 12:35:52 Finished clumping.\n"
     ]
    }
   ],
   "source": [
    "pair.clump(vcf = gl.get_path(\"1kg_eas_hg19\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "36934df9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF_1</th>\n",
       "      <th>BETA_1</th>\n",
       "      <th>SE_1</th>\n",
       "      <th>P_1</th>\n",
       "      <th>STATUS_1</th>\n",
       "      <th>EAF_2</th>\n",
       "      <th>BETA_2</th>\n",
       "      <th>SE_2</th>\n",
       "      <th>P_2</th>\n",
       "      <th>STATUS_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>386282</th>\n",
       "      <td>1:177894591:T:G</td>\n",
       "      <td>1</td>\n",
       "      <td>177894591</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.280</td>\n",
       "      <td>0.065</td>\n",
       "      <td>0.0057</td>\n",
       "      <td>7.800000e-31</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.2650</td>\n",
       "      <td>0.0382</td>\n",
       "      <td>0.0102</td>\n",
       "      <td>1.725000e-04</td>\n",
       "      <td>9960019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593792</th>\n",
       "      <td>2:527882:C:G</td>\n",
       "      <td>2</td>\n",
       "      <td>527882</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.917</td>\n",
       "      <td>-0.053</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>2.100000e-08</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.9286</td>\n",
       "      <td>-0.0476</td>\n",
       "      <td>0.0175</td>\n",
       "      <td>6.409000e-03</td>\n",
       "      <td>9960019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>594310</th>\n",
       "      <td>2:632723:C:T</td>\n",
       "      <td>2</td>\n",
       "      <td>632723</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.907</td>\n",
       "      <td>0.070</td>\n",
       "      <td>0.0088</td>\n",
       "      <td>1.900000e-15</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.9021</td>\n",
       "      <td>0.0937</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>3.940000e-10</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  SNPID  CHR        POS EA NEA  EAF_1  BETA_1    SE_1  \\\n",
       "386282  1:177894591:T:G    1  177894591  G   T  0.280   0.065  0.0057   \n",
       "593792     2:527882:C:G    2     527882  G   C  0.917  -0.053  0.0095   \n",
       "594310     2:632723:C:T    2     632723  T   C  0.907   0.070  0.0088   \n",
       "\n",
       "                 P_1 STATUS_1   EAF_2  BETA_2    SE_2           P_2 STATUS_2  \n",
       "386282  7.800000e-31  9960099  0.2650  0.0382  0.0102  1.725000e-04  9960019  \n",
       "593792  2.100000e-08  9960099  0.9286 -0.0476  0.0175  6.409000e-03  9960019  \n",
       "594310  1.900000e-15  9960099  0.9021  0.0937  0.0150  3.940000e-10  9960099  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pair.clumps[\"clumps\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc040cca",
   "metadata": {},
   "source": [
    "### Run two sample MR by calling twosampleMR  in R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "dfca0b6c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024/12/24 12:35:52  Start to run MR using twosampleMR from command line:\n",
      "2024/12/24 12:35:52  - ncase2, ncontrol2, prevalence2:(40000, 120000, 0.1)\n",
      "2024/12/24 12:35:52 Start to calculate per-SNP heritibility...\n",
      "2024/12/24 12:35:52  -Calculating per-SNP rsq by 2 * (BETA**2) * AF * (1-AF) / Var(y)...\n",
      "2024/12/24 12:35:52  -Var(y) is provided: 1...\n",
      "2024/12/24 12:35:52  -Calculating F-statistic: F = [(N-k-1)/k] * (r2/1-r2)... where k = 1\n",
      "2024/12/24 12:35:52  -For r2, SNPR2 is used.\n",
      "2024/12/24 12:35:52 Finished calculating per-SNP heritability!\n",
      "2024/12/24 12:35:52 Filtered out 0 variants with F < 10\n",
      "2024/12/24 12:35:52  -R version: R scripting front-end version 4.1.3 (2022-03-10)\n",
      "2024/12/24 12:35:52  Running TwoSampleMR from command line...\n",
      "2024/12/24 12:35:55 TwoSampleMR version 0.5.8 \n",
      "[>] New: Option to use non-European LD reference panels for clumping etc\n",
      "[>] Some studies temporarily quarantined to verify effect allele\n",
      "[>] See news(package='TwoSampleMR') and https://gwas.mrcieu.ac.uk for further details\n",
      "\n",
      "\n",
      "Warning:\n",
      "You are running an old version of the TwoSampleMR package.\n",
      "This version:   0.5.8\n",
      "Latest version: 0.6.8\n",
      "Please consider updating using remotes::install_github('MRCIEU/TwoSampleMR')\n",
      "Harmonising STUDY1 (H4lJZt) and STUDY2 (aLOCCI)\n",
      "Analysing 'H4lJZt' on 'aLOCCI'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pair.run_two_sample_mr(clump=True, \n",
    "                       f_check=10,\n",
    "                       r=\"/home/yunye/anaconda3/envs/r41/bin/Rscript\",\n",
    "                       n1=72000,\n",
    "                       cck2=(40000,120000,0.1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "37dbf1fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id.exposure</th>\n",
       "      <th>id.outcome</th>\n",
       "      <th>outcome</th>\n",
       "      <th>exposure</th>\n",
       "      <th>method</th>\n",
       "      <th>nsnp</th>\n",
       "      <th>b</th>\n",
       "      <th>se</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>STUDY1</td>\n",
       "      <td>Inverse variance weighted</td>\n",
       "      <td>3</td>\n",
       "      <td>0.855997</td>\n",
       "      <td>0.236592</td>\n",
       "      <td>0.000297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>STUDY1</td>\n",
       "      <td>Simple mode</td>\n",
       "      <td>3</td>\n",
       "      <td>0.816022</td>\n",
       "      <td>0.320336</td>\n",
       "      <td>0.125697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>STUDY1</td>\n",
       "      <td>Weighted median</td>\n",
       "      <td>3</td>\n",
       "      <td>0.734788</td>\n",
       "      <td>0.164824</td>\n",
       "      <td>0.000008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>STUDY1</td>\n",
       "      <td>MR Egger</td>\n",
       "      <td>3</td>\n",
       "      <td>2.012733</td>\n",
       "      <td>3.582989</td>\n",
       "      <td>0.674167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>STUDY1</td>\n",
       "      <td>Inverse variance weighted (multiplicative rand...</td>\n",
       "      <td>3</td>\n",
       "      <td>0.855997</td>\n",
       "      <td>0.236592</td>\n",
       "      <td>0.000297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>STUDY1</td>\n",
       "      <td>Weighted mode</td>\n",
       "      <td>3</td>\n",
       "      <td>0.629216</td>\n",
       "      <td>0.167361</td>\n",
       "      <td>0.064027</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  id.exposure id.outcome outcome exposure  \\\n",
       "0      H4lJZt     aLOCCI  STUDY2   STUDY1   \n",
       "1      H4lJZt     aLOCCI  STUDY2   STUDY1   \n",
       "2      H4lJZt     aLOCCI  STUDY2   STUDY1   \n",
       "3      H4lJZt     aLOCCI  STUDY2   STUDY1   \n",
       "4      H4lJZt     aLOCCI  STUDY2   STUDY1   \n",
       "5      H4lJZt     aLOCCI  STUDY2   STUDY1   \n",
       "\n",
       "                                              method  nsnp         b  \\\n",
       "0                          Inverse variance weighted     3  0.855997   \n",
       "1                                        Simple mode     3  0.816022   \n",
       "2                                    Weighted median     3  0.734788   \n",
       "3                                           MR Egger     3  2.012733   \n",
       "4  Inverse variance weighted (multiplicative rand...     3  0.855997   \n",
       "5                                      Weighted mode     3  0.629216   \n",
       "\n",
       "         se      pval  \n",
       "0  0.236592  0.000297  \n",
       "1  0.320336  0.125697  \n",
       "2  0.164824  0.000008  \n",
       "3  3.582989  0.674167  \n",
       "4  0.236592  0.000297  \n",
       "5  0.167361  0.064027  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pair.mr[\"mr\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "57399b8d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id.exposure</th>\n",
       "      <th>id.outcome</th>\n",
       "      <th>outcome</th>\n",
       "      <th>exposure</th>\n",
       "      <th>egger_intercept</th>\n",
       "      <th>se</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>STUDY1</td>\n",
       "      <td>-0.074644</td>\n",
       "      <td>0.230296</td>\n",
       "      <td>0.800459</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  id.exposure id.outcome outcome exposure  egger_intercept        se      pval\n",
       "0      H4lJZt     aLOCCI  STUDY2   STUDY1        -0.074644  0.230296  0.800459"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pair.mr[\"pleiotropy\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0e98327f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id.exposure</th>\n",
       "      <th>id.outcome</th>\n",
       "      <th>outcome</th>\n",
       "      <th>exposure</th>\n",
       "      <th>method</th>\n",
       "      <th>Q</th>\n",
       "      <th>Q_df</th>\n",
       "      <th>Q_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>STUDY1</td>\n",
       "      <td>MR Egger</td>\n",
       "      <td>7.249592</td>\n",
       "      <td>1</td>\n",
       "      <td>0.007092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>STUDY1</td>\n",
       "      <td>Inverse variance weighted</td>\n",
       "      <td>8.011201</td>\n",
       "      <td>2</td>\n",
       "      <td>0.018213</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  id.exposure id.outcome outcome exposure                     method  \\\n",
       "0      H4lJZt     aLOCCI  STUDY2   STUDY1                   MR Egger   \n",
       "1      H4lJZt     aLOCCI  STUDY2   STUDY1  Inverse variance weighted   \n",
       "\n",
       "          Q  Q_df    Q_pval  \n",
       "0  7.249592     1  0.007092  \n",
       "1  8.011201     2  0.018213  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pair.mr[\"heterogeneity\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "354c0fbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id.exposure</th>\n",
       "      <th>id.outcome</th>\n",
       "      <th>exposure</th>\n",
       "      <th>outcome</th>\n",
       "      <th>snp_r2.exposure</th>\n",
       "      <th>snp_r2.outcome</th>\n",
       "      <th>correct_causal_direction</th>\n",
       "      <th>steiger_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>STUDY1</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>0.003113</td>\n",
       "      <td>0.00037</td>\n",
       "      <td>True</td>\n",
       "      <td>3.341027e-16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  id.exposure id.outcome exposure outcome  snp_r2.exposure  snp_r2.outcome  \\\n",
       "0      H4lJZt     aLOCCI   STUDY1  STUDY2         0.003113         0.00037   \n",
       "\n",
       "   correct_causal_direction  steiger_pval  \n",
       "0                      True  3.341027e-16  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pair.mr[\"directionality\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "cf341818",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exposure</th>\n",
       "      <th>outcome</th>\n",
       "      <th>id.exposure</th>\n",
       "      <th>id.outcome</th>\n",
       "      <th>samplesize</th>\n",
       "      <th>SNP</th>\n",
       "      <th>b</th>\n",
       "      <th>se</th>\n",
       "      <th>p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STUDY1</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>160000</td>\n",
       "      <td>1:177894591:t:g</td>\n",
       "      <td>0.587692</td>\n",
       "      <td>0.156923</td>\n",
       "      <td>1.803233e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STUDY1</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>160000</td>\n",
       "      <td>2:527882:c:g</td>\n",
       "      <td>0.898113</td>\n",
       "      <td>0.330189</td>\n",
       "      <td>6.528192e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STUDY1</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>160000</td>\n",
       "      <td>2:632723:c:t</td>\n",
       "      <td>1.338571</td>\n",
       "      <td>0.214286</td>\n",
       "      <td>4.193046e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STUDY1</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>160000</td>\n",
       "      <td>All - Inverse variance weighted</td>\n",
       "      <td>0.855997</td>\n",
       "      <td>0.236592</td>\n",
       "      <td>2.968568e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STUDY1</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>160000</td>\n",
       "      <td>All - MR Egger</td>\n",
       "      <td>2.012733</td>\n",
       "      <td>3.582989</td>\n",
       "      <td>6.741670e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  exposure outcome id.exposure id.outcome  samplesize  \\\n",
       "0   STUDY1  STUDY2      H4lJZt     aLOCCI      160000   \n",
       "1   STUDY1  STUDY2      H4lJZt     aLOCCI      160000   \n",
       "2   STUDY1  STUDY2      H4lJZt     aLOCCI      160000   \n",
       "3   STUDY1  STUDY2      H4lJZt     aLOCCI      160000   \n",
       "4   STUDY1  STUDY2      H4lJZt     aLOCCI      160000   \n",
       "\n",
       "                               SNP         b        se             p  \n",
       "0                  1:177894591:t:g  0.587692  0.156923  1.803233e-04  \n",
       "1                     2:527882:c:g  0.898113  0.330189  6.528192e-03  \n",
       "2                     2:632723:c:t  1.338571  0.214286  4.193046e-10  \n",
       "3  All - Inverse variance weighted  0.855997  0.236592  2.968568e-04  \n",
       "4                   All - MR Egger  2.012733  3.582989  6.741670e-01  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pair.mr[\"singlesnp\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "58749061",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exposure</th>\n",
       "      <th>outcome</th>\n",
       "      <th>id.exposure</th>\n",
       "      <th>id.outcome</th>\n",
       "      <th>samplesize</th>\n",
       "      <th>SNP</th>\n",
       "      <th>b</th>\n",
       "      <th>se</th>\n",
       "      <th>p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STUDY1</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>160000</td>\n",
       "      <td>1:177894591:t:g</td>\n",
       "      <td>1.208039</td>\n",
       "      <td>0.201135</td>\n",
       "      <td>1.900400e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STUDY1</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>160000</td>\n",
       "      <td>2:527882:c:g</td>\n",
       "      <td>0.849805</td>\n",
       "      <td>0.357927</td>\n",
       "      <td>1.758512e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STUDY1</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>160000</td>\n",
       "      <td>2:632723:c:t</td>\n",
       "      <td>0.644887</td>\n",
       "      <td>0.141731</td>\n",
       "      <td>5.362743e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STUDY1</td>\n",
       "      <td>STUDY2</td>\n",
       "      <td>H4lJZt</td>\n",
       "      <td>aLOCCI</td>\n",
       "      <td>160000</td>\n",
       "      <td>All</td>\n",
       "      <td>0.855997</td>\n",
       "      <td>0.236592</td>\n",
       "      <td>2.968568e-04</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  exposure outcome id.exposure id.outcome  samplesize              SNP  \\\n",
       "0   STUDY1  STUDY2      H4lJZt     aLOCCI      160000  1:177894591:t:g   \n",
       "1   STUDY1  STUDY2      H4lJZt     aLOCCI      160000     2:527882:c:g   \n",
       "2   STUDY1  STUDY2      H4lJZt     aLOCCI      160000     2:632723:c:t   \n",
       "3   STUDY1  STUDY2      H4lJZt     aLOCCI      160000              All   \n",
       "\n",
       "          b        se             p  \n",
       "0  1.208039  0.201135  1.900400e-09  \n",
       "1  0.849805  0.357927  1.758512e-02  \n",
       "2  0.644887  0.141731  5.362743e-06  \n",
       "3  0.855997  0.236592  2.968568e-04  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pair.mr[\"leaveoneout\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6e57a977",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"TwoSampleMR version 0.5.8 \\n[>] New: Option to use non-European LD reference panels for clumping etc\\n[>] Some studies temporarily quarantined to verify effect allele\\n[>] See news(package='TwoSampleMR') and https://gwas.mrcieu.ac.uk for further details\\n\\n\\nWarning:\\nYou are running an old version of the TwoSampleMR package.\\nThis version:   0.5.8\\nLatest version: 0.6.8\\nPlease consider updating using remotes::install_github('MRCIEU/TwoSampleMR')\\nHarmonising STUDY1 (H4lJZt) and STUDY2 (aLOCCI)\\nAnalysing 'H4lJZt' on 'aLOCCI'\\n\\n\""
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pair.mr[\"r_log\"]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
