
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="HE Yunye">
      
      
      
        <link rel="prev" href="../tutorial_3.4/">
      
      
        <link rel="next" href="../StatusCode/">
      
      
      <link rel="icon" href="../assets/gwaslab.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.2">
    
    
      
        <title>Sumstats Object - GWASLab</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.50c56a3b.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue" data-md-color-accent="blue">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#sumstats-object-in-gwaslab" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="GWASLab" class="md-header__button md-logo" aria-label="GWASLab" data-md-component="logo">
      
  <img src="../assets/gwaslab.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            GWASLab
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Sumstats Object
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/Cloufield/gwaslab/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href=".." class="md-tabs__link">
          
  
  Home

        </a>
      </li>
    
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../tutorial_3.4/" class="md-tabs__link">
        
  
    
  
  Tutorial for gwaslab

      </a>
    </li>
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="./" class="md-tabs__link">
          
  
  Sumstats manipulation

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Visualization/" class="md-tabs__link">
          
  
  Visualization

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../ExtractLead/" class="md-tabs__link">
          
  
  Utilities

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Download/" class="md-tabs__link">
          
  
  References

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../standardization_workflow/" class="md-tabs__link">
          
  
  Examples

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="GWASLab" class="md-nav__button md-logo" aria-label="GWASLab" data-md-component="logo">
      
  <img src="../assets/gwaslab.png" alt="logo">

    </a>
    GWASLab
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Cloufield/gwaslab/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
    
    
    
      
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" >
        
          
          <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Home
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            Home
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../UpdateLogs/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Update logs
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../KnownIssues/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Known Issues
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../tutorial_3.4/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Tutorial for gwaslab
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
      
      
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Sumstats manipulation
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Sumstats manipulation
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Sumstats Object
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Sumstats Object
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1-usage" class="md-nav__link">
    <span class="md-ellipsis">
      1. Usage
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-loading-options" class="md-nav__link">
    <span class="md-ellipsis">
      2. Loading options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3-loading-sumstats" class="md-nav__link">
    <span class="md-ellipsis">
      3 Loading sumstats
    </span>
  </a>
  
    <nav class="md-nav" aria-label="3 Loading sumstats">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#31-loading-by-specifying-columns" class="md-nav__link">
    <span class="md-ellipsis">
      3.1 Loading by specifying columns
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#32-loading-by-specifying-formats" class="md-nav__link">
    <span class="md-ellipsis">
      3.2 Loading by specifying formats
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#33-load-sumstats-with-auto-mode" class="md-nav__link">
    <span class="md-ellipsis">
      3.3 Load sumstats with auto mode
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#34-loading-sumstats-from-chromosome-separated-files" class="md-nav__link">
    <span class="md-ellipsis">
      3.4 Loading sumstats from chromosome-separated files
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-check-and-save-sumstats" class="md-nav__link">
    <span class="md-ellipsis">
      4. Check and save sumstats
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#5-saving-half-finished-sumstats-object" class="md-nav__link">
    <span class="md-ellipsis">
      5. Saving half-finished Sumstats Object
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#6-logging" class="md-nav__link">
    <span class="md-ellipsis">
      6. Logging
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#7-sumstats-summary" class="md-nav__link">
    <span class="md-ellipsis">
      7. Sumstats summary
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#8-other-functions" class="md-nav__link">
    <span class="md-ellipsis">
      8. Other functions
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../StatusCode/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Status Code
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Standardization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Standardization and Normalization
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../QC%26Filtering/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    QC and Filtering
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Harmonization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Harmonization
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../AssignrsID/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assigning rsID
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../AssignCHRPOS/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assign CHR/POS using rsID
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../LiftOver/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Liftover
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Conversion/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Data conversion
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
      
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Visualization
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Visualization
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Visualization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Manhattan and QQ plot
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../RegionalPlot/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Regional Plot
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../BrisbanePlot/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Brisbane Plot
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../MiamiPlot/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Miami Plot
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../TrumpetPlot/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Trumpet Plot (testing)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../EffectSize/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Effect Size Comparison
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../AlleleFrequency/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Allele Frequency Comparison
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../GeneticCorrelation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Genetic correlation heatmap
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Gallery/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Gallery
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
      
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Utilities
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Utilities
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ExtractLead/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Extract Lead Variants
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ExtractNovel/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Extract Novel Variants
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Format/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Format and Save
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Pickle/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Save Sumstats Object
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../LoadLDSC/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Load LDSC log
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../HeritabilityConversion/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Convert Heritability
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../PerSNPh2/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Per SNP Heritability and F statistics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../InferBuild/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Infer Genome Build
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
      
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="">
            
  
  <span class="md-ellipsis">
    References
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            References
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Download/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Download
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Reference/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Reference data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../CommonData/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Common data
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
      
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Examples
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            Examples
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../standardization_workflow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Standardization
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../harmonization_workflow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Harmonization
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../utility_data_conversion/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Data conversion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../visualization_regional/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Regional plot
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../visualization_miami2/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Miami plot
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1-usage" class="md-nav__link">
    <span class="md-ellipsis">
      1. Usage
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-loading-options" class="md-nav__link">
    <span class="md-ellipsis">
      2. Loading options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3-loading-sumstats" class="md-nav__link">
    <span class="md-ellipsis">
      3 Loading sumstats
    </span>
  </a>
  
    <nav class="md-nav" aria-label="3 Loading sumstats">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#31-loading-by-specifying-columns" class="md-nav__link">
    <span class="md-ellipsis">
      3.1 Loading by specifying columns
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#32-loading-by-specifying-formats" class="md-nav__link">
    <span class="md-ellipsis">
      3.2 Loading by specifying formats
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#33-load-sumstats-with-auto-mode" class="md-nav__link">
    <span class="md-ellipsis">
      3.3 Load sumstats with auto mode
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#34-loading-sumstats-from-chromosome-separated-files" class="md-nav__link">
    <span class="md-ellipsis">
      3.4 Loading sumstats from chromosome-separated files
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-check-and-save-sumstats" class="md-nav__link">
    <span class="md-ellipsis">
      4. Check and save sumstats
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#5-saving-half-finished-sumstats-object" class="md-nav__link">
    <span class="md-ellipsis">
      5. Saving half-finished Sumstats Object
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#6-logging" class="md-nav__link">
    <span class="md-ellipsis">
      6. Logging
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#7-sumstats-summary" class="md-nav__link">
    <span class="md-ellipsis">
      7. Sumstats summary
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#8-other-functions" class="md-nav__link">
    <span class="md-ellipsis">
      8. Other functions
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="sumstats-object-in-gwaslab">Sumstats Object in GWASLab</h1>
<p>In GWASLab, sumstats were stored in a <code>Sumstats Object</code>, which is built on <code>pandas Dataframe</code>. All other functions are designed as methods of this Sumstats Object. </p>
<p>To load any sumstats into the object, simply specify the column name and load the raw GWAS summary statistics from a pandas DataFrame or specify a file path. All raw data will be loaded as <code>"string"</code> datatype. </p>
<h2 id="1-usage">1. Usage</h2>
<div class="language-py highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="n">mysumstats</span> <span class="o">=</span> <span class="n">gl</span><span class="o">.</span><span class="n">Sumstats</span><span class="p">(</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>             <span class="n">sumstats</span><span class="p">,</span>
</span><span id="__span-0-3"><a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>             <span class="n">fmt</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-4"><a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>             <span class="n">snpid</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-5"><a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>             <span class="n">rsid</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-6"><a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>             <span class="n">chrom</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-7"><a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>             <span class="n">pos</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-8"><a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>             <span class="n">ea</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-9"><a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>             <span class="n">nea</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-10"><a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>             <span class="o">...</span>
</span><span id="__span-0-11"><a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="p">)</span>
</span></code></pre></div>
<h2 id="2-loading-options">2. Loading options</h2>
<p><code>sumstats</code>: either a file path <code>string</code> or a pandas <code>DataFrame</code></p>
<p>Currently, GWASLab supports the following columns:</p>
<table>
<thead>
<tr>
<th>Option</th>
<th>DataType</th>
<th>Description</th>
<th>Header in GWASLab</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>snpid</code></td>
<td><code>string</code></td>
<td>variant ID column name, preferably in chr:pos:ea:nea format.</td>
<td><code>SNPID</code></td>
</tr>
<tr>
<td><code>rsid</code></td>
<td><code>string</code></td>
<td>dbSNP rsID column name</td>
<td><code>rsID</code></td>
</tr>
</tbody>
</table>
<p>The minimum required columns are just either <code>rsid</code>or <code>snpid</code>. 
All other columns and options are optional.</p>
<table>
<thead>
<tr>
<th>Option</th>
<th>DataType</th>
<th>Description</th>
<th>Header in GWASLab</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>fmt</code></td>
<td><code>string</code></td>
<td>input sumstats format. For formats supported by GWASLab, please check <a href="https://github.com/Cloufield/formatbook">https://github.com/Cloufield/formatbook</a></td>
<td>-</td>
</tr>
<tr>
<td><code>chrom</code></td>
<td><code>string</code></td>
<td>chromosome column name</td>
<td><code>CHR</code></td>
</tr>
<tr>
<td><code>pos</code></td>
<td><code>string</code></td>
<td>basepair position column name</td>
<td><code>POS</code></td>
</tr>
<tr>
<td><code>ea</code></td>
<td><code>string</code></td>
<td>effect allele column name; BETA, OR, HR, EAF are in reference to EA...</td>
<td><code>EA</code></td>
</tr>
<tr>
<td><code>nea</code></td>
<td><code>string</code></td>
<td>non-effect allele column name</td>
<td><code>NEA</code></td>
</tr>
<tr>
<td><code>ref</code></td>
<td><code>string</code></td>
<td>reference allele column name; the allele on reference genome</td>
<td><code>REF</code></td>
</tr>
<tr>
<td><code>alt</code></td>
<td><code>string</code></td>
<td>alternative allele column name; the allele that is not on reference genome; when <code>ea</code>,<code>ref</code> and <code>alt</code> are specified, <code>nea</code> will be inferred.</td>
<td><code>ALT</code></td>
</tr>
<tr>
<td><code>eaf</code></td>
<td><code>string</code></td>
<td>effect allele frequency</td>
<td><code>EAF</code></td>
</tr>
<tr>
<td><code>neaf</code></td>
<td><code>string</code></td>
<td>non-effect allele frequency. NEAF will be converted to EAF (EAF = 1 - NEAF) while loading.</td>
<td><code>EAF</code></td>
</tr>
<tr>
<td><code>n</code></td>
<td><code>string</code> or <code>integer</code></td>
<td>sample size column name or just input a single <code>integer</code> as sample size for all variants</td>
<td><code>N</code></td>
</tr>
<tr>
<td><code>beta</code></td>
<td><code>string</code></td>
<td>effect size beta column name; in reference to EA</td>
<td><code>BETA</code></td>
</tr>
<tr>
<td><code>se</code></td>
<td><code>string</code></td>
<td>standard error column name</td>
<td><code>SE</code></td>
</tr>
<tr>
<td><code>chisq</code></td>
<td><code>string</code></td>
<td>chi square column name</td>
<td><code>CHISQ</code></td>
</tr>
<tr>
<td><code>z</code></td>
<td><code>string</code></td>
<td>z score column name</td>
<td><code>Z</code></td>
</tr>
<tr>
<td><code>p</code></td>
<td><code>string</code></td>
<td>p value column name</td>
<td><code>P</code></td>
</tr>
<tr>
<td><code>mlog10p</code></td>
<td><code>string</code></td>
<td>-log10(P) column name</td>
<td><code>MLOG10P</code></td>
</tr>
<tr>
<td><code>info</code></td>
<td><code>string</code></td>
<td>imputation info or rsq column name</td>
<td><code>INFO</code></td>
</tr>
<tr>
<td><code>OR</code></td>
<td><code>string</code></td>
<td>odds ratio column name; in reference to EA</td>
<td><code>OR</code></td>
</tr>
<tr>
<td><code>OR_95L</code></td>
<td><code>string</code></td>
<td>odds ratio lower 95% CI column name</td>
<td><code>OR_95L</code></td>
</tr>
<tr>
<td><code>OR_95U</code></td>
<td><code>string</code></td>
<td>odds ratio upper 95% CI column name</td>
<td><code>OR_95U</code></td>
</tr>
<tr>
<td><code>direction</code></td>
<td><code>string</code></td>
<td>direction column name. GWASLab uses METAL format (e.g. <code>++--+?+0</code>)</td>
<td><code>DIRECTION</code></td>
</tr>
<tr>
<td><code>other</code></td>
<td><code>list</code></td>
<td>a list  of other column names you want to keep with the core columns (probably some annotations).</td>
<td>-</td>
</tr>
<tr>
<td><code>status</code></td>
<td><code>string</code></td>
<td>status code column name. GWASLab uses a 7-digit vairant status code. For details, please check status code page.</td>
<td><code>STATUS</code></td>
</tr>
<tr>
<td><code>verbose</code></td>
<td><code>boolean</code></td>
<td>if True, print log.</td>
<td>-</td>
</tr>
<tr>
<td><code>build</code></td>
<td><code>string</code></td>
<td>genome build. <code>19</code> for hg19, <code>38</code> for hg38 and <code>99</code> for unknown.</td>
<td>The first two digits of <code>STATUS</code></td>
</tr>
<tr>
<td><code>**arg</code></td>
<td><code>string</code></td>
<td>additional parameters for <a href="https://pandas.pydata.org/docs/reference/api/pandas.read_table.html">pd.read_table()</a> function. Some common options include : <code>sep</code>,<code>nrows</code>, <code>skiprows</code> and <code>na_values</code>.</td>
<td>-</td>
</tr>
</tbody>
</table>
<p>Other columns GWASLab uses (these will be implemented soon)</p>
<table>
<thead>
<tr>
<th>Option</th>
<th>DataType</th>
<th>Description</th>
<th>Header in GWASLab</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>ncontrol</code></td>
<td><code>string</code> or <code>integer</code></td>
<td>sample size column name for controls or just input a single <code>integer</code> as sample size for all variants</td>
<td><code>N_CONTROL</code></td>
</tr>
<tr>
<td><code>ncase</code></td>
<td><code>string</code> or <code>integer</code></td>
<td>sample size column name for cases or just input a single <code>integer</code> as sample size for all variants</td>
<td><code>N_CASE</code></td>
</tr>
<tr>
<td><code>HR</code></td>
<td><code>string</code></td>
<td>hazrad ratio column name</td>
<td><code>HR</code></td>
</tr>
<tr>
<td><code>HR_95U</code></td>
<td><code>string</code></td>
<td>hazrad ratio upper 95% CI column name</td>
<td><code>HR_95U</code></td>
</tr>
<tr>
<td><code>HR_95L</code></td>
<td><code>string</code></td>
<td>hazrad ratio lower 95% CI column name</td>
<td><code>HR_95L</code></td>
</tr>
<tr>
<td><code>beta_95L</code></td>
<td><code>string</code></td>
<td>beta upper 95% CI column name</td>
<td><code>BETA_95L</code></td>
</tr>
<tr>
<td><code>beta_95U</code></td>
<td><code>string</code></td>
<td>beta upper 95% CI column name</td>
<td><code>BETA_95U</code></td>
</tr>
<tr>
<td><code>i2</code></td>
<td><code>string</code></td>
<td>I2 column name</td>
<td><code>I2_HET</code></td>
</tr>
<tr>
<td><code>phet</code></td>
<td><code>string</code></td>
<td>heterogeneity test P value</td>
<td><code>P_HET</code></td>
</tr>
<tr>
<td><code>dof</code></td>
<td><code>string</code> or <code>integer</code></td>
<td>degree of freedom</td>
<td><code>DOF</code></td>
</tr>
<tr>
<td><code>snpr2</code></td>
<td><code>string</code></td>
<td>column name for proportion of phenotypic variance explained by each variant</td>
<td><code>SNPR2</code></td>
</tr>
<tr>
<td><code>maf</code></td>
<td><code>string</code></td>
<td>minor allele frequency column header</td>
<td><code>MAF</code></td>
</tr>
<tr>
<td><code>f</code></td>
<td><code>string</code></td>
<td>F statistics column header</td>
<td><code>F</code></td>
</tr>
</tbody>
</table>
<h2 id="3-loading-sumstats">3 Loading sumstats</h2>
<h3 id="31-loading-by-specifying-columns">3.1 Loading by specifying columns</h3>
<p>You can load the sumstats by specifying the columns like:</p>
<div class="admonition example">
<p class="admonition-title">Load sumstats by manually specifying columns</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>mysumstats = gl.Sumstats(&quot;t2d_bbj.txt.gz&quot;,
</span><span id="__span-1-2"><a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>             snpid=&quot;SNPID&quot;,
</span><span id="__span-1-3"><a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>             chrom=&quot;CHR&quot;,
</span><span id="__span-1-4"><a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a>             pos=&quot;POS&quot;,
</span><span id="__span-1-5"><a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a>             ea=&quot;Allele2&quot;,
</span><span id="__span-1-6"><a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a>             nea=&quot;Allele1&quot;,
</span><span id="__span-1-7"><a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a>             eaf=&quot;AF_Allele2&quot;,
</span><span id="__span-1-8"><a id="__codelineno-1-8" name="__codelineno-1-8" href="#__codelineno-1-8"></a>             beta=&quot;BETA&quot;,
</span><span id="__span-1-9"><a id="__codelineno-1-9" name="__codelineno-1-9" href="#__codelineno-1-9"></a>             se=&quot;SE&quot;,
</span><span id="__span-1-10"><a id="__codelineno-1-10" name="__codelineno-1-10" href="#__codelineno-1-10"></a>             p=&quot;p.value&quot;,
</span><span id="__span-1-11"><a id="__codelineno-1-11" name="__codelineno-1-11" href="#__codelineno-1-11"></a>             n=&quot;N&quot;)
</span></code></pre></div>
</div>
<h3 id="32-loading-by-specifying-formats">3.2 Loading by specifying formats</h3>
<p>GWASLab supports common sumstats formats and you can load sumstats by specifying <code>fmt</code>.  </p>
<p>GWASLab uses a manually curated format conversion dictionary in <a href="https:/github.com/Cloufield/formatbook">https://github.com/Cloufield/formatbook</a>. Currently, it supports the following formats:</p>
<table>
<thead>
<tr>
<th><code>Keyword</code></th>
<th><code>Description</code></th>
</tr>
</thead>
<tbody>
<tr>
<td><code>ssf</code></td>
<td>GWAS-SSF</td>
</tr>
<tr>
<td><code>gwascatalog</code></td>
<td>GWAS Catalog format</td>
</tr>
<tr>
<td><code>pgscatalog</code></td>
<td>PGS Catalog format</td>
</tr>
<tr>
<td><code>plink</code></td>
<td>PLINK output format</td>
</tr>
<tr>
<td><code>plink2</code></td>
<td>PLINK2 output format</td>
</tr>
<tr>
<td><code>saige</code></td>
<td>SAIGE output format</td>
</tr>
<tr>
<td><code>regenie</code></td>
<td>output format</td>
</tr>
<tr>
<td><code>fastgwa</code></td>
<td>output format</td>
</tr>
<tr>
<td><code>metal</code></td>
<td>output format</td>
</tr>
<tr>
<td><code>mrmega</code></td>
<td>output format</td>
</tr>
<tr>
<td><code>fuma</code></td>
<td>input format</td>
</tr>
<tr>
<td><code>ldsc</code></td>
<td>input format</td>
</tr>
<tr>
<td><code>locuszoom</code></td>
<td>input format</td>
</tr>
<tr>
<td><code>vcf</code></td>
<td>gwas-vcf format</td>
</tr>
<tr>
<td><code>bolt_lmm</code></td>
<td>output format</td>
</tr>
</tbody>
</table>
<div class="admonition info">
<p class="admonition-title">Update Formatbook using <code>gl.update_formaybook()</code></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>gl.update_formatbook()
</span><span id="__span-2-2"><a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>Mon Jul 17 17:38:11 2023 Updating formatbook from: https://raw.github.com/Cloufield/formatbook/main/formatbook.json
</span><span id="__span-2-3"><a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a>Mon Jul 17 17:38:12 2023 Overwrite formatbook to :  /home/yunye/gwaslab/gwaslab/src/gwaslab/data/formatbook.json
</span><span id="__span-2-4"><a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a>Mon Jul 17 17:38:12 2023 Available formats: auto,bolt_lmm,fastgwa,gwascatalog,gwascatalog_hm,gwaslab,ldsc,metal,mrmega,mtag,pgscatalog,pgscatalog_hm,pheweb,plink,plink2,regenie,saige,ssf,template,vcf
</span><span id="__span-2-5"><a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a>Mon Jul 17 17:38:12 2023 Formatbook has been updated!
</span></code></pre></div>
</div>
<div class="admonition example">
<p class="admonition-title">Load sumstats by simply specifying the format</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>mysumstats = gl.Sumstats(&quot;t2d_bbj.txt.gz&quot;, fmt=&quot;saige&quot;)
</span></code></pre></div>
</div>
<h3 id="33-load-sumstats-with-auto-mode">3.3 Load sumstats with <code>auto</code> mode</h3>
<p>GWASLab also provides an auto mode (<code>fmt="auto"</code>; available since v3.4.21) which assumes A1 or alternative allele (ALT) is the effect allele (EA) and Frq refers to the allele frequency of effect allele (EAF). Common headers will be detected. You can find the conversion table <a href="https://github.com/Cloufield/formatbook/blob/main/formats/auto.json">here</a></p>
<div class="admonition example">
<p class="admonition-title">Load sumstats with auto mode</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>mysumstats = gl.Sumstats(&quot;t2d_bbj.txt.gz&quot;, fmt=&quot;auto&quot;)
</span></code></pre></div>
</div>
<h3 id="34-loading-sumstats-from-chromosome-separated-files">3.4 Loading sumstats from chromosome-separated files</h3>
<p>GWASLab supports loading sumstats from chromosome-separated files (file names need to be in the same pattern.). Just use @ to replace the chromosome numbers. </p>
<div class="admonition example">
<p class="admonition-title">Example</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>mysumstats = gl.Sumstats(&quot;t2d.chr@.txt.gz&quot;,fmt=&quot;metal&quot;)
</span></code></pre></div>
</div>
<h2 id="4-check-and-save-sumstats">4. Check and save sumstats</h2>
<p>After loading, the raw data columns will be renamed to new columns without ambiguity and the dataframe is stored in <code>.data</code> :</p>
<div class="admonition example">
<p class="admonition-title">Example</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-6-1"><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a><span class="n">mysumstats</span><span class="o">.</span><span class="n">data</span>
</span></code></pre></div>
</div>
<p>You can simply save the processed data using pandas saving functions, for example:</p>
<div class="admonition example">
<p class="admonition-title">Example</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-7-1"><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>mysumstats.data.to_csv(&quot;./mysumstats.csv&quot;)
</span></code></pre></div>
</div>
<p>or convert the sumstats to other sumstats using GWASLab <code>to_format()</code> function (recommended):</p>
<div class="admonition example">
<p class="admonition-title">Example</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-8-1"><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>mysumstats.to_format(&quot;./mysumstats&quot;, fmt=&quot;ldsc&quot;,hapmap3=True, exclude_hla=True, build=&quot;19&quot;)
</span></code></pre></div>
</div>
<p>Please check <a href="https://cloufield.github.io/gwaslab/Format/">GWASLab - Format</a> for more details.</p>
<h2 id="5-saving-half-finished-sumstats-object">5. Saving half-finished Sumstats Object</h2>
<p>If the pipeline is very long, and you need to temporarily save the Sumstats Object, you can use the <code>.dump_pickle()</code> method to temporarily save the Sumstats Object.</p>
<p>Please check <a href="https://cloufield.github.io/gwaslab/Pickle/">GWASLab - Pickle</a> for more details.</p>
<h2 id="6-logging">6. Logging</h2>
<p>All manipulation conducted to the sumstats will be logged for reproducibility and traceability. </p>
<p>The log is stored in a <code>gl.Log()</code> object. You can check it by <code>.log.show()</code>and save it using <code>.log.save()</code></p>
<div class="admonition example">
<p class="admonition-title">Example</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-9-1"><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>mysumstats.log.show()
</span><span id="__span-9-2"><a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a>
</span><span id="__span-9-3"><a id="__codelineno-9-3" name="__codelineno-9-3" href="#__codelineno-9-3"></a>mysumstats.log.save()
</span></code></pre></div>
</div>
<h2 id="7-sumstats-summary">7. Sumstats summary</h2>
<p>You can check the meta information of sumstats by:</p>
<div class="admonition example">
<p class="admonition-title">Example</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-10-1"><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a><span class="n">mysumstats</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
</span></code></pre></div>
</div>
<h2 id="8-other-functions">8. Other functions</h2>
<p>Other functions of GWASLab are implemented as the methods of Sumstats Object.</p>
<div class="admonition example">
<p class="admonition-title">Example</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-11-1"><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a><span class="n">mysumstats</span><span class="o">.</span><span class="n">basic_check</span><span class="p">()</span>
</span><span id="__span-11-2"><a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a>
</span><span id="__span-11-3"><a id="__codelineno-11-3" name="__codelineno-11-3" href="#__codelineno-11-3"></a><span class="n">mysumstats</span><span class="o">.</span><span class="n">plot_mqq</span><span class="p">()</span>
</span><span id="__span-11-4"><a id="__codelineno-11-4" name="__codelineno-11-4" href="#__codelineno-11-4"></a>
</span><span id="__span-11-5"><a id="__codelineno-11-5" name="__codelineno-11-5" href="#__codelineno-11-5"></a><span class="o">...</span>
</span></code></pre></div>
</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      GWASLab is licensed under the MIT license
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs"], "search": "../assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.d7c377c4.min.js"></script>
      
        <script src="../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>