
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="HE Yunye">
      
      
      <link rel="icon" href="../assets/gwaslab.png">
      <meta name="generator" content="mkdocs-1.3.0, mkdocs-material-8.2.15">
    
    
      
        <title>Reference data - GWASLab</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.c382b1dc.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.cc9b2e1e.min.css">
        
          
          
          <meta name="theme-color" content="#2094f3">
        
      
    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="blue" data-md-color-accent="blue">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#reference-data-for-handling-sumstats" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="GWASLab" class="md-header__button md-logo" aria-label="GWASLab" data-md-component="logo">
      
  <img src="../assets/gwaslab.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            GWASLab
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Reference data
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/Cloufield/gwaslab/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-tabs__inner md-grid">
    <ul class="md-tabs__list">
      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        Home
      </a>
    </li>
  

      
        
  
  


  <li class="md-tabs__item">
    <a href="../tutorial_3.4/" class="md-tabs__link">
      Tutorial for gwaslab
    </a>
  </li>

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../SumstatsObject/" class="md-tabs__link">
        Sumstats manipulation
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../Visualization/" class="md-tabs__link">
        Visualization
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../ExtractLead/" class="md-tabs__link">
        Utilities
      </a>
    </li>
  

      
        
  
  
    
  


  
  
  
    <li class="md-tabs__item">
      <a href="../Download/" class="md-tabs__link md-tabs__link--active">
        References
      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="GWASLab" class="md-nav__button md-logo" aria-label="GWASLab" data-md-component="logo">
      
  <img src="../assets/gwaslab.png" alt="logo">

    </a>
    GWASLab
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Cloufield/gwaslab/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_1" type="checkbox" id="__nav_1" >
      
      
      
      
        <label class="md-nav__link" for="__nav_1">
          Home
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Home" data-md-level="1">
        <label class="md-nav__title" for="__nav_1">
          <span class="md-nav__icon md-icon"></span>
          Home
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Home
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../UpdateLogs/" class="md-nav__link">
        Update logs
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../tutorial_3.4/" class="md-nav__link">
        Tutorial for gwaslab
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3" type="checkbox" id="__nav_3" >
      
      
      
      
        <label class="md-nav__link" for="__nav_3">
          Sumstats manipulation
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Sumstats manipulation" data-md-level="1">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          Sumstats manipulation
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../SumstatsObject/" class="md-nav__link">
        Sumstats Object
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../StatusCode/" class="md-nav__link">
        Status Code
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Standardization/" class="md-nav__link">
        Standardization and Normalization
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../QC%26Filtering/" class="md-nav__link">
        QC and Filtering
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Harmonization/" class="md-nav__link">
        Harmonization
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../AssignrsID/" class="md-nav__link">
        Assigning rsID
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../AssignCHRPOS/" class="md-nav__link">
        Assign CHR/POS using rsID
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../LiftOver/" class="md-nav__link">
        Liftover
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Conversion/" class="md-nav__link">
        Data conversion
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4" type="checkbox" id="__nav_4" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4">
          Visualization
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Visualization" data-md-level="1">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Visualization
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Visualization/" class="md-nav__link">
        Manhattan and QQ plot
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../RegionalPlot/" class="md-nav__link">
        Regional Plot
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../BrisbanePlot/" class="md-nav__link">
        Brisbane Plot
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../MiamiPlot/" class="md-nav__link">
        Miami Plot
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../EffectSize/" class="md-nav__link">
        Effect Size Comparison
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5" type="checkbox" id="__nav_5" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5">
          Utilities
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Utilities" data-md-level="1">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          Utilities
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ExtractLead/" class="md-nav__link">
        Extract Lead Variants
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ExtractNovel/" class="md-nav__link">
        Extract Novel Variants
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Format/" class="md-nav__link">
        Format and Save
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Pickle/" class="md-nav__link">
        Save Sumstats Object
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../LoadLDSC/" class="md-nav__link">
        Load LDSC log
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../HeritabilityConversion/" class="md-nav__link">
        Convert Heritability
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../PerSNPh2/" class="md-nav__link">
        Per SNP Heritability and F statistics
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../InferBuild/" class="md-nav__link">
        Infer Genome Build
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6" type="checkbox" id="__nav_6" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_6">
          References
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="References" data-md-level="1">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          References
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Download/" class="md-nav__link">
        Download
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Reference data
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Reference data
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#reference-genome-sequence-for-variant-allele-alignment" class="md-nav__link">
    Reference genome sequence for variant allele alignment
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processed-reference-files-for-harmonization" class="md-nav__link">
    Processed Reference files for harmonization
  </a>
  
    <nav class="md-nav" aria-label="Processed Reference files for harmonization">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1000-genome-projecthg19" class="md-nav__link">
    1000 Genome Project(hg19)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sample-code" class="md-nav__link">
    Sample code:
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#rsid-conversion-table-for-quick-annotation" class="md-nav__link">
    rsID conversion table for quick annotation
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dbsnp-database-for-annotation-of-rsid" class="md-nav__link">
    dbsnp: database for annotation of rsID
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomad" class="md-nav__link">
    gnomad
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reference-library-for-variants-annotation" class="md-nav__link">
    Reference Library for variants annotation
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CommonData/" class="md-nav__link">
        Common data
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#reference-genome-sequence-for-variant-allele-alignment" class="md-nav__link">
    Reference genome sequence for variant allele alignment
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processed-reference-files-for-harmonization" class="md-nav__link">
    Processed Reference files for harmonization
  </a>
  
    <nav class="md-nav" aria-label="Processed Reference files for harmonization">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1000-genome-projecthg19" class="md-nav__link">
    1000 Genome Project(hg19)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sample-code" class="md-nav__link">
    Sample code:
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#rsid-conversion-table-for-quick-annotation" class="md-nav__link">
    rsID conversion table for quick annotation
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dbsnp-database-for-annotation-of-rsid" class="md-nav__link">
    dbsnp: database for annotation of rsID
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomad" class="md-nav__link">
    gnomad
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reference-library-for-variants-annotation" class="md-nav__link">
    Reference Library for variants annotation
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                


<h1 id="reference-data-for-handling-sumstats">Reference data for handling Sumstats</h1>
<h2 id="reference-genome-sequence-for-variant-allele-alignment">Reference genome sequence for variant allele alignment</h2>
<ul>
<li>GRCh37 / hg19 : <a href="http://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/">ucsc_hg19</a></li>
<li>GRCh38 / hg38 : <a href="https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/">ucsc_hg38</a></li>
</ul>
<p>For details about reference genome, please check <a href="https://cloufield.github.io/CTGCatalog/Reference_data_Genome_README/">https://cloufield.github.io/CTGCatalog/Reference_data_Genome_README/</a></p>
<h2 id="processed-reference-files-for-harmonization">Processed Reference files for harmonization</h2>
<h3 id="1000-genome-projecthg19">1000 Genome Project(hg19)</h3>
<p>Download:
<a href="http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/">Index of /vol1/ftp/release/20130502/</a></p>
<p>Process the 1000 genome vcf file for EAS sample:</p>
<ul>
<li>Extract EAS sample</li>
<li>Split multiallelic variants</li>
<li>Normalize variants</li>
<li>Rename variants as CHR:POS:REF:ALT</li>
<li>Remove duplicated variants</li>
<li>Calculate Alternative allele frequency</li>
<li>Tabix index</li>
</ul>
<h3 id="sample-code">Sample code:</h3>
<p><div class="highlight"><pre><span></span><code><span class="ch">#!/bin/bash</span>
<span class="c1"># extract EAS sample ID</span>
awk<span class="w"> </span><span class="s1">&#39;$3==&quot;EAS&quot;{print $1}&#39;</span><span class="w"> </span>integrated_call_samples_v3.20130502.ALL.panel<span class="w"> </span>&gt;EAS.sample

<span class="c1"># process</span>
<span class="k">for</span><span class="w"> </span>chr<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="o">{</span><span class="m">1</span>..22<span class="o">}</span>
<span class="k">do</span>
<span class="w">    </span>bcftools<span class="w"> </span>view<span class="w"> </span>-S<span class="w"> </span>EAS.sample<span class="w"> </span>ALL.chr<span class="s2">&quot;</span><span class="si">${</span><span class="nv">chr</span><span class="si">}</span><span class="s2">&quot;</span>.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="se">\</span>
<span class="w">        </span>bcftools<span class="w"> </span>norm<span class="w"> </span>-m-any<span class="w"> </span>--check-ref<span class="w"> </span>w<span class="w"> </span>-f<span class="w"> </span>human_g1k_v37.fasta<span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="se">\</span>
<span class="w">        </span>bcftools<span class="w"> </span>annotate<span class="w"> </span>-x<span class="w"> </span>ID,INFO<span class="w"> </span>-I<span class="w"> </span>+<span class="s1">&#39;%CHROM:%POS:%REF:%ALT&#39;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="se">\</span>
<span class="w">        </span>bcftools<span class="w"> </span>norm<span class="w"> </span>--rm-dup<span class="w"> </span>both<span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="se">\</span>
<span class="w">        </span>bcftools<span class="w"> </span>+fill-tags<span class="w"> </span>-Oz<span class="w">  </span>--<span class="w"> </span>-t<span class="w"> </span>AF<span class="w">  </span><span class="se">\</span>
<span class="w">          </span>&gt;<span class="w"> </span>EAS.chr<span class="s2">&quot;</span><span class="si">${</span><span class="nv">chr</span><span class="si">}</span><span class="s2">&quot;</span>.split_norm_af.vcf.gz
<span class="w">    </span>tabix<span class="w"> </span>-p<span class="w"> </span>vcf<span class="w"> </span>EAS.chr<span class="s2">&quot;</span><span class="si">${</span><span class="nv">chr</span><span class="si">}</span><span class="s2">&quot;</span>.split_norm_af.vcf.gz
<span class="w">    </span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;EAS.chr&quot;</span><span class="si">${</span><span class="nv">chr</span><span class="si">}</span><span class="s2">&quot;.split_norm_af.vcf.gz&quot;</span><span class="w"> </span>&gt;&gt;concat_list.txt<span class="w"> </span>
<span class="k">done</span>

<span class="c1"># merge</span>
bcftools<span class="w"> </span>concat<span class="w"> </span>-a<span class="w"> </span>-d<span class="w"> </span>both<span class="w"> </span>-f<span class="w"> </span>concat_list.txt<span class="w"> </span>-Ob<span class="w"> </span><span class="p">|</span><span class="w"> </span>bcftools<span class="w"> </span>sort<span class="w"> </span>-Oz<span class="w">  </span>&gt;<span class="w"> </span>EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz
tabix<span class="w"> </span>-p<span class="w"> </span>vcf<span class="w"> </span>EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz
</code></pre></div>
The processed dataset can then be used for:</p>
<ul>
<li>infer palindromic SNPs / indels</li>
<li>check allele frequency difference </li>
<li>regional plot</li>
</ul>
<h2 id="rsid-conversion-table-for-quick-annotation">rsID conversion table for quick annotation</h2>
<p>Since GWASLab will check chr:pos and ea/nea to match rsID, it would take a little bit longer if we only use vcf. </p>
<p>But we can use a pre-annotated conversion table for common SNPs, and then annotate the rest of SNPs using large VCF file from dbSNP. 
<div class="highlight"><pre><span></span><code><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">23</span><span class="p">):</span>
    <span class="n">sumstats</span> <span class="o">=</span> <span class="n">gl</span><span class="o">.</span><span class="n">Sumstats</span><span class="p">(</span><span class="s2">&quot;./EAS.chr&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">+</span><span class="s2">&quot;.split_norm_af.vcf.gz&quot;</span><span class="p">,</span><span class="n">snpid</span><span class="o">=</span><span class="s2">&quot;ID&quot;</span><span class="p">,</span><span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;vcf&quot;</span><span class="p">)</span>
    <span class="n">sumstats</span><span class="o">.</span><span class="n">harmonize</span><span class="p">(</span> <span class="n">basic_check</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                        <span class="n">ref_seq</span><span class="o">=</span><span class="s2">&quot;./reference_genome/hg19/human_g1k_v37_decoy.fasta&quot;</span><span class="p">,</span>
                        <span class="n">ref_rsid_vcf</span><span class="o">=</span><span class="s2">&quot;./All_20180423.vcf.gz&quot;</span><span class="p">,</span> 
                        <span class="n">n_cores</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
    <span class="n">sumstats</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,[</span><span class="s2">&quot;SNPID&quot;</span><span class="p">,</span><span class="s2">&quot;rsID&quot;</span><span class="p">,</span><span class="s2">&quot;CHR&quot;</span><span class="p">,</span><span class="s2">&quot;POS&quot;</span><span class="p">,</span><span class="s2">&quot;NEA&quot;</span><span class="p">,</span><span class="s2">&quot;EA&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;./1kg_af_dbsnp151.&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">+</span><span class="s2">&quot;.txt.gz&quot;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</code></pre></div></p>
<p>In terminal, combine the files:
<div class="highlight"><pre><span></span><code># get header
zcat 1kg_af_dbsnp151.1.txt.gz | head -1 &gt; 1kg_af_dbsnp151_auto.txt

for i in $(seq 1 22)
do
# get complete SNPID-rsID pairs
zcat 1kg_af_dbsnp151.${i}.txt.gz | awk -v FS=&quot;\t&quot; -v OFS=&quot;\t&quot; &#39;NR&gt;1 &amp;&amp; $2!=&quot;&quot; {print $0}&#39; &gt;&gt;1kg_af_dbsnp151_auto.txt
done
</code></pre></div></p>
<h2 id="dbsnp-database-for-annotation-of-rsid">dbsnp: database for annotation of rsID</h2>
<ul>
<li>dbsnp v151 (GRCh37): <a href="https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh37p13/VCF/00-All.vcf.gz">https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh37p13/VCF/00-All.vcf.gz</a></li>
<li>dbsnp v151 (GRCh37): <a href="https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh37p13/VCF/00-All.vcf.gz.tbi">https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh37p13/VCF/00-All.vcf.gz.tbi</a></li>
<li>dbsnp v151 (GRCh38): <a href="https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF//00-All.vcf.gz">https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF//00-All.vcf.gz</a></li>
<li>dbsnp v151 (GRCh38): <a href="https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF//00-All.vcf.gz.tbi">https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF//00-All.vcf.gz.tbi</a></li>
<li>latest release: <a href="https://ftp.ncbi.nih.gov/snp/latest_release/VCF/">Index of /snp/latest_release/VCF</a></li>
</ul>
<h2 id="gnomad">gnomad</h2>
<ul>
<li>Allele frequency for major ancestries and rsID</li>
<li>gnomAD v2 &amp; gnomAD v2 liftover &amp; gnomAD v3:   <a href="https://gnomad.broadinstitute.org/downloads">gnomAD</a>    </li>
</ul>
<h2 id="reference-library-for-variants-annotation">Reference Library for variants annotation</h2>
<p>GWASLab uses:</p>
<ul>
<li>ensembl release 87 (hg19): https://ftp.ensembl.org/pub/grch37/release-109/gtf/homo_sapiens/</li>
<li>ensembl release 109 (hg38):  https://ftp.ensembl.org/pub/release-109/gtf/homo_sapiens/</li>
<li>NCBI refseq GRCh37 : https://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/GRCh37_latest/refseq_identifiers/GRCh37_latest_genomic.gtf.gz</li>
<li>NCBI refseq GRCh38 : https://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/GRCh38_latest/refseq_identifiers/GRCh38_latest_genomic.gtf.gz</li>
</ul>
<p>Currently, GWASLab could use either ensembl or refseq gtf reference data to annotate lead SNPs with the nearest gene name.</p>

              
            </article>
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../Download/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Download" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Download
            </div>
          </div>
        </a>
      
      
        
        <a href="../CommonData/" class="md-footer__link md-footer__link--next" aria-label="Next: Common data" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Common data
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      GWASLab is licensed under the MIT license
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs"], "search": "../assets/javascripts/workers/search.2a1c317c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.a6c66575.min.js"></script>
      
        <script src="../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>