<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    
    <meta name="author" content="HE Yunye">
    
    <link rel="shortcut icon" href="../img/favicon.ico">

    
    <title>Reference data - GWASLab</title>
    

    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.12.0/css/all.css">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.12.0/css/v4-shims.css">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/npm/hack-font@3.3.0/build/web/hack.min.css">
    <link href='//rsms.me/inter/inter.css' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,700italic,400,300,600,700&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href="../css/bootstrap-custom.min.css" rel="stylesheet">
    <link href="../css/base.min.css" rel="stylesheet">
    <link href="../css/cinder.min.css" rel="stylesheet">

    
        
        <link rel="stylesheet" href="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/styles/github.min.css">
        
    
    <link href="../stylesheets/mystyle.css" rel="stylesheet">

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js"></script>
            <script src="https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js"></script>
        <![endif]-->

    

     
</head>

<body>

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->

            
              <a class="navbar-brand" href="..">GWASLab</a>
            
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="..">Home</a>
                    </li>
                
                
                
                    <li >
                        <a href="../Tutorial/">Tutorial</a>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Sumstats manipulation <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../SumstatsObject/">SumstatsObject</a>
</li>

                        
                            
<li >
    <a href="../StatusCode/">StatusCode</a>
</li>

                        
                            
<li >
    <a href="../Standardization/">Standardization&normalization</a>
</li>

                        
                            
<li >
    <a href="../QC%26Filtering/">QC&Filtering</a>
</li>

                        
                            
<li >
    <a href="../Harmonization/">Harmonization</a>
</li>

                        
                            
<li >
    <a href="../LiftOver/">Liftover</a>
</li>

                        
                            
<li >
    <a href="../Conversion/">Data conversion</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Visualization <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../Visualization/">Manhattan & QQ plot</a>
</li>

                        
                            
<li >
    <a href="../RegionalPlot/">Regional Plot</a>
</li>

                        
                            
<li >
    <a href="../BrisbanePlot/">Brisbane Plot</a>
</li>

                        
                            
<li >
    <a href="../MiamiPlot/">Miami Plot</a>
</li>

                        
                            
<li >
    <a href="../EffectSize/">Scatter: effect size comparison</a>
</li>

                        
                            
<li >
    <a href="../AlleleFrequency/">Scatter: allele frequency comparison</a>
</li>

                        
                            
<li >
    <a href="../GeneticCorrelation/">Heatmap: genetic correlation</a>
</li>

                        
                            
<li >
    <a href="../ForestPlot/">Forest Plot</a>
</li>

                        
                            
<li >
    <a href="../Gallery/">Gallery</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Utilities <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../ExtractLead/">Extract Lead Variants</a>
</li>

                        
                            
<li >
    <a href="../ExtractNovel/">Extract Novel Variants</a>
</li>

                        
                            
<li >
    <a href="../Format/">Format & Save</a>
</li>

                        
                            
<li >
    <a href="../LoadLDSC/">Load LDSC log</a>
</li>

                        
                            
<li >
    <a href="../HeritabilityConversion/">Convert Heritability</a>
</li>

                        
                            
<li >
    <a href="../InferBuild/">Infer Genome Build</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown active">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">References <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li class="active">
    <a href="./">Reference data</a>
</li>

                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Standardization and QC example</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../standardization_workflow/">Workflow</a>
</li>

        
            
<li >
    <a href="../quality_control_and_filtering/">QC &amp; Filtering</a>
</li>

        
    </ul>
  </li>

                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Harmonization example</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../harmonization_workflow/">Workflow</a>
</li>

        
            
<li >
    <a href="../harmonization_liftover/">Liftover</a>
</li>

        
    </ul>
  </li>

                        
                            
<li >
    <a href="../format_load_save/">Formatting and saving</a>
</li>

                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Utility example</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../utility_data_conversion/">Data conversion</a>
</li>

        
            
<li >
    <a href="../utility_get_novel/">Get novel variants</a>
</li>

        
    </ul>
  </li>

                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Visualization example</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../visualization_mqq/">Manhattan and Q-Q plot</a>
</li>

        
            
<li >
    <a href="../visualization_miami/">Miami plot</a>
</li>

        
            
<li >
    <a href="../visualization_regional/">Regional plot</a>
</li>

        
            
<li >
    <a href="../visualization_brisbane/">Brisbane plot</a>
</li>

        
    </ul>
  </li>

                        
                            
<li >
    <a href="../CommonData/">Common data</a>
</li>

                        
                            
<li >
    <a href="../UpdateLogs/">Update logs</a>
</li>

                        
                        </ul>
                    </li>
                
                
                </ul>

            <ul class="nav navbar-nav navbar-right">
                    <li >
                        <a rel="prev" href="../InferBuild/">
                            <i class="fas fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="next" href="../standardization_workflow/">
                            Next <i class="fas fa-arrow-right"></i>
                        </a>
                    </li>
                    <li>
                        <a href="https://github.com/Cloufield/gwaslab/"><i class="fab fa-github"></i> GitHub</a>
                    </li>
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
        <li class="first-level active"><a href="#reference-data-for-handling-sumstats">Reference data for handling Sumstats</a></li>
            <li class="second-level"><a href="#reference-genome-sequence-for-variant-allele-alignment">Reference genome sequence for variant allele alignment</a></li>
                
            <li class="second-level"><a href="#processed-reference-files-for-harmonization">Processed Reference files for harmonization</a></li>
                
                <li class="third-level"><a href="#1000-genome-projecthg19">1000 Genome Project(hg19)</a></li>
                <li class="third-level"><a href="#sample-code">Sample code:</a></li>
            <li class="second-level"><a href="#rsid-conversion-table-for-quick-annotation">rsID conversion table for quick annotation</a></li>
                
            <li class="second-level"><a href="#dbsnp-database-for-annotation-of-rsid">dbsnp: database for annotation of rsID</a></li>
                
            <li class="second-level"><a href="#gnomad">gnomad</a></li>
                
            <li class="second-level"><a href="#reference-library-for-variants-annotation-pyensembl">Reference Library for variants annotation : pyensembl</a></li>
                
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<h1 id="reference-data-for-handling-sumstats">Reference data for handling Sumstats</h1>
<h2 id="reference-genome-sequence-for-variant-allele-alignment">Reference genome sequence for variant allele alignment</h2>
<ul>
<li>GRCh37 / hg19 : <a href="http://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/">ucsc_hg19</a></li>
<li>GRCh38 / hg38 : <a href="https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/">ucsc_hg38</a></li>
</ul>
<p>For details about reference genome, please check <a href="https://cloufield.github.io/CTGCatalog/Reference_data_Genome_README/">https://cloufield.github.io/CTGCatalog/Reference_data_Genome_README/</a></p>
<h2 id="processed-reference-files-for-harmonization">Processed Reference files for harmonization</h2>
<h3 id="1000-genome-projecthg19">1000 Genome Project(hg19)</h3>
<p>Download:
<a href="http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/">Index of /vol1/ftp/release/20130502/</a></p>
<p>Process the 1000 genome vcf file for EAS sample:</p>
<ul>
<li>Extract EAS sample</li>
<li>Split multiallelic variants</li>
<li>Normalize variants</li>
<li>Rename variants as CHR:POS:REF:ALT</li>
<li>Remove duplicated variants</li>
<li>Calculate Alternative allele frequency</li>
<li>Tabix index</li>
</ul>
<h3 id="sample-code">Sample code:</h3>
<pre><code class="language-bash">#!/bin/bash
# extract EAS sample ID
awk '$3==&quot;EAS&quot;{print $1}' integrated_call_samples_v3.20130502.ALL.panel &gt;EAS.sample

# process
for chr in {1..22}
do
    bcftools view -S EAS.sample ALL.chr&quot;${chr}&quot;.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz | \
        bcftools norm -m-any --check-ref w -f human_g1k_v37.fasta | \
        bcftools annotate -x ID,INFO -I +'%CHROM:%POS:%REF:%ALT' | \
        bcftools norm --rm-dup both | \
        bcftools +fill-tags -Oz  -- -t AF  \
          &gt; EAS.chr&quot;${chr}&quot;.split_norm_af.vcf.gz
    tabix -p vcf EAS.chr&quot;${chr}&quot;.split_norm_af.vcf.gz
    echo &quot;EAS.chr&quot;${chr}&quot;.split_norm_af.vcf.gz&quot; &gt;&gt;concat_list.txt 
done

# merge
bcftools concat -a -d both -f concat_list.txt -Ob | bcftools sort -Oz  &gt; EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz
tabix -p vcf EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz
</code></pre>
<p>The processed dataset can then be used for:</p>
<ul>
<li>infer palindromic SNPs / indels</li>
<li>check allele frequency difference </li>
<li>regional plot</li>
</ul>
<h2 id="rsid-conversion-table-for-quick-annotation">rsID conversion table for quick annotation</h2>
<p>Since gwaslab will check chr:pos and ea/nea to match rsID, it would take a little bit longer if we only use vcf. </p>
<p>But we can use a pre-annotated conversion table for common SNPs, and then annotate the rest of SNPs using large VCF file from dbSNP. </p>
<pre><code class="language-python">for i in range(1,23):
    sumstats = gl.Sumstats(&quot;./EAS.chr&quot;+str(i)+&quot;.split_norm_af.vcf.gz&quot;,snpid=&quot;ID&quot;,fmt=&quot;vcf&quot;)
    sumstats.harmonize( basic_check=True,
                        ref_seq=&quot;./reference_genome/hg19/human_g1k_v37_decoy.fasta&quot;,
                        ref_rsid_vcf=&quot;./All_20180423.vcf.gz&quot;, 
                        n_cores=4)
    sumstats.data.loc[:,[&quot;SNPID&quot;,&quot;rsID&quot;,&quot;CHR&quot;,&quot;POS&quot;,&quot;NEA&quot;,&quot;EA&quot;]].to_csv(&quot;./1kg_af_dbsnp151.&quot;+str(i)+&quot;.txt.gz&quot;,&quot;\t&quot;,index=None)
</code></pre>
<p>In terminal, combine the files:</p>
<pre><code># get header
zcat 1kg_af_dbsnp151.1.txt.gz | head -1 &gt; 1kg_af_dbsnp151_auto.txt

for i in $(seq 1 22)
do
# get complete SNPID-rsID pairs
zcat 1kg_af_dbsnp151.${i}.txt.gz | awk -v FS=&quot;\t&quot; -v OFS=&quot;\t&quot; 'NR&gt;1 &amp;&amp; $2!=&quot;&quot; {print $0}' &gt;&gt;1kg_af_dbsnp151_auto.txt
done
</code></pre>
<h2 id="dbsnp-database-for-annotation-of-rsid">dbsnp: database for annotation of rsID</h2>
<ul>
<li>dbsnp v151 (GRCh37): <a href="https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh37p13/VCF/00-All.vcf.gz">https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh37p13/VCF/00-All.vcf.gz</a></li>
<li>dbsnp v151 (GRCh37): <a href="https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh37p13/VCF/00-All.vcf.gz.tbi">https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh37p13/VCF/00-All.vcf.gz.tbi</a></li>
<li>dbsnp v151 (GRCh38): <a href="https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF//00-All.vcf.gz">https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF//00-All.vcf.gz</a></li>
<li>dbsnp v151 (GRCh38): <a href="https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF//00-All.vcf.gz.tbi">https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF//00-All.vcf.gz.tbi</a></li>
<li>latest release: <a href="https://ftp.ncbi.nih.gov/snp/latest_release/VCF/">Index of /snp/latest_release/VCF</a></li>
</ul>
<h2 id="gnomad">gnomad</h2>
<ul>
<li>Allele frequency for major ancestries and rsID</li>
<li>gnomAD v2 &amp; gnomAD v2 liftover &amp; gnomAD v3:   <a href="https://gnomad.broadinstitute.org/downloads">gnomAD</a>    </li>
</ul>
<h2 id="reference-library-for-variants-annotation-pyensembl">Reference Library for variants annotation : pyensembl</h2>
<p>Install pyensembl and download reference:</p>
<pre><code># install  pyensembl if not
pip install pyensembl

# syntax for download reference for pyensembl
pyensembl install --release &lt;list of Ensembl release numbers&gt; --species &lt;species-name&gt;
</code></pre>
<p>GWASlab uses:
- ensembl release 75 : hg19
- ensembl release 107 : hg38 
- NCBI refseq GRCh37
- NCBI refseq GRCh38</p>
<p>Currently, gwaslab could use ensembl or refseq gtf reference data to annotate lead SNPs with the nearest gene name.
For details about pyensembl, please check <a href="https://github.com/openvax/pyensembl">https://github.com/openvax/pyensembl</a></p></div>
        
        
    </div>

    
      <footer class="col-md-12 text-center">
          
          
            <hr>
            <p>
            <small>GWASLab is licensed under the MIT license</small><br>
            
            <small>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</small>
            </p>
          

          
          
      </footer>
    
    <script src="//ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="../js/bootstrap-3.0.3.min.js"></script>

    
    <script src="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/highlight.min.js"></script>
        
    <script>hljs.initHighlightingOnLoad();</script>
    

    <script>var base_url = ".."</script>
    
    <script src="../js/base.js"></script>

    <div class="modal" id="mkdocs_keyboard_modal" tabindex="-1" role="dialog" aria-labelledby="keyboardModalLabel" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
            </div>
            <div class="modal-body">
              <table class="table">
                <thead>
                  <tr>
                    <th style="width: 20%;">Keys</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td class="help shortcut"><kbd>?</kbd></td>
                    <td>Open this help</td>
                  </tr>
                  <tr>
                    <td class="next shortcut"><kbd>n</kbd></td>
                    <td>Next page</td>
                  </tr>
                  <tr>
                    <td class="prev shortcut"><kbd>p</kbd></td>
                    <td>Previous page</td>
                  </tr>
                  <tr>
                    <td class="search shortcut"><kbd>s</kbd></td>
                    <td>Search</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>
    </body>

</html>
