
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="HE Yunye">
      
      
      <link rel="icon" href="../assets/gwaslab.png">
      <meta name="generator" content="mkdocs-1.3.0, mkdocs-material-8.2.15">
    
    
      
        <title>Effect Size Comparison - GWASLab</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.c382b1dc.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.cc9b2e1e.min.css">
        
          
          
          <meta name="theme-color" content="#2094f3">
        
      
    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="blue" data-md-color-accent="blue">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#comparing-effect-sizes" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="GWASLab" class="md-header__button md-logo" aria-label="GWASLab" data-md-component="logo">
      
  <img src="../assets/gwaslab.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            GWASLab
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Effect Size Comparison
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/Cloufield/gwaslab/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-tabs__inner md-grid">
    <ul class="md-tabs__list">
      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        Home
      </a>
    </li>
  

      
        
  
  


  <li class="md-tabs__item">
    <a href="../tutorial_3.4/" class="md-tabs__link">
      Tutorial for gwaslab
    </a>
  </li>

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../SumstatsObject/" class="md-tabs__link">
        Sumstats manipulation
      </a>
    </li>
  

      
        
  
  
    
  


  
  
  
    <li class="md-tabs__item">
      <a href="../Visualization/" class="md-tabs__link md-tabs__link--active">
        Visualization
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../ExtractLead/" class="md-tabs__link">
        Utilities
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../Download/" class="md-tabs__link">
        References
      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="GWASLab" class="md-nav__button md-logo" aria-label="GWASLab" data-md-component="logo">
      
  <img src="../assets/gwaslab.png" alt="logo">

    </a>
    GWASLab
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Cloufield/gwaslab/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_1" type="checkbox" id="__nav_1" >
      
      
      
      
        <label class="md-nav__link" for="__nav_1">
          Home
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Home" data-md-level="1">
        <label class="md-nav__title" for="__nav_1">
          <span class="md-nav__icon md-icon"></span>
          Home
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Home
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../UpdateLogs/" class="md-nav__link">
        Update logs
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../tutorial_3.4/" class="md-nav__link">
        Tutorial for gwaslab
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3" type="checkbox" id="__nav_3" >
      
      
      
      
        <label class="md-nav__link" for="__nav_3">
          Sumstats manipulation
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Sumstats manipulation" data-md-level="1">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          Sumstats manipulation
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../SumstatsObject/" class="md-nav__link">
        Sumstats Object
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../StatusCode/" class="md-nav__link">
        Status Code
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Standardization/" class="md-nav__link">
        Standardization and Normalization
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../QC%26Filtering/" class="md-nav__link">
        QC and Filtering
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Harmonization/" class="md-nav__link">
        Harmonization
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../AssignrsID/" class="md-nav__link">
        Assigning rsID
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../AssignCHRPOS/" class="md-nav__link">
        Assign CHR/POS using rsID
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../LiftOver/" class="md-nav__link">
        Liftover
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Conversion/" class="md-nav__link">
        Data conversion
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4" type="checkbox" id="__nav_4" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_4">
          Visualization
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Visualization" data-md-level="1">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Visualization
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Visualization/" class="md-nav__link">
        Manhattan and QQ plot
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../RegionalPlot/" class="md-nav__link">
        Regional Plot
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../BrisbanePlot/" class="md-nav__link">
        Brisbane Plot
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../MiamiPlot/" class="md-nav__link">
        Miami Plot
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Effect Size Comparison
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Effect Size Comparison
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#scatter-plot-effect-size-comparison" class="md-nav__link">
    Scatter plot : effect size comparison
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#warning-this-function-is-still-being-tested-will-be-available-from-v3417" class="md-nav__link">
    (WARNING: this function is still being tested!!! Will be available from v3.4.17)
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#options" class="md-nav__link">
    Options
  </a>
  
    <nav class="md-nav" aria-label="Options">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#path-and-column" class="md-nav__link">
    Path and column
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#save-figures" class="md-nav__link">
    Save figures
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#snplist" class="md-nav__link">
    Snplist
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filter-by-maf" class="md-nav__link">
    Filter by maf:
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#label" class="md-nav__link">
    Label
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#annotation" class="md-nav__link">
    Annotation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#heterogeneity-test" class="md-nav__link">
    Heterogeneity test
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#r-se" class="md-nav__link">
    R SE
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#example" class="md-nav__link">
    Example:
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../AlleleFrequency/" class="md-nav__link">
        Allele Frequency Comparison
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../GeneticCorrelation/" class="md-nav__link">
        Genetic correlation heatmap
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Gallery/" class="md-nav__link">
        Gallery
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5" type="checkbox" id="__nav_5" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5">
          Utilities
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Utilities" data-md-level="1">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          Utilities
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ExtractLead/" class="md-nav__link">
        Extract Lead Variants
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ExtractNovel/" class="md-nav__link">
        Extract Novel Variants
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Format/" class="md-nav__link">
        Format and Save
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Pickle/" class="md-nav__link">
        Save Sumstats Object
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../LoadLDSC/" class="md-nav__link">
        Load LDSC log
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../HeritabilityConversion/" class="md-nav__link">
        Convert Heritability
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../PerSNPh2/" class="md-nav__link">
        Per SNP Heritability and F statistics
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../InferBuild/" class="md-nav__link">
        Infer Genome Build
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6" type="checkbox" id="__nav_6" >
      
      
      
      
        <label class="md-nav__link" for="__nav_6">
          References
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="References" data-md-level="1">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          References
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Download/" class="md-nav__link">
        Download
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Reference/" class="md-nav__link">
        Reference data
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CommonData/" class="md-nav__link">
        Common data
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#scatter-plot-effect-size-comparison" class="md-nav__link">
    Scatter plot : effect size comparison
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#warning-this-function-is-still-being-tested-will-be-available-from-v3417" class="md-nav__link">
    (WARNING: this function is still being tested!!! Will be available from v3.4.17)
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#options" class="md-nav__link">
    Options
  </a>
  
    <nav class="md-nav" aria-label="Options">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#path-and-column" class="md-nav__link">
    Path and column
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#save-figures" class="md-nav__link">
    Save figures
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#snplist" class="md-nav__link">
    Snplist
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filter-by-maf" class="md-nav__link">
    Filter by maf:
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#label" class="md-nav__link">
    Label
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#annotation" class="md-nav__link">
    Annotation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#heterogeneity-test" class="md-nav__link">
    Heterogeneity test
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#r-se" class="md-nav__link">
    R SE
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#example" class="md-nav__link">
    Example:
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                


<h1 id="comparing-effect-sizes">Comparing effect sizes</h1>
<h2 id="scatter-plot-effect-size-comparison">Scatter plot : effect size comparison</h2>
<h2 id="warning-this-function-is-still-being-tested-will-be-available-from-v3417">(WARNING: this function is still being tested!!! Will be available from v3.4.17)</h2>
<p><code>gl.compare_effect()</code> will plot effect size comparison plot using two sets of sumstats. Alleles will be aligned to effect alleles in sumstats1.</p>
<div class="highlight"><pre><span></span><code><span class="n">gl</span><span class="o">.</span><span class="n">compare_effect</span> <span class="p">(</span><span class="n">path1</span><span class="p">,</span>
                   <span class="n">path2</span><span class="p">,</span>
                   <span class="n">cols_name_list_1</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">effect_cols_list_1</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                   <span class="n">cols_name_list_2</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">effect_cols_list_2</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                   <span class="n">eaf</span><span class="o">=</span><span class="p">[],</span>
                   <span class="n">maf_level</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                   <span class="n">label</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                   <span class="n">snplist</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                   <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;beta&quot;</span><span class="p">,</span>
                   <span class="n">anno</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                   <span class="n">anno_het</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                   <span class="n">anno_min</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                   <span class="n">anno_min1</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                   <span class="n">anno_min2</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                   <span class="n">anno_diff</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                   <span class="n">scaled</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                   <span class="n">scaled1</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                   <span class="n">scaled2</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                   <span class="n">wc_correction</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> 
                   <span class="n">null_beta</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                   <span class="n">is_q</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                   <span class="n">include_all</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                   <span class="n">q_level</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
                   <span class="n">sig_level</span><span class="o">=</span><span class="mf">5e-8</span><span class="p">,</span>
                   <span class="n">drop</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                   <span class="n">wc_sig_level</span><span class="o">=</span><span class="mf">5e-8</span><span class="p">,</span>
                   <span class="c1"># reg</span>
                   <span class="n">reg_box</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                   <span class="n">is_reg</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                   <span class="n">fdr</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                   <span class="n">allele_match</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                   <span class="n">r_se</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                   <span class="n">is_45_helper_line</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                   <span class="n">legend_mode</span><span class="o">=</span><span class="s2">&quot;full&quot;</span><span class="p">,</span>
                   <span class="n">legend_title</span><span class="o">=</span><span class="sa">r</span><span class="s1">&#39;$ P &lt; 5 x 10^{-8}$ in:&#39;</span><span class="p">,</span>
                   <span class="n">legend_title2</span><span class="o">=</span><span class="sa">r</span><span class="s1">&#39;Heterogeneity test:&#39;</span><span class="p">,</span>
                   <span class="n">legend_pos</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span>
                   <span class="n">scatterargs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                   <span class="n">plt_args</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                   <span class="n">xylabel_prefix</span><span class="o">=</span><span class="s2">&quot;Per-allele effect size in &quot;</span><span class="p">,</span>
                   <span class="n">helper_line_args</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                   <span class="n">fontargs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                   <span class="n">errargs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                   <span class="n">legend_args</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                   <span class="n">sep</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">],</span>
                   <span class="n">log</span> <span class="o">=</span> <span class="n">Log</span><span class="p">(),</span>
                   <span class="n">save</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                   <span class="n">saveargs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                   <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div>
<h2 id="options">Options</h2>
<h3 id="path-and-column">Path and column</h3>
<ul>
<li><code>path1</code> and <code>path2</code> : the paths to the sumstats. Can also be <code>gl.Sumstats</code> Object or <code>pd.DataFrame</code> (from v3.4.17). If <code>gl.Sumstats</code> Objects are provided, there is no need to set cols_name_list and effect_cols_list.</li>
<li><code>cols_name_list_1</code> and <code>cols_name_list_2</code> : list of column names for variants basic information</li>
<li><code>effect_cols_list_1</code> and <code>effect_cols_list_2</code> : list of column names for effect size-related columns</li>
<li><code>mode</code> : use beta or OR </li>
<li><code>cols_name_list_x</code> and <code>effect_cols_list_x</code> examples:<ul>
<li><code>[snpid,p,ea,nea]</code>        ,<code>[effect,se]</code></li>
<li><code>[snpid,p,ea,nea,chr,pos]</code>,<code>[effect,se]</code></li>
<li><code>[snpid,p,ea,nea,chr,pos]</code>,<code>[OR,OR_l,OR_h]</code></li>
</ul>
</li>
</ul>
<table>
<thead>
<tr>
<th>Option</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>path1</code></td>
<td><code>string</code>,<code>gl.Sumstats</code>, or <code>pd.DataFrame</code></td>
<td>path to the sumstats file, or gwaslab Sumstats object or pandas Dataframe</td>
<td>None</td>
</tr>
<tr>
<td><code>path2</code></td>
<td><code>string</code>,<code>gl.Sumstats</code>, or <code>pd.DataFrame</code></td>
<td>path to the sumstats file, or gwaslab Sumstats object or pandas Dataframe</td>
<td>None</td>
</tr>
<tr>
<td><code>cols_name_list_1</code></td>
<td><code>list</code></td>
<td>"[snpid,p,ea,nea]" or "[snpid,p,ea,nea,chr,pos]"</td>
<td>None</td>
</tr>
<tr>
<td><code>cols_name_list_2</code></td>
<td><code>list</code></td>
<td>"[snpid,p,ea,nea]" or "[snpid,p,ea,nea,chr,pos]"</td>
<td>None</td>
</tr>
<tr>
<td><code>effect_cols_list_1</code></td>
<td><code>list</code></td>
<td>"[effect,se]" or "[OR,OR_95l,OR_95h]"</td>
<td>None</td>
</tr>
<tr>
<td><code>effect_cols_list_2</code></td>
<td><code>list</code></td>
<td>"[effect,se]" or "[OR,OR_95l,OR_95h]"</td>
<td>None</td>
</tr>
<tr>
<td><code>mode</code></td>
<td><code>beta</code> or <code>OR</code></td>
<td>plot beta or OR</td>
<td><code>beta</code></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>from v3.4.17, you need to specify the parameters using keywords instead of using them as positional arguments for <code>cols_name_list_1</code> and <code>cols_name_list_2</code>, <code>effect_cols_list_1</code> and <code>effect_cols_list_1</code> .</p>
</div>
<h3 id="save-figures">Save figures</h3>
<table>
<thead>
<tr>
<th>Option</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>save</code></td>
<td><code>string</code></td>
<td>path to the saved file</td>
<td><code>./Sumstats1_Sumstats2_effect_comparison_plot.png</code></td>
</tr>
<tr>
<td><code>saveargs</code></td>
<td><code>dict</code></td>
<td>parametrs for plt.savefig()</td>
<td><code>{"dpi":300,"facecolor":"white"}</code></td>
</tr>
</tbody>
</table>
<h3 id="snplist">Snplist</h3>
<table>
<thead>
<tr>
<th>Option</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>snplist</code></td>
<td><code>list</code></td>
<td>optional, specify the variants you want to compare. If None, GWASLab will automatically extract lead variants from both sumstats.</td>
<td>None</td>
</tr>
</tbody>
</table>
<h3 id="filter-by-maf">Filter by maf:</h3>
<table>
<thead>
<tr>
<th>Option</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>eaf</code></td>
<td><code>list</code></td>
<td>optional, a list column names for effect allele frequency, in the order of [sumstats1_eaf, sumstats2_eaf]. It is required when you need to filter by maf using <code>maf_level</code>.</td>
<td><code>None</code></td>
</tr>
<tr>
<td><code>maf_level</code></td>
<td><code>float</code></td>
<td>the maf filter for variants. Vairants with maf &lt; maf_level will be removed from comparison.</td>
<td><code>None</code></td>
</tr>
</tbody>
</table>
<h3 id="label">Label</h3>
<table>
<thead>
<tr>
<th>Option</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>label</code></td>
<td><code>list</code></td>
<td>a list of labels for the legend , in the order of ["Sumstats_1","Sumstats_2","Both","None"]</td>
<td><code>["Sumstats_1","Sumstats_2","Both","None"]</code></td>
</tr>
<tr>
<td><code>sig_level</code></td>
<td><code>float</code></td>
<td>the significance level for auto-extracting lead variants. If <code>snplist</code> is provided, the auto-extraction will be skipped.</td>
<td><code>5e-8</code></td>
</tr>
<tr>
<td><code>legend_title</code></td>
<td><code>string</code></td>
<td>legend title</td>
<td><code>'$ P &lt; 5 x 10^{-8}$ in:'</code></td>
</tr>
<tr>
<td><code>legend_pos</code></td>
<td><code>string</code></td>
<td>legend position</td>
<td><code>upper left</code></td>
</tr>
<tr>
<td><code>xylabel_prefix</code></td>
<td><code>string</code></td>
<td>-</td>
<td><code>"Per-allele effect size in "</code></td>
</tr>
</tbody>
</table>
<h3 id="annotation">Annotation</h3>
<table>
<thead>
<tr>
<th>Option</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>is_reg</code></td>
<td><code>boolean</code></td>
<td>if true, draw regression line.</td>
<td><code>True</code></td>
</tr>
<tr>
<td><code>is_45_helper_line</code></td>
<td><code>boolean</code></td>
<td>if true, draw 45 degree line.</td>
<td><code>True</code></td>
</tr>
<tr>
<td><code>anno</code></td>
<td><code>boolean</code></td>
<td>if true, annotate the variants with ID.</td>
<td><code>False</code></td>
</tr>
<tr>
<td><code>anno_diff</code></td>
<td><code>float</code></td>
<td>threshold of effect size difference for annotation.</td>
<td><code>0</code></td>
</tr>
<tr>
<td><code>anno_min1</code></td>
<td><code>float</code></td>
<td>threshold of sumstats1 minimum absolute effect size for annotation.</td>
<td><code>0</code></td>
</tr>
<tr>
<td><code>anno_min2</code></td>
<td><code>float</code></td>
<td>threshold of sumstats2 minimum absolute effect size for annotation.</td>
<td><code>0</code></td>
</tr>
</tbody>
</table>
<h3 id="heterogeneity-test">Heterogeneity test</h3>
<table>
<thead>
<tr>
<th>Option</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>is_q</code></td>
<td><code>boolean</code></td>
<td>if true, apply the heterogeneity tests by Cochran's Q test.</td>
<td><code>True</code></td>
</tr>
<tr>
<td><code>q_level</code></td>
<td><code>float</code></td>
<td>the significance threshold for Cochran's Q test (raw p value).</td>
<td><code>0.05</code></td>
</tr>
<tr>
<td><code>anno_het</code></td>
<td></td>
<td>annotate only variants with Phet &lt; <code>q_level</code></td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h3 id="r-se">R SE</h3>
<table>
<thead>
<tr>
<th>Option</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>r_se</code></td>
<td><code>boolean</code></td>
<td>If True, SE for r will be estimated using the jackknife method. (Note: available from v3.4.17)</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<div class="arithmatex">\[ s.e.(\hat{r}_{jack}) = \sqrt{ {{n-1}\over{n}} \sum_{i=1}^n(\hat{r_i} -\bar{r}_{jack} )^2 } \]</div>
<h3 id="_1"></h3>
<h2 id="example">Example:</h2>
<div class="admonition example">
<p class="admonition-title">Use gl.Sumstats object</p>
<p><div class="highlight"><pre><span></span><code>gl1 = gl.Sumstats(&quot;bbj_bmi_female.txt.gz&quot;,fmt=&quot;gwaslab&quot;,snpid=&quot;SNP&quot;,ea=&quot;REF&quot;,nea=&quot;ALT&quot;,sep=&quot;\t&quot;)
gl2 = gl.Sumstats(&quot;bbj_bmi_male.txt.gz&quot;,fmt=&quot;gwaslab&quot;,snpid=&quot;SNP&quot;,ea=&quot;REF&quot;,nea=&quot;ALT&quot;,sep=&quot;\t&quot;)

# auto extract lead SNPs and compare the effect sizes
a = gl.compare_effect(path1= gl1,
                      path2= gl2
)
</code></pre></div>
<img width="500" alt="image" src="https://github.com/Cloufield/gwaslab/assets/40289485/258d3316-7d73-408a-8e20-e560b831861c"></p>
</div>
<div class="admonition example">
<p class="admonition-title">Use pandas DataFrame</p>
<div class="highlight"><pre><span></span><code>pd1 = pd.read_table(&quot;bbj_bmi_female.txt.gz&quot;,sep=&quot;\t&quot;)
pd2 = pd.read_table(&quot;bbj_bmi_male.txt.gz&quot;,sep=&quot;\t&quot;)

# cols_name_list should be SNPID, P, Effect Allele, Non-Effect allele, Chromosome and Position
# effect_cols_list should be BETA,SE

a = gl.compare_effect(path1 = pd1,
                    cols_name_list_1 = [&quot;SNP&quot;,&quot;P&quot;,&quot;REF&quot;,&quot;ALT&quot;,&quot;CHR&quot;,&quot;POS&quot;],effect_cols_list_1= [&quot;BETA&quot;,&quot;SE&quot;],
                    path2 = pd2,
                    cols_name_list_2 = [&quot;SNP&quot;,&quot;P&quot;,&quot;REF&quot;,&quot;ALT&quot;,&quot;CHR&quot;,&quot;POS&quot;],effect_cols_list_2= [&quot;BETA&quot;,&quot;SE&quot;]
                    )
</code></pre></div>
</div>
<div class="admonition example">
<p class="admonition-title">Heterogeneity test</p>
<p><div class="highlight"><pre><span></span><code>pd1 = pd.read_table(&quot;bbj_bmi_female.txt.gz&quot;,sep=&quot;\t&quot;)
pd2 = pd.read_table(&quot;bbj_bmi_male.txt.gz&quot;,sep=&quot;\t&quot;)

# cols_name_list should be SNPID, P, Effect Allele, Non-Effect allele, Chromosome and Position
# effect_cols_list should be BETA,SE
# is_q : if true, apply the heterogeneity tests by Cochran&#39;s Q test.
# q_level` : `float`, the significance threshold for Cochran&#39;s Q test (raw p value).

a = gl.compare_effect(path1 = pd1,
                      cols_name_list_1 = [&quot;SNP&quot;,&quot;P&quot;,&quot;REF&quot;,&quot;ALT&quot;,&quot;CHR&quot;,&quot;POS&quot;],effect_cols_list_1= [&quot;BETA&quot;,&quot;SE&quot;],
                      path2 = pd2,
                      cols_name_list_2 = [&quot;SNP&quot;,&quot;P&quot;,&quot;REF&quot;,&quot;ALT&quot;,&quot;CHR&quot;,&quot;POS&quot;],effect_cols_list_2= [&quot;BETA&quot;,&quot;SE&quot;],
                      is_q=True,
                      q_level=0.05,
                      legend_mode=&quot;full&quot;
                      )
</code></pre></div>
 <img width="500" alt="image" src="https://github.com/Cloufield/gwaslab/assets/40289485/b9c4511c-c1eb-4c36-8b72-788d2b21b790"></p>
</div>
<div class="admonition example">
<p class="admonition-title">Annotation</p>
<p><div class="highlight"><pre><span></span><code>a = gl.compare_effect(path1=&quot;bbj_bmi_female.txt.gz&quot;,
                  cols_name_list_1=[&quot;SNP&quot;,&quot;P&quot;,&quot;REF&quot;,&quot;ALT&quot;,&quot;CHR&quot;,&quot;POS&quot;],effect_cols_list_1=[&quot;BETA&quot;,&quot;SE&quot;],
                  path2=&quot;bbj_bmi_male.txt.gz&quot;,
                  cols_name_list_2=[&quot;SNP&quot;,&quot;P&quot;,&quot;REF&quot;,&quot;ALT&quot;,&quot;CHR&quot;,&quot;POS&quot;],effect_cols_list_2=[&quot;BETA&quot;,&quot;SE&quot;],
                  label=[&quot;Female&quot;,&quot;Male&quot;,&quot;Both&quot;,&quot;None&quot;],
                  xylabel_prefix=&quot;Per-allele effect size for &quot;,
                  r_se=True, 
                  is_q=True, 
                  anno=True, 
                  anno_het=True, # only annotate variants with significant heterogeneity
                  anno_diff=0.015,    # only annotate variants with a difference in effect size &gt; 0.015
                  sig_level=5e-8,
                  legend_title=r&#39;$ P &lt; 5 x 10^{-8}$ in:&#39;,
                  verbose=True)
</code></pre></div>
<img width="500" alt="image" src="https://github.com/Cloufield/gwaslab/assets/40289485/9e2ac224-63d6-44da-9dbc-1cd43596055f"></p>
</div>
<div class="admonition example">
<p class="admonition-title">Male-specific and female-specific BMI Sumstats from JENGER</p>
<div class="highlight"><pre><span></span><code>!wget -O bbj_bmi_male.txt.gz http://jenger.riken.jp/2analysisresult_qtl_download/
!wget -O bbj_bmi_female.txt.gz http://jenger.riken.jp/4analysisresult_qtl_download/

Headers of the files : # SNP    CHR POS REF ALT Frq Rsq BETA    SE  P
</code></pre></div>
<p><div class="highlight"><pre><span></span><code><span class="c1"># GWASLab will automatically extract significant variants from both sumstats. </span>
<span class="n">a</span> <span class="o">=</span> <span class="n">gl</span><span class="o">.</span><span class="n">compare_effect</span><span class="p">(</span><span class="n">path1</span><span class="o">=</span><span class="s2">&quot;bbj_bmi_female.txt.gz&quot;</span><span class="p">,</span>
                  <span class="n">cols_name_list_1</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SNP&quot;</span><span class="p">,</span><span class="s2">&quot;P&quot;</span><span class="p">,</span><span class="s2">&quot;REF&quot;</span><span class="p">,</span><span class="s2">&quot;ALT&quot;</span><span class="p">,</span><span class="s2">&quot;CHR&quot;</span><span class="p">,</span><span class="s2">&quot;POS&quot;</span><span class="p">],</span><span class="n">effect_cols_list_1</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;BETA&quot;</span><span class="p">,</span><span class="s2">&quot;SE&quot;</span><span class="p">],</span>
                  <span class="n">path2</span><span class="o">=</span><span class="s2">&quot;bbj_bmi_male.txt.gz&quot;</span><span class="p">,</span>
                  <span class="n">cols_name_list_2</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SNP&quot;</span><span class="p">,</span><span class="s2">&quot;P&quot;</span><span class="p">,</span><span class="s2">&quot;REF&quot;</span><span class="p">,</span><span class="s2">&quot;ALT&quot;</span><span class="p">,</span><span class="s2">&quot;CHR&quot;</span><span class="p">,</span><span class="s2">&quot;POS&quot;</span><span class="p">],</span><span class="n">effect_cols_list_2</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;BETA&quot;</span><span class="p">,</span><span class="s2">&quot;SE&quot;</span><span class="p">],</span>
                  <span class="n">label</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Female&quot;</span><span class="p">,</span><span class="s2">&quot;Male&quot;</span><span class="p">,</span><span class="s2">&quot;Both&quot;</span><span class="p">,</span><span class="s2">&quot;None&quot;</span><span class="p">],</span>
                  <span class="n">xylabel_prefix</span><span class="o">=</span><span class="s2">&quot;Per-allele effect size for &quot;</span><span class="p">,</span>
                  <span class="n">anno</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                  <span class="n">anno_het</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                  <span class="n">anno_diff</span><span class="o">=</span><span class="mf">0.015</span><span class="p">,</span>
                  <span class="n">sig_level</span><span class="o">=</span><span class="mf">5e-8</span><span class="p">,</span>
                  <span class="n">legend_title</span><span class="o">=</span><span class="sa">r</span><span class="s1">&#39;$ P &lt; 5 x 10^{-8}$ in:&#39;</span><span class="p">,</span>
                  <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                  <span class="n">save</span> <span class="o">=</span> <span class="s2">&quot;myplot.png&quot;</span><span class="p">,</span>
                  <span class="n">saveargs</span><span class="o">=</span> <span class="p">{</span><span class="s2">&quot;dpi&quot;</span><span class="p">:</span><span class="mi">300</span><span class="p">,</span><span class="s2">&quot;facecolor&quot;</span><span class="p">:</span><span class="s2">&quot;white&quot;</span><span class="p">}</span>
<span class="p">)</span>
</code></pre></div>
<img width="500" alt="image" src="https://github.com/Cloufield/gwaslab/assets/40289485/ed462a3d-64cf-4f06-bdb5-13294b4182c8"></p>
<p>Reference: Akiyama, M., Okada, Y., Kanai, M., Takahashi, A., Momozawa, Y., Ikeda, M., ... &amp; Kamatani, Y. (2017). Genome-wide association study identifies 112 new loci for body mass index in the Japanese population. Nature genetics, 49(10), 1458-1467.</p>
</div>

              
            </article>
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../MiamiPlot/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Miami Plot" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Miami Plot
            </div>
          </div>
        </a>
      
      
        
        <a href="../AlleleFrequency/" class="md-footer__link md-footer__link--next" aria-label="Next: Allele Frequency Comparison" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Allele Frequency Comparison
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      GWASLab is licensed under the MIT license
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs"], "search": "../assets/javascripts/workers/search.2a1c317c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.a6c66575.min.js"></script>
      
        <script src="../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>