{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0,\"/Users/he/work/gwaslab/src\")\n",
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading raw sumstats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Nov  4 14:17:34 2022 Start to initiate from file :t2d_bbj.txt.gz\n",
      "Fri Nov  4 14:17:59 2022  -Reading columns          : Frq,Dir,N,P,ALT,POS,CHR,BETA,REF,SNP,SE\n",
      "Fri Nov  4 14:17:59 2022  -Renaming columns to      : EAF,DIRECTION,N,P,EA,POS,CHR,BETA,NEA,SNPID,SE\n",
      "Fri Nov  4 14:17:59 2022  -Current dataframe shape  : Rows  12557761  x  11  Columns\n",
      "Fri Nov  4 14:18:01 2022  -Initiating a status column ...\n",
      "Fri Nov  4 14:18:08 2022  -NEAF is specified...\n",
      "Fri Nov  4 14:18:08 2022  -Checking if 0<= NEAF <=1 ...\n",
      "Fri Nov  4 14:18:10 2022  -Converted NEAF to EAF.\n",
      "Fri Nov  4 14:18:10 2022  -Removed 0 variants with bad NEAF.\n",
      "Fri Nov  4 14:18:10 2022  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,EAF,BETA,SE,P,N,DIRECTION,STATUS\n",
      "Fri Nov  4 14:18:14 2022 Finished loading data successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(\"t2d_bbj.txt.gz\",\n",
    "             snpid=\"SNP\",\n",
    "             chrom=\"CHR\",\n",
    "             pos=\"POS\",\n",
    "             ea=\"ALT\",\n",
    "             nea=\"REF\",\n",
    "             neaf=\"Frq\",\n",
    "             beta=\"BETA\",\n",
    "             se=\"SE\",\n",
    "             p=\"P\",\n",
    "             direction=\"Dir\",\n",
    "             n=\"N\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>DIRECTION</th>\n",
       "      <th>STATUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:725932_G_A</td>\n",
       "      <td>1</td>\n",
       "      <td>725932</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9960</td>\n",
       "      <td>-0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.5970</td>\n",
       "      <td>166718</td>\n",
       "      <td>-?+-</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:725933_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>725933</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.5973</td>\n",
       "      <td>166718</td>\n",
       "      <td>+?-+</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:737801_T_C</td>\n",
       "      <td>1</td>\n",
       "      <td>737801</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.0490</td>\n",
       "      <td>0.1231</td>\n",
       "      <td>0.6908</td>\n",
       "      <td>166718</td>\n",
       "      <td>+?-+</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:749963_T_TAA</td>\n",
       "      <td>1</td>\n",
       "      <td>749963</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8374</td>\n",
       "      <td>0.0213</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.2846</td>\n",
       "      <td>166718</td>\n",
       "      <td>-?++</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:751343_T_A</td>\n",
       "      <td>1</td>\n",
       "      <td>751343</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8593</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.2705</td>\n",
       "      <td>166718</td>\n",
       "      <td>-?++</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12557756</th>\n",
       "      <td>X:154874837_A_G</td>\n",
       "      <td>X</td>\n",
       "      <td>154874837</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7478</td>\n",
       "      <td>-0.0064</td>\n",
       "      <td>0.0117</td>\n",
       "      <td>0.5840</td>\n",
       "      <td>191764</td>\n",
       "      <td>-+-+</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12557757</th>\n",
       "      <td>X:154875192_GTACTC_G</td>\n",
       "      <td>X</td>\n",
       "      <td>154875192</td>\n",
       "      <td>GTACTC</td>\n",
       "      <td>G</td>\n",
       "      <td>0.2525</td>\n",
       "      <td>0.0071</td>\n",
       "      <td>0.0122</td>\n",
       "      <td>0.5612</td>\n",
       "      <td>191764</td>\n",
       "      <td>+-+-</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12557758</th>\n",
       "      <td>X:154879115_A_G</td>\n",
       "      <td>X</td>\n",
       "      <td>154879115</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7463</td>\n",
       "      <td>-0.0070</td>\n",
       "      <td>0.0122</td>\n",
       "      <td>0.5646</td>\n",
       "      <td>191764</td>\n",
       "      <td>-+-+</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12557759</th>\n",
       "      <td>X:154880669_T_A</td>\n",
       "      <td>X</td>\n",
       "      <td>154880669</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.2558</td>\n",
       "      <td>0.0071</td>\n",
       "      <td>0.0122</td>\n",
       "      <td>0.5618</td>\n",
       "      <td>191764</td>\n",
       "      <td>+-+-</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12557760</th>\n",
       "      <td>X:154880917_C_T</td>\n",
       "      <td>X</td>\n",
       "      <td>154880917</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.2558</td>\n",
       "      <td>0.0072</td>\n",
       "      <td>0.0122</td>\n",
       "      <td>0.5570</td>\n",
       "      <td>191764</td>\n",
       "      <td>+-+-</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12557761 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         SNPID CHR        POS      EA NEA     EAF    BETA  \\\n",
       "0                 1:725932_G_A   1     725932       G   A  0.9960 -0.0737   \n",
       "1                 1:725933_A_G   1     725933       G   A  0.0040  0.0737   \n",
       "2                 1:737801_T_C   1     737801       C   T  0.0051  0.0490   \n",
       "3               1:749963_T_TAA   1     749963     TAA   T  0.8374  0.0213   \n",
       "4                 1:751343_T_A   1     751343       T   A  0.8593  0.0172   \n",
       "...                        ...  ..        ...     ...  ..     ...     ...   \n",
       "12557756       X:154874837_A_G   X  154874837       G   A  0.7478 -0.0064   \n",
       "12557757  X:154875192_GTACTC_G   X  154875192  GTACTC   G  0.2525  0.0071   \n",
       "12557758       X:154879115_A_G   X  154879115       G   A  0.7463 -0.0070   \n",
       "12557759       X:154880669_T_A   X  154880669       T   A  0.2558  0.0071   \n",
       "12557760       X:154880917_C_T   X  154880917       C   T  0.2558  0.0072   \n",
       "\n",
       "              SE       P       N DIRECTION   STATUS  \n",
       "0         0.1394  0.5970  166718      -?+-  9999999  \n",
       "1         0.1394  0.5973  166718      +?-+  9999999  \n",
       "2         0.1231  0.6908  166718      +?-+  9999999  \n",
       "3         0.0199  0.2846  166718      -?++  9999999  \n",
       "4         0.0156  0.2705  166718      -?++  9999999  \n",
       "...          ...     ...     ...       ...      ...  \n",
       "12557756  0.0117  0.5840  191764      -+-+  9999999  \n",
       "12557757  0.0122  0.5612  191764      +-+-  9999999  \n",
       "12557758  0.0122  0.5646  191764      -+-+  9999999  \n",
       "12557759  0.0122  0.5618  191764      +-+-  9999999  \n",
       "12557760  0.0122  0.5570  191764      +-+-  9999999  \n",
       "\n",
       "[12557761 rows x 12 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Nov  4 14:18:14 2022 Start to check IDs...\n",
      "Fri Nov  4 14:18:14 2022  -Current Dataframe shape : 12557761  x  12\n",
      "Fri Nov  4 14:18:14 2022  -Checking if SNPID is chr:pos:ref:alt...(separator: - ,: , _)\n",
      "Fri Nov  4 14:18:35 2022  -Replacing [_-] in SNPID with \":\" ...\n",
      "Fri Nov  4 14:18:51 2022 Finished checking IDs successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.fix_id(fixsep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Nov  4 14:18:51 2022 Start to randomly select variants from the sumstats...\n",
      "Fri Nov  4 14:18:51 2022  -Number of variants selected from the sumstats: 100000\n",
      "Fri Nov  4 14:18:56 2022 Finished sampling...\n"
     ]
    }
   ],
   "source": [
    "mysumstats.random_variants(n=100000,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Nov  4 14:18:56 2022 Start to check IDs...\n",
      "Fri Nov  4 14:18:56 2022  -Current Dataframe shape : 100000  x  12\n",
      "Fri Nov  4 14:18:56 2022  -Checking if SNPID is chr:pos:ref:alt...(separator: - ,: , _)\n",
      "Fri Nov  4 14:18:57 2022 Finished checking IDs successfully!\n",
      "Fri Nov  4 14:18:57 2022 Start to fix chromosome notation...\n",
      "Fri Nov  4 14:18:57 2022  -Current Dataframe shape : 100000  x  12\n",
      "Fri Nov  4 14:18:57 2022  -Vairants with fixable chromosome notations: 2506\n",
      "Fri Nov  4 14:18:57 2022  -Converting to string datatype and UPPERCASE...\n",
      "Fri Nov  4 14:18:57 2022  -Stripping chr prefix if exists : CHR_-.0...\n",
      "Fri Nov  4 14:18:57 2022  -Identified  2506  variants on sex chromosomes...\n",
      "Fri Nov  4 14:18:57 2022  -Standardizing sex chromosome notations: X Y MT  to 23,24,25...\n",
      "Fri Nov  4 14:18:57 2022  -No unrecognized chromosome notations...\n",
      "Fri Nov  4 14:18:57 2022 Finished fixing chromosome notation successfully!\n",
      "Fri Nov  4 14:18:57 2022 Start to fix basepair positions...\n",
      "Fri Nov  4 14:18:57 2022  -Current Dataframe shape : 100000  x  12\n",
      "Fri Nov  4 14:18:57 2022  -Position upper_bound is: 250,000,000\n",
      "Fri Nov  4 14:18:58 2022  -Remove outliers: 0\n",
      "Fri Nov  4 14:18:58 2022  -Converted all position to datatype Int64.\n",
      "Fri Nov  4 14:18:58 2022 Finished fixing basepair position successfully!\n",
      "Fri Nov  4 14:18:58 2022 Start to fix alleles...\n",
      "Fri Nov  4 14:18:58 2022  -Current Dataframe shape : 100000  x  12\n",
      "Fri Nov  4 14:18:58 2022  -Detected 0 variants with alleles that contain bases other than A/C/T/G .\n",
      "Fri Nov  4 14:18:58 2022  -Converted all bases to string datatype and UPPERCASE.\n",
      "Fri Nov  4 14:18:58 2022 Finished fixing allele successfully!\n",
      "Fri Nov  4 14:18:58 2022 Start sanity check for statistics ...\n",
      "Fri Nov  4 14:18:58 2022  -Current Dataframe shape : 100000  x  12\n",
      "Fri Nov  4 14:18:58 2022  -Checking if  0 <=N<= inf  ...\n",
      "Fri Nov  4 14:18:58 2022  -Removed 0 variants with bad N.\n",
      "Fri Nov  4 14:18:58 2022  -Checking if  0 <=EAF<= 1  ...\n",
      "Fri Nov  4 14:18:58 2022  -Removed 0 variants with bad EAF.\n",
      "Fri Nov  4 14:18:58 2022  -Checking if  5 <=MAC<= inf  ...\n",
      "Fri Nov  4 14:18:58 2022  -Removed 0 variants with bad MAC.\n",
      "Fri Nov  4 14:18:58 2022  -Checking if  5e-300 <= P <= 1  ...\n",
      "Fri Nov  4 14:18:58 2022  -Removed 0 variants with bad P.\n",
      "Fri Nov  4 14:18:58 2022  -Checking if  -10 <BETA)< 10  ...\n",
      "Fri Nov  4 14:18:58 2022  -Removed 0 variants with bad BETA.\n",
      "Fri Nov  4 14:18:58 2022  -Checking if  0 <SE< inf  ...\n",
      "Fri Nov  4 14:18:58 2022  -Removed 0 variants with bad SE.\n",
      "Fri Nov  4 14:18:58 2022  -Checking STATUS...\n",
      "Fri Nov  4 14:18:59 2022  -Coverting STAUTUS to category.\n",
      "Fri Nov  4 14:18:59 2022  -Removed 0 variants with bad statistics in total.\n",
      "Fri Nov  4 14:18:59 2022 Finished sanity check successfully!\n",
      "Fri Nov  4 14:18:59 2022 Start to normalize variants...\n",
      "Fri Nov  4 14:18:59 2022  -Current Dataframe shape : 100000  x  12\n",
      "Fri Nov  4 14:18:59 2022  -No available variants to normalize..\n",
      "Fri Nov  4 14:18:59 2022 Finished normalizing variants successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.basic_check()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Nov  4 14:18:59 2022 Start to infer genome build version using hapmap3 SNPs...\n",
      "Fri Nov  4 14:18:59 2022  -Loading Hapmap3 variants data...\n",
      "Fri Nov  4 14:19:01 2022  -CHR:POS will be used for matching...\n",
      "Fri Nov  4 14:19:03 2022  -Matching variants for hg19: num_hg19 =  8797\n",
      "Fri Nov  4 14:19:03 2022  -Matching variants for hg38: num_hg38 =  115\n",
      "Fri Nov  4 14:19:03 2022  -Warning: please be cautious due to the limited number of variants.\n",
      "Fri Nov  4 14:19:03 2022  -Since num_hg19 >> num_hg38, assigning genome build hg19...\n"
     ]
    }
   ],
   "source": [
    "mysumstats.infer_build()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Harmonization worflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ref_seq : reference genome fasta file for allele alignment\n",
    "#ref_rsid_tsv : tsv file for annotation of common used variants\n",
    "#ref_rsid_vcf : vcf file for annotation of other variants\n",
    "#ref_infer : vcf file with allele frequency information for inferring strand and comparing allele frequency \n",
    "#ref_alt_freq : field in INFO of vcf file for alternative allele frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Nov  4 14:19:22 2022 Start to check available reference files...\n",
      "Fri Nov  4 14:19:22 2022  - 1kg_eas_hg19  :  https://www.dropbox.com/s/lztaxqhy2o6dpxw/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz?dl=1\n",
      "Fri Nov  4 14:19:22 2022  - 1kg_eas_hg19_tbi  :  https://www.dropbox.com/s/k9klefl8m9fcfo1/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz.tbi?dl=1\n",
      "Fri Nov  4 14:19:22 2022  - 1kg_eur_hg19  :  https://www.dropbox.com/s/1nbgqshknevseks/EUR.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz?dl=1\n",
      "Fri Nov  4 14:19:22 2022  - 1kg_eur_hg19_tbi  :  https://www.dropbox.com/s/vscvkrflh6fc5a0/EUR.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz.tbi?dl=1\n",
      "Fri Nov  4 14:19:22 2022  - 1kg_eas_hg38  :  https://www.dropbox.com/s/3dstbbb1el9r3au/EAS.ALL.split_norm_af.1kg_30x.hg38.vcf.gz?dl=1\n",
      "Fri Nov  4 14:19:22 2022  - 1kg_eas_hg38_tbi  :  https://www.dropbox.com/s/vwnp5vd8dcqksn4/EAS.ALL.split_norm_af.1kg_30x.hg38.vcf.gz.tbi?dl=1\n",
      "Fri Nov  4 14:19:22 2022  - 1kg_eur_hg38  :  https://www.dropbox.com/s/z0mkehg17lryapv/EUR.ALL.split_norm_af.1kg_30x.hg38.vcf.gz?dl=1\n",
      "Fri Nov  4 14:19:22 2022  - 1kg_eur_hg38_tbi  :  https://www.dropbox.com/s/ze8g58x75x9qbf0/EUR.ALL.split_norm_af.1kg_30x.hg38.vcf.gz.tbi?dl=1\n",
      "Fri Nov  4 14:19:22 2022  - dbsnp_v151_hg19  :  https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh37p13/VCF/00-All.vcf.gz\n",
      "Fri Nov  4 14:19:22 2022  - dbsnp_v151_hg38  :  https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF/00-All.vcf.gz\n",
      "Fri Nov  4 14:19:22 2022  - ucsc_genome_hg19  :  http://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/hg19.fa.gz\n",
      "Fri Nov  4 14:19:22 2022  - ucsc_genome_hg38  :  https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz\n",
      "Fri Nov  4 14:19:22 2022  - 1kg_dbsnp151_hg19_auto  :  https://www.dropbox.com/s/37p2u1xwmux4gwo/1kg_dbsnp151_hg19_auto.txt.gz?dl=1\n",
      "Fri Nov  4 14:19:22 2022  - 1kg_dbsnp151_hg38_auto  :  https://www.dropbox.com/s/ouf60n7gdz6cm0g/1kg_dbsnp151_hg38_auto.txt.gz?dl=1\n",
      "Fri Nov  4 14:19:22 2022  - testlink  :  https://www.dropbox.com/s/8u7capwge0ihshu/EAS.chr22.split_norm_af.1kgp3v5.vcf.gz?dl=1\n",
      "Fri Nov  4 14:19:22 2022  - testlink_tbi  :  https://www.dropbox.com/s/hdneg53t6u1j6ib/EAS.chr22.split_norm_af.1kgp3v5.vcf.gz.tbi?dl=1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'1kg_eas_hg19': 'https://www.dropbox.com/s/lztaxqhy2o6dpxw/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz?dl=1',\n",
       " '1kg_eas_hg19_tbi': 'https://www.dropbox.com/s/k9klefl8m9fcfo1/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz.tbi?dl=1',\n",
       " '1kg_eur_hg19': 'https://www.dropbox.com/s/1nbgqshknevseks/EUR.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz?dl=1',\n",
       " '1kg_eur_hg19_tbi': 'https://www.dropbox.com/s/vscvkrflh6fc5a0/EUR.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz.tbi?dl=1',\n",
       " '1kg_eas_hg38': 'https://www.dropbox.com/s/3dstbbb1el9r3au/EAS.ALL.split_norm_af.1kg_30x.hg38.vcf.gz?dl=1',\n",
       " '1kg_eas_hg38_tbi': 'https://www.dropbox.com/s/vwnp5vd8dcqksn4/EAS.ALL.split_norm_af.1kg_30x.hg38.vcf.gz.tbi?dl=1',\n",
       " '1kg_eur_hg38': 'https://www.dropbox.com/s/z0mkehg17lryapv/EUR.ALL.split_norm_af.1kg_30x.hg38.vcf.gz?dl=1',\n",
       " '1kg_eur_hg38_tbi': 'https://www.dropbox.com/s/ze8g58x75x9qbf0/EUR.ALL.split_norm_af.1kg_30x.hg38.vcf.gz.tbi?dl=1',\n",
       " 'dbsnp_v151_hg19': 'https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh37p13/VCF/00-All.vcf.gz',\n",
       " 'dbsnp_v151_hg38': 'https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF/00-All.vcf.gz',\n",
       " 'ucsc_genome_hg19': 'http://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/hg19.fa.gz',\n",
       " 'ucsc_genome_hg38': 'https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz',\n",
       " '1kg_dbsnp151_hg19_auto': 'https://www.dropbox.com/s/37p2u1xwmux4gwo/1kg_dbsnp151_hg19_auto.txt.gz?dl=1',\n",
       " '1kg_dbsnp151_hg38_auto': 'https://www.dropbox.com/s/ouf60n7gdz6cm0g/1kg_dbsnp151_hg38_auto.txt.gz?dl=1',\n",
       " 'testlink': 'https://www.dropbox.com/s/8u7capwge0ihshu/EAS.chr22.split_norm_af.1kgp3v5.vcf.gz?dl=1',\n",
       " 'testlink_tbi': 'https://www.dropbox.com/s/hdneg53t6u1j6ib/EAS.chr22.split_norm_af.1kgp3v5.vcf.gz.tbi?dl=1'}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gl.check_available_ref()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Nov  4 14:19:54 2022 Start to download  ucsc_genome_hg19  ...\n",
      "Fri Nov  4 14:19:54 2022  -Downloading to: /Users/he/work/gwaslab/src/gwaslab/data/hg19.fa.gz\n",
      "Fri Nov  4 14:24:17 2022  -Updating record in config file...\n",
      "Fri Nov  4 14:24:17 2022  -gunzip : /Users/he/work/gwaslab/src/gwaslab/data/hg19.fa.gz\n",
      "Fri Nov  4 14:24:42 2022  -Updating record in config file...\n",
      "Fri Nov  4 14:24:42 2022 Downloaded  ucsc_genome_hg19  successfully!\n",
      "Fri Nov  4 14:24:42 2022 Start to download  1kg_dbsnp151_hg19_auto  ...\n",
      "Fri Nov  4 14:24:42 2022  -Downloading to: /Users/he/work/gwaslab/src/gwaslab/data/1kg_dbsnp151_hg19_auto.txt.gz\n",
      "Fri Nov  4 14:25:41 2022  -Updating record in config file...\n",
      "Fri Nov  4 14:25:41 2022 Downloaded  1kg_dbsnp151_hg19_auto  successfully!\n"
     ]
    }
   ],
   "source": [
    "gl.download_ref(\"ucsc_genome_hg19\")\n",
    "gl.download_ref(\"1kg_dbsnp151_hg19_auto\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "get_path() can retrieve the path "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/he/work/gwaslab/src/gwaslab/data/1kg_dbsnp151_hg19_auto.txt.gz'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gl.get_path(\"1kg_dbsnp151_hg19_auto\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "you can simply the file downloaded using gwaslab like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Nov  4 14:27:45 2022 Start to check if NEA is aligned with reference sequence...\n",
      "Fri Nov  4 14:27:45 2022  -Current Dataframe shape : 100000  x  12\n",
      "Fri Nov  4 14:27:45 2022  -Reference genome fasta file: /Users/he/work/gwaslab/src/gwaslab/data/hg19.fa\n",
      "Fri Nov  4 14:27:45 2022  -Checking records: 1  2  3  4  5  6  7  X  8  9  10  11  12  13  14  15  16  17  18  20  Y  19  22  21  M  \n",
      "Fri Nov  4 14:28:37 2022  -Variants allele on given reference sequence :  91744\n",
      "Fri Nov  4 14:28:37 2022  -Variants flipped :  0\n",
      "Fri Nov  4 14:28:37 2022   -Raw Matching rate :  91.74%\n",
      "Fri Nov  4 14:28:37 2022  -Variants inferred reverse_complement :  0\n",
      "Fri Nov  4 14:28:37 2022  -Variants inferred reverse_complement_flipped :  0\n",
      "Fri Nov  4 14:28:37 2022  -Both allele on genome + unable to distinguish :  8256\n",
      "Fri Nov  4 14:28:37 2022  -Variants not on given reference sequence :  0\n",
      "Fri Nov  4 14:28:38 2022  -Current Dataframe shape : 100000  x  12\n",
      "Fri Nov  4 14:28:41 2022 Finished converting successfully!\n",
      "Fri Nov  4 14:28:42 2022 Start to annotate rsID based on chromosome and position information...\n",
      "Fri Nov  4 14:28:42 2022  -Current Dataframe shape : 100000  x  12\n",
      "Fri Nov  4 14:28:42 2022  -SNPID-rsID text file: /Users/he/work/gwaslab/src/gwaslab/data/1kg_dbsnp151_hg19_auto.txt.gz\n",
      "Fri Nov  4 14:28:42 2022  -100000 rsID could be possibly fixed...\n",
      "Fri Nov  4 14:28:42 2022  -Setting block size:  5000000\n",
      "Fri Nov  4 14:28:42 2022  -Loading block: 0   1   2   3   4   5   6   7   8   9   10   11   12   13   14   15   \n",
      "Fri Nov  4 14:35:53 2022  -rsID Annotation for 3086 need to be fixed!\n",
      "Fri Nov  4 14:35:53 2022  -Annotated 96914 rsID successfully!\n",
      "Fri Nov  4 14:35:53 2022 Start to sort the genome coordinates...\n",
      "Fri Nov  4 14:35:53 2022  -Current Dataframe shape : 100000  x  13\n",
      "Fri Nov  4 14:35:53 2022  -Force converting POS to integers...\n",
      "Fri Nov  4 14:35:53 2022  -Sorting genome coordinates...\n",
      "Fri Nov  4 14:35:53 2022 Finished sorting genome coordinates successfully!\n",
      "Fri Nov  4 14:35:53 2022 Start to reorder the columns...\n",
      "Fri Nov  4 14:35:53 2022  -Current Dataframe shape : 100000  x  13\n",
      "Fri Nov  4 14:35:53 2022  -Reordering columns to    : SNPID,rsID,CHR,POS,EA,NEA,EAF,BETA,SE,P,N,DIRECTION,STATUS\n",
      "Fri Nov  4 14:35:53 2022 Finished sorting columns successfully!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<gwaslab.Sumstats.Sumstats at 0x7fe85d020130>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.harmonize(   basic_check=False,\n",
    "                        n_cores=3,\n",
    "                        ref_seq = gl.get_path(\"ucsc_genome_hg19\"),\n",
    "                        ref_rsid_tsv = gl.get_path(\"1kg_dbsnp151_hg19_auto\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>rsID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>DIRECTION</th>\n",
       "      <th>STATUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:754182:A:G</td>\n",
       "      <td>rs3131969</td>\n",
       "      <td>1</td>\n",
       "      <td>754182</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7495</td>\n",
       "      <td>0.0222</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>0.1817</td>\n",
       "      <td>166718</td>\n",
       "      <td>-?++</td>\n",
       "      <td>1960000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:766007:A:C</td>\n",
       "      <td>rs61768174</td>\n",
       "      <td>1</td>\n",
       "      <td>766007</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0766</td>\n",
       "      <td>-0.0228</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0.4506</td>\n",
       "      <td>166718</td>\n",
       "      <td>+?--</td>\n",
       "      <td>1960000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:867151:G:A</td>\n",
       "      <td>rs79837947</td>\n",
       "      <td>1</td>\n",
       "      <td>867151</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0.1293</td>\n",
       "      <td>0.0672</td>\n",
       "      <td>0.0544</td>\n",
       "      <td>166718</td>\n",
       "      <td>+?++</td>\n",
       "      <td>1960000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:920648:T:C</td>\n",
       "      <td>rs6677020</td>\n",
       "      <td>1</td>\n",
       "      <td>920648</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.9881</td>\n",
       "      <td>-0.0104</td>\n",
       "      <td>0.0466</td>\n",
       "      <td>0.8236</td>\n",
       "      <td>191764</td>\n",
       "      <td>+--+</td>\n",
       "      <td>1960000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:923977:A:G</td>\n",
       "      <td>rs202005511</td>\n",
       "      <td>1</td>\n",
       "      <td>923977</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0.0355</td>\n",
       "      <td>0.0864</td>\n",
       "      <td>0.6809</td>\n",
       "      <td>191764</td>\n",
       "      <td>-++-</td>\n",
       "      <td>1960000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99995</th>\n",
       "      <td>X:154412526:C:T</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>23</td>\n",
       "      <td>154412526</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.1941</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>0.0089</td>\n",
       "      <td>0.3827</td>\n",
       "      <td>191764</td>\n",
       "      <td>+-+-</td>\n",
       "      <td>1960000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99996</th>\n",
       "      <td>X:154560189:G:T</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>23</td>\n",
       "      <td>154560189</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0061</td>\n",
       "      <td>0.0142</td>\n",
       "      <td>0.0858</td>\n",
       "      <td>0.8685</td>\n",
       "      <td>166718</td>\n",
       "      <td>+?+-</td>\n",
       "      <td>1960000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99997</th>\n",
       "      <td>X:154640027:T:A</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>23</td>\n",
       "      <td>154640027</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.7841</td>\n",
       "      <td>-0.0082</td>\n",
       "      <td>0.0084</td>\n",
       "      <td>0.3264</td>\n",
       "      <td>191764</td>\n",
       "      <td>-+-+</td>\n",
       "      <td>1960009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99998</th>\n",
       "      <td>X:154671432:T:A</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>23</td>\n",
       "      <td>154671432</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.7850</td>\n",
       "      <td>-0.0081</td>\n",
       "      <td>0.0084</td>\n",
       "      <td>0.3330</td>\n",
       "      <td>191764</td>\n",
       "      <td>-+-+</td>\n",
       "      <td>1960009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99999</th>\n",
       "      <td>X:154858649:C:T</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>23</td>\n",
       "      <td>154858649</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.7510</td>\n",
       "      <td>-0.0069</td>\n",
       "      <td>0.0113</td>\n",
       "      <td>0.5432</td>\n",
       "      <td>191764</td>\n",
       "      <td>-+-+</td>\n",
       "      <td>1960000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100000 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 SNPID         rsID  CHR        POS EA NEA     EAF    BETA  \\\n",
       "0         1:754182:A:G    rs3131969    1     754182  G   A  0.7495  0.0222   \n",
       "1         1:766007:A:C   rs61768174    1     766007  C   A  0.0766 -0.0228   \n",
       "2         1:867151:G:A   rs79837947    1     867151  A   G  0.0151  0.1293   \n",
       "3         1:920648:T:C    rs6677020    1     920648  C   T  0.9881 -0.0104   \n",
       "4         1:923977:A:G  rs202005511    1     923977  G   A  0.0058  0.0355   \n",
       "...                ...          ...  ...        ... ..  ..     ...     ...   \n",
       "99995  X:154412526:C:T         <NA>   23  154412526  T   C  0.1941  0.0078   \n",
       "99996  X:154560189:G:T         <NA>   23  154560189  T   G  0.0061  0.0142   \n",
       "99997  X:154640027:T:A         <NA>   23  154640027  A   T  0.7841 -0.0082   \n",
       "99998  X:154671432:T:A         <NA>   23  154671432  A   T  0.7850 -0.0081   \n",
       "99999  X:154858649:C:T         <NA>   23  154858649  T   C  0.7510 -0.0069   \n",
       "\n",
       "           SE       P       N DIRECTION   STATUS  \n",
       "0      0.0166  0.1817  166718      -?++  1960000  \n",
       "1      0.0302  0.4506  166718      +?--  1960000  \n",
       "2      0.0672  0.0544  166718      +?++  1960000  \n",
       "3      0.0466  0.8236  191764      +--+  1960000  \n",
       "4      0.0864  0.6809  191764      -++-  1960000  \n",
       "...       ...     ...     ...       ...      ...  \n",
       "99995  0.0089  0.3827  191764      +-+-  1960000  \n",
       "99996  0.0858  0.8685  166718      +?+-  1960000  \n",
       "99997  0.0084  0.3264  191764      -+-+  1960009  \n",
       "99998  0.0084  0.3330  191764      -+-+  1960009  \n",
       "99999  0.0113  0.5432  191764      -+-+  1960000  \n",
       "\n",
       "[100000 rows x 13 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "or you can use your own reference file "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sun Oct 30 11:58:59 2022 Start to check if NEA is aligned with reference sequence...\n",
      "Sun Oct 30 11:58:59 2022  -Current Dataframe shape : 12557761  x  12\n",
      "Sun Oct 30 11:58:59 2022  -Reference genome fasta file: /home/yunye/mydata/d_disk/human_g1k_v37.fasta\n",
      "Sun Oct 30 11:58:59 2022  -Checking records: 1  2  3  4  5  6  7  8  9  10  11  12  13  14  15  16  17  18  19  20  21  22  X  Y  MT  \n",
      "Sun Oct 30 12:07:47 2022  -Variants allele on given reference sequence :  5201808\n",
      "Sun Oct 30 12:07:47 2022  -Variants flipped :  6312911\n",
      "Sun Oct 30 12:07:47 2022   -Raw Matching rate :  91.69%\n",
      "Sun Oct 30 12:07:47 2022  -Variants inferred reverse_complement :  0\n",
      "Sun Oct 30 12:07:47 2022  -Variants inferred reverse_complement_flipped :  0\n",
      "Sun Oct 30 12:07:47 2022  -Both allele on genome + unable to distinguish :  1043042\n",
      "Sun Oct 30 12:07:47 2022  -Variants not on given reference sequence :  0\n",
      "Sun Oct 30 12:07:47 2022  -Current Dataframe shape : 12557761  x  12\n",
      "Sun Oct 30 12:08:43 2022 Start to flip allele-specific stats for SNPs with status xxxxx[35]x: alt->ea , ref->nea ... \n",
      "Sun Oct 30 12:09:08 2022  -Flipping 6312911 variants...\n",
      "Sun Oct 30 12:09:08 2022  -Swapping column: NEA <=> EA...\n",
      "Sun Oct 30 12:09:11 2022  -Flipping column: BETA = - BETA...\n",
      "Sun Oct 30 12:09:11 2022  -Flipping column: EAF = 1 - EAF...\n",
      "Sun Oct 30 12:09:11 2022  -Flipping column: DIRECTION +-? <=> -+? ...\n",
      "Sun Oct 30 12:09:15 2022  -Changed the status for flipped variants : xxxxx[35]x -> xxxxx[12]x\n",
      "Sun Oct 30 12:10:19 2022 Finished converting successfully!\n",
      "Sun Oct 30 12:10:19 2022 Start to infer strand for palindromic SNPs...\n",
      "Sun Oct 30 12:10:19 2022  -Current Dataframe shape : 12557761  x  12\n",
      "Sun Oct 30 12:10:19 2022  -Reference vcf file: /home/yunye/mydata/d_disk/eas_1kg_af/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz\n",
      "Sun Oct 30 12:10:19 2022  -Alternative allele frequency in INFO: AF\n",
      "Sun Oct 30 12:10:59 2022  -Identified  1780081  palindromic SNPs...\n",
      "Sun Oct 30 12:11:24 2022  -After filtering by MAF<  0.4  , the strand of  1627569  palindromic SNPs will be inferred...\n",
      "Sun Oct 30 12:27:22 2022   -Non-palindromic :  9641135\n",
      "Sun Oct 30 12:27:46 2022   -Palindromic SNPs on + strand:  1582262\n",
      "Sun Oct 30 12:28:10 2022   -Palindromic SNPs on - strand and need to be flipped: 1951\n",
      "Sun Oct 30 12:28:34 2022   -Palindromic SNPs with maf not availble to infer :  152512\n",
      "Sun Oct 30 12:28:58 2022   -Palindromic SNPs with no macthes or no information :  21741\n",
      "Sun Oct 30 12:29:26 2022  -Identified  1043042  indistinguishable Indels...\n",
      "Sun Oct 30 12:29:51 2022  -Indistinguishable Indels will be inferred from reference vcf ref and alt...\n",
      "Sun Oct 30 12:40:04 2022   -Indels ea/nea match reference :  454542\n",
      "Sun Oct 30 12:40:28 2022   -Indels ea/nea need to be flipped :  553573\n",
      "Sun Oct 30 12:40:52 2022   -Indels with no macthes or no information :  34927\n",
      "Sun Oct 30 12:40:52 2022  -Current Dataframe shape : 12557761  x  12\n",
      "Sun Oct 30 12:42:16 2022 Start to flip allele-specific stats for standardized indels with status xxxx[123][67][6]: alt->ea , ref->nea ... \n",
      "Sun Oct 30 12:42:40 2022  -Flipping 553573 variants...\n",
      "Sun Oct 30 12:42:40 2022  -Swapping column: NEA <=> EA...\n",
      "Sun Oct 30 12:42:41 2022  -Flipping column: BETA = - BETA...\n",
      "Sun Oct 30 12:42:41 2022  -Flipping column: EAF = 1 - EAF...\n",
      "Sun Oct 30 12:42:41 2022  -Flipping column: DIRECTION +-? <=> -+? ...\n",
      "Sun Oct 30 12:42:41 2022  -Changed the status for flipped variants xxxx[123][67]6 -> xxxx[123][67]4\n",
      "Sun Oct 30 12:43:11 2022 Start to flip allele-specific stats for palindromic SNPs with status xxxxx[12]5: (-)strand <=> (+)strand ... \n",
      "Sun Oct 30 12:43:36 2022  -Flipping 1951 variants...\n",
      "Sun Oct 30 12:43:36 2022  -Flipping column: BETA = - BETA...\n",
      "Sun Oct 30 12:43:36 2022  -Flipping column: EAF = 1 - EAF...\n",
      "Sun Oct 30 12:43:36 2022  -Flipping column: DIRECTION +-? <=> -+? ...\n",
      "Sun Oct 30 12:43:36 2022  -Changed the status for flipped variants:  xxxxx[012]5: ->  xxxxx[012]2\n",
      "Sun Oct 30 12:43:36 2022 Finished converting successfully!\n",
      "Sun Oct 30 12:43:36 2022 Start to annotate rsID based on chromosome and position information...\n",
      "Sun Oct 30 12:43:36 2022  -Current Dataframe shape : 12557761  x  12\n",
      "Sun Oct 30 12:43:36 2022  -SNPID-rsID text file: /home/yunye/mydata/d_disk/1kg_norm_unique.dbsnp151.auto.tsv.gz\n",
      "Sun Oct 30 12:43:42 2022  -12557761 rsID could be possibly fixed...\n",
      "Sun Oct 30 12:43:44 2022  -Setting block size:  5000000\n",
      "Sun Oct 30 12:43:44 2022  -Loading block: 0   1   2   3   4   5   6   7   8   9   10   11   12   13   14   15   \n",
      "Sun Oct 30 12:51:08 2022  -rsID Annotation for 406539 need to be fixed!\n",
      "Sun Oct 30 12:51:08 2022  -Annotated 12151222 rsID successfully!\n",
      "Sun Oct 30 12:51:08 2022 Start to assign rsID using vcf...\n",
      "Sun Oct 30 12:51:08 2022  -Current Dataframe shape : 12557761  x  13\n",
      "Sun Oct 30 12:51:08 2022  -CPU Cores to use : 3\n",
      "Sun Oct 30 12:51:08 2022  -Reference VCF file: /home/yunye/mydata/d_disk/dbsnp/00-All.vcf.gz\n",
      "Sun Oct 30 12:51:08 2022  -Assigning rsID based on chr:pos and ref:alt/alt:ref...\n",
      "Sun Oct 30 12:56:16 2022  -rsID Annotation for 75796 need to be fixed!\n",
      "Sun Oct 30 12:56:16 2022  -Annotated 330743 rsID successfully!\n",
      "Sun Oct 30 12:56:16 2022 Start to sort the genome coordinates...\n",
      "Sun Oct 30 12:56:16 2022  -Current Dataframe shape : 12557761  x  13\n",
      "Sun Oct 30 12:56:16 2022  -Force converting POS to integers...\n",
      "Sun Oct 30 12:56:17 2022  -Sorting genome coordinates...\n",
      "Sun Oct 30 12:56:32 2022 Finished sorting genome coordinates successfully!\n",
      "Sun Oct 30 12:56:32 2022 Start to reorder the columns...\n",
      "Sun Oct 30 12:56:32 2022  -Current Dataframe shape : 12557761  x  13\n",
      "Sun Oct 30 12:56:32 2022  -Reordering columns to    : SNPID,rsID,CHR,POS,EA,NEA,EAF,BETA,SE,P,N,DIRECTION,STATUS\n",
      "Sun Oct 30 12:56:33 2022 Finished sorting columns successfully!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<gwaslab.Sumstats.Sumstats at 0x7fcde4820a30>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.harmonize(   basic_check=False,\n",
    "                        n_cores=3,\n",
    "                        ref_seq=\"/home/yunye/mydata/d_disk/human_g1k_v37.fasta\",\n",
    "                        ref_rsid_tsv=\"/home/yunye/mydata/d_disk/1kg_norm_unique.dbsnp151.auto.tsv.gz\",\n",
    "                        ref_rsid_vcf=\"/home/yunye/mydata/d_disk/dbsnp/00-All.vcf.gz\",\n",
    "                        ref_infer=\"/home/yunye/mydata/d_disk/eas_1kg_af/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz\",ref_alt_freq=\"AF\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>rsID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>DIRECTION</th>\n",
       "      <th>STATUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:725932:G:A</td>\n",
       "      <td>rs565766235</td>\n",
       "      <td>1</td>\n",
       "      <td>725932</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.5970</td>\n",
       "      <td>166718</td>\n",
       "      <td>+?-+</td>\n",
       "      <td>1960010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:725933:A:G</td>\n",
       "      <td>rs534711480</td>\n",
       "      <td>1</td>\n",
       "      <td>725933</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.5973</td>\n",
       "      <td>166718</td>\n",
       "      <td>+?-+</td>\n",
       "      <td>1960000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:737801:T:C</td>\n",
       "      <td>rs540210562</td>\n",
       "      <td>1</td>\n",
       "      <td>737801</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.0490</td>\n",
       "      <td>0.1231</td>\n",
       "      <td>0.6908</td>\n",
       "      <td>166718</td>\n",
       "      <td>+?-+</td>\n",
       "      <td>1960000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:749963:T:TAA</td>\n",
       "      <td>rs529266287</td>\n",
       "      <td>1</td>\n",
       "      <td>749963</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8374</td>\n",
       "      <td>0.0213</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.2846</td>\n",
       "      <td>166718</td>\n",
       "      <td>-?++</td>\n",
       "      <td>1960363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:751343:T:A</td>\n",
       "      <td>rs28544273</td>\n",
       "      <td>1</td>\n",
       "      <td>751343</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1407</td>\n",
       "      <td>-0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.2705</td>\n",
       "      <td>166718</td>\n",
       "      <td>+?--</td>\n",
       "      <td>1960011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12557756</th>\n",
       "      <td>X:154874837:A:G</td>\n",
       "      <td>rs642362</td>\n",
       "      <td>23</td>\n",
       "      <td>154874837</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7478</td>\n",
       "      <td>-0.0064</td>\n",
       "      <td>0.0117</td>\n",
       "      <td>0.5840</td>\n",
       "      <td>191764</td>\n",
       "      <td>-+-+</td>\n",
       "      <td>1960000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12557757</th>\n",
       "      <td>X:154875192:GTACTC:G</td>\n",
       "      <td>rs4013152</td>\n",
       "      <td>23</td>\n",
       "      <td>154875192</td>\n",
       "      <td>GTACTC</td>\n",
       "      <td>G</td>\n",
       "      <td>0.2525</td>\n",
       "      <td>0.0071</td>\n",
       "      <td>0.0122</td>\n",
       "      <td>0.5612</td>\n",
       "      <td>191764</td>\n",
       "      <td>+-+-</td>\n",
       "      <td>1960368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12557758</th>\n",
       "      <td>X:154879115:A:G</td>\n",
       "      <td>rs9779788</td>\n",
       "      <td>23</td>\n",
       "      <td>154879115</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7463</td>\n",
       "      <td>-0.0070</td>\n",
       "      <td>0.0122</td>\n",
       "      <td>0.5646</td>\n",
       "      <td>191764</td>\n",
       "      <td>-+-+</td>\n",
       "      <td>1960000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12557759</th>\n",
       "      <td>X:154880669:T:A</td>\n",
       "      <td>rs491661</td>\n",
       "      <td>23</td>\n",
       "      <td>154880669</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.7442</td>\n",
       "      <td>-0.0071</td>\n",
       "      <td>0.0122</td>\n",
       "      <td>0.5618</td>\n",
       "      <td>191764</td>\n",
       "      <td>-+-+</td>\n",
       "      <td>1960018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12557760</th>\n",
       "      <td>X:154880917:C:T</td>\n",
       "      <td>rs494282</td>\n",
       "      <td>23</td>\n",
       "      <td>154880917</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.7442</td>\n",
       "      <td>-0.0072</td>\n",
       "      <td>0.0122</td>\n",
       "      <td>0.5570</td>\n",
       "      <td>191764</td>\n",
       "      <td>-+-+</td>\n",
       "      <td>1960010</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12557761 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         SNPID         rsID  CHR        POS      EA NEA  \\\n",
       "0                 1:725932:G:A  rs565766235    1     725932       A   G   \n",
       "1                 1:725933:A:G  rs534711480    1     725933       G   A   \n",
       "2                 1:737801:T:C  rs540210562    1     737801       C   T   \n",
       "3               1:749963:T:TAA  rs529266287    1     749963     TAA   T   \n",
       "4                 1:751343:T:A   rs28544273    1     751343       A   T   \n",
       "...                        ...          ...  ...        ...     ...  ..   \n",
       "12557756       X:154874837:A:G     rs642362   23  154874837       G   A   \n",
       "12557757  X:154875192:GTACTC:G    rs4013152   23  154875192  GTACTC   G   \n",
       "12557758       X:154879115:A:G    rs9779788   23  154879115       G   A   \n",
       "12557759       X:154880669:T:A     rs491661   23  154880669       A   T   \n",
       "12557760       X:154880917:C:T     rs494282   23  154880917       T   C   \n",
       "\n",
       "             EAF    BETA      SE       P       N DIRECTION   STATUS  \n",
       "0         0.0040  0.0737  0.1394  0.5970  166718      +?-+  1960010  \n",
       "1         0.0040  0.0737  0.1394  0.5973  166718      +?-+  1960000  \n",
       "2         0.0051  0.0490  0.1231  0.6908  166718      +?-+  1960000  \n",
       "3         0.8374  0.0213  0.0199  0.2846  166718      -?++  1960363  \n",
       "4         0.1407 -0.0172  0.0156  0.2705  166718      +?--  1960011  \n",
       "...          ...     ...     ...     ...     ...       ...      ...  \n",
       "12557756  0.7478 -0.0064  0.0117  0.5840  191764      -+-+  1960000  \n",
       "12557757  0.2525  0.0071  0.0122  0.5612  191764      +-+-  1960368  \n",
       "12557758  0.7463 -0.0070  0.0122  0.5646  191764      -+-+  1960000  \n",
       "12557759  0.7442 -0.0071  0.0122  0.5618  191764      -+-+  1960018  \n",
       "12557760  0.7442 -0.0072  0.0122  0.5570  191764      -+-+  1960010  \n",
       "\n",
       "[12557761 rows x 13 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Values</th>\n",
       "      <th>Percentage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Category</th>\n",
       "      <th>Items</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">META</th>\n",
       "      <th>Row_num</th>\n",
       "      <td>12557761</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Column_num</th>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Column_names</th>\n",
       "      <td>SNPID,rsID,EAF,P,N,STATUS</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Last_checked_time</th>\n",
       "      <td>Sun Oct 30 12:56:35 2022</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">MISSING</th>\n",
       "      <th>Missing_total</th>\n",
       "      <td>75796</td>\n",
       "      <td>0.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Missing_rsID</th>\n",
       "      <td>75796</td>\n",
       "      <td>0.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">MAF</th>\n",
       "      <th>Common</th>\n",
       "      <td>6359357</td>\n",
       "      <td>50.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Low_frequency</th>\n",
       "      <td>2131089</td>\n",
       "      <td>16.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rare</th>\n",
       "      <td>4067315</td>\n",
       "      <td>32.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">P</th>\n",
       "      <th>Minimum</th>\n",
       "      <td>2.58e-168</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Significant</th>\n",
       "      <td>9461</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Suggestive</th>\n",
       "      <td>18641</td>\n",
       "      <td>0.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"15\" valign=\"top\">STATUS</th>\n",
       "      <th>1960010</th>\n",
       "      <td>5400022</td>\n",
       "      <td>43.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960000</th>\n",
       "      <td>4241113</td>\n",
       "      <td>33.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960001</th>\n",
       "      <td>791358</td>\n",
       "      <td>6.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960011</th>\n",
       "      <td>790904</td>\n",
       "      <td>6.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960364</th>\n",
       "      <td>553573</td>\n",
       "      <td>4.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960363</th>\n",
       "      <td>454542</td>\n",
       "      <td>3.62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960007</th>\n",
       "      <td>76292</td>\n",
       "      <td>0.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960017</th>\n",
       "      <td>76220</td>\n",
       "      <td>0.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960309</th>\n",
       "      <td>70443</td>\n",
       "      <td>0.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960368</th>\n",
       "      <td>34927</td>\n",
       "      <td>0.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960319</th>\n",
       "      <td>23060</td>\n",
       "      <td>0.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960018</th>\n",
       "      <td>21741</td>\n",
       "      <td>0.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960008</th>\n",
       "      <td>21615</td>\n",
       "      <td>0.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960002</th>\n",
       "      <td>987</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960012</th>\n",
       "      <td>964</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Values  Percentage\n",
       "Category Items                                                   \n",
       "META     Row_num                             12557761         NaN\n",
       "         Column_num                                 6         NaN\n",
       "         Column_names       SNPID,rsID,EAF,P,N,STATUS         NaN\n",
       "         Last_checked_time   Sun Oct 30 12:56:35 2022         NaN\n",
       "MISSING  Missing_total                          75796        0.60\n",
       "         Missing_rsID                           75796        0.60\n",
       "MAF      Common                               6359357       50.64\n",
       "         Low_frequency                        2131089       16.97\n",
       "         Rare                                 4067315       32.39\n",
       "P        Minimum                            2.58e-168        0.00\n",
       "         Significant                             9461        0.08\n",
       "         Suggestive                             18641        0.15\n",
       "STATUS   1960010                              5400022       43.00\n",
       "         1960000                              4241113       33.77\n",
       "         1960001                               791358        6.30\n",
       "         1960011                               790904        6.30\n",
       "         1960364                               553573        4.41\n",
       "         1960363                               454542        3.62\n",
       "         1960007                                76292        0.61\n",
       "         1960017                                76220        0.61\n",
       "         1960309                                70443        0.56\n",
       "         1960368                                34927        0.28\n",
       "         1960319                                23060        0.18\n",
       "         1960018                                21741        0.17\n",
       "         1960008                                21615        0.17\n",
       "         1960002                                  987        0.01\n",
       "         1960012                                  964        0.01"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome_Build</th>\n",
       "      <th>rsID&amp;SNPID</th>\n",
       "      <th>CHR&amp;POS</th>\n",
       "      <th>Stadardize&amp;Normalize</th>\n",
       "      <th>Align</th>\n",
       "      <th>Panlidromic_SNP&amp;Indel</th>\n",
       "      <th>Count</th>\n",
       "      <th>Percentage(%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1960000</th>\n",
       "      <td>hg19</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Match: NEA=REF</td>\n",
       "      <td>Not_palindromic_SNPs</td>\n",
       "      <td>4241113</td>\n",
       "      <td>33.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960001</th>\n",
       "      <td>hg19</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Match: NEA=REF</td>\n",
       "      <td>Palindromic+strand</td>\n",
       "      <td>791358</td>\n",
       "      <td>6.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960002</th>\n",
       "      <td>hg19</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Match: NEA=REF</td>\n",
       "      <td>Palindromic-strand_fixed</td>\n",
       "      <td>987</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960007</th>\n",
       "      <td>hg19</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Match: NEA=REF</td>\n",
       "      <td>Indistinguishable</td>\n",
       "      <td>76292</td>\n",
       "      <td>0.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960008</th>\n",
       "      <td>hg19</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Match: NEA=REF</td>\n",
       "      <td>No_matching_or_no_info</td>\n",
       "      <td>21615</td>\n",
       "      <td>0.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960010</th>\n",
       "      <td>hg19</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Flipped_fixed</td>\n",
       "      <td>Not_palindromic_SNPs</td>\n",
       "      <td>5400022</td>\n",
       "      <td>43.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960011</th>\n",
       "      <td>hg19</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Flipped_fixed</td>\n",
       "      <td>Palindromic+strand</td>\n",
       "      <td>790904</td>\n",
       "      <td>6.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960012</th>\n",
       "      <td>hg19</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Flipped_fixed</td>\n",
       "      <td>Palindromic-strand_fixed</td>\n",
       "      <td>964</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960017</th>\n",
       "      <td>hg19</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Flipped_fixed</td>\n",
       "      <td>Indistinguishable</td>\n",
       "      <td>76220</td>\n",
       "      <td>0.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960018</th>\n",
       "      <td>hg19</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Flipped_fixed</td>\n",
       "      <td>No_matching_or_no_info</td>\n",
       "      <td>21741</td>\n",
       "      <td>0.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960309</th>\n",
       "      <td>hg19</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized &amp; normalized indel</td>\n",
       "      <td>Match: NEA=REF</td>\n",
       "      <td>Unchecked</td>\n",
       "      <td>70443</td>\n",
       "      <td>0.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960319</th>\n",
       "      <td>hg19</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized &amp; normalized indel</td>\n",
       "      <td>Flipped_fixed</td>\n",
       "      <td>Unchecked</td>\n",
       "      <td>23060</td>\n",
       "      <td>0.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960363</th>\n",
       "      <td>hg19</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized &amp; normalized indel</td>\n",
       "      <td>Both_alleles_on_ref+indistinguishable</td>\n",
       "      <td>Indel_match</td>\n",
       "      <td>454542</td>\n",
       "      <td>3.62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960364</th>\n",
       "      <td>hg19</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized &amp; normalized indel</td>\n",
       "      <td>Both_alleles_on_ref+indistinguishable</td>\n",
       "      <td>Indel_flipped_fixed</td>\n",
       "      <td>553573</td>\n",
       "      <td>4.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960368</th>\n",
       "      <td>hg19</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized &amp; normalized indel</td>\n",
       "      <td>Both_alleles_on_ref+indistinguishable</td>\n",
       "      <td>No_matching_or_no_info</td>\n",
       "      <td>34927</td>\n",
       "      <td>0.28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Genome_Build                  rsID&SNPID                CHR&POS  \\\n",
       "1960000         hg19  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "1960001         hg19  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "1960002         hg19  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "1960007         hg19  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "1960008         hg19  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "1960010         hg19  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "1960011         hg19  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "1960012         hg19  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "1960017         hg19  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "1960018         hg19  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "1960309         hg19  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "1960319         hg19  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "1960363         hg19  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "1960364         hg19  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "1960368         hg19  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "\n",
       "                    Stadardize&Normalize  \\\n",
       "1960000                 standardized SNP   \n",
       "1960001                 standardized SNP   \n",
       "1960002                 standardized SNP   \n",
       "1960007                 standardized SNP   \n",
       "1960008                 standardized SNP   \n",
       "1960010                 standardized SNP   \n",
       "1960011                 standardized SNP   \n",
       "1960012                 standardized SNP   \n",
       "1960017                 standardized SNP   \n",
       "1960018                 standardized SNP   \n",
       "1960309  standardized & normalized indel   \n",
       "1960319  standardized & normalized indel   \n",
       "1960363  standardized & normalized indel   \n",
       "1960364  standardized & normalized indel   \n",
       "1960368  standardized & normalized indel   \n",
       "\n",
       "                                         Align     Panlidromic_SNP&Indel  \\\n",
       "1960000                         Match: NEA=REF      Not_palindromic_SNPs   \n",
       "1960001                         Match: NEA=REF        Palindromic+strand   \n",
       "1960002                         Match: NEA=REF  Palindromic-strand_fixed   \n",
       "1960007                         Match: NEA=REF         Indistinguishable   \n",
       "1960008                         Match: NEA=REF    No_matching_or_no_info   \n",
       "1960010                          Flipped_fixed      Not_palindromic_SNPs   \n",
       "1960011                          Flipped_fixed        Palindromic+strand   \n",
       "1960012                          Flipped_fixed  Palindromic-strand_fixed   \n",
       "1960017                          Flipped_fixed         Indistinguishable   \n",
       "1960018                          Flipped_fixed    No_matching_or_no_info   \n",
       "1960309                         Match: NEA=REF                 Unchecked   \n",
       "1960319                          Flipped_fixed                 Unchecked   \n",
       "1960363  Both_alleles_on_ref+indistinguishable               Indel_match   \n",
       "1960364  Both_alleles_on_ref+indistinguishable       Indel_flipped_fixed   \n",
       "1960368  Both_alleles_on_ref+indistinguishable    No_matching_or_no_info   \n",
       "\n",
       "           Count Percentage(%)  \n",
       "1960000  4241113         33.77  \n",
       "1960001   791358           6.3  \n",
       "1960002      987          0.01  \n",
       "1960007    76292          0.61  \n",
       "1960008    21615          0.17  \n",
       "1960010  5400022          43.0  \n",
       "1960011   790904           6.3  \n",
       "1960012      964          0.01  \n",
       "1960017    76220          0.61  \n",
       "1960018    21741          0.17  \n",
       "1960309    70443          0.56  \n",
       "1960319    23060          0.18  \n",
       "1960363   454542          3.62  \n",
       "1960364   553573          4.41  \n",
       "1960368    34927          0.28  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.lookup_status()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mysumstats.to_format(\"clean_sumstats\",fmt=\"ssf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## VCF : NCBI Sequence Identifiers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For some reference file the chromosome notation is not in the form of 1, or chr1, instead, they are using NCBI Sequence Identifiers (for example: NC_000001.11)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See https://www.ncbi.nlm.nih.gov/genbank/sequenceids/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For example: GCF_000001405.25.vcf.gz\n",
    "\n",
    "NC_000001.10    10059   rs1570391745    C       G       .       .       RS=1570391745;dbSNPBuildID=154;SSR=0;PSEUDOGENEINFO=DDX11L1:100287102;VC=SNV;R5;GNO;FREQ=KOREAN:0.9997,0.0003425|dbGaP_PopFreq:1,0\n",
    "\n",
    "NC_000001.10    10060   rs1639544146    C       CT      .       .       RS=1639544146;SSR=0;PSEUDOGENEINFO=DDX11L1:100287102;VC=INDEL;R5;GNO;FREQ=dbGaP_PopFreq:1,0\n",
    "\n",
    "NC_000001.10    10060   rs1639544159    CT      C       .       .       RS=1639544159;SSR=0;PSEUDOGENEINFO=DDX11L1:100287102;VC=DEL;R5;GNO;FREQ=dbGaP_PopFreq:1,0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " In such case, gwaslab provide built-in conversion table. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: 'NC_000001.10',\n",
       " 2: 'NC_000002.11',\n",
       " 3: 'NC_000003.11',\n",
       " 4: 'NC_000004.11',\n",
       " 5: 'NC_000005.9',\n",
       " 6: 'NC_000006.11',\n",
       " 7: 'NC_000007.13',\n",
       " 8: 'NC_000008.10',\n",
       " 9: 'NC_000009.11',\n",
       " 10: 'NC_000010.10',\n",
       " 11: 'NC_000011.9',\n",
       " 12: 'NC_000012.11',\n",
       " 13: 'NC_000013.10',\n",
       " 14: 'NC_000014.8',\n",
       " 15: 'NC_000015.9',\n",
       " 16: 'NC_000016.9',\n",
       " 17: 'NC_000017.10',\n",
       " 18: 'NC_000018.9',\n",
       " 19: 'NC_000019.9',\n",
       " 20: 'NC_000020.10',\n",
       " 21: 'NC_000021.8',\n",
       " 22: 'NC_000022.10',\n",
       " 23: 'NC_000023.10',\n",
       " 24: 'NC_000024.9',\n",
       " 25: 'NC_012920.1'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gl.get_number_to_NC(build=\"19\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "and pass this conversion table to corresponding functions by specifying `chr_dict` like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mysumstats.assign_rsid(ref_rsid_vcf=\"/home/yunye/mydata/d_disk/dbsnp/00-All.vcf.gz\", chr_dict=gl.get_number_to_NC(build=\"19\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "or specify it in assignrsid_args and inferstrand_args for the all-in-one function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Thu Nov  3 13:00:16 2022 Start to check if NEA is aligned with reference sequence...\n",
      "Thu Nov  3 13:00:16 2022  -Current Dataframe shape : 100000  x  12\n",
      "Thu Nov  3 13:00:16 2022  -Reference genome fasta file: /home/yunye/mydata/d_disk/human_g1k_v37.fasta\n",
      "Thu Nov  3 13:00:16 2022  -Checking records: 1  2  3  4  5  6  7  8  9  10  11  12  13  14  15  16  17  18  19  20  21  22  X  Y  MT  \n",
      "Thu Nov  3 13:00:51 2022  -Variants allele on given reference sequence :  41324\n",
      "Thu Nov  3 13:00:51 2022  -Variants flipped :  50404\n",
      "Thu Nov  3 13:00:51 2022   -Raw Matching rate :  91.73%\n",
      "Thu Nov  3 13:00:51 2022  -Variants inferred reverse_complement :  0\n",
      "Thu Nov  3 13:00:51 2022  -Variants inferred reverse_complement_flipped :  0\n",
      "Thu Nov  3 13:00:51 2022  -Both allele on genome + unable to distinguish :  8272\n",
      "Thu Nov  3 13:00:51 2022  -Variants not on given reference sequence :  0\n",
      "Thu Nov  3 13:00:51 2022  -Current Dataframe shape : 100000  x  12\n",
      "Thu Nov  3 13:00:52 2022 Start to flip allele-specific stats for SNPs with status xxxxx[35]x: alt->ea , ref->nea ... \n",
      "Thu Nov  3 13:00:52 2022  -Flipping 50404 variants...\n",
      "Thu Nov  3 13:00:52 2022  -Swapping column: NEA <=> EA...\n",
      "Thu Nov  3 13:00:52 2022  -Flipping column: BETA = - BETA...\n",
      "Thu Nov  3 13:00:52 2022  -Flipping column: EAF = 1 - EAF...\n",
      "Thu Nov  3 13:00:52 2022  -Flipping column: DIRECTION +-? <=> -+? ...\n",
      "Thu Nov  3 13:00:52 2022  -Changed the status for flipped variants : xxxxx[35]x -> xxxxx[12]x\n",
      "Thu Nov  3 13:00:52 2022 Finished converting successfully!\n",
      "Thu Nov  3 13:00:52 2022 Start to annotate rsID based on chromosome and position information...\n",
      "Thu Nov  3 13:00:52 2022  -Current Dataframe shape : 100000  x  12\n",
      "Thu Nov  3 13:00:52 2022  -SNPID-rsID text file: /home/yunye/mydata/d_disk/1kg_norm_unique.dbsnp151.auto.tsv.gz\n",
      "Thu Nov  3 13:00:52 2022  -100000 rsID could be possibly fixed...\n",
      "Thu Nov  3 13:00:53 2022  -Setting block size:  5000000\n",
      "Thu Nov  3 13:00:53 2022  -Loading block: 0   1   2   3   4   5   6   7   8   9   10   11   12   13   14   15   \n",
      "Thu Nov  3 13:04:56 2022  -rsID Annotation for 3282 need to be fixed!\n",
      "Thu Nov  3 13:04:56 2022  -Annotated 96718 rsID successfully!\n",
      "Thu Nov  3 13:04:57 2022 Start to assign rsID using vcf...\n",
      "Thu Nov  3 13:04:57 2022  -Current Dataframe shape : 100000  x  13\n",
      "Thu Nov  3 13:04:57 2022  -CPU Cores to use : 3\n",
      "Thu Nov  3 13:04:57 2022  -Reference VCF file: /home/yunye/mydata/d_disk/dbsnp/GCF_000001405.25.vcf.gz\n",
      "Thu Nov  3 13:04:57 2022  -Assigning rsID based on chr:pos and ref:alt/alt:ref...\n",
      "Thu Nov  3 13:05:04 2022  -rsID Annotation for 118 need to be fixed!\n",
      "Thu Nov  3 13:05:04 2022  -Annotated 3164 rsID successfully!\n",
      "Thu Nov  3 13:05:04 2022 Start to sort the genome coordinates...\n",
      "Thu Nov  3 13:05:04 2022  -Current Dataframe shape : 100000  x  13\n",
      "Thu Nov  3 13:05:04 2022  -Force converting POS to integers...\n",
      "Thu Nov  3 13:05:04 2022  -Sorting genome coordinates...\n",
      "Thu Nov  3 13:05:05 2022 Finished sorting genome coordinates successfully!\n",
      "Thu Nov  3 13:05:05 2022 Start to reorder the columns...\n",
      "Thu Nov  3 13:05:05 2022  -Current Dataframe shape : 100000  x  13\n",
      "Thu Nov  3 13:05:05 2022  -Reordering columns to    : SNPID,rsID,CHR,POS,EA,NEA,EAF,BETA,SE,P,N,DIRECTION,STATUS\n",
      "Thu Nov  3 13:05:05 2022 Finished sorting columns successfully!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<gwaslab.Sumstats.Sumstats at 0x7f8b68dac340>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.harmonize(   basic_check=False,\n",
    "                        n_cores=3,\n",
    "                        ref_seq=\"/home/yunye/mydata/d_disk/human_g1k_v37.fasta\",\n",
    "                        ref_rsid_tsv=\"/home/yunye/mydata/d_disk/1kg_norm_unique.dbsnp151.auto.tsv.gz\",\n",
    "                        ref_rsid_vcf=\"/home/yunye/mydata/d_disk/dbsnp/GCF_000001405.25.vcf.gz\",\n",
    "                        assignrsid_args= {\"chr_dict\" : gl.get_number_to_NC(build=\"19\")}\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
