{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1ce41337",
   "metadata": {},
   "source": [
    "# Standardization Workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0242fb7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0,\"/home/yunye/work/gwaslab/src/\")\n",
    "import gwaslab as gl\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9553c53",
   "metadata": {},
   "source": [
    "# Load sample data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "73f7b157",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wed Jan 24 12:31:50 2024 GWASLab v3.4.36 https://cloufield.github.io/gwaslab/\n",
      "Wed Jan 24 12:31:50 2024 (C) 2022-2024, Yunye He, Kamatani Lab, MIT License, gwaslab@gmail.com\n",
      "Wed Jan 24 12:31:50 2024 Start to load format from formatbook....\n",
      "Wed Jan 24 12:31:50 2024  -gwaslab format meta info:\n",
      "Wed Jan 24 12:31:50 2024   - format_name  : gwaslab\n",
      "Wed Jan 24 12:31:50 2024   - format_source  : https://cloufield.github.io/gwaslab/\n",
      "Wed Jan 24 12:31:50 2024   - format_version  : 20231220_v4\n",
      "Wed Jan 24 12:31:50 2024 Start to initiate from file :/home/yunye/work/gwaslab/examples/toy_data/dirty_sumstats.tsv\n",
      "Wed Jan 24 12:31:50 2024  -Reading columns          : CHR,NEA,N_CASE,DIRECTION,OR_95U,SE,POS,OR_95L,Z,BETA,NOTE,SNPID,EAF,MLOG10P,N,P,EA,OR,CHISQ,N_CONTROL\n",
      "Wed Jan 24 12:31:50 2024  -Renaming columns to      : CHR,NEA,N_CASE,DIRECTION,OR_95U,SE,POS,OR_95L,Z,BETA,NOTE,SNPID,EAF,MLOG10P,N,P,EA,OR,CHISQ,N_CONTROL\n",
      "Wed Jan 24 12:31:50 2024  -Current Dataframe shape : 63  x  20\n",
      "Wed Jan 24 12:31:50 2024  -Initiating a status column: STATUS ...\n",
      "Wed Jan 24 12:31:50 2024  -Version of genomic coordinates are unknown...\n",
      "Wed Jan 24 12:31:51 2024 Start to reorder the columns...v3.4.36\n",
      "Wed Jan 24 12:31:51 2024  -Current Dataframe shape : 63  x  21\n",
      "Wed Jan 24 12:31:51 2024  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,EAF,BETA,SE,Z,CHISQ,P,MLOG10P,OR,OR_95L,OR_95U,N,N_CASE,N_CONTROL,DIRECTION,STATUS,NOTE\n",
      "Wed Jan 24 12:31:51 2024 Finished sorting columns successfully!\n",
      "Wed Jan 24 12:31:51 2024  -Column  : SNPID  CHR    POS    EA       NEA      EAF     BETA    SE      Z       CHISQ   P       MLOG10P OR      OR_95L  OR_95U  N       N_CASE N_CONTROL DIRECTION STATUS   NOTE  \n",
      "Wed Jan 24 12:31:51 2024  -DType   : object string object category category float64 float64 float64 float64 float64 float64 float64 float64 float64 float64 float64 int64  int64     object    category object\n",
      "Wed Jan 24 12:31:51 2024  -Verified: T      F      F      T        T        T       T       T       T       T       T       T       T       T       T       F       T      T         T         T        NA    \n",
      "Wed Jan 24 12:31:51 2024  -WARNING! Columns with possibly incompatable dtypes: CHR,POS,N\n",
      "Wed Jan 24 12:31:51 2024 Finished loading data successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(\"/home/yunye/work/gwaslab/examples/toy_data/dirty_sumstats.tsv\",fmt=\"gwaslab\",eaf=\"EAF\",other=[\"NOTE\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "79d31788",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wed Jan 24 12:31:51 2024 Start to check IDs...v3.4.36\n",
      "Wed Jan 24 12:31:51 2024  -Current Dataframe shape : 63  x  21\n",
      "Wed Jan 24 12:31:51 2024  -Checking if SNPID is chr:pos:ref:alt...(separator: - ,: , _)\n",
      "Wed Jan 24 12:31:52 2024 Finished checking IDs successfully!\n",
      "Wed Jan 24 12:31:52 2024 Start to fix chromosome notation...v3.4.36\n",
      "Wed Jan 24 12:31:52 2024  -Current Dataframe shape : 63  x  21\n",
      "Wed Jan 24 12:31:52 2024  -Checking CHR data type...\n",
      "Wed Jan 24 12:31:52 2024  -Variants with standardized chromosome notation: 56\n",
      "Wed Jan 24 12:31:52 2024  -Variants with fixable chromosome notations: 4\n",
      "Wed Jan 24 12:31:52 2024  -Variants with NA chromosome notations: 1\n",
      "Wed Jan 24 12:31:52 2024  -Variants with invalid chromosome notations: 2\n",
      "Wed Jan 24 12:31:52 2024  -A look at invalid chromosome notations: {'1.0001', '-1'}\n",
      "Wed Jan 24 12:31:52 2024  -Identifying non-autosomal chromosomes : X, Y, and MT ...\n",
      "Wed Jan 24 12:31:52 2024  -Identified  1  variants on sex chromosomes...\n",
      "Wed Jan 24 12:31:52 2024  -Standardizing sex chromosome notations: X to 23...\n",
      "Wed Jan 24 12:31:54 2024  -Valid CHR list: 1 - 25\n",
      "Wed Jan 24 12:31:54 2024  -Removed 5 variants with chromosome notations not in CHR list.\n",
      "Wed Jan 24 12:31:54 2024  -A look at chromosome notations not in CHR list: {'300', '0', <NA>}\n",
      "Wed Jan 24 12:31:54 2024 Finished fixing chromosome notation successfully!\n",
      "Wed Jan 24 12:31:54 2024 Start to fix basepair positions...v3.4.36\n",
      "Wed Jan 24 12:31:54 2024  -Current Dataframe shape : 58  x  21\n",
      "Wed Jan 24 12:31:54 2024  -Removing thousands separator \",\" or underbar \"_\" ...\n",
      "Wed Jan 24 12:31:54 2024  -Converting to Int64 data type ...\n",
      "Wed Jan 24 12:31:54 2024  -Force converting to Int64 data type ...\n",
      "Wed Jan 24 12:31:56 2024  -Position bound:(0 , 250,000,000)\n",
      "Wed Jan 24 12:31:56 2024  -Removed outliers: 2\n",
      "Wed Jan 24 12:31:56 2024  -Removed 4 variants with bad positions.\n",
      "Wed Jan 24 12:31:56 2024  -Converted all position to datatype Int64.\n",
      "Wed Jan 24 12:31:56 2024 Finished fixing basepair position successfully!\n",
      "Wed Jan 24 12:31:56 2024 Start to fix alleles...v3.4.36\n",
      "Wed Jan 24 12:31:56 2024  -Current Dataframe shape : 54  x  21\n",
      "Wed Jan 24 12:31:56 2024  -Converted all bases to string datatype and UPPERCASE.\n",
      "Wed Jan 24 12:31:56 2024  -Variants with bad EA  : 1\n",
      "Wed Jan 24 12:31:56 2024  -Variants with bad NEA : 5\n",
      "Wed Jan 24 12:31:56 2024  -Variants with NA for EA or NEA: 1\n",
      "Wed Jan 24 12:31:56 2024  -Variants with same EA and NEA: 1\n",
      "Wed Jan 24 12:31:56 2024  -A look at the non-ATCG EA: {'<CN0>'} ...\n",
      "Wed Jan 24 12:31:56 2024  -A look at the non-ATCG NEA: {nan, 'N', '<CN1>', '*'} ...\n",
      "Wed Jan 24 12:31:56 2024  -Removed 5 variants with NA alleles or alleles that contain bases other than A/C/T/G.\n",
      "Wed Jan 24 12:31:56 2024  -Removed 1 variants with same allele for EA and NEA.\n",
      "Wed Jan 24 12:32:00 2024 Finished fixing allele successfully!\n",
      "Wed Jan 24 12:32:00 2024 Start sanity check for statistics...v3.4.36\n",
      "Wed Jan 24 12:32:00 2024  -Current Dataframe shape : 48  x  21\n",
      "Wed Jan 24 12:32:00 2024  -Comparison tolerance for floats: 1e-07\n",
      "Wed Jan 24 12:32:00 2024  -Checking if  0 <=N<= 2147483647  ...\n",
      "Wed Jan 24 12:32:00 2024  -Removed 3 variants with bad N.\n",
      "Wed Jan 24 12:32:00 2024  -Checking if  0 <=N_CASE<= 2147483647  ...\n",
      "Wed Jan 24 12:32:00 2024  -Removed 1 variants with bad N_CASE.\n",
      "Wed Jan 24 12:32:00 2024  -Checking if  0 <=N_CONTROL<= 2147483647  ...\n",
      "Wed Jan 24 12:32:00 2024  -Removed 1 variants with bad N_CONTROL.\n",
      "Wed Jan 24 12:32:00 2024  -Checking if N = N_CASE + N_CONTROL ...\n",
      "Wed Jan 24 12:32:00 2024  -Removed 1 variants with N != N_CASE + N_CONTROL.\n",
      "Wed Jan 24 12:32:00 2024  -Checking if  -1e-07 <EAF< 1.0000001  ...\n",
      "Wed Jan 24 12:32:00 2024  -Removed 3 variants with bad EAF.\n",
      "Wed Jan 24 12:32:00 2024  -Checking if  0 <=MAC<= 2147483647  ...\n",
      "Wed Jan 24 12:32:00 2024  -Removed 0 variants with bad MAC.\n",
      "Wed Jan 24 12:32:00 2024  -Checking if  -1e-07 <CHISQ< inf  ...\n",
      "Wed Jan 24 12:32:00 2024  -Removed 2 variants with bad CHISQ.\n",
      "Wed Jan 24 12:32:00 2024  -Checking if  -9999.0000001 <Z< 9999.0000001  ...\n",
      "Wed Jan 24 12:32:00 2024  -Removed 2 variants with bad Z.\n",
      "Wed Jan 24 12:32:00 2024  -Checking if  -1e-07 < P < 1.0000001  ...\n",
      "Wed Jan 24 12:32:00 2024  -Removed 3 variants with bad P.\n",
      "Wed Jan 24 12:32:00 2024  -Checking if  -1e-07 <MLOG10P< 9999.0000001  ...\n",
      "Wed Jan 24 12:32:00 2024  -Removed 3 variants with bad MLOG10P.\n",
      "Wed Jan 24 12:32:00 2024  -Checking if  -100.0000001 <BETA< 100.0000001  ...\n",
      "Wed Jan 24 12:32:00 2024  -Removed 2 variants with bad BETA.\n",
      "Wed Jan 24 12:32:00 2024  -Checking if  -1e-07 <SE< inf  ...\n",
      "Wed Jan 24 12:32:00 2024  -Removed 2 variants with bad SE.\n",
      "Wed Jan 24 12:32:00 2024  -Checking if  -100.0000001 <log(OR)< 100.0000001  ...\n",
      "Wed Jan 24 12:32:00 2024  -Removed 1 variants with bad OR.\n",
      "Wed Jan 24 12:32:00 2024  -Checking if  -1e-07 <OR_95L< inf  ...\n",
      "Wed Jan 24 12:32:00 2024  -Removed 2 variants with bad OR_95L.\n",
      "Wed Jan 24 12:32:00 2024  -Checking if  -1e-07 <OR_95U< inf  ...\n",
      "Wed Jan 24 12:32:00 2024  -Removed 2 variants with bad OR_95U.\n",
      "Wed Jan 24 12:32:00 2024  -Checking STATUS and converting STATUS to categories....\n",
      "Wed Jan 24 12:32:01 2024  -Removed 28 variants with bad statistics in total.\n",
      "Wed Jan 24 12:32:01 2024  -Data types for each column:\n",
      "Wed Jan 24 12:32:01 2024  -Column  : SNPID  CHR   POS   EA       NEA      EAF     BETA    SE      Z       CHISQ   P       MLOG10P OR      OR_95L  OR_95U  N     N_CASE N_CONTROL DIRECTION STATUS   NOTE  \n",
      "Wed Jan 24 12:32:01 2024  -DType   : object Int64 Int64 category category float32 float64 float64 float64 float64 float64 float64 float64 float64 float64 Int64 Int64  Int64     object    category object\n",
      "Wed Jan 24 12:32:01 2024  -Verified: T      T     T     T        T        T       T       T       T       T       T       T       T       T       T       T     T      T         T         T        NA    \n",
      "Wed Jan 24 12:32:01 2024 Finished sanity check successfully!\n",
      "Wed Jan 24 12:32:01 2024 Start to normalize variants...v3.4.36\n",
      "Wed Jan 24 12:32:01 2024  -Current Dataframe shape : 20  x  21\n",
      "Wed Jan 24 12:32:02 2024  -Not normalized allele IDs:22 ... \n",
      "Wed Jan 24 12:32:02 2024  -Not normalized allele:['AT' 'GT']... \n",
      "Wed Jan 24 12:32:02 2024  -Modified 1 variants according to parsimony and left alignment principal.\n",
      "Wed Jan 24 12:32:02 2024 Finished normalizing variants successfully!\n",
      "Wed Jan 24 12:32:02 2024 Start to sort the sumstats using P...\n",
      "Wed Jan 24 12:32:02 2024 Start to remove duplicated variants based on snpid...v3.4.36\n",
      "Wed Jan 24 12:32:02 2024  -Current Dataframe shape : 20  x  21\n",
      "Wed Jan 24 12:32:02 2024  -Which variant to keep:  first\n",
      "Wed Jan 24 12:32:02 2024  -Removed  2  based on SNPID...\n",
      "Wed Jan 24 12:32:02 2024 Start to remove duplicated variants based on CHR,POS,EA and NEA...\n",
      "Wed Jan 24 12:32:02 2024  -Current Dataframe shape : 18  x  21\n",
      "Wed Jan 24 12:32:02 2024  -Which variant to keep:  first\n",
      "Wed Jan 24 12:32:02 2024  -Removed  1  based on CHR,POS,EA and NEA...\n",
      "Wed Jan 24 12:32:02 2024 Start to remove multiallelic variants based on chr:pos...\n",
      "Wed Jan 24 12:32:02 2024  -Which variant to keep:  first\n",
      "Wed Jan 24 12:32:02 2024  -Removed  1  multiallelic variants...\n",
      "Wed Jan 24 12:32:02 2024  -Removed  4  variants in total.\n",
      "Wed Jan 24 12:32:02 2024  -Sort the coordinates...\n",
      "Wed Jan 24 12:32:02 2024 Finished removing successfully!\n",
      "Wed Jan 24 12:32:02 2024 Start to sort the genome coordinates...v3.4.36\n",
      "Wed Jan 24 12:32:02 2024  -Current Dataframe shape : 16  x  21\n",
      "Wed Jan 24 12:32:02 2024  -Sorting genome coordinates...\n",
      "Wed Jan 24 12:32:02 2024 Finished sorting genome coordinates successfully!\n",
      "Wed Jan 24 12:32:02 2024 Start to reorder the columns...v3.4.36\n",
      "Wed Jan 24 12:32:02 2024  -Current Dataframe shape : 16  x  21\n",
      "Wed Jan 24 12:32:02 2024  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,EAF,BETA,SE,Z,CHISQ,P,MLOG10P,OR,OR_95L,OR_95U,N,N_CASE,N_CONTROL,DIRECTION,STATUS,NOTE\n",
      "Wed Jan 24 12:32:02 2024 Finished sorting columns successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.basic_check(remove=True,remove_dup=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2f2e58b",
   "metadata": {},
   "source": [
    "## fix id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "54c21788",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wed Jan 24 01:24:03 2024 Start to check IDs...v3.4.36\n",
      "Wed Jan 24 01:24:03 2024  -Current Dataframe shape : 16  x  21\n",
      "Wed Jan 24 01:24:03 2024  -Checking if SNPID is chr:pos:ref:alt...(separator: - ,: , _)\n",
      "Wed Jan 24 01:24:04 2024  -Replacing [_-] in SNPID with \":\" ...\n",
      "Wed Jan 24 01:24:04 2024 Finished checking IDs successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.fix_id(fixsep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "90ca2f0e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>Z</th>\n",
       "      <th>CHISQ</th>\n",
       "      <th>P</th>\n",
       "      <th>MLOG10P</th>\n",
       "      <th>OR</th>\n",
       "      <th>OR_95L</th>\n",
       "      <th>OR_95U</th>\n",
       "      <th>N</th>\n",
       "      <th>N_CASE</th>\n",
       "      <th>N_CONTROL</th>\n",
       "      <th>DIRECTION</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>NOTE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>POS float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>Lowercase alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990399</td>\n",
       "      <td>Not normalizated allelels</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990399</td>\n",
       "      <td>Both Uppercase and lowercases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>N float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>SE out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>42</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>999999.000000</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>OR out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>54</td>\n",
       "      <td>AG</td>\n",
       "      <td>A</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990399</td>\n",
       "      <td>Clean sumstats</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>55</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>Clean sumstats</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>23</td>\n",
       "      <td>4</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>Sex chromosomes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    CHR  POS  EA NEA    EAF    BETA      SE         Z      CHISQ  \\\n",
       "0     1    1   A   G  0.004  0.0603  0.0103  5.854369  34.273636   \n",
       "1     1    2   T   C  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "2     1    3   A   T  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "3     1    6   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "4     1    7   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "5     1    8   T   C  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "6     1   13   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "7     1   16   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "8     1   22   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "9     1   23  AT   A  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "10    1   24   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "11    1   36   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "12    1   42   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "13    1   54  AG   A  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "14    1   55   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "15   23    4   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "\n",
       "               P   MLOG10P             OR    OR_95L    OR_95U       N  N_CASE  \\\n",
       "0   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "1   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "2   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "3   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "4   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "5   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "6   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "7   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "8   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "9   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "10  4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "11  4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "12  4.710000e-09  8.327348  999999.000000  1.040927  1.083816  160000  120000   \n",
       "13  4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "14  4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "15  4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "\n",
       "    N_CONTROL DIRECTION   STATUS                           NOTE  \n",
       "0       40000      --++  9990099                     Duplicated  \n",
       "1       40000      --++  9990099                   Multiallelic  \n",
       "2       40000      --++  9990099                   Multiallelic  \n",
       "3       40000      --++  9990099                CHR with prefix  \n",
       "4       40000      --++  9990099                CHR with prefix  \n",
       "5       40000      --++  9990099                CHR with prefix  \n",
       "6       40000      --++  9990099                      POS float  \n",
       "7       40000      --++  9990099              Lowercase alleles  \n",
       "8       40000      --++  9990399      Not normalizated allelels  \n",
       "9       40000      --++  9990399  Both Uppercase and lowercases  \n",
       "10      40000      --++  9990099                        N float  \n",
       "11      40000      --++  9990099                SE out of range  \n",
       "12      40000      --++  9990099                OR out of range  \n",
       "13      40000      --++  9990399                 Clean sumstats  \n",
       "14      40000      --++  9990099                 Clean sumstats  \n",
       "15      40000      --++  9990099                Sex chromosomes  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea78e4dc",
   "metadata": {},
   "source": [
    "## fix chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e83d1fa5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wed Jan 24 00:50:00 2024 Start to fix chromosome notation...v3.4.36\n",
      "Wed Jan 24 00:50:00 2024  -Current Dataframe shape : 16  x  20\n",
      "Wed Jan 24 00:50:00 2024  -Checking CHR data type...\n",
      "Wed Jan 24 00:50:00 2024  -Variants with standardized chromosome notation: 16\n",
      "Wed Jan 24 00:50:00 2024  -All CHR are already fixed...\n",
      "Wed Jan 24 00:50:01 2024  -Sanity check for CHR...\n",
      "Wed Jan 24 00:50:01 2024  -Removed 0 variants with CHR < 1...\n",
      "Wed Jan 24 00:50:01 2024 Finished fixing chromosome notation successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.fix_chr(remove=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0ebd3b0e",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>Z</th>\n",
       "      <th>CHISQ</th>\n",
       "      <th>P</th>\n",
       "      <th>MLOG10P</th>\n",
       "      <th>OR</th>\n",
       "      <th>OR_95L</th>\n",
       "      <th>OR_95U</th>\n",
       "      <th>N</th>\n",
       "      <th>N_CASE</th>\n",
       "      <th>N_CONTROL</th>\n",
       "      <th>DIRECTION</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>NOTE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>POS float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>Lowercase alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990399</td>\n",
       "      <td>Not normalizated allelels</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990399</td>\n",
       "      <td>Both Uppercase and lowercases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>N float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>SE out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>42</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>999999.000000</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>OR out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>54</td>\n",
       "      <td>AG</td>\n",
       "      <td>A</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990399</td>\n",
       "      <td>Clean sumstats</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>55</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>Clean sumstats</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>23</td>\n",
       "      <td>4</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>5.854369</td>\n",
       "      <td>34.273636</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>8.327348</td>\n",
       "      <td>1.062155</td>\n",
       "      <td>1.040927</td>\n",
       "      <td>1.083816</td>\n",
       "      <td>160000</td>\n",
       "      <td>120000</td>\n",
       "      <td>40000</td>\n",
       "      <td>--++</td>\n",
       "      <td>9990099</td>\n",
       "      <td>Sex chromosomes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    CHR  POS  EA NEA    EAF    BETA      SE         Z      CHISQ  \\\n",
       "0     1    1   A   G  0.004  0.0603  0.0103  5.854369  34.273636   \n",
       "1     1    2   T   C  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "2     1    3   A   T  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "3     1    6   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "4     1    7   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "5     1    8   T   C  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "6     1   13   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "7     1   16   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "8     1   22   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "9     1   23  AT   A  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "10    1   24   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "11    1   36   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "12    1   42   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "13    1   54  AG   A  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "14    1   55   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "15   23    4   A   G  0.996  0.0603  0.0103  5.854369  34.273636   \n",
       "\n",
       "               P   MLOG10P             OR    OR_95L    OR_95U       N  N_CASE  \\\n",
       "0   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "1   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "2   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "3   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "4   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "5   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "6   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "7   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "8   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "9   4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "10  4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "11  4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "12  4.710000e-09  8.327348  999999.000000  1.040927  1.083816  160000  120000   \n",
       "13  4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "14  4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "15  4.710000e-09  8.327348       1.062155  1.040927  1.083816  160000  120000   \n",
       "\n",
       "    N_CONTROL DIRECTION   STATUS                           NOTE  \n",
       "0       40000      --++  9990099                     Duplicated  \n",
       "1       40000      --++  9990099                   Multiallelic  \n",
       "2       40000      --++  9990099                   Multiallelic  \n",
       "3       40000      --++  9990099                CHR with prefix  \n",
       "4       40000      --++  9990099                CHR with prefix  \n",
       "5       40000      --++  9990099                CHR with prefix  \n",
       "6       40000      --++  9990099                      POS float  \n",
       "7       40000      --++  9990099              Lowercase alleles  \n",
       "8       40000      --++  9990399      Not normalizated allelels  \n",
       "9       40000      --++  9990399  Both Uppercase and lowercases  \n",
       "10      40000      --++  9990099                        N float  \n",
       "11      40000      --++  9990099                SE out of range  \n",
       "12      40000      --++  9990099                OR out of range  \n",
       "13      40000      --++  9990399                 Clean sumstats  \n",
       "14      40000      --++  9990099                 Clean sumstats  \n",
       "15      40000      --++  9990099                Sex chromosomes  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd76afda",
   "metadata": {},
   "source": [
    "## fix position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a9063d5c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 23 18:40:19 2024 Start to fix basepair positions...v3.4.36\n",
      "Tue Jan 23 18:40:19 2024  -Current Dataframe shape : 38  x  12\n",
      "Tue Jan 23 18:40:19 2024  -Converting to Int64 data type ...\n",
      "Tue Jan 23 18:40:19 2024  -Force converting to Int64 data type ...\n",
      "Tue Jan 23 18:40:20 2024  -Position upper_bound is: 250,000,000\n",
      "Tue Jan 23 18:40:20 2024  -Remove outliers: 1\n",
      "Tue Jan 23 18:40:20 2024  -Removed 1 variants with bad positions.\n",
      "Tue Jan 23 18:40:20 2024  -Converted all position to datatype Int64.\n",
      "Tue Jan 23 18:40:20 2024 Finished fixing basepair position successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.fix_pos(remove=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "96c5a234",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9960</td>\n",
       "      <td>-0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59700</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59730</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.0490</td>\n",
       "      <td>0.1231</td>\n",
       "      <td>0.69080</td>\n",
       "      <td>166718</td>\n",
       "      <td>9980999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8374</td>\n",
       "      <td>0.0213</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.28460</td>\n",
       "      <td>166718</td>\n",
       "      <td>9980999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:3:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8593</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1:3:T:GA</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>GA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1470</td>\n",
       "      <td>-0.0187</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0.28300</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1:4:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1400</td>\n",
       "      <td>-0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1:5:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1578</td>\n",
       "      <td>-0.0155</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.23500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Flipped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1:6:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1:7:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1:8:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9866</td>\n",
       "      <td>-0.1439</td>\n",
       "      <td>0.0748</td>\n",
       "      <td>0.05451</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1:9:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1327</td>\n",
       "      <td>0.0035</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.80870</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1:10:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8522</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0.0157</td>\n",
       "      <td>0.05421</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1:11:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0.1858</td>\n",
       "      <td>0.1546</td>\n",
       "      <td>0.22930</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1:12:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.1430</td>\n",
       "      <td>0.0747</td>\n",
       "      <td>0.05575</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1:13:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.7445</td>\n",
       "      <td>-0.0135</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>0.32450</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1:14:C:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4544</td>\n",
       "      <td>-0.0170</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.21910</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1:15:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8516</td>\n",
       "      <td>0.0292</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.06173</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1:16:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>16</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1642</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.49750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Sex chromosomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1:17:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7428</td>\n",
       "      <td>-0.0119</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.36210</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8699</td>\n",
       "      <td>-0.0038</td>\n",
       "      <td>0.0161</td>\n",
       "      <td>0.81540</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1:25:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8220</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.0139</td>\n",
       "      <td>0.15350</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>POS float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1:864002:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>864002</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8666</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0155</td>\n",
       "      <td>0.93090</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Same alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1:864615:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864615</td>\n",
       "      <td>g</td>\n",
       "      <td>a</td>\n",
       "      <td>0.8725</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>0.0158</td>\n",
       "      <td>0.08092</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Lowercase alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1:864701:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>864701</td>\n",
       "      <td>&lt;CN1&gt;</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>0.1306</td>\n",
       "      <td>0.0680</td>\n",
       "      <td>0.05466</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Unrecognized alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1:864755:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>864755</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8242</td>\n",
       "      <td>0.0163</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.32220</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Unrecognized alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1:864757:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864757</td>\n",
       "      <td>GT</td>\n",
       "      <td>AT</td>\n",
       "      <td>0.1758</td>\n",
       "      <td>-0.0164</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.31990</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Not normalizated allelels</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1:865219:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>865219</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8685</td>\n",
       "      <td>300.0000</td>\n",
       "      <td>0.0161</td>\n",
       "      <td>0.79480</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>BETA out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1:865367:G:T</td>\n",
       "      <td>1</td>\n",
       "      <td>865367</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8486</td>\n",
       "      <td>-300.0000</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.05270</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>BETA out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1:865438:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>865438</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.9849</td>\n",
       "      <td>-0.1299</td>\n",
       "      <td>-0.0680</td>\n",
       "      <td>0.05601</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>SE out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>1:865694:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>865694</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8489</td>\n",
       "      <td>0.0265</td>\n",
       "      <td>0.0132</td>\n",
       "      <td>-0.04424</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>P out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>1:866893:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>866893</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.3104</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>1.67220</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>P out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>1:866938:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>866938</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.6028</td>\n",
       "      <td>-0.0104</td>\n",
       "      <td>0.0121</td>\n",
       "      <td>0.38960</td>\n",
       "      <td>-1</td>\n",
       "      <td>9960999</td>\n",
       "      <td>N out of range</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNPID  CHR     POS     EA NEA     EAF      BETA      SE        P  \\\n",
       "0        1:1:G:A    1       1      G   A  0.9960   -0.0737  0.1394  0.59700   \n",
       "1        1:1:A:G    1       1      G   A  0.0040    0.0737  0.1394  0.59730   \n",
       "2        1:1:A:G    1       1      G   A  0.0040    0.0737  0.1394  0.59750   \n",
       "3            1:2    1       2      C   T  0.0051    0.0490  0.1231  0.69080   \n",
       "4            1:2    1       2    TAA   T  0.8374    0.0213  0.0199  0.28460   \n",
       "5        1:3:T:A    1       3      T   A  0.8593    0.0172  0.0156  0.27050   \n",
       "6       1:3:T:GA    1       3     GA   T  0.1470   -0.0187  0.0174  0.28300   \n",
       "7        1:4:T:C    1       4      C   T  0.1400   -0.0172  0.0156  0.27050   \n",
       "8        1:5:G:A    1       5      G   A  0.1578   -0.0155  0.0131  0.23500   \n",
       "9        1:6:A:G    1       6      G   A  0.7493    0.0204  0.0147  0.16500   \n",
       "10       1:7:A:G    1       7      G   A  0.7493    0.0204  0.0147  0.16500   \n",
       "11       1:8:G:C    1       8      G   C  0.9866   -0.1439  0.0748  0.05451   \n",
       "12       1:9:A:G    1       9      G   A  0.1327    0.0035  0.0146  0.80870   \n",
       "13      1:10:G:A    1      10      G   A  0.8522    0.0302  0.0157  0.05421   \n",
       "14      1:11:T:C    1      11      C   T  0.0027    0.1858  0.1546  0.22930   \n",
       "15      1:12:A:G    1      12      G   A  0.0131    0.1430  0.0747  0.05575   \n",
       "16      1:13:C:T    1      13      C   T  0.7445   -0.0135  0.0137  0.32450   \n",
       "17      1:14:C:A    1      14      C   A  0.4544   -0.0170  0.0138  0.21910   \n",
       "18      1:15:C:T    1      15      C   T  0.8516    0.0292  0.0156  0.06173   \n",
       "19      1:16:A:G   23      16      G   A  0.1642    0.0099  0.0146  0.49750   \n",
       "20      1:17:G:A    1      17      G   A  0.7428   -0.0119  0.0131  0.36210   \n",
       "21       1:1:G:A    1      18      G   A  0.8699   -0.0038  0.0161  0.81540   \n",
       "23      1:18:A:G    1      20      G   A  0.7316    0.0013  0.0147  0.93200   \n",
       "24      1:18:A:G    1      21      G   A  0.7316    0.0013  0.0147  0.93200   \n",
       "25      1:18:A:G    1      22      G   A  0.7316    0.0013  0.0147  0.93200   \n",
       "27      1:25:A:G    1      25      G   A  0.8220    0.0199  0.0139  0.15350   \n",
       "28  1:864002:G:C    1  864002      A   A  0.8666    0.0013  0.0155  0.93090   \n",
       "29  1:864615:G:A    1  864615      g   a  0.8725    0.0275  0.0158  0.08092   \n",
       "30  1:864701:A:G    1  864701  <CN1>   A  0.0150    0.1306  0.0680  0.05466   \n",
       "31  1:864755:A:G    1  864755      N   A  0.8242    0.0163  0.0165  0.32220   \n",
       "32  1:864757:G:A    1  864757     GT  AT  0.1758   -0.0164  0.0165  0.31990   \n",
       "35  1:865219:G:A    1  865219      G   A  0.8685  300.0000  0.0161  0.79480   \n",
       "36  1:865367:G:T    1  865367      G   T  0.8486 -300.0000  0.0138  0.05270   \n",
       "37  1:865438:C:T    1  865438      C   T  0.9849   -0.1299 -0.0680  0.05601   \n",
       "38  1:865694:C:T    1  865694      C   T  0.8489    0.0265  0.0132 -0.04424   \n",
       "39  1:866893:T:C    1  866893      C   T  0.3104    0.0058  0.0137  1.67220   \n",
       "40  1:866938:G:A    1  866938      G   A  0.6028   -0.0104  0.0121  0.38960   \n",
       "\n",
       "         N   STATUS                       Note  \n",
       "0   191764  9960999                 Duplicated  \n",
       "1   166718  9960999                 Duplicated  \n",
       "2   166718  9960999                 Duplicated  \n",
       "3   166718  9980999               Multiallelic  \n",
       "4   166718  9980999               Multiallelic  \n",
       "5   166718  9960999               Multiallelic  \n",
       "6   166718  9960999               Multiallelic  \n",
       "7   191764  9960999                        NaN  \n",
       "8   166718  9960999                    Flipped  \n",
       "9   166718  9960999                        NaN  \n",
       "10  166718  9960999                        NaN  \n",
       "11  166718  9960999                        NaN  \n",
       "12  166718  9960999                        NaN  \n",
       "13  166718  9960999                        NaN  \n",
       "14  166718  9960999                        NaN  \n",
       "15  166718  9960999                        NaN  \n",
       "16  191764  9960999                        NaN  \n",
       "17  166718  9960999                        NaN  \n",
       "18  166718  9960999                        NaN  \n",
       "19  166718  9960999            Sex chromosomes  \n",
       "20  166718  9960999            CHR with prefix  \n",
       "21  166718  9960999            CHR with prefix  \n",
       "23  191764  9960999                        NaN  \n",
       "24  166718  9960999                        NaN  \n",
       "25  166718  9960999                        NaN  \n",
       "27  166718  9960999                  POS float  \n",
       "28  166718  9960999               Same alleles  \n",
       "29  166718  9960999          Lowercase alleles  \n",
       "30  166718  9960999       Unrecognized alleles  \n",
       "31  166718  9960999       Unrecognized alleles  \n",
       "32  166718  9960999  Not normalizated allelels  \n",
       "35  166718  9960999          BETA out of range  \n",
       "36  166718  9960999          BETA out of range  \n",
       "37  166718  9960999            SE out of range  \n",
       "38  166718  9960999             P out of range  \n",
       "39  166718  9960999             P out of range  \n",
       "40      -1  9960999             N out of range  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45c72b40",
   "metadata": {},
   "source": [
    "## fix allele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3edf8219",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 23 18:40:42 2024 Start to fix alleles...v3.4.36\n",
      "Tue Jan 23 18:40:42 2024  -Current Dataframe shape : 37  x  12\n",
      "Tue Jan 23 18:40:42 2024  -Converted all bases to string datatype and UPPERCASE.\n",
      "Tue Jan 23 18:40:42 2024  -Variants with bad EA  : 2\n",
      "Tue Jan 23 18:40:42 2024  -Variants with bad NEA : 0\n",
      "Tue Jan 23 18:40:42 2024  -Variants with NA for EA or NEA: 0\n",
      "Tue Jan 23 18:40:42 2024  -Variants with same EA and NEA: 1\n",
      "Tue Jan 23 18:40:42 2024  -A look at the non-ATCG EA: {'<CN1>', 'N'} ...\n",
      "Tue Jan 23 18:40:42 2024  -Removed 2 variants with NA alleles or alleles that contain bases other than A/C/T/G.\n",
      "Tue Jan 23 18:40:42 2024  -Removed 1 variants with same allele for EA and NEA.\n",
      "Tue Jan 23 18:40:42 2024 Finished fixing allele successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.fix_allele(remove=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d7e0cafe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9960</td>\n",
       "      <td>-0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59700</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59730</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.0490</td>\n",
       "      <td>0.1231</td>\n",
       "      <td>0.69080</td>\n",
       "      <td>166718</td>\n",
       "      <td>9980099</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8374</td>\n",
       "      <td>0.0213</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.28460</td>\n",
       "      <td>166718</td>\n",
       "      <td>9980399</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:3:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8593</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1:3:T:GA</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>GA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1470</td>\n",
       "      <td>-0.0187</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0.28300</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960399</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1:4:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1400</td>\n",
       "      <td>-0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1:5:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1578</td>\n",
       "      <td>-0.0155</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.23500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Flipped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1:6:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1:7:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1:8:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9866</td>\n",
       "      <td>-0.1439</td>\n",
       "      <td>0.0748</td>\n",
       "      <td>0.05451</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1:9:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1327</td>\n",
       "      <td>0.0035</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.80870</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1:10:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8522</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0.0157</td>\n",
       "      <td>0.05421</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1:11:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0.1858</td>\n",
       "      <td>0.1546</td>\n",
       "      <td>0.22930</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1:12:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.1430</td>\n",
       "      <td>0.0747</td>\n",
       "      <td>0.05575</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1:13:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.7445</td>\n",
       "      <td>-0.0135</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>0.32450</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1:14:C:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4544</td>\n",
       "      <td>-0.0170</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.21910</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1:15:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8516</td>\n",
       "      <td>0.0292</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.06173</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1:16:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>16</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1642</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.49750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Sex chromosomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1:17:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7428</td>\n",
       "      <td>-0.0119</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.36210</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8699</td>\n",
       "      <td>-0.0038</td>\n",
       "      <td>0.0161</td>\n",
       "      <td>0.81540</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1:25:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8220</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.0139</td>\n",
       "      <td>0.15350</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>POS float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1:864615:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864615</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8725</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>0.0158</td>\n",
       "      <td>0.08092</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Lowercase alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1:864757:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864757</td>\n",
       "      <td>GT</td>\n",
       "      <td>AT</td>\n",
       "      <td>0.1758</td>\n",
       "      <td>-0.0164</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.31990</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960599</td>\n",
       "      <td>Not normalizated allelels</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1:865219:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>865219</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8685</td>\n",
       "      <td>300.0000</td>\n",
       "      <td>0.0161</td>\n",
       "      <td>0.79480</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>BETA out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1:865367:G:T</td>\n",
       "      <td>1</td>\n",
       "      <td>865367</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8486</td>\n",
       "      <td>-300.0000</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.05270</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>BETA out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1:865438:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>865438</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.9849</td>\n",
       "      <td>-0.1299</td>\n",
       "      <td>-0.0680</td>\n",
       "      <td>0.05601</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>SE out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>1:865694:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>865694</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8489</td>\n",
       "      <td>0.0265</td>\n",
       "      <td>0.0132</td>\n",
       "      <td>-0.04424</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>P out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>1:866893:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>866893</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.3104</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>1.67220</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>P out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>1:866938:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>866938</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.6028</td>\n",
       "      <td>-0.0104</td>\n",
       "      <td>0.0121</td>\n",
       "      <td>0.38960</td>\n",
       "      <td>-1</td>\n",
       "      <td>9960099</td>\n",
       "      <td>N out of range</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNPID  CHR     POS   EA NEA     EAF      BETA      SE        P  \\\n",
       "0        1:1:G:A    1       1    G   A  0.9960   -0.0737  0.1394  0.59700   \n",
       "1        1:1:A:G    1       1    G   A  0.0040    0.0737  0.1394  0.59730   \n",
       "2        1:1:A:G    1       1    G   A  0.0040    0.0737  0.1394  0.59750   \n",
       "3            1:2    1       2    C   T  0.0051    0.0490  0.1231  0.69080   \n",
       "4            1:2    1       2  TAA   T  0.8374    0.0213  0.0199  0.28460   \n",
       "5        1:3:T:A    1       3    T   A  0.8593    0.0172  0.0156  0.27050   \n",
       "6       1:3:T:GA    1       3   GA   T  0.1470   -0.0187  0.0174  0.28300   \n",
       "7        1:4:T:C    1       4    C   T  0.1400   -0.0172  0.0156  0.27050   \n",
       "8        1:5:G:A    1       5    G   A  0.1578   -0.0155  0.0131  0.23500   \n",
       "9        1:6:A:G    1       6    G   A  0.7493    0.0204  0.0147  0.16500   \n",
       "10       1:7:A:G    1       7    G   A  0.7493    0.0204  0.0147  0.16500   \n",
       "11       1:8:G:C    1       8    G   C  0.9866   -0.1439  0.0748  0.05451   \n",
       "12       1:9:A:G    1       9    G   A  0.1327    0.0035  0.0146  0.80870   \n",
       "13      1:10:G:A    1      10    G   A  0.8522    0.0302  0.0157  0.05421   \n",
       "14      1:11:T:C    1      11    C   T  0.0027    0.1858  0.1546  0.22930   \n",
       "15      1:12:A:G    1      12    G   A  0.0131    0.1430  0.0747  0.05575   \n",
       "16      1:13:C:T    1      13    C   T  0.7445   -0.0135  0.0137  0.32450   \n",
       "17      1:14:C:A    1      14    C   A  0.4544   -0.0170  0.0138  0.21910   \n",
       "18      1:15:C:T    1      15    C   T  0.8516    0.0292  0.0156  0.06173   \n",
       "19      1:16:A:G   23      16    G   A  0.1642    0.0099  0.0146  0.49750   \n",
       "20      1:17:G:A    1      17    G   A  0.7428   -0.0119  0.0131  0.36210   \n",
       "21       1:1:G:A    1      18    G   A  0.8699   -0.0038  0.0161  0.81540   \n",
       "23      1:18:A:G    1      20    G   A  0.7316    0.0013  0.0147  0.93200   \n",
       "24      1:18:A:G    1      21    G   A  0.7316    0.0013  0.0147  0.93200   \n",
       "25      1:18:A:G    1      22    G   A  0.7316    0.0013  0.0147  0.93200   \n",
       "27      1:25:A:G    1      25    G   A  0.8220    0.0199  0.0139  0.15350   \n",
       "29  1:864615:G:A    1  864615    G   A  0.8725    0.0275  0.0158  0.08092   \n",
       "32  1:864757:G:A    1  864757   GT  AT  0.1758   -0.0164  0.0165  0.31990   \n",
       "35  1:865219:G:A    1  865219    G   A  0.8685  300.0000  0.0161  0.79480   \n",
       "36  1:865367:G:T    1  865367    G   T  0.8486 -300.0000  0.0138  0.05270   \n",
       "37  1:865438:C:T    1  865438    C   T  0.9849   -0.1299 -0.0680  0.05601   \n",
       "38  1:865694:C:T    1  865694    C   T  0.8489    0.0265  0.0132 -0.04424   \n",
       "39  1:866893:T:C    1  866893    C   T  0.3104    0.0058  0.0137  1.67220   \n",
       "40  1:866938:G:A    1  866938    G   A  0.6028   -0.0104  0.0121  0.38960   \n",
       "\n",
       "         N   STATUS                       Note  \n",
       "0   191764  9960099                 Duplicated  \n",
       "1   166718  9960099                 Duplicated  \n",
       "2   166718  9960099                 Duplicated  \n",
       "3   166718  9980099               Multiallelic  \n",
       "4   166718  9980399               Multiallelic  \n",
       "5   166718  9960099               Multiallelic  \n",
       "6   166718  9960399               Multiallelic  \n",
       "7   191764  9960099                        NaN  \n",
       "8   166718  9960099                    Flipped  \n",
       "9   166718  9960099                        NaN  \n",
       "10  166718  9960099                        NaN  \n",
       "11  166718  9960099                        NaN  \n",
       "12  166718  9960099                        NaN  \n",
       "13  166718  9960099                        NaN  \n",
       "14  166718  9960099                        NaN  \n",
       "15  166718  9960099                        NaN  \n",
       "16  191764  9960099                        NaN  \n",
       "17  166718  9960099                        NaN  \n",
       "18  166718  9960099                        NaN  \n",
       "19  166718  9960099            Sex chromosomes  \n",
       "20  166718  9960099            CHR with prefix  \n",
       "21  166718  9960099            CHR with prefix  \n",
       "23  191764  9960099                        NaN  \n",
       "24  166718  9960099                        NaN  \n",
       "25  166718  9960099                        NaN  \n",
       "27  166718  9960099                  POS float  \n",
       "29  166718  9960099          Lowercase alleles  \n",
       "32  166718  9960599  Not normalizated allelels  \n",
       "35  166718  9960099          BETA out of range  \n",
       "36  166718  9960099          BETA out of range  \n",
       "37  166718  9960099            SE out of range  \n",
       "38  166718  9960099             P out of range  \n",
       "39  166718  9960099             P out of range  \n",
       "40      -1  9960099             N out of range  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54345b77",
   "metadata": {},
   "source": [
    "## sanity check for statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "819cf4bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 23 18:41:08 2024 Start sanity check for statistics...v3.4.36\n",
      "Tue Jan 23 18:41:08 2024  -Current Dataframe shape : 34  x  12\n",
      "Tue Jan 23 18:41:08 2024  -Comparison tolerance for floats: 1e-07\n",
      "Tue Jan 23 18:41:08 2024  -Checking if  0 <=N<= 2147483647  ...\n",
      "Tue Jan 23 18:41:08 2024  -Removed 1 variants with bad N.\n",
      "Tue Jan 23 18:41:08 2024  -Checking if  -1e-07 <EAF< 1.0000001  ...\n",
      "Tue Jan 23 18:41:08 2024  -Removed 0 variants with bad EAF.\n",
      "Tue Jan 23 18:41:08 2024  -Checking if  0 <=MAC<= 2147483647  ...\n",
      "Tue Jan 23 18:41:08 2024  -Removed 0 variants with bad MAC.\n",
      "Tue Jan 23 18:41:08 2024  -Checking if  -1e-07 < P < 1.0000001  ...\n",
      "Tue Jan 23 18:41:08 2024  -Removed 2 variants with bad P.\n",
      "Tue Jan 23 18:41:08 2024  -Checking if  -100.0000001 <BETA< 100.0000001  ...\n",
      "Tue Jan 23 18:41:08 2024  -Removed 2 variants with bad BETA.\n",
      "Tue Jan 23 18:41:08 2024  -Checking if  -1e-07 <SE< inf  ...\n",
      "Tue Jan 23 18:41:08 2024  -Removed 1 variants with bad SE.\n",
      "Tue Jan 23 18:41:08 2024  -Checking STATUS and converting STATUS to categories....\n",
      "Tue Jan 23 18:41:08 2024  -Removed 0 variants with NAs in the checked columns...\n",
      "Tue Jan 23 18:41:08 2024  -Removed 6 variants with bad statistics in total.\n",
      "Tue Jan 23 18:41:08 2024  -Data types for each column:\n",
      "Tue Jan 23 18:41:08 2024  -Column  : SNPID  CHR   POS   EA       NEA      EAF     BETA    SE      P       N     STATUS   Note  \n",
      "Tue Jan 23 18:41:08 2024  -DType   : object Int64 Int64 category category float32 float64 float64 float64 Int32 category object\n",
      "Tue Jan 23 18:41:08 2024  -Verified: T      T     T     T        T        T       T       T       T       T     T        NA    \n",
      "Tue Jan 23 18:41:08 2024 Finished sanity check successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.check_sanity()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4518a81d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9960</td>\n",
       "      <td>-0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59700</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59730</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.0490</td>\n",
       "      <td>0.1231</td>\n",
       "      <td>0.69080</td>\n",
       "      <td>166718</td>\n",
       "      <td>9980099</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8374</td>\n",
       "      <td>0.0213</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.28460</td>\n",
       "      <td>166718</td>\n",
       "      <td>9980399</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:3:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8593</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1:3:T:GA</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>GA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1470</td>\n",
       "      <td>-0.0187</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0.28300</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960399</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1:4:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1400</td>\n",
       "      <td>-0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1:5:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1578</td>\n",
       "      <td>-0.0155</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.23500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Flipped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1:6:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1:7:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1:8:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9866</td>\n",
       "      <td>-0.1439</td>\n",
       "      <td>0.0748</td>\n",
       "      <td>0.05451</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1:9:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1327</td>\n",
       "      <td>0.0035</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.80870</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1:10:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8522</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0.0157</td>\n",
       "      <td>0.05421</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1:11:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0.1858</td>\n",
       "      <td>0.1546</td>\n",
       "      <td>0.22930</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1:12:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.1430</td>\n",
       "      <td>0.0747</td>\n",
       "      <td>0.05575</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1:13:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.7445</td>\n",
       "      <td>-0.0135</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>0.32450</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1:14:C:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4544</td>\n",
       "      <td>-0.0170</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.21910</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1:15:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8516</td>\n",
       "      <td>0.0292</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.06173</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1:16:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>16</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1642</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.49750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Sex chromosomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1:17:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7428</td>\n",
       "      <td>-0.0119</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.36210</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8699</td>\n",
       "      <td>-0.0038</td>\n",
       "      <td>0.0161</td>\n",
       "      <td>0.81540</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1:25:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8220</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.0139</td>\n",
       "      <td>0.15350</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>POS float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1:864615:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864615</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8725</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>0.0158</td>\n",
       "      <td>0.08092</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Lowercase alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1:864757:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864757</td>\n",
       "      <td>GT</td>\n",
       "      <td>AT</td>\n",
       "      <td>0.1758</td>\n",
       "      <td>-0.0164</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.31990</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960599</td>\n",
       "      <td>Not normalizated allelels</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNPID  CHR     POS   EA NEA     EAF    BETA      SE        P  \\\n",
       "0        1:1:G:A    1       1    G   A  0.9960 -0.0737  0.1394  0.59700   \n",
       "1        1:1:A:G    1       1    G   A  0.0040  0.0737  0.1394  0.59730   \n",
       "2        1:1:A:G    1       1    G   A  0.0040  0.0737  0.1394  0.59750   \n",
       "3            1:2    1       2    C   T  0.0051  0.0490  0.1231  0.69080   \n",
       "4            1:2    1       2  TAA   T  0.8374  0.0213  0.0199  0.28460   \n",
       "5        1:3:T:A    1       3    T   A  0.8593  0.0172  0.0156  0.27050   \n",
       "6       1:3:T:GA    1       3   GA   T  0.1470 -0.0187  0.0174  0.28300   \n",
       "7        1:4:T:C    1       4    C   T  0.1400 -0.0172  0.0156  0.27050   \n",
       "8        1:5:G:A    1       5    G   A  0.1578 -0.0155  0.0131  0.23500   \n",
       "9        1:6:A:G    1       6    G   A  0.7493  0.0204  0.0147  0.16500   \n",
       "10       1:7:A:G    1       7    G   A  0.7493  0.0204  0.0147  0.16500   \n",
       "11       1:8:G:C    1       8    G   C  0.9866 -0.1439  0.0748  0.05451   \n",
       "12       1:9:A:G    1       9    G   A  0.1327  0.0035  0.0146  0.80870   \n",
       "13      1:10:G:A    1      10    G   A  0.8522  0.0302  0.0157  0.05421   \n",
       "14      1:11:T:C    1      11    C   T  0.0027  0.1858  0.1546  0.22930   \n",
       "15      1:12:A:G    1      12    G   A  0.0131  0.1430  0.0747  0.05575   \n",
       "16      1:13:C:T    1      13    C   T  0.7445 -0.0135  0.0137  0.32450   \n",
       "17      1:14:C:A    1      14    C   A  0.4544 -0.0170  0.0138  0.21910   \n",
       "18      1:15:C:T    1      15    C   T  0.8516  0.0292  0.0156  0.06173   \n",
       "19      1:16:A:G   23      16    G   A  0.1642  0.0099  0.0146  0.49750   \n",
       "20      1:17:G:A    1      17    G   A  0.7428 -0.0119  0.0131  0.36210   \n",
       "21       1:1:G:A    1      18    G   A  0.8699 -0.0038  0.0161  0.81540   \n",
       "23      1:18:A:G    1      20    G   A  0.7316  0.0013  0.0147  0.93200   \n",
       "24      1:18:A:G    1      21    G   A  0.7316  0.0013  0.0147  0.93200   \n",
       "25      1:18:A:G    1      22    G   A  0.7316  0.0013  0.0147  0.93200   \n",
       "27      1:25:A:G    1      25    G   A  0.8220  0.0199  0.0139  0.15350   \n",
       "29  1:864615:G:A    1  864615    G   A  0.8725  0.0275  0.0158  0.08092   \n",
       "32  1:864757:G:A    1  864757   GT  AT  0.1758 -0.0164  0.0165  0.31990   \n",
       "\n",
       "         N   STATUS                       Note  \n",
       "0   191764  9960099                 Duplicated  \n",
       "1   166718  9960099                 Duplicated  \n",
       "2   166718  9960099                 Duplicated  \n",
       "3   166718  9980099               Multiallelic  \n",
       "4   166718  9980399               Multiallelic  \n",
       "5   166718  9960099               Multiallelic  \n",
       "6   166718  9960399               Multiallelic  \n",
       "7   191764  9960099                        NaN  \n",
       "8   166718  9960099                    Flipped  \n",
       "9   166718  9960099                        NaN  \n",
       "10  166718  9960099                        NaN  \n",
       "11  166718  9960099                        NaN  \n",
       "12  166718  9960099                        NaN  \n",
       "13  166718  9960099                        NaN  \n",
       "14  166718  9960099                        NaN  \n",
       "15  166718  9960099                        NaN  \n",
       "16  191764  9960099                        NaN  \n",
       "17  166718  9960099                        NaN  \n",
       "18  166718  9960099                        NaN  \n",
       "19  166718  9960099            Sex chromosomes  \n",
       "20  166718  9960099            CHR with prefix  \n",
       "21  166718  9960099            CHR with prefix  \n",
       "23  191764  9960099                        NaN  \n",
       "24  166718  9960099                        NaN  \n",
       "25  166718  9960099                        NaN  \n",
       "27  166718  9960099                  POS float  \n",
       "29  166718  9960099          Lowercase alleles  \n",
       "32  166718  9960599  Not normalizated allelels  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b773946",
   "metadata": {},
   "source": [
    "## normalize variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "86561df4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 23 18:41:20 2024 Start to normalize variants...v3.4.36\n",
      "Tue Jan 23 18:41:20 2024  -Current Dataframe shape : 28  x  12\n",
      "Tue Jan 23 18:41:21 2024  -Not normalized allele IDs:1:864757:G:A ... \n",
      "Tue Jan 23 18:41:21 2024  -Not normalized allele:['GT' 'AT']... \n",
      "Tue Jan 23 18:41:21 2024  -Modified 1 variants according to parsimony and left alignment principal.\n",
      "Tue Jan 23 18:41:21 2024 Finished normalizing variants successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.normalize_allele()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "bea1ddbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9960</td>\n",
       "      <td>-0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59700</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59730</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.0490</td>\n",
       "      <td>0.1231</td>\n",
       "      <td>0.69080</td>\n",
       "      <td>166718</td>\n",
       "      <td>9980099</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8374</td>\n",
       "      <td>0.0213</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.28460</td>\n",
       "      <td>166718</td>\n",
       "      <td>9980399</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:3:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8593</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1:3:T:GA</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>GA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1470</td>\n",
       "      <td>-0.0187</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0.28300</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960399</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1:4:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1400</td>\n",
       "      <td>-0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1:5:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1578</td>\n",
       "      <td>-0.0155</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.23500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Flipped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1:6:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1:7:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1:8:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9866</td>\n",
       "      <td>-0.1439</td>\n",
       "      <td>0.0748</td>\n",
       "      <td>0.05451</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1:9:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1327</td>\n",
       "      <td>0.0035</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.80870</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1:10:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8522</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0.0157</td>\n",
       "      <td>0.05421</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1:11:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0.1858</td>\n",
       "      <td>0.1546</td>\n",
       "      <td>0.22930</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1:12:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.1430</td>\n",
       "      <td>0.0747</td>\n",
       "      <td>0.05575</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1:13:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.7445</td>\n",
       "      <td>-0.0135</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>0.32450</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1:14:C:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4544</td>\n",
       "      <td>-0.0170</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.21910</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1:15:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8516</td>\n",
       "      <td>0.0292</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.06173</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1:16:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>16</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1642</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.49750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Sex chromosomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1:17:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7428</td>\n",
       "      <td>-0.0119</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.36210</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8699</td>\n",
       "      <td>-0.0038</td>\n",
       "      <td>0.0161</td>\n",
       "      <td>0.81540</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1:25:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8220</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.0139</td>\n",
       "      <td>0.15350</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>POS float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1:864615:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864615</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8725</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>0.0158</td>\n",
       "      <td>0.08092</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Lowercase alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1:864757:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864757</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1758</td>\n",
       "      <td>-0.0164</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.31990</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960399</td>\n",
       "      <td>Not normalizated allelels</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNPID  CHR     POS   EA NEA     EAF    BETA      SE        P  \\\n",
       "0        1:1:G:A    1       1    G   A  0.9960 -0.0737  0.1394  0.59700   \n",
       "1        1:1:A:G    1       1    G   A  0.0040  0.0737  0.1394  0.59730   \n",
       "2        1:1:A:G    1       1    G   A  0.0040  0.0737  0.1394  0.59750   \n",
       "3            1:2    1       2    C   T  0.0051  0.0490  0.1231  0.69080   \n",
       "4            1:2    1       2  TAA   T  0.8374  0.0213  0.0199  0.28460   \n",
       "5        1:3:T:A    1       3    T   A  0.8593  0.0172  0.0156  0.27050   \n",
       "6       1:3:T:GA    1       3   GA   T  0.1470 -0.0187  0.0174  0.28300   \n",
       "7        1:4:T:C    1       4    C   T  0.1400 -0.0172  0.0156  0.27050   \n",
       "8        1:5:G:A    1       5    G   A  0.1578 -0.0155  0.0131  0.23500   \n",
       "9        1:6:A:G    1       6    G   A  0.7493  0.0204  0.0147  0.16500   \n",
       "10       1:7:A:G    1       7    G   A  0.7493  0.0204  0.0147  0.16500   \n",
       "11       1:8:G:C    1       8    G   C  0.9866 -0.1439  0.0748  0.05451   \n",
       "12       1:9:A:G    1       9    G   A  0.1327  0.0035  0.0146  0.80870   \n",
       "13      1:10:G:A    1      10    G   A  0.8522  0.0302  0.0157  0.05421   \n",
       "14      1:11:T:C    1      11    C   T  0.0027  0.1858  0.1546  0.22930   \n",
       "15      1:12:A:G    1      12    G   A  0.0131  0.1430  0.0747  0.05575   \n",
       "16      1:13:C:T    1      13    C   T  0.7445 -0.0135  0.0137  0.32450   \n",
       "17      1:14:C:A    1      14    C   A  0.4544 -0.0170  0.0138  0.21910   \n",
       "18      1:15:C:T    1      15    C   T  0.8516  0.0292  0.0156  0.06173   \n",
       "19      1:16:A:G   23      16    G   A  0.1642  0.0099  0.0146  0.49750   \n",
       "20      1:17:G:A    1      17    G   A  0.7428 -0.0119  0.0131  0.36210   \n",
       "21       1:1:G:A    1      18    G   A  0.8699 -0.0038  0.0161  0.81540   \n",
       "23      1:18:A:G    1      20    G   A  0.7316  0.0013  0.0147  0.93200   \n",
       "24      1:18:A:G    1      21    G   A  0.7316  0.0013  0.0147  0.93200   \n",
       "25      1:18:A:G    1      22    G   A  0.7316  0.0013  0.0147  0.93200   \n",
       "27      1:25:A:G    1      25    G   A  0.8220  0.0199  0.0139  0.15350   \n",
       "29  1:864615:G:A    1  864615    G   A  0.8725  0.0275  0.0158  0.08092   \n",
       "32  1:864757:G:A    1  864757    G   A  0.1758 -0.0164  0.0165  0.31990   \n",
       "\n",
       "         N   STATUS                       Note  \n",
       "0   191764  9960099                 Duplicated  \n",
       "1   166718  9960099                 Duplicated  \n",
       "2   166718  9960099                 Duplicated  \n",
       "3   166718  9980099               Multiallelic  \n",
       "4   166718  9980399               Multiallelic  \n",
       "5   166718  9960099               Multiallelic  \n",
       "6   166718  9960399               Multiallelic  \n",
       "7   191764  9960099                        NaN  \n",
       "8   166718  9960099                    Flipped  \n",
       "9   166718  9960099                        NaN  \n",
       "10  166718  9960099                        NaN  \n",
       "11  166718  9960099                        NaN  \n",
       "12  166718  9960099                        NaN  \n",
       "13  166718  9960099                        NaN  \n",
       "14  166718  9960099                        NaN  \n",
       "15  166718  9960099                        NaN  \n",
       "16  191764  9960099                        NaN  \n",
       "17  166718  9960099                        NaN  \n",
       "18  166718  9960099                        NaN  \n",
       "19  166718  9960099            Sex chromosomes  \n",
       "20  166718  9960099            CHR with prefix  \n",
       "21  166718  9960099            CHR with prefix  \n",
       "23  191764  9960099                        NaN  \n",
       "24  166718  9960099                        NaN  \n",
       "25  166718  9960099                        NaN  \n",
       "27  166718  9960099                  POS float  \n",
       "29  166718  9960099          Lowercase alleles  \n",
       "32  166718  9960399  Not normalizated allelels  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6437b3c",
   "metadata": {},
   "source": [
    "## remove duplicated / multiallelic variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ff597de1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 23 18:41:37 2024 Start to sort the sumstats using P...\n",
      "Tue Jan 23 18:41:37 2024 Start to remove duplicated variants based on snpid...v3.4.36\n",
      "Tue Jan 23 18:41:37 2024  -Current Dataframe shape : 28  x  12\n",
      "Tue Jan 23 18:41:37 2024  -Which variant to keep:  first\n",
      "Tue Jan 23 18:41:37 2024  -Removed  5  based on SNPID...\n",
      "Tue Jan 23 18:41:37 2024 Start to remove duplicated variants based on CHR,POS,EA and NEA...\n",
      "Tue Jan 23 18:41:37 2024  -Current Dataframe shape : 23  x  12\n",
      "Tue Jan 23 18:41:37 2024  -Which variant to keep:  first\n",
      "Tue Jan 23 18:41:37 2024  -Removed  1  based on CHR,POS,EA and NEA...\n",
      "Tue Jan 23 18:41:37 2024 Start to remove multiallelic variants based on chr:pos...\n",
      "Tue Jan 23 18:41:37 2024  -Which variant to keep:  first\n",
      "Tue Jan 23 18:41:37 2024  -Removed  1  multiallelic variants...\n",
      "Tue Jan 23 18:41:37 2024  -Removed  7  variants in total.\n",
      "Tue Jan 23 18:41:37 2024  -Sort the coordinates...\n",
      "Tue Jan 23 18:41:37 2024 Finished removing successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.remove_dup(mode=\"md\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9f2a4fe1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9960</td>\n",
       "      <td>-0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59700</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8374</td>\n",
       "      <td>0.0213</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.28460</td>\n",
       "      <td>166718</td>\n",
       "      <td>9980399</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:3:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8593</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:4:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1400</td>\n",
       "      <td>-0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:5:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1578</td>\n",
       "      <td>-0.0155</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.23500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Flipped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:6:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1:7:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1:8:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9866</td>\n",
       "      <td>-0.1439</td>\n",
       "      <td>0.0748</td>\n",
       "      <td>0.05451</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1:9:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1327</td>\n",
       "      <td>0.0035</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.80870</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1:10:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8522</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0.0157</td>\n",
       "      <td>0.05421</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1:11:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0.1858</td>\n",
       "      <td>0.1546</td>\n",
       "      <td>0.22930</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1:12:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.1430</td>\n",
       "      <td>0.0747</td>\n",
       "      <td>0.05575</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1:13:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.7445</td>\n",
       "      <td>-0.0135</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>0.32450</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1:14:C:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4544</td>\n",
       "      <td>-0.0170</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.21910</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1:15:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8516</td>\n",
       "      <td>0.0292</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.06173</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1:17:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7428</td>\n",
       "      <td>-0.0119</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.36210</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1:25:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8220</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.0139</td>\n",
       "      <td>0.15350</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>POS float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1:864615:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864615</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8725</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>0.0158</td>\n",
       "      <td>0.08092</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Lowercase alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1:864757:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864757</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1758</td>\n",
       "      <td>-0.0164</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.31990</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960399</td>\n",
       "      <td>Not normalizated allelels</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1:16:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>16</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1642</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.49750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Sex chromosomes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNPID  CHR     POS   EA NEA     EAF    BETA      SE        P  \\\n",
       "0        1:1:G:A    1       1    G   A  0.9960 -0.0737  0.1394  0.59700   \n",
       "1            1:2    1       2  TAA   T  0.8374  0.0213  0.0199  0.28460   \n",
       "2        1:3:T:A    1       3    T   A  0.8593  0.0172  0.0156  0.27050   \n",
       "3        1:4:T:C    1       4    C   T  0.1400 -0.0172  0.0156  0.27050   \n",
       "4        1:5:G:A    1       5    G   A  0.1578 -0.0155  0.0131  0.23500   \n",
       "5        1:6:A:G    1       6    G   A  0.7493  0.0204  0.0147  0.16500   \n",
       "6        1:7:A:G    1       7    G   A  0.7493  0.0204  0.0147  0.16500   \n",
       "7        1:8:G:C    1       8    G   C  0.9866 -0.1439  0.0748  0.05451   \n",
       "8        1:9:A:G    1       9    G   A  0.1327  0.0035  0.0146  0.80870   \n",
       "9       1:10:G:A    1      10    G   A  0.8522  0.0302  0.0157  0.05421   \n",
       "10      1:11:T:C    1      11    C   T  0.0027  0.1858  0.1546  0.22930   \n",
       "11      1:12:A:G    1      12    G   A  0.0131  0.1430  0.0747  0.05575   \n",
       "12      1:13:C:T    1      13    C   T  0.7445 -0.0135  0.0137  0.32450   \n",
       "13      1:14:C:A    1      14    C   A  0.4544 -0.0170  0.0138  0.21910   \n",
       "14      1:15:C:T    1      15    C   T  0.8516  0.0292  0.0156  0.06173   \n",
       "15      1:17:G:A    1      17    G   A  0.7428 -0.0119  0.0131  0.36210   \n",
       "16      1:18:A:G    1      20    G   A  0.7316  0.0013  0.0147  0.93200   \n",
       "17      1:25:A:G    1      25    G   A  0.8220  0.0199  0.0139  0.15350   \n",
       "18  1:864615:G:A    1  864615    G   A  0.8725  0.0275  0.0158  0.08092   \n",
       "19  1:864757:G:A    1  864757    G   A  0.1758 -0.0164  0.0165  0.31990   \n",
       "20      1:16:A:G   23      16    G   A  0.1642  0.0099  0.0146  0.49750   \n",
       "\n",
       "         N   STATUS                       Note  \n",
       "0   191764  9960099                 Duplicated  \n",
       "1   166718  9980399               Multiallelic  \n",
       "2   166718  9960099               Multiallelic  \n",
       "3   191764  9960099                        NaN  \n",
       "4   166718  9960099                    Flipped  \n",
       "5   166718  9960099                        NaN  \n",
       "6   166718  9960099                        NaN  \n",
       "7   166718  9960099                        NaN  \n",
       "8   166718  9960099                        NaN  \n",
       "9   166718  9960099                        NaN  \n",
       "10  166718  9960099                        NaN  \n",
       "11  166718  9960099                        NaN  \n",
       "12  191764  9960099                        NaN  \n",
       "13  166718  9960099                        NaN  \n",
       "14  166718  9960099                        NaN  \n",
       "15  166718  9960099            CHR with prefix  \n",
       "16  191764  9960099                        NaN  \n",
       "17  166718  9960099                  POS float  \n",
       "18  166718  9960099          Lowercase alleles  \n",
       "19  166718  9960399  Not normalizated allelels  \n",
       "20  166718  9960099            Sex chromosomes  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27c29690",
   "metadata": {},
   "source": [
    "## sort genome coordinate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "91ee7b0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 23 18:41:56 2024 Start to sort the genome coordinates...v3.4.36\n",
      "Tue Jan 23 18:41:56 2024  -Current Dataframe shape : 21  x  12\n",
      "Tue Jan 23 18:41:56 2024  -Sorting genome coordinates...\n",
      "Tue Jan 23 18:41:56 2024 Finished sorting genome coordinates successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.sort_coordinate()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a631ece",
   "metadata": {},
   "source": [
    "## sort column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "61d8d9c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 23 18:41:57 2024 Start to reorder the columns...v3.4.36\n",
      "Tue Jan 23 18:41:57 2024  -Current Dataframe shape : 21  x  12\n",
      "Tue Jan 23 18:41:57 2024  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,EAF,BETA,SE,P,N,STATUS,Note\n",
      "Tue Jan 23 18:41:57 2024 Finished sorting columns successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.sort_column()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed14018e",
   "metadata": {},
   "source": [
    "# All in one function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "45ad8af1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mon Oct 31 12:42:47 2022 Start to check IDs...\n",
      "Mon Oct 31 12:42:47 2022  -Current Dataframe shape : 12557759  x  12\n",
      "Mon Oct 31 12:42:47 2022  -Checking if SNPID is chr:pos:ref:alt...(separator: - ,: , _)\n",
      "Mon Oct 31 12:43:06 2022 Finished checking IDs successfully!\n",
      "Mon Oct 31 12:43:06 2022 Start to fix chromosome notation...\n",
      "Mon Oct 31 12:43:06 2022  -Current Dataframe shape : 12557759  x  12\n",
      "Mon Oct 31 12:43:58 2022  -All CHR are already fixed...\n",
      "Mon Oct 31 12:44:30 2022 Finished fixing chromosome notation successfully!\n",
      "Mon Oct 31 12:44:30 2022 Start to fix basepair positions...\n",
      "Mon Oct 31 12:44:30 2022  -Current Dataframe shape : 12557759  x  12\n",
      "Mon Oct 31 12:44:55 2022  -Position upper_bound is: 250,000,000\n",
      "Mon Oct 31 12:45:29 2022  -Remove outliers: 0\n",
      "Mon Oct 31 12:45:40 2022  -Converted all position to datatype Int64.\n",
      "Mon Oct 31 12:45:40 2022 Finished fixing basepair position successfully!\n",
      "Mon Oct 31 12:45:40 2022 Start to fix alleles...\n",
      "Mon Oct 31 12:45:40 2022  -Current Dataframe shape : 12557759  x  12\n",
      "Mon Oct 31 12:45:57 2022  -Converted all bases to string datatype and UPPERCASE.\n",
      "Mon Oct 31 12:46:25 2022 Finished fixing allele successfully!\n",
      "Mon Oct 31 12:46:25 2022 Start sanity check for statistics ...\n",
      "Mon Oct 31 12:46:25 2022  -Current Dataframe shape : 12557759  x  12\n",
      "Mon Oct 31 12:46:31 2022  -Checking if  0 <=N<= inf  ...\n",
      "Mon Oct 31 12:46:37 2022  -Removed 0 variants with bad N.\n",
      "Mon Oct 31 12:46:37 2022  -Checking if  0 <=EAF<= 1  ...\n",
      "Mon Oct 31 12:46:40 2022  -Removed 0 variants with bad EAF.\n",
      "Mon Oct 31 12:46:40 2022  -Checking if  5 <=MAC<= inf  ...\n",
      "Mon Oct 31 12:46:44 2022  -Removed 0 variants with bad MAC.\n",
      "Mon Oct 31 12:46:44 2022  -Checking if  5e-300 <= P <= 1  ...\n",
      "Mon Oct 31 12:46:46 2022  -Removed 0 variants with bad P.\n",
      "Mon Oct 31 12:46:46 2022  -Checking if  -10 <BETA)< 10  ...\n",
      "Mon Oct 31 12:46:47 2022  -Removed 0 variants with bad BETA.\n",
      "Mon Oct 31 12:46:47 2022  -Checking if  0 <SE< inf  ...\n",
      "Mon Oct 31 12:46:49 2022  -Removed 0 variants with bad SE.\n",
      "Mon Oct 31 12:46:49 2022  -Checking STATUS...\n",
      "Mon Oct 31 12:46:51 2022  -Coverting STAUTUS to category.\n",
      "Mon Oct 31 12:46:53 2022  -Removed 0 variants with bad statistics in total.\n",
      "Mon Oct 31 12:46:53 2022 Finished sanity check successfully!\n",
      "Mon Oct 31 12:46:55 2022 Start to normalize variants...\n",
      "Mon Oct 31 12:46:55 2022  -Current Dataframe shape : 12557759  x  12\n",
      "Mon Oct 31 12:46:56 2022  -No available variants to normalize..\n",
      "Mon Oct 31 12:46:56 2022 Finished normalizing variants successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.basic_check()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
