{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1ce41337",
   "metadata": {},
   "source": [
    "# Standardization Workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0242fb7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0,\"/home/yunye/work/gwaslab/src/\")\n",
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9553c53",
   "metadata": {},
   "source": [
    "# Load sample data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "73f7b157",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 23 18:39:44 2024 GWASLab v3.4.36 https://cloufield.github.io/gwaslab/\n",
      "Tue Jan 23 18:39:44 2024 (C) 2022-2024, Yunye He, Kamatani Lab, MIT License, gwaslab@gmail.com\n",
      "Tue Jan 23 18:39:44 2024 Start to initiate from file :/home/yunye/work/gwaslab/examples/toy_data/dirty_sumstats.tsv\n",
      "Tue Jan 23 18:39:44 2024  -Reading columns          : BETA,Note,REF,N,Frq,P,ALT,POS,CHR,SNP,SE\n",
      "Tue Jan 23 18:39:44 2024  -Renaming columns to      : BETA,Note,NEA,N,EAF,P,EA,POS,CHR,SNPID,SE\n",
      "Tue Jan 23 18:39:44 2024  -Current Dataframe shape : 41  x  11\n",
      "Tue Jan 23 18:39:44 2024  -Initiating a status column: STATUS ...\n",
      "Tue Jan 23 18:39:44 2024  -Version of genomic coordinates are unknown...\n",
      "Tue Jan 23 18:39:44 2024  -NEAF is specified...\n",
      "Tue Jan 23 18:39:44 2024  -Checking if 0<= NEAF <=1 ...\n",
      "Tue Jan 23 18:39:44 2024  -Converted NEAF to EAF.\n",
      "Tue Jan 23 18:39:44 2024  -Removed 2 variants with bad NEAF.\n",
      "Tue Jan 23 18:39:44 2024 Start to reorder the columns...v3.4.36\n",
      "Tue Jan 23 18:39:44 2024  -Current Dataframe shape : 39  x  12\n",
      "Tue Jan 23 18:39:44 2024  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,EAF,BETA,SE,P,N,STATUS,Note\n",
      "Tue Jan 23 18:39:44 2024 Finished sorting columns successfully!\n",
      "Tue Jan 23 18:39:44 2024  -Column  : SNPID  CHR    POS     EA       NEA      EAF     BETA    SE      P       N     STATUS   Note  \n",
      "Tue Jan 23 18:39:44 2024  -DType   : object string float64 category category float64 float64 float64 float64 int64 category object\n",
      "Tue Jan 23 18:39:44 2024  -Verified: T      F      F       T        T        T       T       T       T       T     T        NA    \n",
      "Tue Jan 23 18:39:44 2024  -WARNING! Columns with possibly incompatable dtypes: CHR,POS\n",
      "Tue Jan 23 18:39:44 2024 Finished loading data successfully!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/yunye/work/gwaslab/src/gwaslab/io_preformat_input.py:453: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sumstats.loc[:,\"EAF\"] = 1- sumstats.loc[:,\"EAF\"]\n"
     ]
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(\"/home/yunye/work/gwaslab/examples/toy_data/dirty_sumstats.tsv\",\n",
    "             snpid=\"SNP\",\n",
    "             chrom=\"CHR\",\n",
    "             pos=\"POS\",\n",
    "             ea=\"ALT\",\n",
    "             nea=\"REF\",\n",
    "             neaf=\"Frq\",\n",
    "             beta=\"BETA\",\n",
    "             se=\"SE\",\n",
    "             p=\"P\",\n",
    "             n=\"N\",\n",
    "             other=[\"Note\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1a6a7ace",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:1_G_A</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9960</td>\n",
       "      <td>-0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59700</td>\n",
       "      <td>191764</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:1_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59730</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:1_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2.000000e+00</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.0490</td>\n",
       "      <td>0.1231</td>\n",
       "      <td>0.69080</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2.000000e+00</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8374</td>\n",
       "      <td>0.0213</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.28460</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:3_T_A</td>\n",
       "      <td>1</td>\n",
       "      <td>3.000000e+00</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8593</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1:3_T_GA</td>\n",
       "      <td>1</td>\n",
       "      <td>3.000000e+00</td>\n",
       "      <td>GA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1470</td>\n",
       "      <td>-0.0187</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0.28300</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1:4_T_C</td>\n",
       "      <td>1</td>\n",
       "      <td>4.000000e+00</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1400</td>\n",
       "      <td>-0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>191764</td>\n",
       "      <td>9999999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1:5_G_A</td>\n",
       "      <td>1</td>\n",
       "      <td>5.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1578</td>\n",
       "      <td>-0.0155</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.23500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Flipped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1:6_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>6.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1:7_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>7.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1:8_G_C</td>\n",
       "      <td>1</td>\n",
       "      <td>8.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9866</td>\n",
       "      <td>-0.1439</td>\n",
       "      <td>0.0748</td>\n",
       "      <td>0.05451</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1:9_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>9.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1327</td>\n",
       "      <td>0.0035</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.80870</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1:10_G_A</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8522</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0.0157</td>\n",
       "      <td>0.05421</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1:11_T_C</td>\n",
       "      <td>1</td>\n",
       "      <td>1.100000e+01</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0.1858</td>\n",
       "      <td>0.1546</td>\n",
       "      <td>0.22930</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1:12_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>1.200000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.1430</td>\n",
       "      <td>0.0747</td>\n",
       "      <td>0.05575</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1:13_C_T</td>\n",
       "      <td>1</td>\n",
       "      <td>1.300000e+01</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.7445</td>\n",
       "      <td>-0.0135</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>0.32450</td>\n",
       "      <td>191764</td>\n",
       "      <td>9999999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1:14_C_A</td>\n",
       "      <td>1</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4544</td>\n",
       "      <td>-0.0170</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.21910</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1:15_C_T</td>\n",
       "      <td>1</td>\n",
       "      <td>1.500000e+01</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8516</td>\n",
       "      <td>0.0292</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.06173</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1:16_A_G</td>\n",
       "      <td>X</td>\n",
       "      <td>1.600000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1642</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.49750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Sex chromosomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1:17_G_A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1.700000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7428</td>\n",
       "      <td>-0.0119</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.36210</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1:1_G_A</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>1.800000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8699</td>\n",
       "      <td>-0.0038</td>\n",
       "      <td>0.0161</td>\n",
       "      <td>0.81540</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1:19:T:C</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.900000e+01</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0361</td>\n",
       "      <td>0.0670</td>\n",
       "      <td>0.0453</td>\n",
       "      <td>0.13880</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>CHR &lt; 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1:18_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>2.000000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>191764</td>\n",
       "      <td>9999999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1:18_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>2.100000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1:18_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>2.200000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1:18_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>1.232140e+13</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>POS out of normal range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1:25:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>2.500000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8220</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.0139</td>\n",
       "      <td>0.15350</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>POS float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1:864002_G_C</td>\n",
       "      <td>1</td>\n",
       "      <td>8.640020e+05</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8666</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0155</td>\n",
       "      <td>0.93090</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Same alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1:864615_G_A</td>\n",
       "      <td>1</td>\n",
       "      <td>8.646150e+05</td>\n",
       "      <td>g</td>\n",
       "      <td>a</td>\n",
       "      <td>0.8725</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>0.0158</td>\n",
       "      <td>0.08092</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Lowercase alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1:864701_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>8.647010e+05</td>\n",
       "      <td>&lt;CN1&gt;</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>0.1306</td>\n",
       "      <td>0.0680</td>\n",
       "      <td>0.05466</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Unrecognized alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1:864755_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>8.647550e+05</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8242</td>\n",
       "      <td>0.0163</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.32220</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Unrecognized alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1:864757_G_A</td>\n",
       "      <td>1</td>\n",
       "      <td>8.647570e+05</td>\n",
       "      <td>GT</td>\n",
       "      <td>AT</td>\n",
       "      <td>0.1758</td>\n",
       "      <td>-0.0164</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.31990</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Not normalizated allelels</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1:865219_G_A</td>\n",
       "      <td>1</td>\n",
       "      <td>8.652190e+05</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8685</td>\n",
       "      <td>300.0000</td>\n",
       "      <td>0.0161</td>\n",
       "      <td>0.79480</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>BETA out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1:865367_G_T</td>\n",
       "      <td>1</td>\n",
       "      <td>8.653670e+05</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8486</td>\n",
       "      <td>-300.0000</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.05270</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>BETA out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1:865438_C_T</td>\n",
       "      <td>1</td>\n",
       "      <td>8.654380e+05</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.9849</td>\n",
       "      <td>-0.1299</td>\n",
       "      <td>-0.0680</td>\n",
       "      <td>0.05601</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>SE out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>1:865694_C_T</td>\n",
       "      <td>1</td>\n",
       "      <td>8.656940e+05</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8489</td>\n",
       "      <td>0.0265</td>\n",
       "      <td>0.0132</td>\n",
       "      <td>-0.04424</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>P out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>1:866893_T_C</td>\n",
       "      <td>1</td>\n",
       "      <td>8.668930e+05</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.3104</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>1.67220</td>\n",
       "      <td>166718</td>\n",
       "      <td>9999999</td>\n",
       "      <td>P out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>1:866938_G_A</td>\n",
       "      <td>1</td>\n",
       "      <td>8.669380e+05</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.6028</td>\n",
       "      <td>-0.0104</td>\n",
       "      <td>0.0121</td>\n",
       "      <td>0.38960</td>\n",
       "      <td>-1</td>\n",
       "      <td>9999999</td>\n",
       "      <td>N out of range</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNPID   CHR           POS     EA NEA     EAF      BETA      SE  \\\n",
       "0        1:1_G_A     1  1.000000e+00      G   A  0.9960   -0.0737  0.1394   \n",
       "1        1:1_A_G     1  1.000000e+00      G   A  0.0040    0.0737  0.1394   \n",
       "2        1:1_A_G     1  1.000000e+00      G   A  0.0040    0.0737  0.1394   \n",
       "3            1:2     1  2.000000e+00      C   T  0.0051    0.0490  0.1231   \n",
       "4            1:2     1  2.000000e+00    TAA   T  0.8374    0.0213  0.0199   \n",
       "5        1:3_T_A     1  3.000000e+00      T   A  0.8593    0.0172  0.0156   \n",
       "6       1:3_T_GA     1  3.000000e+00     GA   T  0.1470   -0.0187  0.0174   \n",
       "7        1:4_T_C     1  4.000000e+00      C   T  0.1400   -0.0172  0.0156   \n",
       "8        1:5_G_A     1  5.000000e+00      G   A  0.1578   -0.0155  0.0131   \n",
       "9        1:6_A_G     1  6.000000e+00      G   A  0.7493    0.0204  0.0147   \n",
       "10       1:7_A_G     1  7.000000e+00      G   A  0.7493    0.0204  0.0147   \n",
       "11       1:8_G_C     1  8.000000e+00      G   C  0.9866   -0.1439  0.0748   \n",
       "12       1:9_A_G     1  9.000000e+00      G   A  0.1327    0.0035  0.0146   \n",
       "13      1:10_G_A     1  1.000000e+01      G   A  0.8522    0.0302  0.0157   \n",
       "14      1:11_T_C     1  1.100000e+01      C   T  0.0027    0.1858  0.1546   \n",
       "15      1:12_A_G     1  1.200000e+01      G   A  0.0131    0.1430  0.0747   \n",
       "16      1:13_C_T     1  1.300000e+01      C   T  0.7445   -0.0135  0.0137   \n",
       "17      1:14_C_A     1  1.400000e+01      C   A  0.4544   -0.0170  0.0138   \n",
       "18      1:15_C_T     1  1.500000e+01      C   T  0.8516    0.0292  0.0156   \n",
       "19      1:16_A_G     X  1.600000e+01      G   A  0.1642    0.0099  0.0146   \n",
       "20      1:17_G_A  chr1  1.700000e+01      G   A  0.7428   -0.0119  0.0131   \n",
       "21       1:1_G_A  Chr1  1.800000e+01      G   A  0.8699   -0.0038  0.0161   \n",
       "22      1:19:T:C    -1  1.900000e+01      C   T  0.0361    0.0670  0.0453   \n",
       "23      1:18_A_G     1  2.000000e+01      G   A  0.7316    0.0013  0.0147   \n",
       "24      1:18_A_G     1  2.100000e+01      G   A  0.7316    0.0013  0.0147   \n",
       "25      1:18_A_G     1  2.200000e+01      G   A  0.7316    0.0013  0.0147   \n",
       "26      1:18_A_G     1  1.232140e+13      G   A  0.7316    0.0013  0.0147   \n",
       "27      1:25:A:G     1  2.500000e+01      G   A  0.8220    0.0199  0.0139   \n",
       "28  1:864002_G_C     1  8.640020e+05      A   A  0.8666    0.0013  0.0155   \n",
       "29  1:864615_G_A     1  8.646150e+05      g   a  0.8725    0.0275  0.0158   \n",
       "30  1:864701_A_G     1  8.647010e+05  <CN1>   A  0.0150    0.1306  0.0680   \n",
       "31  1:864755_A_G     1  8.647550e+05      N   A  0.8242    0.0163  0.0165   \n",
       "32  1:864757_G_A     1  8.647570e+05     GT  AT  0.1758   -0.0164  0.0165   \n",
       "35  1:865219_G_A     1  8.652190e+05      G   A  0.8685  300.0000  0.0161   \n",
       "36  1:865367_G_T     1  8.653670e+05      G   T  0.8486 -300.0000  0.0138   \n",
       "37  1:865438_C_T     1  8.654380e+05      C   T  0.9849   -0.1299 -0.0680   \n",
       "38  1:865694_C_T     1  8.656940e+05      C   T  0.8489    0.0265  0.0132   \n",
       "39  1:866893_T_C     1  8.668930e+05      C   T  0.3104    0.0058  0.0137   \n",
       "40  1:866938_G_A     1  8.669380e+05      G   A  0.6028   -0.0104  0.0121   \n",
       "\n",
       "          P       N   STATUS                       Note  \n",
       "0   0.59700  191764  9999999                 Duplicated  \n",
       "1   0.59730  166718  9999999                 Duplicated  \n",
       "2   0.59750  166718  9999999                 Duplicated  \n",
       "3   0.69080  166718  9999999               Multiallelic  \n",
       "4   0.28460  166718  9999999               Multiallelic  \n",
       "5   0.27050  166718  9999999               Multiallelic  \n",
       "6   0.28300  166718  9999999               Multiallelic  \n",
       "7   0.27050  191764  9999999                        NaN  \n",
       "8   0.23500  166718  9999999                    Flipped  \n",
       "9   0.16500  166718  9999999                        NaN  \n",
       "10  0.16500  166718  9999999                        NaN  \n",
       "11  0.05451  166718  9999999                        NaN  \n",
       "12  0.80870  166718  9999999                        NaN  \n",
       "13  0.05421  166718  9999999                        NaN  \n",
       "14  0.22930  166718  9999999                        NaN  \n",
       "15  0.05575  166718  9999999                        NaN  \n",
       "16  0.32450  191764  9999999                        NaN  \n",
       "17  0.21910  166718  9999999                        NaN  \n",
       "18  0.06173  166718  9999999                        NaN  \n",
       "19  0.49750  166718  9999999            Sex chromosomes  \n",
       "20  0.36210  166718  9999999            CHR with prefix  \n",
       "21  0.81540  166718  9999999            CHR with prefix  \n",
       "22  0.13880  166718  9999999                    CHR < 0  \n",
       "23  0.93200  191764  9999999                        NaN  \n",
       "24  0.93200  166718  9999999                        NaN  \n",
       "25  0.93200  166718  9999999                        NaN  \n",
       "26  0.93200  166718  9999999    POS out of normal range  \n",
       "27  0.15350  166718  9999999                  POS float  \n",
       "28  0.93090  166718  9999999               Same alleles  \n",
       "29  0.08092  166718  9999999          Lowercase alleles  \n",
       "30  0.05466  166718  9999999       Unrecognized alleles  \n",
       "31  0.32220  166718  9999999       Unrecognized alleles  \n",
       "32  0.31990  166718  9999999  Not normalizated allelels  \n",
       "35  0.79480  166718  9999999          BETA out of range  \n",
       "36  0.05270  166718  9999999          BETA out of range  \n",
       "37  0.05601  166718  9999999            SE out of range  \n",
       "38 -0.04424  166718  9999999             P out of range  \n",
       "39  1.67220  166718  9999999             P out of range  \n",
       "40  0.38960      -1  9999999             N out of range  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2f2e58b",
   "metadata": {},
   "source": [
    "## fix id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "54c21788",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 23 18:39:45 2024 Start to check IDs...v3.4.36\n",
      "Tue Jan 23 18:39:45 2024  -Current Dataframe shape : 39  x  12\n",
      "Tue Jan 23 18:39:45 2024  -Checking if SNPID is chr:pos:ref:alt...(separator: - ,: , _)\n",
      "Tue Jan 23 18:39:46 2024  -Replacing [_-] in SNPID with \":\" ...\n",
      "Tue Jan 23 18:39:46 2024 Finished checking IDs successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.fix_id(fixsep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "90ca2f0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9960</td>\n",
       "      <td>-0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59700</td>\n",
       "      <td>191764</td>\n",
       "      <td>9969999</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59730</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2.000000e+00</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.0490</td>\n",
       "      <td>0.1231</td>\n",
       "      <td>0.69080</td>\n",
       "      <td>166718</td>\n",
       "      <td>9989999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2.000000e+00</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8374</td>\n",
       "      <td>0.0213</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.28460</td>\n",
       "      <td>166718</td>\n",
       "      <td>9989999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:3:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>3.000000e+00</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8593</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1:3:T:GA</td>\n",
       "      <td>1</td>\n",
       "      <td>3.000000e+00</td>\n",
       "      <td>GA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1470</td>\n",
       "      <td>-0.0187</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0.28300</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1:4:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>4.000000e+00</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1400</td>\n",
       "      <td>-0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>191764</td>\n",
       "      <td>9969999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1:5:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>5.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1578</td>\n",
       "      <td>-0.0155</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.23500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>Flipped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1:6:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>6.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1:7:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>7.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1:8:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>8.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9866</td>\n",
       "      <td>-0.1439</td>\n",
       "      <td>0.0748</td>\n",
       "      <td>0.05451</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1:9:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>9.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1327</td>\n",
       "      <td>0.0035</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.80870</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1:10:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8522</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0.0157</td>\n",
       "      <td>0.05421</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1:11:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>1.100000e+01</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0.1858</td>\n",
       "      <td>0.1546</td>\n",
       "      <td>0.22930</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1:12:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1.200000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.1430</td>\n",
       "      <td>0.0747</td>\n",
       "      <td>0.05575</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1:13:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>1.300000e+01</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.7445</td>\n",
       "      <td>-0.0135</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>0.32450</td>\n",
       "      <td>191764</td>\n",
       "      <td>9969999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1:14:C:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4544</td>\n",
       "      <td>-0.0170</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.21910</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1:15:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>1.500000e+01</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8516</td>\n",
       "      <td>0.0292</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.06173</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1:16:A:G</td>\n",
       "      <td>X</td>\n",
       "      <td>1.600000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1642</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.49750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>Sex chromosomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1:17:G:A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1.700000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7428</td>\n",
       "      <td>-0.0119</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.36210</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>1.800000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8699</td>\n",
       "      <td>-0.0038</td>\n",
       "      <td>0.0161</td>\n",
       "      <td>0.81540</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1:19:T:C</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.900000e+01</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0361</td>\n",
       "      <td>0.0670</td>\n",
       "      <td>0.0453</td>\n",
       "      <td>0.13880</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>CHR &lt; 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>2.000000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>191764</td>\n",
       "      <td>9969999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>2.100000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>2.200000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1.232140e+13</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>POS out of normal range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1:25:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>2.500000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8220</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.0139</td>\n",
       "      <td>0.15350</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>POS float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1:864002:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>8.640020e+05</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8666</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0155</td>\n",
       "      <td>0.93090</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>Same alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1:864615:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>8.646150e+05</td>\n",
       "      <td>g</td>\n",
       "      <td>a</td>\n",
       "      <td>0.8725</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>0.0158</td>\n",
       "      <td>0.08092</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>Lowercase alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1:864701:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>8.647010e+05</td>\n",
       "      <td>&lt;CN1&gt;</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>0.1306</td>\n",
       "      <td>0.0680</td>\n",
       "      <td>0.05466</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>Unrecognized alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1:864755:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>8.647550e+05</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8242</td>\n",
       "      <td>0.0163</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.32220</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>Unrecognized alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1:864757:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>8.647570e+05</td>\n",
       "      <td>GT</td>\n",
       "      <td>AT</td>\n",
       "      <td>0.1758</td>\n",
       "      <td>-0.0164</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.31990</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>Not normalizated allelels</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1:865219:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>8.652190e+05</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8685</td>\n",
       "      <td>300.0000</td>\n",
       "      <td>0.0161</td>\n",
       "      <td>0.79480</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>BETA out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1:865367:G:T</td>\n",
       "      <td>1</td>\n",
       "      <td>8.653670e+05</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8486</td>\n",
       "      <td>-300.0000</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.05270</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>BETA out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1:865438:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>8.654380e+05</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.9849</td>\n",
       "      <td>-0.1299</td>\n",
       "      <td>-0.0680</td>\n",
       "      <td>0.05601</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>SE out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>1:865694:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>8.656940e+05</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8489</td>\n",
       "      <td>0.0265</td>\n",
       "      <td>0.0132</td>\n",
       "      <td>-0.04424</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>P out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>1:866893:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>8.668930e+05</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.3104</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>1.67220</td>\n",
       "      <td>166718</td>\n",
       "      <td>9969999</td>\n",
       "      <td>P out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>1:866938:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>8.669380e+05</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.6028</td>\n",
       "      <td>-0.0104</td>\n",
       "      <td>0.0121</td>\n",
       "      <td>0.38960</td>\n",
       "      <td>-1</td>\n",
       "      <td>9969999</td>\n",
       "      <td>N out of range</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNPID   CHR           POS     EA NEA     EAF      BETA      SE  \\\n",
       "0        1:1:G:A     1  1.000000e+00      G   A  0.9960   -0.0737  0.1394   \n",
       "1        1:1:A:G     1  1.000000e+00      G   A  0.0040    0.0737  0.1394   \n",
       "2        1:1:A:G     1  1.000000e+00      G   A  0.0040    0.0737  0.1394   \n",
       "3            1:2     1  2.000000e+00      C   T  0.0051    0.0490  0.1231   \n",
       "4            1:2     1  2.000000e+00    TAA   T  0.8374    0.0213  0.0199   \n",
       "5        1:3:T:A     1  3.000000e+00      T   A  0.8593    0.0172  0.0156   \n",
       "6       1:3:T:GA     1  3.000000e+00     GA   T  0.1470   -0.0187  0.0174   \n",
       "7        1:4:T:C     1  4.000000e+00      C   T  0.1400   -0.0172  0.0156   \n",
       "8        1:5:G:A     1  5.000000e+00      G   A  0.1578   -0.0155  0.0131   \n",
       "9        1:6:A:G     1  6.000000e+00      G   A  0.7493    0.0204  0.0147   \n",
       "10       1:7:A:G     1  7.000000e+00      G   A  0.7493    0.0204  0.0147   \n",
       "11       1:8:G:C     1  8.000000e+00      G   C  0.9866   -0.1439  0.0748   \n",
       "12       1:9:A:G     1  9.000000e+00      G   A  0.1327    0.0035  0.0146   \n",
       "13      1:10:G:A     1  1.000000e+01      G   A  0.8522    0.0302  0.0157   \n",
       "14      1:11:T:C     1  1.100000e+01      C   T  0.0027    0.1858  0.1546   \n",
       "15      1:12:A:G     1  1.200000e+01      G   A  0.0131    0.1430  0.0747   \n",
       "16      1:13:C:T     1  1.300000e+01      C   T  0.7445   -0.0135  0.0137   \n",
       "17      1:14:C:A     1  1.400000e+01      C   A  0.4544   -0.0170  0.0138   \n",
       "18      1:15:C:T     1  1.500000e+01      C   T  0.8516    0.0292  0.0156   \n",
       "19      1:16:A:G     X  1.600000e+01      G   A  0.1642    0.0099  0.0146   \n",
       "20      1:17:G:A  chr1  1.700000e+01      G   A  0.7428   -0.0119  0.0131   \n",
       "21       1:1:G:A  Chr1  1.800000e+01      G   A  0.8699   -0.0038  0.0161   \n",
       "22      1:19:T:C    -1  1.900000e+01      C   T  0.0361    0.0670  0.0453   \n",
       "23      1:18:A:G     1  2.000000e+01      G   A  0.7316    0.0013  0.0147   \n",
       "24      1:18:A:G     1  2.100000e+01      G   A  0.7316    0.0013  0.0147   \n",
       "25      1:18:A:G     1  2.200000e+01      G   A  0.7316    0.0013  0.0147   \n",
       "26      1:18:A:G     1  1.232140e+13      G   A  0.7316    0.0013  0.0147   \n",
       "27      1:25:A:G     1  2.500000e+01      G   A  0.8220    0.0199  0.0139   \n",
       "28  1:864002:G:C     1  8.640020e+05      A   A  0.8666    0.0013  0.0155   \n",
       "29  1:864615:G:A     1  8.646150e+05      g   a  0.8725    0.0275  0.0158   \n",
       "30  1:864701:A:G     1  8.647010e+05  <CN1>   A  0.0150    0.1306  0.0680   \n",
       "31  1:864755:A:G     1  8.647550e+05      N   A  0.8242    0.0163  0.0165   \n",
       "32  1:864757:G:A     1  8.647570e+05     GT  AT  0.1758   -0.0164  0.0165   \n",
       "35  1:865219:G:A     1  8.652190e+05      G   A  0.8685  300.0000  0.0161   \n",
       "36  1:865367:G:T     1  8.653670e+05      G   T  0.8486 -300.0000  0.0138   \n",
       "37  1:865438:C:T     1  8.654380e+05      C   T  0.9849   -0.1299 -0.0680   \n",
       "38  1:865694:C:T     1  8.656940e+05      C   T  0.8489    0.0265  0.0132   \n",
       "39  1:866893:T:C     1  8.668930e+05      C   T  0.3104    0.0058  0.0137   \n",
       "40  1:866938:G:A     1  8.669380e+05      G   A  0.6028   -0.0104  0.0121   \n",
       "\n",
       "          P       N   STATUS                       Note  \n",
       "0   0.59700  191764  9969999                 Duplicated  \n",
       "1   0.59730  166718  9969999                 Duplicated  \n",
       "2   0.59750  166718  9969999                 Duplicated  \n",
       "3   0.69080  166718  9989999               Multiallelic  \n",
       "4   0.28460  166718  9989999               Multiallelic  \n",
       "5   0.27050  166718  9969999               Multiallelic  \n",
       "6   0.28300  166718  9969999               Multiallelic  \n",
       "7   0.27050  191764  9969999                        NaN  \n",
       "8   0.23500  166718  9969999                    Flipped  \n",
       "9   0.16500  166718  9969999                        NaN  \n",
       "10  0.16500  166718  9969999                        NaN  \n",
       "11  0.05451  166718  9969999                        NaN  \n",
       "12  0.80870  166718  9969999                        NaN  \n",
       "13  0.05421  166718  9969999                        NaN  \n",
       "14  0.22930  166718  9969999                        NaN  \n",
       "15  0.05575  166718  9969999                        NaN  \n",
       "16  0.32450  191764  9969999                        NaN  \n",
       "17  0.21910  166718  9969999                        NaN  \n",
       "18  0.06173  166718  9969999                        NaN  \n",
       "19  0.49750  166718  9969999            Sex chromosomes  \n",
       "20  0.36210  166718  9969999            CHR with prefix  \n",
       "21  0.81540  166718  9969999            CHR with prefix  \n",
       "22  0.13880  166718  9969999                    CHR < 0  \n",
       "23  0.93200  191764  9969999                        NaN  \n",
       "24  0.93200  166718  9969999                        NaN  \n",
       "25  0.93200  166718  9969999                        NaN  \n",
       "26  0.93200  166718  9969999    POS out of normal range  \n",
       "27  0.15350  166718  9969999                  POS float  \n",
       "28  0.93090  166718  9969999               Same alleles  \n",
       "29  0.08092  166718  9969999          Lowercase alleles  \n",
       "30  0.05466  166718  9969999       Unrecognized alleles  \n",
       "31  0.32220  166718  9969999       Unrecognized alleles  \n",
       "32  0.31990  166718  9969999  Not normalizated allelels  \n",
       "35  0.79480  166718  9969999          BETA out of range  \n",
       "36  0.05270  166718  9969999          BETA out of range  \n",
       "37  0.05601  166718  9969999            SE out of range  \n",
       "38 -0.04424  166718  9969999             P out of range  \n",
       "39  1.67220  166718  9969999             P out of range  \n",
       "40  0.38960      -1  9969999             N out of range  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea78e4dc",
   "metadata": {},
   "source": [
    "## fix chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e83d1fa5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 23 18:40:01 2024 Start to fix chromosome notation...v3.4.36\n",
      "Tue Jan 23 18:40:01 2024  -Current Dataframe shape : 39  x  12\n",
      "Tue Jan 23 18:40:01 2024  -Checking CHR data type...\n",
      "Tue Jan 23 18:40:01 2024  -Variants with standardized chromosome notation: 38\n",
      "Tue Jan 23 18:40:01 2024  -Variants with fixable chromosome notations: 0\n",
      "Tue Jan 23 18:40:01 2024  -Variants with NA chromosome notations: 1\n",
      "Tue Jan 23 18:40:01 2024  -No unrecognized chromosome notations...\n",
      "Tue Jan 23 18:40:03 2024  -Removed 1 variants with unrecognized chromosome notations.\n",
      "Tue Jan 23 18:40:03 2024  -Sanity check for CHR...\n",
      "Tue Jan 23 18:40:03 2024  -Removed 0 variants with CHR < 1...\n",
      "Tue Jan 23 18:40:03 2024 Finished fixing chromosome notation successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.fix_chr(remove=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0ebd3b0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9960</td>\n",
       "      <td>-0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59700</td>\n",
       "      <td>191764</td>\n",
       "      <td>9965999</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59730</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2.000000e+00</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.0490</td>\n",
       "      <td>0.1231</td>\n",
       "      <td>0.69080</td>\n",
       "      <td>166718</td>\n",
       "      <td>9985999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2.000000e+00</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8374</td>\n",
       "      <td>0.0213</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.28460</td>\n",
       "      <td>166718</td>\n",
       "      <td>9985999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:3:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>3.000000e+00</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8593</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1:3:T:GA</td>\n",
       "      <td>1</td>\n",
       "      <td>3.000000e+00</td>\n",
       "      <td>GA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1470</td>\n",
       "      <td>-0.0187</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0.28300</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1:4:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>4.000000e+00</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1400</td>\n",
       "      <td>-0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>191764</td>\n",
       "      <td>9965999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1:5:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>5.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1578</td>\n",
       "      <td>-0.0155</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.23500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>Flipped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1:6:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>6.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1:7:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>7.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1:8:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>8.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9866</td>\n",
       "      <td>-0.1439</td>\n",
       "      <td>0.0748</td>\n",
       "      <td>0.05451</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1:9:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>9.000000e+00</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1327</td>\n",
       "      <td>0.0035</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.80870</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1:10:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8522</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0.0157</td>\n",
       "      <td>0.05421</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1:11:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>1.100000e+01</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0.1858</td>\n",
       "      <td>0.1546</td>\n",
       "      <td>0.22930</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1:12:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1.200000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.1430</td>\n",
       "      <td>0.0747</td>\n",
       "      <td>0.05575</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1:13:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>1.300000e+01</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.7445</td>\n",
       "      <td>-0.0135</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>0.32450</td>\n",
       "      <td>191764</td>\n",
       "      <td>9965999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1:14:C:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4544</td>\n",
       "      <td>-0.0170</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.21910</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1:15:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>1.500000e+01</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8516</td>\n",
       "      <td>0.0292</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.06173</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1:16:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>1.600000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1642</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.49750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>Sex chromosomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1:17:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1.700000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7428</td>\n",
       "      <td>-0.0119</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.36210</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1.800000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8699</td>\n",
       "      <td>-0.0038</td>\n",
       "      <td>0.0161</td>\n",
       "      <td>0.81540</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>2.000000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>191764</td>\n",
       "      <td>9965999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>2.100000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>2.200000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1.232140e+13</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>POS out of normal range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1:25:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>2.500000e+01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8220</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.0139</td>\n",
       "      <td>0.15350</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>POS float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1:864002:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>8.640020e+05</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8666</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0155</td>\n",
       "      <td>0.93090</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>Same alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1:864615:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>8.646150e+05</td>\n",
       "      <td>g</td>\n",
       "      <td>a</td>\n",
       "      <td>0.8725</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>0.0158</td>\n",
       "      <td>0.08092</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>Lowercase alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1:864701:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>8.647010e+05</td>\n",
       "      <td>&lt;CN1&gt;</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>0.1306</td>\n",
       "      <td>0.0680</td>\n",
       "      <td>0.05466</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>Unrecognized alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1:864755:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>8.647550e+05</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8242</td>\n",
       "      <td>0.0163</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.32220</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>Unrecognized alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1:864757:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>8.647570e+05</td>\n",
       "      <td>GT</td>\n",
       "      <td>AT</td>\n",
       "      <td>0.1758</td>\n",
       "      <td>-0.0164</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.31990</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>Not normalizated allelels</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1:865219:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>8.652190e+05</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8685</td>\n",
       "      <td>300.0000</td>\n",
       "      <td>0.0161</td>\n",
       "      <td>0.79480</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>BETA out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1:865367:G:T</td>\n",
       "      <td>1</td>\n",
       "      <td>8.653670e+05</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8486</td>\n",
       "      <td>-300.0000</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.05270</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>BETA out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1:865438:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>8.654380e+05</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.9849</td>\n",
       "      <td>-0.1299</td>\n",
       "      <td>-0.0680</td>\n",
       "      <td>0.05601</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>SE out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>1:865694:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>8.656940e+05</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8489</td>\n",
       "      <td>0.0265</td>\n",
       "      <td>0.0132</td>\n",
       "      <td>-0.04424</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>P out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>1:866893:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>8.668930e+05</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.3104</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>1.67220</td>\n",
       "      <td>166718</td>\n",
       "      <td>9965999</td>\n",
       "      <td>P out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>1:866938:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>8.669380e+05</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.6028</td>\n",
       "      <td>-0.0104</td>\n",
       "      <td>0.0121</td>\n",
       "      <td>0.38960</td>\n",
       "      <td>-1</td>\n",
       "      <td>9965999</td>\n",
       "      <td>N out of range</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNPID  CHR           POS     EA NEA     EAF      BETA      SE  \\\n",
       "0        1:1:G:A    1  1.000000e+00      G   A  0.9960   -0.0737  0.1394   \n",
       "1        1:1:A:G    1  1.000000e+00      G   A  0.0040    0.0737  0.1394   \n",
       "2        1:1:A:G    1  1.000000e+00      G   A  0.0040    0.0737  0.1394   \n",
       "3            1:2    1  2.000000e+00      C   T  0.0051    0.0490  0.1231   \n",
       "4            1:2    1  2.000000e+00    TAA   T  0.8374    0.0213  0.0199   \n",
       "5        1:3:T:A    1  3.000000e+00      T   A  0.8593    0.0172  0.0156   \n",
       "6       1:3:T:GA    1  3.000000e+00     GA   T  0.1470   -0.0187  0.0174   \n",
       "7        1:4:T:C    1  4.000000e+00      C   T  0.1400   -0.0172  0.0156   \n",
       "8        1:5:G:A    1  5.000000e+00      G   A  0.1578   -0.0155  0.0131   \n",
       "9        1:6:A:G    1  6.000000e+00      G   A  0.7493    0.0204  0.0147   \n",
       "10       1:7:A:G    1  7.000000e+00      G   A  0.7493    0.0204  0.0147   \n",
       "11       1:8:G:C    1  8.000000e+00      G   C  0.9866   -0.1439  0.0748   \n",
       "12       1:9:A:G    1  9.000000e+00      G   A  0.1327    0.0035  0.0146   \n",
       "13      1:10:G:A    1  1.000000e+01      G   A  0.8522    0.0302  0.0157   \n",
       "14      1:11:T:C    1  1.100000e+01      C   T  0.0027    0.1858  0.1546   \n",
       "15      1:12:A:G    1  1.200000e+01      G   A  0.0131    0.1430  0.0747   \n",
       "16      1:13:C:T    1  1.300000e+01      C   T  0.7445   -0.0135  0.0137   \n",
       "17      1:14:C:A    1  1.400000e+01      C   A  0.4544   -0.0170  0.0138   \n",
       "18      1:15:C:T    1  1.500000e+01      C   T  0.8516    0.0292  0.0156   \n",
       "19      1:16:A:G   23  1.600000e+01      G   A  0.1642    0.0099  0.0146   \n",
       "20      1:17:G:A    1  1.700000e+01      G   A  0.7428   -0.0119  0.0131   \n",
       "21       1:1:G:A    1  1.800000e+01      G   A  0.8699   -0.0038  0.0161   \n",
       "23      1:18:A:G    1  2.000000e+01      G   A  0.7316    0.0013  0.0147   \n",
       "24      1:18:A:G    1  2.100000e+01      G   A  0.7316    0.0013  0.0147   \n",
       "25      1:18:A:G    1  2.200000e+01      G   A  0.7316    0.0013  0.0147   \n",
       "26      1:18:A:G    1  1.232140e+13      G   A  0.7316    0.0013  0.0147   \n",
       "27      1:25:A:G    1  2.500000e+01      G   A  0.8220    0.0199  0.0139   \n",
       "28  1:864002:G:C    1  8.640020e+05      A   A  0.8666    0.0013  0.0155   \n",
       "29  1:864615:G:A    1  8.646150e+05      g   a  0.8725    0.0275  0.0158   \n",
       "30  1:864701:A:G    1  8.647010e+05  <CN1>   A  0.0150    0.1306  0.0680   \n",
       "31  1:864755:A:G    1  8.647550e+05      N   A  0.8242    0.0163  0.0165   \n",
       "32  1:864757:G:A    1  8.647570e+05     GT  AT  0.1758   -0.0164  0.0165   \n",
       "35  1:865219:G:A    1  8.652190e+05      G   A  0.8685  300.0000  0.0161   \n",
       "36  1:865367:G:T    1  8.653670e+05      G   T  0.8486 -300.0000  0.0138   \n",
       "37  1:865438:C:T    1  8.654380e+05      C   T  0.9849   -0.1299 -0.0680   \n",
       "38  1:865694:C:T    1  8.656940e+05      C   T  0.8489    0.0265  0.0132   \n",
       "39  1:866893:T:C    1  8.668930e+05      C   T  0.3104    0.0058  0.0137   \n",
       "40  1:866938:G:A    1  8.669380e+05      G   A  0.6028   -0.0104  0.0121   \n",
       "\n",
       "          P       N   STATUS                       Note  \n",
       "0   0.59700  191764  9965999                 Duplicated  \n",
       "1   0.59730  166718  9965999                 Duplicated  \n",
       "2   0.59750  166718  9965999                 Duplicated  \n",
       "3   0.69080  166718  9985999               Multiallelic  \n",
       "4   0.28460  166718  9985999               Multiallelic  \n",
       "5   0.27050  166718  9965999               Multiallelic  \n",
       "6   0.28300  166718  9965999               Multiallelic  \n",
       "7   0.27050  191764  9965999                        NaN  \n",
       "8   0.23500  166718  9965999                    Flipped  \n",
       "9   0.16500  166718  9965999                        NaN  \n",
       "10  0.16500  166718  9965999                        NaN  \n",
       "11  0.05451  166718  9965999                        NaN  \n",
       "12  0.80870  166718  9965999                        NaN  \n",
       "13  0.05421  166718  9965999                        NaN  \n",
       "14  0.22930  166718  9965999                        NaN  \n",
       "15  0.05575  166718  9965999                        NaN  \n",
       "16  0.32450  191764  9965999                        NaN  \n",
       "17  0.21910  166718  9965999                        NaN  \n",
       "18  0.06173  166718  9965999                        NaN  \n",
       "19  0.49750  166718  9965999            Sex chromosomes  \n",
       "20  0.36210  166718  9965999            CHR with prefix  \n",
       "21  0.81540  166718  9965999            CHR with prefix  \n",
       "23  0.93200  191764  9965999                        NaN  \n",
       "24  0.93200  166718  9965999                        NaN  \n",
       "25  0.93200  166718  9965999                        NaN  \n",
       "26  0.93200  166718  9965999    POS out of normal range  \n",
       "27  0.15350  166718  9965999                  POS float  \n",
       "28  0.93090  166718  9965999               Same alleles  \n",
       "29  0.08092  166718  9965999          Lowercase alleles  \n",
       "30  0.05466  166718  9965999       Unrecognized alleles  \n",
       "31  0.32220  166718  9965999       Unrecognized alleles  \n",
       "32  0.31990  166718  9965999  Not normalizated allelels  \n",
       "35  0.79480  166718  9965999          BETA out of range  \n",
       "36  0.05270  166718  9965999          BETA out of range  \n",
       "37  0.05601  166718  9965999            SE out of range  \n",
       "38 -0.04424  166718  9965999             P out of range  \n",
       "39  1.67220  166718  9965999             P out of range  \n",
       "40  0.38960      -1  9965999             N out of range  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd76afda",
   "metadata": {},
   "source": [
    "## fix position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a9063d5c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 23 18:40:19 2024 Start to fix basepair positions...v3.4.36\n",
      "Tue Jan 23 18:40:19 2024  -Current Dataframe shape : 38  x  12\n",
      "Tue Jan 23 18:40:19 2024  -Converting to Int64 data type ...\n",
      "Tue Jan 23 18:40:19 2024  -Force converting to Int64 data type ...\n",
      "Tue Jan 23 18:40:20 2024  -Position upper_bound is: 250,000,000\n",
      "Tue Jan 23 18:40:20 2024  -Remove outliers: 1\n",
      "Tue Jan 23 18:40:20 2024  -Removed 1 variants with bad positions.\n",
      "Tue Jan 23 18:40:20 2024  -Converted all position to datatype Int64.\n",
      "Tue Jan 23 18:40:20 2024 Finished fixing basepair position successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.fix_pos(remove=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "96c5a234",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9960</td>\n",
       "      <td>-0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59700</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59730</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.0490</td>\n",
       "      <td>0.1231</td>\n",
       "      <td>0.69080</td>\n",
       "      <td>166718</td>\n",
       "      <td>9980999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8374</td>\n",
       "      <td>0.0213</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.28460</td>\n",
       "      <td>166718</td>\n",
       "      <td>9980999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:3:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8593</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1:3:T:GA</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>GA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1470</td>\n",
       "      <td>-0.0187</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0.28300</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1:4:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1400</td>\n",
       "      <td>-0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1:5:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1578</td>\n",
       "      <td>-0.0155</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.23500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Flipped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1:6:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1:7:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1:8:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9866</td>\n",
       "      <td>-0.1439</td>\n",
       "      <td>0.0748</td>\n",
       "      <td>0.05451</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1:9:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1327</td>\n",
       "      <td>0.0035</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.80870</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1:10:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8522</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0.0157</td>\n",
       "      <td>0.05421</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1:11:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0.1858</td>\n",
       "      <td>0.1546</td>\n",
       "      <td>0.22930</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1:12:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.1430</td>\n",
       "      <td>0.0747</td>\n",
       "      <td>0.05575</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1:13:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.7445</td>\n",
       "      <td>-0.0135</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>0.32450</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1:14:C:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4544</td>\n",
       "      <td>-0.0170</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.21910</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1:15:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8516</td>\n",
       "      <td>0.0292</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.06173</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1:16:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>16</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1642</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.49750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Sex chromosomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1:17:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7428</td>\n",
       "      <td>-0.0119</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.36210</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8699</td>\n",
       "      <td>-0.0038</td>\n",
       "      <td>0.0161</td>\n",
       "      <td>0.81540</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1:25:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8220</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.0139</td>\n",
       "      <td>0.15350</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>POS float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1:864002:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>864002</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8666</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0155</td>\n",
       "      <td>0.93090</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Same alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1:864615:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864615</td>\n",
       "      <td>g</td>\n",
       "      <td>a</td>\n",
       "      <td>0.8725</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>0.0158</td>\n",
       "      <td>0.08092</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Lowercase alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1:864701:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>864701</td>\n",
       "      <td>&lt;CN1&gt;</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>0.1306</td>\n",
       "      <td>0.0680</td>\n",
       "      <td>0.05466</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Unrecognized alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1:864755:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>864755</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8242</td>\n",
       "      <td>0.0163</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.32220</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Unrecognized alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1:864757:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864757</td>\n",
       "      <td>GT</td>\n",
       "      <td>AT</td>\n",
       "      <td>0.1758</td>\n",
       "      <td>-0.0164</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.31990</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>Not normalizated allelels</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1:865219:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>865219</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8685</td>\n",
       "      <td>300.0000</td>\n",
       "      <td>0.0161</td>\n",
       "      <td>0.79480</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>BETA out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1:865367:G:T</td>\n",
       "      <td>1</td>\n",
       "      <td>865367</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8486</td>\n",
       "      <td>-300.0000</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.05270</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>BETA out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1:865438:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>865438</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.9849</td>\n",
       "      <td>-0.1299</td>\n",
       "      <td>-0.0680</td>\n",
       "      <td>0.05601</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>SE out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>1:865694:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>865694</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8489</td>\n",
       "      <td>0.0265</td>\n",
       "      <td>0.0132</td>\n",
       "      <td>-0.04424</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>P out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>1:866893:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>866893</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.3104</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>1.67220</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960999</td>\n",
       "      <td>P out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>1:866938:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>866938</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.6028</td>\n",
       "      <td>-0.0104</td>\n",
       "      <td>0.0121</td>\n",
       "      <td>0.38960</td>\n",
       "      <td>-1</td>\n",
       "      <td>9960999</td>\n",
       "      <td>N out of range</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNPID  CHR     POS     EA NEA     EAF      BETA      SE        P  \\\n",
       "0        1:1:G:A    1       1      G   A  0.9960   -0.0737  0.1394  0.59700   \n",
       "1        1:1:A:G    1       1      G   A  0.0040    0.0737  0.1394  0.59730   \n",
       "2        1:1:A:G    1       1      G   A  0.0040    0.0737  0.1394  0.59750   \n",
       "3            1:2    1       2      C   T  0.0051    0.0490  0.1231  0.69080   \n",
       "4            1:2    1       2    TAA   T  0.8374    0.0213  0.0199  0.28460   \n",
       "5        1:3:T:A    1       3      T   A  0.8593    0.0172  0.0156  0.27050   \n",
       "6       1:3:T:GA    1       3     GA   T  0.1470   -0.0187  0.0174  0.28300   \n",
       "7        1:4:T:C    1       4      C   T  0.1400   -0.0172  0.0156  0.27050   \n",
       "8        1:5:G:A    1       5      G   A  0.1578   -0.0155  0.0131  0.23500   \n",
       "9        1:6:A:G    1       6      G   A  0.7493    0.0204  0.0147  0.16500   \n",
       "10       1:7:A:G    1       7      G   A  0.7493    0.0204  0.0147  0.16500   \n",
       "11       1:8:G:C    1       8      G   C  0.9866   -0.1439  0.0748  0.05451   \n",
       "12       1:9:A:G    1       9      G   A  0.1327    0.0035  0.0146  0.80870   \n",
       "13      1:10:G:A    1      10      G   A  0.8522    0.0302  0.0157  0.05421   \n",
       "14      1:11:T:C    1      11      C   T  0.0027    0.1858  0.1546  0.22930   \n",
       "15      1:12:A:G    1      12      G   A  0.0131    0.1430  0.0747  0.05575   \n",
       "16      1:13:C:T    1      13      C   T  0.7445   -0.0135  0.0137  0.32450   \n",
       "17      1:14:C:A    1      14      C   A  0.4544   -0.0170  0.0138  0.21910   \n",
       "18      1:15:C:T    1      15      C   T  0.8516    0.0292  0.0156  0.06173   \n",
       "19      1:16:A:G   23      16      G   A  0.1642    0.0099  0.0146  0.49750   \n",
       "20      1:17:G:A    1      17      G   A  0.7428   -0.0119  0.0131  0.36210   \n",
       "21       1:1:G:A    1      18      G   A  0.8699   -0.0038  0.0161  0.81540   \n",
       "23      1:18:A:G    1      20      G   A  0.7316    0.0013  0.0147  0.93200   \n",
       "24      1:18:A:G    1      21      G   A  0.7316    0.0013  0.0147  0.93200   \n",
       "25      1:18:A:G    1      22      G   A  0.7316    0.0013  0.0147  0.93200   \n",
       "27      1:25:A:G    1      25      G   A  0.8220    0.0199  0.0139  0.15350   \n",
       "28  1:864002:G:C    1  864002      A   A  0.8666    0.0013  0.0155  0.93090   \n",
       "29  1:864615:G:A    1  864615      g   a  0.8725    0.0275  0.0158  0.08092   \n",
       "30  1:864701:A:G    1  864701  <CN1>   A  0.0150    0.1306  0.0680  0.05466   \n",
       "31  1:864755:A:G    1  864755      N   A  0.8242    0.0163  0.0165  0.32220   \n",
       "32  1:864757:G:A    1  864757     GT  AT  0.1758   -0.0164  0.0165  0.31990   \n",
       "35  1:865219:G:A    1  865219      G   A  0.8685  300.0000  0.0161  0.79480   \n",
       "36  1:865367:G:T    1  865367      G   T  0.8486 -300.0000  0.0138  0.05270   \n",
       "37  1:865438:C:T    1  865438      C   T  0.9849   -0.1299 -0.0680  0.05601   \n",
       "38  1:865694:C:T    1  865694      C   T  0.8489    0.0265  0.0132 -0.04424   \n",
       "39  1:866893:T:C    1  866893      C   T  0.3104    0.0058  0.0137  1.67220   \n",
       "40  1:866938:G:A    1  866938      G   A  0.6028   -0.0104  0.0121  0.38960   \n",
       "\n",
       "         N   STATUS                       Note  \n",
       "0   191764  9960999                 Duplicated  \n",
       "1   166718  9960999                 Duplicated  \n",
       "2   166718  9960999                 Duplicated  \n",
       "3   166718  9980999               Multiallelic  \n",
       "4   166718  9980999               Multiallelic  \n",
       "5   166718  9960999               Multiallelic  \n",
       "6   166718  9960999               Multiallelic  \n",
       "7   191764  9960999                        NaN  \n",
       "8   166718  9960999                    Flipped  \n",
       "9   166718  9960999                        NaN  \n",
       "10  166718  9960999                        NaN  \n",
       "11  166718  9960999                        NaN  \n",
       "12  166718  9960999                        NaN  \n",
       "13  166718  9960999                        NaN  \n",
       "14  166718  9960999                        NaN  \n",
       "15  166718  9960999                        NaN  \n",
       "16  191764  9960999                        NaN  \n",
       "17  166718  9960999                        NaN  \n",
       "18  166718  9960999                        NaN  \n",
       "19  166718  9960999            Sex chromosomes  \n",
       "20  166718  9960999            CHR with prefix  \n",
       "21  166718  9960999            CHR with prefix  \n",
       "23  191764  9960999                        NaN  \n",
       "24  166718  9960999                        NaN  \n",
       "25  166718  9960999                        NaN  \n",
       "27  166718  9960999                  POS float  \n",
       "28  166718  9960999               Same alleles  \n",
       "29  166718  9960999          Lowercase alleles  \n",
       "30  166718  9960999       Unrecognized alleles  \n",
       "31  166718  9960999       Unrecognized alleles  \n",
       "32  166718  9960999  Not normalizated allelels  \n",
       "35  166718  9960999          BETA out of range  \n",
       "36  166718  9960999          BETA out of range  \n",
       "37  166718  9960999            SE out of range  \n",
       "38  166718  9960999             P out of range  \n",
       "39  166718  9960999             P out of range  \n",
       "40      -1  9960999             N out of range  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45c72b40",
   "metadata": {},
   "source": [
    "## fix allele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3edf8219",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 23 18:40:42 2024 Start to fix alleles...v3.4.36\n",
      "Tue Jan 23 18:40:42 2024  -Current Dataframe shape : 37  x  12\n",
      "Tue Jan 23 18:40:42 2024  -Converted all bases to string datatype and UPPERCASE.\n",
      "Tue Jan 23 18:40:42 2024  -Variants with bad EA  : 2\n",
      "Tue Jan 23 18:40:42 2024  -Variants with bad NEA : 0\n",
      "Tue Jan 23 18:40:42 2024  -Variants with NA for EA or NEA: 0\n",
      "Tue Jan 23 18:40:42 2024  -Variants with same EA and NEA: 1\n",
      "Tue Jan 23 18:40:42 2024  -A look at the non-ATCG EA: {'<CN1>', 'N'} ...\n",
      "Tue Jan 23 18:40:42 2024  -Removed 2 variants with NA alleles or alleles that contain bases other than A/C/T/G.\n",
      "Tue Jan 23 18:40:42 2024  -Removed 1 variants with same allele for EA and NEA.\n",
      "Tue Jan 23 18:40:42 2024 Finished fixing allele successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.fix_allele(remove=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d7e0cafe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9960</td>\n",
       "      <td>-0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59700</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59730</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.0490</td>\n",
       "      <td>0.1231</td>\n",
       "      <td>0.69080</td>\n",
       "      <td>166718</td>\n",
       "      <td>9980099</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8374</td>\n",
       "      <td>0.0213</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.28460</td>\n",
       "      <td>166718</td>\n",
       "      <td>9980399</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:3:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8593</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1:3:T:GA</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>GA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1470</td>\n",
       "      <td>-0.0187</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0.28300</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960399</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1:4:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1400</td>\n",
       "      <td>-0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1:5:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1578</td>\n",
       "      <td>-0.0155</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.23500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Flipped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1:6:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1:7:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1:8:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9866</td>\n",
       "      <td>-0.1439</td>\n",
       "      <td>0.0748</td>\n",
       "      <td>0.05451</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1:9:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1327</td>\n",
       "      <td>0.0035</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.80870</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1:10:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8522</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0.0157</td>\n",
       "      <td>0.05421</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1:11:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0.1858</td>\n",
       "      <td>0.1546</td>\n",
       "      <td>0.22930</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1:12:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.1430</td>\n",
       "      <td>0.0747</td>\n",
       "      <td>0.05575</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1:13:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.7445</td>\n",
       "      <td>-0.0135</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>0.32450</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1:14:C:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4544</td>\n",
       "      <td>-0.0170</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.21910</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1:15:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8516</td>\n",
       "      <td>0.0292</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.06173</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1:16:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>16</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1642</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.49750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Sex chromosomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1:17:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7428</td>\n",
       "      <td>-0.0119</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.36210</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8699</td>\n",
       "      <td>-0.0038</td>\n",
       "      <td>0.0161</td>\n",
       "      <td>0.81540</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1:25:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8220</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.0139</td>\n",
       "      <td>0.15350</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>POS float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1:864615:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864615</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8725</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>0.0158</td>\n",
       "      <td>0.08092</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Lowercase alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1:864757:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864757</td>\n",
       "      <td>GT</td>\n",
       "      <td>AT</td>\n",
       "      <td>0.1758</td>\n",
       "      <td>-0.0164</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.31990</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960599</td>\n",
       "      <td>Not normalizated allelels</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1:865219:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>865219</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8685</td>\n",
       "      <td>300.0000</td>\n",
       "      <td>0.0161</td>\n",
       "      <td>0.79480</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>BETA out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1:865367:G:T</td>\n",
       "      <td>1</td>\n",
       "      <td>865367</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8486</td>\n",
       "      <td>-300.0000</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.05270</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>BETA out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1:865438:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>865438</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.9849</td>\n",
       "      <td>-0.1299</td>\n",
       "      <td>-0.0680</td>\n",
       "      <td>0.05601</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>SE out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>1:865694:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>865694</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8489</td>\n",
       "      <td>0.0265</td>\n",
       "      <td>0.0132</td>\n",
       "      <td>-0.04424</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>P out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>1:866893:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>866893</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.3104</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>1.67220</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>P out of range</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>1:866938:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>866938</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.6028</td>\n",
       "      <td>-0.0104</td>\n",
       "      <td>0.0121</td>\n",
       "      <td>0.38960</td>\n",
       "      <td>-1</td>\n",
       "      <td>9960099</td>\n",
       "      <td>N out of range</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNPID  CHR     POS   EA NEA     EAF      BETA      SE        P  \\\n",
       "0        1:1:G:A    1       1    G   A  0.9960   -0.0737  0.1394  0.59700   \n",
       "1        1:1:A:G    1       1    G   A  0.0040    0.0737  0.1394  0.59730   \n",
       "2        1:1:A:G    1       1    G   A  0.0040    0.0737  0.1394  0.59750   \n",
       "3            1:2    1       2    C   T  0.0051    0.0490  0.1231  0.69080   \n",
       "4            1:2    1       2  TAA   T  0.8374    0.0213  0.0199  0.28460   \n",
       "5        1:3:T:A    1       3    T   A  0.8593    0.0172  0.0156  0.27050   \n",
       "6       1:3:T:GA    1       3   GA   T  0.1470   -0.0187  0.0174  0.28300   \n",
       "7        1:4:T:C    1       4    C   T  0.1400   -0.0172  0.0156  0.27050   \n",
       "8        1:5:G:A    1       5    G   A  0.1578   -0.0155  0.0131  0.23500   \n",
       "9        1:6:A:G    1       6    G   A  0.7493    0.0204  0.0147  0.16500   \n",
       "10       1:7:A:G    1       7    G   A  0.7493    0.0204  0.0147  0.16500   \n",
       "11       1:8:G:C    1       8    G   C  0.9866   -0.1439  0.0748  0.05451   \n",
       "12       1:9:A:G    1       9    G   A  0.1327    0.0035  0.0146  0.80870   \n",
       "13      1:10:G:A    1      10    G   A  0.8522    0.0302  0.0157  0.05421   \n",
       "14      1:11:T:C    1      11    C   T  0.0027    0.1858  0.1546  0.22930   \n",
       "15      1:12:A:G    1      12    G   A  0.0131    0.1430  0.0747  0.05575   \n",
       "16      1:13:C:T    1      13    C   T  0.7445   -0.0135  0.0137  0.32450   \n",
       "17      1:14:C:A    1      14    C   A  0.4544   -0.0170  0.0138  0.21910   \n",
       "18      1:15:C:T    1      15    C   T  0.8516    0.0292  0.0156  0.06173   \n",
       "19      1:16:A:G   23      16    G   A  0.1642    0.0099  0.0146  0.49750   \n",
       "20      1:17:G:A    1      17    G   A  0.7428   -0.0119  0.0131  0.36210   \n",
       "21       1:1:G:A    1      18    G   A  0.8699   -0.0038  0.0161  0.81540   \n",
       "23      1:18:A:G    1      20    G   A  0.7316    0.0013  0.0147  0.93200   \n",
       "24      1:18:A:G    1      21    G   A  0.7316    0.0013  0.0147  0.93200   \n",
       "25      1:18:A:G    1      22    G   A  0.7316    0.0013  0.0147  0.93200   \n",
       "27      1:25:A:G    1      25    G   A  0.8220    0.0199  0.0139  0.15350   \n",
       "29  1:864615:G:A    1  864615    G   A  0.8725    0.0275  0.0158  0.08092   \n",
       "32  1:864757:G:A    1  864757   GT  AT  0.1758   -0.0164  0.0165  0.31990   \n",
       "35  1:865219:G:A    1  865219    G   A  0.8685  300.0000  0.0161  0.79480   \n",
       "36  1:865367:G:T    1  865367    G   T  0.8486 -300.0000  0.0138  0.05270   \n",
       "37  1:865438:C:T    1  865438    C   T  0.9849   -0.1299 -0.0680  0.05601   \n",
       "38  1:865694:C:T    1  865694    C   T  0.8489    0.0265  0.0132 -0.04424   \n",
       "39  1:866893:T:C    1  866893    C   T  0.3104    0.0058  0.0137  1.67220   \n",
       "40  1:866938:G:A    1  866938    G   A  0.6028   -0.0104  0.0121  0.38960   \n",
       "\n",
       "         N   STATUS                       Note  \n",
       "0   191764  9960099                 Duplicated  \n",
       "1   166718  9960099                 Duplicated  \n",
       "2   166718  9960099                 Duplicated  \n",
       "3   166718  9980099               Multiallelic  \n",
       "4   166718  9980399               Multiallelic  \n",
       "5   166718  9960099               Multiallelic  \n",
       "6   166718  9960399               Multiallelic  \n",
       "7   191764  9960099                        NaN  \n",
       "8   166718  9960099                    Flipped  \n",
       "9   166718  9960099                        NaN  \n",
       "10  166718  9960099                        NaN  \n",
       "11  166718  9960099                        NaN  \n",
       "12  166718  9960099                        NaN  \n",
       "13  166718  9960099                        NaN  \n",
       "14  166718  9960099                        NaN  \n",
       "15  166718  9960099                        NaN  \n",
       "16  191764  9960099                        NaN  \n",
       "17  166718  9960099                        NaN  \n",
       "18  166718  9960099                        NaN  \n",
       "19  166718  9960099            Sex chromosomes  \n",
       "20  166718  9960099            CHR with prefix  \n",
       "21  166718  9960099            CHR with prefix  \n",
       "23  191764  9960099                        NaN  \n",
       "24  166718  9960099                        NaN  \n",
       "25  166718  9960099                        NaN  \n",
       "27  166718  9960099                  POS float  \n",
       "29  166718  9960099          Lowercase alleles  \n",
       "32  166718  9960599  Not normalizated allelels  \n",
       "35  166718  9960099          BETA out of range  \n",
       "36  166718  9960099          BETA out of range  \n",
       "37  166718  9960099            SE out of range  \n",
       "38  166718  9960099             P out of range  \n",
       "39  166718  9960099             P out of range  \n",
       "40      -1  9960099             N out of range  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54345b77",
   "metadata": {},
   "source": [
    "## sanity check for statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "819cf4bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 23 18:41:08 2024 Start sanity check for statistics...v3.4.36\n",
      "Tue Jan 23 18:41:08 2024  -Current Dataframe shape : 34  x  12\n",
      "Tue Jan 23 18:41:08 2024  -Comparison tolerance for floats: 1e-07\n",
      "Tue Jan 23 18:41:08 2024  -Checking if  0 <=N<= 2147483647  ...\n",
      "Tue Jan 23 18:41:08 2024  -Removed 1 variants with bad N.\n",
      "Tue Jan 23 18:41:08 2024  -Checking if  -1e-07 <EAF< 1.0000001  ...\n",
      "Tue Jan 23 18:41:08 2024  -Removed 0 variants with bad EAF.\n",
      "Tue Jan 23 18:41:08 2024  -Checking if  0 <=MAC<= 2147483647  ...\n",
      "Tue Jan 23 18:41:08 2024  -Removed 0 variants with bad MAC.\n",
      "Tue Jan 23 18:41:08 2024  -Checking if  -1e-07 < P < 1.0000001  ...\n",
      "Tue Jan 23 18:41:08 2024  -Removed 2 variants with bad P.\n",
      "Tue Jan 23 18:41:08 2024  -Checking if  -100.0000001 <BETA< 100.0000001  ...\n",
      "Tue Jan 23 18:41:08 2024  -Removed 2 variants with bad BETA.\n",
      "Tue Jan 23 18:41:08 2024  -Checking if  -1e-07 <SE< inf  ...\n",
      "Tue Jan 23 18:41:08 2024  -Removed 1 variants with bad SE.\n",
      "Tue Jan 23 18:41:08 2024  -Checking STATUS and converting STATUS to categories....\n",
      "Tue Jan 23 18:41:08 2024  -Removed 0 variants with NAs in the checked columns...\n",
      "Tue Jan 23 18:41:08 2024  -Removed 6 variants with bad statistics in total.\n",
      "Tue Jan 23 18:41:08 2024  -Data types for each column:\n",
      "Tue Jan 23 18:41:08 2024  -Column  : SNPID  CHR   POS   EA       NEA      EAF     BETA    SE      P       N     STATUS   Note  \n",
      "Tue Jan 23 18:41:08 2024  -DType   : object Int64 Int64 category category float32 float64 float64 float64 Int32 category object\n",
      "Tue Jan 23 18:41:08 2024  -Verified: T      T     T     T        T        T       T       T       T       T     T        NA    \n",
      "Tue Jan 23 18:41:08 2024 Finished sanity check successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.check_sanity()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4518a81d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9960</td>\n",
       "      <td>-0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59700</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59730</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.0490</td>\n",
       "      <td>0.1231</td>\n",
       "      <td>0.69080</td>\n",
       "      <td>166718</td>\n",
       "      <td>9980099</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8374</td>\n",
       "      <td>0.0213</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.28460</td>\n",
       "      <td>166718</td>\n",
       "      <td>9980399</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:3:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8593</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1:3:T:GA</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>GA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1470</td>\n",
       "      <td>-0.0187</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0.28300</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960399</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1:4:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1400</td>\n",
       "      <td>-0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1:5:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1578</td>\n",
       "      <td>-0.0155</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.23500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Flipped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1:6:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1:7:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1:8:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9866</td>\n",
       "      <td>-0.1439</td>\n",
       "      <td>0.0748</td>\n",
       "      <td>0.05451</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1:9:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1327</td>\n",
       "      <td>0.0035</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.80870</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1:10:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8522</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0.0157</td>\n",
       "      <td>0.05421</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1:11:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0.1858</td>\n",
       "      <td>0.1546</td>\n",
       "      <td>0.22930</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1:12:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.1430</td>\n",
       "      <td>0.0747</td>\n",
       "      <td>0.05575</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1:13:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.7445</td>\n",
       "      <td>-0.0135</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>0.32450</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1:14:C:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4544</td>\n",
       "      <td>-0.0170</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.21910</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1:15:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8516</td>\n",
       "      <td>0.0292</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.06173</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1:16:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>16</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1642</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.49750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Sex chromosomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1:17:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7428</td>\n",
       "      <td>-0.0119</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.36210</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8699</td>\n",
       "      <td>-0.0038</td>\n",
       "      <td>0.0161</td>\n",
       "      <td>0.81540</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1:25:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8220</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.0139</td>\n",
       "      <td>0.15350</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>POS float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1:864615:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864615</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8725</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>0.0158</td>\n",
       "      <td>0.08092</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Lowercase alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1:864757:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864757</td>\n",
       "      <td>GT</td>\n",
       "      <td>AT</td>\n",
       "      <td>0.1758</td>\n",
       "      <td>-0.0164</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.31990</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960599</td>\n",
       "      <td>Not normalizated allelels</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNPID  CHR     POS   EA NEA     EAF    BETA      SE        P  \\\n",
       "0        1:1:G:A    1       1    G   A  0.9960 -0.0737  0.1394  0.59700   \n",
       "1        1:1:A:G    1       1    G   A  0.0040  0.0737  0.1394  0.59730   \n",
       "2        1:1:A:G    1       1    G   A  0.0040  0.0737  0.1394  0.59750   \n",
       "3            1:2    1       2    C   T  0.0051  0.0490  0.1231  0.69080   \n",
       "4            1:2    1       2  TAA   T  0.8374  0.0213  0.0199  0.28460   \n",
       "5        1:3:T:A    1       3    T   A  0.8593  0.0172  0.0156  0.27050   \n",
       "6       1:3:T:GA    1       3   GA   T  0.1470 -0.0187  0.0174  0.28300   \n",
       "7        1:4:T:C    1       4    C   T  0.1400 -0.0172  0.0156  0.27050   \n",
       "8        1:5:G:A    1       5    G   A  0.1578 -0.0155  0.0131  0.23500   \n",
       "9        1:6:A:G    1       6    G   A  0.7493  0.0204  0.0147  0.16500   \n",
       "10       1:7:A:G    1       7    G   A  0.7493  0.0204  0.0147  0.16500   \n",
       "11       1:8:G:C    1       8    G   C  0.9866 -0.1439  0.0748  0.05451   \n",
       "12       1:9:A:G    1       9    G   A  0.1327  0.0035  0.0146  0.80870   \n",
       "13      1:10:G:A    1      10    G   A  0.8522  0.0302  0.0157  0.05421   \n",
       "14      1:11:T:C    1      11    C   T  0.0027  0.1858  0.1546  0.22930   \n",
       "15      1:12:A:G    1      12    G   A  0.0131  0.1430  0.0747  0.05575   \n",
       "16      1:13:C:T    1      13    C   T  0.7445 -0.0135  0.0137  0.32450   \n",
       "17      1:14:C:A    1      14    C   A  0.4544 -0.0170  0.0138  0.21910   \n",
       "18      1:15:C:T    1      15    C   T  0.8516  0.0292  0.0156  0.06173   \n",
       "19      1:16:A:G   23      16    G   A  0.1642  0.0099  0.0146  0.49750   \n",
       "20      1:17:G:A    1      17    G   A  0.7428 -0.0119  0.0131  0.36210   \n",
       "21       1:1:G:A    1      18    G   A  0.8699 -0.0038  0.0161  0.81540   \n",
       "23      1:18:A:G    1      20    G   A  0.7316  0.0013  0.0147  0.93200   \n",
       "24      1:18:A:G    1      21    G   A  0.7316  0.0013  0.0147  0.93200   \n",
       "25      1:18:A:G    1      22    G   A  0.7316  0.0013  0.0147  0.93200   \n",
       "27      1:25:A:G    1      25    G   A  0.8220  0.0199  0.0139  0.15350   \n",
       "29  1:864615:G:A    1  864615    G   A  0.8725  0.0275  0.0158  0.08092   \n",
       "32  1:864757:G:A    1  864757   GT  AT  0.1758 -0.0164  0.0165  0.31990   \n",
       "\n",
       "         N   STATUS                       Note  \n",
       "0   191764  9960099                 Duplicated  \n",
       "1   166718  9960099                 Duplicated  \n",
       "2   166718  9960099                 Duplicated  \n",
       "3   166718  9980099               Multiallelic  \n",
       "4   166718  9980399               Multiallelic  \n",
       "5   166718  9960099               Multiallelic  \n",
       "6   166718  9960399               Multiallelic  \n",
       "7   191764  9960099                        NaN  \n",
       "8   166718  9960099                    Flipped  \n",
       "9   166718  9960099                        NaN  \n",
       "10  166718  9960099                        NaN  \n",
       "11  166718  9960099                        NaN  \n",
       "12  166718  9960099                        NaN  \n",
       "13  166718  9960099                        NaN  \n",
       "14  166718  9960099                        NaN  \n",
       "15  166718  9960099                        NaN  \n",
       "16  191764  9960099                        NaN  \n",
       "17  166718  9960099                        NaN  \n",
       "18  166718  9960099                        NaN  \n",
       "19  166718  9960099            Sex chromosomes  \n",
       "20  166718  9960099            CHR with prefix  \n",
       "21  166718  9960099            CHR with prefix  \n",
       "23  191764  9960099                        NaN  \n",
       "24  166718  9960099                        NaN  \n",
       "25  166718  9960099                        NaN  \n",
       "27  166718  9960099                  POS float  \n",
       "29  166718  9960099          Lowercase alleles  \n",
       "32  166718  9960599  Not normalizated allelels  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b773946",
   "metadata": {},
   "source": [
    "## normalize variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "86561df4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 23 18:41:20 2024 Start to normalize variants...v3.4.36\n",
      "Tue Jan 23 18:41:20 2024  -Current Dataframe shape : 28  x  12\n",
      "Tue Jan 23 18:41:21 2024  -Not normalized allele IDs:1:864757:G:A ... \n",
      "Tue Jan 23 18:41:21 2024  -Not normalized allele:['GT' 'AT']... \n",
      "Tue Jan 23 18:41:21 2024  -Modified 1 variants according to parsimony and left alignment principal.\n",
      "Tue Jan 23 18:41:21 2024 Finished normalizing variants successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.normalize_allele()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "bea1ddbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9960</td>\n",
       "      <td>-0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59700</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59730</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:1:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.0490</td>\n",
       "      <td>0.1231</td>\n",
       "      <td>0.69080</td>\n",
       "      <td>166718</td>\n",
       "      <td>9980099</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8374</td>\n",
       "      <td>0.0213</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.28460</td>\n",
       "      <td>166718</td>\n",
       "      <td>9980399</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:3:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8593</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1:3:T:GA</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>GA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1470</td>\n",
       "      <td>-0.0187</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0.28300</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960399</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1:4:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1400</td>\n",
       "      <td>-0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1:5:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1578</td>\n",
       "      <td>-0.0155</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.23500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Flipped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1:6:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1:7:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1:8:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9866</td>\n",
       "      <td>-0.1439</td>\n",
       "      <td>0.0748</td>\n",
       "      <td>0.05451</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1:9:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1327</td>\n",
       "      <td>0.0035</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.80870</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1:10:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8522</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0.0157</td>\n",
       "      <td>0.05421</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1:11:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0.1858</td>\n",
       "      <td>0.1546</td>\n",
       "      <td>0.22930</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1:12:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.1430</td>\n",
       "      <td>0.0747</td>\n",
       "      <td>0.05575</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1:13:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.7445</td>\n",
       "      <td>-0.0135</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>0.32450</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1:14:C:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4544</td>\n",
       "      <td>-0.0170</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.21910</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1:15:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8516</td>\n",
       "      <td>0.0292</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.06173</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1:16:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>16</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1642</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.49750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Sex chromosomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1:17:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7428</td>\n",
       "      <td>-0.0119</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.36210</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8699</td>\n",
       "      <td>-0.0038</td>\n",
       "      <td>0.0161</td>\n",
       "      <td>0.81540</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1:25:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8220</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.0139</td>\n",
       "      <td>0.15350</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>POS float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1:864615:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864615</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8725</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>0.0158</td>\n",
       "      <td>0.08092</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Lowercase alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1:864757:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864757</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1758</td>\n",
       "      <td>-0.0164</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.31990</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960399</td>\n",
       "      <td>Not normalizated allelels</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNPID  CHR     POS   EA NEA     EAF    BETA      SE        P  \\\n",
       "0        1:1:G:A    1       1    G   A  0.9960 -0.0737  0.1394  0.59700   \n",
       "1        1:1:A:G    1       1    G   A  0.0040  0.0737  0.1394  0.59730   \n",
       "2        1:1:A:G    1       1    G   A  0.0040  0.0737  0.1394  0.59750   \n",
       "3            1:2    1       2    C   T  0.0051  0.0490  0.1231  0.69080   \n",
       "4            1:2    1       2  TAA   T  0.8374  0.0213  0.0199  0.28460   \n",
       "5        1:3:T:A    1       3    T   A  0.8593  0.0172  0.0156  0.27050   \n",
       "6       1:3:T:GA    1       3   GA   T  0.1470 -0.0187  0.0174  0.28300   \n",
       "7        1:4:T:C    1       4    C   T  0.1400 -0.0172  0.0156  0.27050   \n",
       "8        1:5:G:A    1       5    G   A  0.1578 -0.0155  0.0131  0.23500   \n",
       "9        1:6:A:G    1       6    G   A  0.7493  0.0204  0.0147  0.16500   \n",
       "10       1:7:A:G    1       7    G   A  0.7493  0.0204  0.0147  0.16500   \n",
       "11       1:8:G:C    1       8    G   C  0.9866 -0.1439  0.0748  0.05451   \n",
       "12       1:9:A:G    1       9    G   A  0.1327  0.0035  0.0146  0.80870   \n",
       "13      1:10:G:A    1      10    G   A  0.8522  0.0302  0.0157  0.05421   \n",
       "14      1:11:T:C    1      11    C   T  0.0027  0.1858  0.1546  0.22930   \n",
       "15      1:12:A:G    1      12    G   A  0.0131  0.1430  0.0747  0.05575   \n",
       "16      1:13:C:T    1      13    C   T  0.7445 -0.0135  0.0137  0.32450   \n",
       "17      1:14:C:A    1      14    C   A  0.4544 -0.0170  0.0138  0.21910   \n",
       "18      1:15:C:T    1      15    C   T  0.8516  0.0292  0.0156  0.06173   \n",
       "19      1:16:A:G   23      16    G   A  0.1642  0.0099  0.0146  0.49750   \n",
       "20      1:17:G:A    1      17    G   A  0.7428 -0.0119  0.0131  0.36210   \n",
       "21       1:1:G:A    1      18    G   A  0.8699 -0.0038  0.0161  0.81540   \n",
       "23      1:18:A:G    1      20    G   A  0.7316  0.0013  0.0147  0.93200   \n",
       "24      1:18:A:G    1      21    G   A  0.7316  0.0013  0.0147  0.93200   \n",
       "25      1:18:A:G    1      22    G   A  0.7316  0.0013  0.0147  0.93200   \n",
       "27      1:25:A:G    1      25    G   A  0.8220  0.0199  0.0139  0.15350   \n",
       "29  1:864615:G:A    1  864615    G   A  0.8725  0.0275  0.0158  0.08092   \n",
       "32  1:864757:G:A    1  864757    G   A  0.1758 -0.0164  0.0165  0.31990   \n",
       "\n",
       "         N   STATUS                       Note  \n",
       "0   191764  9960099                 Duplicated  \n",
       "1   166718  9960099                 Duplicated  \n",
       "2   166718  9960099                 Duplicated  \n",
       "3   166718  9980099               Multiallelic  \n",
       "4   166718  9980399               Multiallelic  \n",
       "5   166718  9960099               Multiallelic  \n",
       "6   166718  9960399               Multiallelic  \n",
       "7   191764  9960099                        NaN  \n",
       "8   166718  9960099                    Flipped  \n",
       "9   166718  9960099                        NaN  \n",
       "10  166718  9960099                        NaN  \n",
       "11  166718  9960099                        NaN  \n",
       "12  166718  9960099                        NaN  \n",
       "13  166718  9960099                        NaN  \n",
       "14  166718  9960099                        NaN  \n",
       "15  166718  9960099                        NaN  \n",
       "16  191764  9960099                        NaN  \n",
       "17  166718  9960099                        NaN  \n",
       "18  166718  9960099                        NaN  \n",
       "19  166718  9960099            Sex chromosomes  \n",
       "20  166718  9960099            CHR with prefix  \n",
       "21  166718  9960099            CHR with prefix  \n",
       "23  191764  9960099                        NaN  \n",
       "24  166718  9960099                        NaN  \n",
       "25  166718  9960099                        NaN  \n",
       "27  166718  9960099                  POS float  \n",
       "29  166718  9960099          Lowercase alleles  \n",
       "32  166718  9960399  Not normalizated allelels  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6437b3c",
   "metadata": {},
   "source": [
    "## remove duplicated / multiallelic variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ff597de1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 23 18:41:37 2024 Start to sort the sumstats using P...\n",
      "Tue Jan 23 18:41:37 2024 Start to remove duplicated variants based on snpid...v3.4.36\n",
      "Tue Jan 23 18:41:37 2024  -Current Dataframe shape : 28  x  12\n",
      "Tue Jan 23 18:41:37 2024  -Which variant to keep:  first\n",
      "Tue Jan 23 18:41:37 2024  -Removed  5  based on SNPID...\n",
      "Tue Jan 23 18:41:37 2024 Start to remove duplicated variants based on CHR,POS,EA and NEA...\n",
      "Tue Jan 23 18:41:37 2024  -Current Dataframe shape : 23  x  12\n",
      "Tue Jan 23 18:41:37 2024  -Which variant to keep:  first\n",
      "Tue Jan 23 18:41:37 2024  -Removed  1  based on CHR,POS,EA and NEA...\n",
      "Tue Jan 23 18:41:37 2024 Start to remove multiallelic variants based on chr:pos...\n",
      "Tue Jan 23 18:41:37 2024  -Which variant to keep:  first\n",
      "Tue Jan 23 18:41:37 2024  -Removed  1  multiallelic variants...\n",
      "Tue Jan 23 18:41:37 2024  -Removed  7  variants in total.\n",
      "Tue Jan 23 18:41:37 2024  -Sort the coordinates...\n",
      "Tue Jan 23 18:41:37 2024 Finished removing successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.remove_dup(mode=\"md\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9f2a4fe1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:1:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9960</td>\n",
       "      <td>-0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59700</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Duplicated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8374</td>\n",
       "      <td>0.0213</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.28460</td>\n",
       "      <td>166718</td>\n",
       "      <td>9980399</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:3:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8593</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Multiallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:4:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1400</td>\n",
       "      <td>-0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:5:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1578</td>\n",
       "      <td>-0.0155</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.23500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Flipped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:6:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1:7:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7493</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1:8:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9866</td>\n",
       "      <td>-0.1439</td>\n",
       "      <td>0.0748</td>\n",
       "      <td>0.05451</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1:9:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1327</td>\n",
       "      <td>0.0035</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.80870</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1:10:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8522</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0.0157</td>\n",
       "      <td>0.05421</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1:11:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0.1858</td>\n",
       "      <td>0.1546</td>\n",
       "      <td>0.22930</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1:12:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.1430</td>\n",
       "      <td>0.0747</td>\n",
       "      <td>0.05575</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1:13:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.7445</td>\n",
       "      <td>-0.0135</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>0.32450</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1:14:C:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4544</td>\n",
       "      <td>-0.0170</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.21910</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1:15:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8516</td>\n",
       "      <td>0.0292</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.06173</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1:17:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7428</td>\n",
       "      <td>-0.0119</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.36210</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>CHR with prefix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1:18:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7316</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.93200</td>\n",
       "      <td>191764</td>\n",
       "      <td>9960099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1:25:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8220</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.0139</td>\n",
       "      <td>0.15350</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>POS float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1:864615:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864615</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8725</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>0.0158</td>\n",
       "      <td>0.08092</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Lowercase alleles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1:864757:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>864757</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1758</td>\n",
       "      <td>-0.0164</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.31990</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960399</td>\n",
       "      <td>Not normalizated allelels</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1:16:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>16</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1642</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.49750</td>\n",
       "      <td>166718</td>\n",
       "      <td>9960099</td>\n",
       "      <td>Sex chromosomes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNPID  CHR     POS   EA NEA     EAF    BETA      SE        P  \\\n",
       "0        1:1:G:A    1       1    G   A  0.9960 -0.0737  0.1394  0.59700   \n",
       "1            1:2    1       2  TAA   T  0.8374  0.0213  0.0199  0.28460   \n",
       "2        1:3:T:A    1       3    T   A  0.8593  0.0172  0.0156  0.27050   \n",
       "3        1:4:T:C    1       4    C   T  0.1400 -0.0172  0.0156  0.27050   \n",
       "4        1:5:G:A    1       5    G   A  0.1578 -0.0155  0.0131  0.23500   \n",
       "5        1:6:A:G    1       6    G   A  0.7493  0.0204  0.0147  0.16500   \n",
       "6        1:7:A:G    1       7    G   A  0.7493  0.0204  0.0147  0.16500   \n",
       "7        1:8:G:C    1       8    G   C  0.9866 -0.1439  0.0748  0.05451   \n",
       "8        1:9:A:G    1       9    G   A  0.1327  0.0035  0.0146  0.80870   \n",
       "9       1:10:G:A    1      10    G   A  0.8522  0.0302  0.0157  0.05421   \n",
       "10      1:11:T:C    1      11    C   T  0.0027  0.1858  0.1546  0.22930   \n",
       "11      1:12:A:G    1      12    G   A  0.0131  0.1430  0.0747  0.05575   \n",
       "12      1:13:C:T    1      13    C   T  0.7445 -0.0135  0.0137  0.32450   \n",
       "13      1:14:C:A    1      14    C   A  0.4544 -0.0170  0.0138  0.21910   \n",
       "14      1:15:C:T    1      15    C   T  0.8516  0.0292  0.0156  0.06173   \n",
       "15      1:17:G:A    1      17    G   A  0.7428 -0.0119  0.0131  0.36210   \n",
       "16      1:18:A:G    1      20    G   A  0.7316  0.0013  0.0147  0.93200   \n",
       "17      1:25:A:G    1      25    G   A  0.8220  0.0199  0.0139  0.15350   \n",
       "18  1:864615:G:A    1  864615    G   A  0.8725  0.0275  0.0158  0.08092   \n",
       "19  1:864757:G:A    1  864757    G   A  0.1758 -0.0164  0.0165  0.31990   \n",
       "20      1:16:A:G   23      16    G   A  0.1642  0.0099  0.0146  0.49750   \n",
       "\n",
       "         N   STATUS                       Note  \n",
       "0   191764  9960099                 Duplicated  \n",
       "1   166718  9980399               Multiallelic  \n",
       "2   166718  9960099               Multiallelic  \n",
       "3   191764  9960099                        NaN  \n",
       "4   166718  9960099                    Flipped  \n",
       "5   166718  9960099                        NaN  \n",
       "6   166718  9960099                        NaN  \n",
       "7   166718  9960099                        NaN  \n",
       "8   166718  9960099                        NaN  \n",
       "9   166718  9960099                        NaN  \n",
       "10  166718  9960099                        NaN  \n",
       "11  166718  9960099                        NaN  \n",
       "12  191764  9960099                        NaN  \n",
       "13  166718  9960099                        NaN  \n",
       "14  166718  9960099                        NaN  \n",
       "15  166718  9960099            CHR with prefix  \n",
       "16  191764  9960099                        NaN  \n",
       "17  166718  9960099                  POS float  \n",
       "18  166718  9960099          Lowercase alleles  \n",
       "19  166718  9960399  Not normalizated allelels  \n",
       "20  166718  9960099            Sex chromosomes  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27c29690",
   "metadata": {},
   "source": [
    "## sort genome coordinate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "91ee7b0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 23 18:41:56 2024 Start to sort the genome coordinates...v3.4.36\n",
      "Tue Jan 23 18:41:56 2024  -Current Dataframe shape : 21  x  12\n",
      "Tue Jan 23 18:41:56 2024  -Sorting genome coordinates...\n",
      "Tue Jan 23 18:41:56 2024 Finished sorting genome coordinates successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.sort_coordinate()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a631ece",
   "metadata": {},
   "source": [
    "## sort column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "61d8d9c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 23 18:41:57 2024 Start to reorder the columns...v3.4.36\n",
      "Tue Jan 23 18:41:57 2024  -Current Dataframe shape : 21  x  12\n",
      "Tue Jan 23 18:41:57 2024  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,EAF,BETA,SE,P,N,STATUS,Note\n",
      "Tue Jan 23 18:41:57 2024 Finished sorting columns successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.sort_column()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed14018e",
   "metadata": {},
   "source": [
    "# All in one function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "45ad8af1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mon Oct 31 12:42:47 2022 Start to check IDs...\n",
      "Mon Oct 31 12:42:47 2022  -Current Dataframe shape : 12557759  x  12\n",
      "Mon Oct 31 12:42:47 2022  -Checking if SNPID is chr:pos:ref:alt...(separator: - ,: , _)\n",
      "Mon Oct 31 12:43:06 2022 Finished checking IDs successfully!\n",
      "Mon Oct 31 12:43:06 2022 Start to fix chromosome notation...\n",
      "Mon Oct 31 12:43:06 2022  -Current Dataframe shape : 12557759  x  12\n",
      "Mon Oct 31 12:43:58 2022  -All CHR are already fixed...\n",
      "Mon Oct 31 12:44:30 2022 Finished fixing chromosome notation successfully!\n",
      "Mon Oct 31 12:44:30 2022 Start to fix basepair positions...\n",
      "Mon Oct 31 12:44:30 2022  -Current Dataframe shape : 12557759  x  12\n",
      "Mon Oct 31 12:44:55 2022  -Position upper_bound is: 250,000,000\n",
      "Mon Oct 31 12:45:29 2022  -Remove outliers: 0\n",
      "Mon Oct 31 12:45:40 2022  -Converted all position to datatype Int64.\n",
      "Mon Oct 31 12:45:40 2022 Finished fixing basepair position successfully!\n",
      "Mon Oct 31 12:45:40 2022 Start to fix alleles...\n",
      "Mon Oct 31 12:45:40 2022  -Current Dataframe shape : 12557759  x  12\n",
      "Mon Oct 31 12:45:57 2022  -Converted all bases to string datatype and UPPERCASE.\n",
      "Mon Oct 31 12:46:25 2022 Finished fixing allele successfully!\n",
      "Mon Oct 31 12:46:25 2022 Start sanity check for statistics ...\n",
      "Mon Oct 31 12:46:25 2022  -Current Dataframe shape : 12557759  x  12\n",
      "Mon Oct 31 12:46:31 2022  -Checking if  0 <=N<= inf  ...\n",
      "Mon Oct 31 12:46:37 2022  -Removed 0 variants with bad N.\n",
      "Mon Oct 31 12:46:37 2022  -Checking if  0 <=EAF<= 1  ...\n",
      "Mon Oct 31 12:46:40 2022  -Removed 0 variants with bad EAF.\n",
      "Mon Oct 31 12:46:40 2022  -Checking if  5 <=MAC<= inf  ...\n",
      "Mon Oct 31 12:46:44 2022  -Removed 0 variants with bad MAC.\n",
      "Mon Oct 31 12:46:44 2022  -Checking if  5e-300 <= P <= 1  ...\n",
      "Mon Oct 31 12:46:46 2022  -Removed 0 variants with bad P.\n",
      "Mon Oct 31 12:46:46 2022  -Checking if  -10 <BETA)< 10  ...\n",
      "Mon Oct 31 12:46:47 2022  -Removed 0 variants with bad BETA.\n",
      "Mon Oct 31 12:46:47 2022  -Checking if  0 <SE< inf  ...\n",
      "Mon Oct 31 12:46:49 2022  -Removed 0 variants with bad SE.\n",
      "Mon Oct 31 12:46:49 2022  -Checking STATUS...\n",
      "Mon Oct 31 12:46:51 2022  -Coverting STAUTUS to category.\n",
      "Mon Oct 31 12:46:53 2022  -Removed 0 variants with bad statistics in total.\n",
      "Mon Oct 31 12:46:53 2022 Finished sanity check successfully!\n",
      "Mon Oct 31 12:46:55 2022 Start to normalize variants...\n",
      "Mon Oct 31 12:46:55 2022  -Current Dataframe shape : 12557759  x  12\n",
      "Mon Oct 31 12:46:56 2022  -No available variants to normalize..\n",
      "Mon Oct 31 12:46:56 2022 Finished normalizing variants successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.basic_check()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
