{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ad1a0847",
   "metadata": {},
   "source": [
    "# Input and output sumstats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "03531f39",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "deae441a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/25 21:52:10 GWASLab v4.0.0 https://cloufield.github.io/gwaslab/\n",
      "2025/12/25 21:52:10 (C) 2022-2025, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2025/12/25 21:52:10 Python version: 3.12.0 | packaged by conda-forge | (main, Oct  3 2023, 08:43:22) [GCC 12.3.0]\n"
     ]
    }
   ],
   "source": [
    "gl.show_version()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6dc237be",
   "metadata": {},
   "source": [
    "# Input"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "665784c7",
   "metadata": {},
   "source": [
    "## Loading data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ce0f142c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>DIRECTION</th>\n",
       "      <th>N</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:2005486_C_T</td>\n",
       "      <td>1</td>\n",
       "      <td>2005486</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.9863</td>\n",
       "      <td>-0.0969</td>\n",
       "      <td>0.0471</td>\n",
       "      <td>0.039820</td>\n",
       "      <td>+---</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:2247939_AAGG_A</td>\n",
       "      <td>1</td>\n",
       "      <td>2247939</td>\n",
       "      <td>AAGG</td>\n",
       "      <td>A</td>\n",
       "      <td>1960399</td>\n",
       "      <td>0.9966</td>\n",
       "      <td>0.0330</td>\n",
       "      <td>0.1249</td>\n",
       "      <td>0.791900</td>\n",
       "      <td>++--</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:3741853_G_A</td>\n",
       "      <td>1</td>\n",
       "      <td>3741853</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.8849</td>\n",
       "      <td>-0.0375</td>\n",
       "      <td>0.0142</td>\n",
       "      <td>0.008282</td>\n",
       "      <td>----</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:5017526_G_A</td>\n",
       "      <td>1</td>\n",
       "      <td>5017526</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.9822</td>\n",
       "      <td>0.0126</td>\n",
       "      <td>0.0373</td>\n",
       "      <td>0.736200</td>\n",
       "      <td>+-++</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:5843475_C_T</td>\n",
       "      <td>1</td>\n",
       "      <td>5843475</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.9857</td>\n",
       "      <td>-0.0011</td>\n",
       "      <td>0.0433</td>\n",
       "      <td>0.980100</td>\n",
       "      <td>--++</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>X:139318378_A_G</td>\n",
       "      <td>23</td>\n",
       "      <td>139318378</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.5252</td>\n",
       "      <td>-0.0132</td>\n",
       "      <td>0.0071</td>\n",
       "      <td>0.061670</td>\n",
       "      <td>+---</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>X:144540038_G_T</td>\n",
       "      <td>23</td>\n",
       "      <td>144540038</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.9866</td>\n",
       "      <td>0.0379</td>\n",
       "      <td>0.0363</td>\n",
       "      <td>0.295800</td>\n",
       "      <td>++++</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>X:145299627_C_T</td>\n",
       "      <td>23</td>\n",
       "      <td>145299627</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.9984</td>\n",
       "      <td>-0.0663</td>\n",
       "      <td>0.1671</td>\n",
       "      <td>0.691400</td>\n",
       "      <td>---+</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>X:146441317_G_A</td>\n",
       "      <td>23</td>\n",
       "      <td>146441317</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.7345</td>\n",
       "      <td>0.0037</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>0.635100</td>\n",
       "      <td>++-+</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>X:152025052_A_G</td>\n",
       "      <td>23</td>\n",
       "      <td>152025052</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.2417</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>0.0082</td>\n",
       "      <td>0.622200</td>\n",
       "      <td>-++-</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x7f10939b73b0>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(\"../0_sample_data/t2d_bbj.txt.gz\",\n",
    "             snpid=\"SNP\",\n",
    "             chrom=\"CHR\",\n",
    "             pos=\"POS\",\n",
    "             ea=\"ALT\",\n",
    "             nea=\"REF\",\n",
    "             neaf=\"Frq\",\n",
    "             beta=\"BETA\",\n",
    "             se=\"SE\",\n",
    "             p=\"P\",\n",
    "             direction=\"Dir\",\n",
    "             build=\"19\",\n",
    "             n=\"N\", verbose=False)\n",
    "\n",
    "# select just 1000 variants for example\n",
    "mysumstats.random_variants(n=1000, inplace=True, random_state=123,verbose=False)\n",
    "\n",
    "# basic_check\n",
    "mysumstats.basic_check(verbose=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0ca29684",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>DIRECTION</th>\n",
       "      <th>N</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:2005486_C_T</td>\n",
       "      <td>1</td>\n",
       "      <td>2005486</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.9863</td>\n",
       "      <td>-0.0969</td>\n",
       "      <td>0.0471</td>\n",
       "      <td>0.039820</td>\n",
       "      <td>+---</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:2247939_AAGG_A</td>\n",
       "      <td>1</td>\n",
       "      <td>2247939</td>\n",
       "      <td>AAGG</td>\n",
       "      <td>A</td>\n",
       "      <td>1960399</td>\n",
       "      <td>0.9966</td>\n",
       "      <td>0.0330</td>\n",
       "      <td>0.1249</td>\n",
       "      <td>0.791900</td>\n",
       "      <td>++--</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:3741853_G_A</td>\n",
       "      <td>1</td>\n",
       "      <td>3741853</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.8849</td>\n",
       "      <td>-0.0375</td>\n",
       "      <td>0.0142</td>\n",
       "      <td>0.008282</td>\n",
       "      <td>----</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:5017526_G_A</td>\n",
       "      <td>1</td>\n",
       "      <td>5017526</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.9822</td>\n",
       "      <td>0.0126</td>\n",
       "      <td>0.0373</td>\n",
       "      <td>0.736200</td>\n",
       "      <td>+-++</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:5843475_C_T</td>\n",
       "      <td>1</td>\n",
       "      <td>5843475</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.9857</td>\n",
       "      <td>-0.0011</td>\n",
       "      <td>0.0433</td>\n",
       "      <td>0.980100</td>\n",
       "      <td>--++</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>X:139318378_A_G</td>\n",
       "      <td>23</td>\n",
       "      <td>139318378</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.5252</td>\n",
       "      <td>-0.0132</td>\n",
       "      <td>0.0071</td>\n",
       "      <td>0.061670</td>\n",
       "      <td>+---</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>X:144540038_G_T</td>\n",
       "      <td>23</td>\n",
       "      <td>144540038</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.9866</td>\n",
       "      <td>0.0379</td>\n",
       "      <td>0.0363</td>\n",
       "      <td>0.295800</td>\n",
       "      <td>++++</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>X:145299627_C_T</td>\n",
       "      <td>23</td>\n",
       "      <td>145299627</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.9984</td>\n",
       "      <td>-0.0663</td>\n",
       "      <td>0.1671</td>\n",
       "      <td>0.691400</td>\n",
       "      <td>---+</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>X:146441317_G_A</td>\n",
       "      <td>23</td>\n",
       "      <td>146441317</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.7345</td>\n",
       "      <td>0.0037</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>0.635100</td>\n",
       "      <td>++-+</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>X:152025052_A_G</td>\n",
       "      <td>23</td>\n",
       "      <td>152025052</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.2417</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>0.0082</td>\n",
       "      <td>0.622200</td>\n",
       "      <td>-++-</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                SNPID  CHR        POS    EA NEA   STATUS     EAF    BETA  \\\n",
       "0       1:2005486_C_T    1    2005486     C   T  1960099  0.9863 -0.0969   \n",
       "1    1:2247939_AAGG_A    1    2247939  AAGG   A  1960399  0.9966  0.0330   \n",
       "2       1:3741853_G_A    1    3741853     G   A  1960099  0.8849 -0.0375   \n",
       "3       1:5017526_G_A    1    5017526     G   A  1960099  0.9822  0.0126   \n",
       "4       1:5843475_C_T    1    5843475     C   T  1960099  0.9857 -0.0011   \n",
       "..                ...  ...        ...   ...  ..      ...     ...     ...   \n",
       "995   X:139318378_A_G   23  139318378     G   A  1960099  0.5252 -0.0132   \n",
       "996   X:144540038_G_T   23  144540038     G   T  1960099  0.9866  0.0379   \n",
       "997   X:145299627_C_T   23  145299627     C   T  1960099  0.9984 -0.0663   \n",
       "998   X:146441317_G_A   23  146441317     G   A  1960099  0.7345  0.0037   \n",
       "999   X:152025052_A_G   23  152025052     G   A  1960099  0.2417  0.0041   \n",
       "\n",
       "         SE         P DIRECTION       N  \n",
       "0    0.0471  0.039820      +---  191764  \n",
       "1    0.1249  0.791900      ++--  191764  \n",
       "2    0.0142  0.008282      ----  191764  \n",
       "3    0.0373  0.736200      +-++  191764  \n",
       "4    0.0433  0.980100      --++  191764  \n",
       "..      ...       ...       ...     ...  \n",
       "995  0.0071  0.061670      +---  191764  \n",
       "996  0.0363  0.295800      ++++  191764  \n",
       "997  0.1671  0.691400      ---+  191764  \n",
       "998  0.0078  0.635100      ++-+  191764  \n",
       "999  0.0082  0.622200      -++-  191764  \n",
       "\n",
       "[1000 rows x 12 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7abf887",
   "metadata": {},
   "source": [
    "## Load and filtering by pattern"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "505a61e5",
   "metadata": {},
   "source": [
    "### Load and filtering by snp pattern"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e2564d3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/25 21:52:37 GWASLab v4.0.0 https://cloufield.github.io/gwaslab/\n",
      "2025/12/25 21:52:37 (C) 2022-2025, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2025/12/25 21:52:37 Python version: 3.12.0 | packaged by conda-forge | (main, Oct  3 2023, 08:43:22) [GCC 12.3.0]\n",
      "2025/12/25 21:52:37 Start to load format from formatbook....\n",
      "2025/12/25 21:52:37  -auto format meta info:\n",
      "2025/12/25 21:52:37   - format_name  : auto\n",
      "2025/12/25 21:52:37   - format_separator  : \\t\n",
      "2025/12/25 21:52:37   - format_na  : #NA\n",
      "2025/12/25 21:52:37   - format_assumption  : Note: auto-detection assumes A1=EA; Alt=EA and Frq=EAF\n",
      "2025/12/25 21:52:37   - format_version  :  20230328\n",
      "2025/12/25 21:52:37   - Auto-detection mode. Note: auto-detection assumes A1=EA; Alt=EA and Frq=EAF...\n",
      "2025/12/25 21:52:37   - Header conversion source: https://github.com/Cloufield/formatbook/blob/main/formats/auto.json\n",
      "2025/12/25 21:52:37 Start to initialize gl.Sumstats from file :../0_sample_data/t2d_bbj.txt.gz\n",
      "2025/12/25 21:52:37  -Columns used to filter variants: SNP\n",
      "2025/12/25 21:52:37  -Loading only variants with pattern :  ^2:123 ...\n",
      "2025/12/25 21:52:53  -Loaded 5722 variants with pattern : ^2:123 ...\n",
      "2025/12/25 21:52:53  -Reading columns          : POS,ALT,N,CHR,BETA,SNP,REF,Frq,P,SE\n",
      "2025/12/25 21:52:53  -Renaming columns to      : POS,EA,N,CHR,BETA,SNPID,NEA,EAF,P,SE\n",
      "2025/12/25 21:52:53  -Current Dataframe shape : 5722  x  10\n",
      "2025/12/25 21:52:53  -Initiating a status column: STATUS ...\n",
      "2025/12/25 21:52:53  -Genomic coordinates are based on GRCh37/hg19...\n",
      "2025/12/25 21:52:53 Start to reorder the columns ...(v4.0.0)\n",
      "2025/12/25 21:52:53  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P,N\n",
      "2025/12/25 21:52:53 Finished reordering the columns.\n",
      "2025/12/25 21:52:53  -Trying to convert datatype for CHR: string -> Int64...Success\n",
      "2025/12/25 21:52:53  -Column  : SNPID  CHR   POS   EA       NEA      STATUS EAF     BETA    SE      P       N    \n",
      "2025/12/25 21:52:53  -DType   : object Int64 int64 category category int64  float64 float64 float64 float64 int64\n",
      "2025/12/25 21:52:53  -Verified: T      T     T     T        T        T      T       T       T       T       T    \n",
      "2025/12/25 21:52:53  -Current Dataframe memory usage: 0.46 MB\n",
      "2025/12/25 21:52:53 Finished loading data successfully!\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>967172</th>\n",
       "      <td>2:12320_AAT_A</td>\n",
       "      <td>2</td>\n",
       "      <td>12320</td>\n",
       "      <td>AAT</td>\n",
       "      <td>A</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.0572</td>\n",
       "      <td>-0.0158</td>\n",
       "      <td>0.0294</td>\n",
       "      <td>0.5907</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>967173</th>\n",
       "      <td>2:12371_G_C</td>\n",
       "      <td>2</td>\n",
       "      <td>12371</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.0350</td>\n",
       "      <td>-0.0190</td>\n",
       "      <td>0.0294</td>\n",
       "      <td>0.5179</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>967529</th>\n",
       "      <td>2:123233_G_A</td>\n",
       "      <td>2</td>\n",
       "      <td>123233</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.8736</td>\n",
       "      <td>0.0191</td>\n",
       "      <td>0.0135</td>\n",
       "      <td>0.1561</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>967530</th>\n",
       "      <td>2:123332_C_G</td>\n",
       "      <td>2</td>\n",
       "      <td>123332</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.1264</td>\n",
       "      <td>-0.0191</td>\n",
       "      <td>0.0135</td>\n",
       "      <td>0.1561</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>967531</th>\n",
       "      <td>2:123554_C_A</td>\n",
       "      <td>2</td>\n",
       "      <td>123554</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.4748</td>\n",
       "      <td>0.0115</td>\n",
       "      <td>0.0088</td>\n",
       "      <td>0.1929</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1490418</th>\n",
       "      <td>2:123998245_G_A</td>\n",
       "      <td>2</td>\n",
       "      <td>123998245</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.1047</td>\n",
       "      <td>-0.0101</td>\n",
       "      <td>0.0144</td>\n",
       "      <td>0.4834</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1490419</th>\n",
       "      <td>2:123998250_T_C</td>\n",
       "      <td>2</td>\n",
       "      <td>123998250</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.8355</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>0.0119</td>\n",
       "      <td>0.4062</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1490420</th>\n",
       "      <td>2:123998340_A_G</td>\n",
       "      <td>2</td>\n",
       "      <td>123998340</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.9566</td>\n",
       "      <td>-0.0082</td>\n",
       "      <td>0.0222</td>\n",
       "      <td>0.7132</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1490421</th>\n",
       "      <td>2:123999824_A_G</td>\n",
       "      <td>2</td>\n",
       "      <td>123999824</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.8953</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>0.0144</td>\n",
       "      <td>0.4760</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1490422</th>\n",
       "      <td>2:123999965_G_A</td>\n",
       "      <td>2</td>\n",
       "      <td>123999965</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.0021</td>\n",
       "      <td>-0.1152</td>\n",
       "      <td>0.1654</td>\n",
       "      <td>0.4861</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5722 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   SNPID  CHR        POS   EA NEA   STATUS     EAF    BETA  \\\n",
       "967172     2:12320_AAT_A    2      12320  AAT   A  1999999  0.0572 -0.0158   \n",
       "967173       2:12371_G_C    2      12371    G   C  1999999  0.0350 -0.0190   \n",
       "967529      2:123233_G_A    2     123233    G   A  1999999  0.8736  0.0191   \n",
       "967530      2:123332_C_G    2     123332    G   C  1999999  0.1264 -0.0191   \n",
       "967531      2:123554_C_A    2     123554    C   A  1999999  0.4748  0.0115   \n",
       "...                  ...  ...        ...  ...  ..      ...     ...     ...   \n",
       "1490418  2:123998245_G_A    2  123998245    G   A  1999999  0.1047 -0.0101   \n",
       "1490419  2:123998250_T_C    2  123998250    C   T  1999999  0.8355  0.0099   \n",
       "1490420  2:123998340_A_G    2  123998340    G   A  1999999  0.9566 -0.0082   \n",
       "1490421  2:123999824_A_G    2  123999824    G   A  1999999  0.8953  0.0103   \n",
       "1490422  2:123999965_G_A    2  123999965    G   A  1999999  0.0021 -0.1152   \n",
       "\n",
       "             SE       P       N  \n",
       "967172   0.0294  0.5907  191764  \n",
       "967173   0.0294  0.5179  191764  \n",
       "967529   0.0135  0.1561  191764  \n",
       "967530   0.0135  0.1561  191764  \n",
       "967531   0.0088  0.1929  191764  \n",
       "...         ...     ...     ...  \n",
       "1490418  0.0144  0.4834  191764  \n",
       "1490419  0.0119  0.4062  191764  \n",
       "1490420  0.0222  0.7132  191764  \n",
       "1490421  0.0144  0.4760  191764  \n",
       "1490422  0.1654  0.4861  191764  \n",
       "\n",
       "[5722 rows x 11 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats_snp2 = gl.Sumstats(\"../0_sample_data/t2d_bbj.txt.gz\",fmt=\"auto\",\n",
    "             snpid_pat=\"^2:123\",\n",
    "             build=\"19\",\n",
    "             n=\"N\", verbose=True)\n",
    "mysumstats_snp2.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91c529cd",
   "metadata": {},
   "source": [
    "### Load and filtering by chr pattern"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "355951cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/25 21:52:53 GWASLab v4.0.0 https://cloufield.github.io/gwaslab/\n",
      "2025/12/25 21:52:53 (C) 2022-2025, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2025/12/25 21:52:53 Python version: 3.12.0 | packaged by conda-forge | (main, Oct  3 2023, 08:43:22) [GCC 12.3.0]\n",
      "2025/12/25 21:52:53 Start to load format from formatbook....\n",
      "2025/12/25 21:52:53  -auto format meta info:\n",
      "2025/12/25 21:52:53   - format_name  : auto\n",
      "2025/12/25 21:52:53   - format_separator  : \\t\n",
      "2025/12/25 21:52:53   - format_na  : #NA\n",
      "2025/12/25 21:52:53   - format_assumption  : Note: auto-detection assumes A1=EA; Alt=EA and Frq=EAF\n",
      "2025/12/25 21:52:53   - format_version  :  20230328\n",
      "2025/12/25 21:52:53   - Auto-detection mode. Note: auto-detection assumes A1=EA; Alt=EA and Frq=EAF...\n",
      "2025/12/25 21:52:53   - Header conversion source: https://github.com/Cloufield/formatbook/blob/main/formats/auto.json\n",
      "2025/12/25 21:52:53 Start to initialize gl.Sumstats from file :../0_sample_data/t2d_bbj.txt.gz\n",
      "2025/12/25 21:52:53  -Columns used to filter variants: CHR\n",
      "2025/12/25 21:52:53  -Loading only variants on chromosome with pattern : ^22 ...\n",
      "2025/12/25 21:53:08  -Loaded 157050 variants on chromosome with pattern :^22 ...\n",
      "2025/12/25 21:53:08  -Reading columns          : POS,ALT,N,CHR,BETA,SNP,REF,Frq,P,SE\n",
      "2025/12/25 21:53:08  -Renaming columns to      : POS,EA,N,CHR,BETA,SNPID,NEA,EAF,P,SE\n",
      "2025/12/25 21:53:08  -Current Dataframe shape : 157050  x  10\n",
      "2025/12/25 21:53:08  -Initiating a status column: STATUS ...\n",
      "2025/12/25 21:53:08  -Genomic coordinates are based on GRCh37/hg19...\n",
      "2025/12/25 21:53:08 Start to reorder the columns ...(v4.0.0)\n",
      "2025/12/25 21:53:08  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P,N\n",
      "2025/12/25 21:53:08 Finished reordering the columns.\n",
      "2025/12/25 21:53:08  -Trying to convert datatype for CHR: string -> Int64...Success\n",
      "2025/12/25 21:53:09  -Column  : SNPID  CHR   POS   EA       NEA      STATUS EAF     BETA    SE      P       N    \n",
      "2025/12/25 21:53:09  -DType   : object Int64 int64 category category int64  float64 float64 float64 float64 int64\n",
      "2025/12/25 21:53:09  -Verified: T      T     T     T        T        T      T       T       T       T       T    \n",
      "2025/12/25 21:53:09  -Current Dataframe memory usage: 12.93 MB\n",
      "2025/12/25 21:53:09 Finished loading data successfully!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/yunye/anaconda3/envs/py312/lib/python3.12/site-packages/gwaslab/io/io_preformat_input.py:860: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  sumstats_filtered = pd.concat([chunk[chunk[chunk_chrom].str.match(chrom_pat, case=False,na=False) ] for chunk in sumstats_iter])\n",
      "/home/yunye/anaconda3/envs/py312/lib/python3.12/site-packages/gwaslab/io/io_preformat_input.py:860: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  sumstats_filtered = pd.concat([chunk[chunk[chunk_chrom].str.match(chrom_pat, case=False,na=False) ] for chunk in sumstats_iter])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12071920</th>\n",
       "      <td>22:16847963_A_G</td>\n",
       "      <td>22</td>\n",
       "      <td>16847963</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.5903</td>\n",
       "      <td>0.0115</td>\n",
       "      <td>0.0163</td>\n",
       "      <td>0.4819</td>\n",
       "      <td>166718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12071921</th>\n",
       "      <td>22:16848015_C_G</td>\n",
       "      <td>22</td>\n",
       "      <td>16848015</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.6942</td>\n",
       "      <td>0.0110</td>\n",
       "      <td>0.0155</td>\n",
       "      <td>0.4795</td>\n",
       "      <td>166718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12071922</th>\n",
       "      <td>22:16848470_A_G</td>\n",
       "      <td>22</td>\n",
       "      <td>16848470</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.5918</td>\n",
       "      <td>0.0118</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.4487</td>\n",
       "      <td>166718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12071923</th>\n",
       "      <td>22:16848520_A_T</td>\n",
       "      <td>22</td>\n",
       "      <td>16848520</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.5918</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.0155</td>\n",
       "      <td>0.4537</td>\n",
       "      <td>166718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12071924</th>\n",
       "      <td>22:16849105_A_G</td>\n",
       "      <td>22</td>\n",
       "      <td>16849105</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.5919</td>\n",
       "      <td>0.0109</td>\n",
       "      <td>0.0154</td>\n",
       "      <td>0.4810</td>\n",
       "      <td>166718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12228965</th>\n",
       "      <td>22:51230048_T_C</td>\n",
       "      <td>22</td>\n",
       "      <td>51230048</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.9988</td>\n",
       "      <td>-0.0645</td>\n",
       "      <td>0.2269</td>\n",
       "      <td>0.7762</td>\n",
       "      <td>166718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12228966</th>\n",
       "      <td>22:51230086_A_T</td>\n",
       "      <td>22</td>\n",
       "      <td>51230086</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.9906</td>\n",
       "      <td>-0.0543</td>\n",
       "      <td>0.0766</td>\n",
       "      <td>0.4780</td>\n",
       "      <td>166718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12228967</th>\n",
       "      <td>22:51237069_T_C</td>\n",
       "      <td>22</td>\n",
       "      <td>51237069</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.9883</td>\n",
       "      <td>-0.0356</td>\n",
       "      <td>0.0720</td>\n",
       "      <td>0.6212</td>\n",
       "      <td>191764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12228968</th>\n",
       "      <td>22:51239678_G_T</td>\n",
       "      <td>22</td>\n",
       "      <td>51239678</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.0592</td>\n",
       "      <td>-0.0038</td>\n",
       "      <td>0.0331</td>\n",
       "      <td>0.9077</td>\n",
       "      <td>166718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12228969</th>\n",
       "      <td>22:51239752_G_A</td>\n",
       "      <td>22</td>\n",
       "      <td>51239752</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.2937</td>\n",
       "      <td>0.2318</td>\n",
       "      <td>0.2051</td>\n",
       "      <td>166718</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>157050 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    SNPID  CHR       POS EA NEA   STATUS     EAF    BETA  \\\n",
       "12071920  22:16847963_A_G   22  16847963  G   A  1999999  0.5903  0.0115   \n",
       "12071921  22:16848015_C_G   22  16848015  G   C  1999999  0.6942  0.0110   \n",
       "12071922  22:16848470_A_G   22  16848470  G   A  1999999  0.5918  0.0118   \n",
       "12071923  22:16848520_A_T   22  16848520  T   A  1999999  0.5918  0.0116   \n",
       "12071924  22:16849105_A_G   22  16849105  G   A  1999999  0.5919  0.0109   \n",
       "...                   ...  ...       ... ..  ..      ...     ...     ...   \n",
       "12228965  22:51230048_T_C   22  51230048  C   T  1999999  0.9988 -0.0645   \n",
       "12228966  22:51230086_A_T   22  51230086  T   A  1999999  0.9906 -0.0543   \n",
       "12228967  22:51237069_T_C   22  51237069  C   T  1999999  0.9883 -0.0356   \n",
       "12228968  22:51239678_G_T   22  51239678  G   T  1999999  0.0592 -0.0038   \n",
       "12228969  22:51239752_G_A   22  51239752  G   A  1999999  0.0013  0.2937   \n",
       "\n",
       "              SE       P       N  \n",
       "12071920  0.0163  0.4819  166718  \n",
       "12071921  0.0155  0.4795  166718  \n",
       "12071922  0.0156  0.4487  166718  \n",
       "12071923  0.0155  0.4537  166718  \n",
       "12071924  0.0154  0.4810  166718  \n",
       "...          ...     ...     ...  \n",
       "12228965  0.2269  0.7762  166718  \n",
       "12228966  0.0766  0.4780  166718  \n",
       "12228967  0.0720  0.6212  191764  \n",
       "12228968  0.0331  0.9077  166718  \n",
       "12228969  0.2318  0.2051  166718  \n",
       "\n",
       "[157050 rows x 11 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats_chr22 = gl.Sumstats(\"../0_sample_data/t2d_bbj.txt.gz\",fmt=\"auto\",\n",
    "             chrom_pat=\"^22\",\n",
    "             build=\"19\",\n",
    "             n=\"N\", verbose=True)\n",
    "mysumstats_chr22.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2484d54",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "81d94826",
   "metadata": {},
   "source": [
    "# Output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "277d2521",
   "metadata": {},
   "source": [
    "## general output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c332e77d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/25 21:53:09 Start to convert the output sumstats in:  gwaslab  format\n",
      "2025/12/25 21:53:09  -Formatting statistics ...\n",
      "2025/12/25 21:53:09  -Float statistics formats:\n",
      "2025/12/25 21:53:09   - Columns       : ['EAF', 'BETA', 'SE', 'P']\n",
      "2025/12/25 21:53:09   - Output formats: ['{:.4g}', '{:.4f}', '{:.4f}', '{:.4e}']\n",
      "2025/12/25 21:53:09  -Start outputting sumstats in gwaslab format...\n",
      "2025/12/25 21:53:09  -gwaslab format will be loaded...\n",
      "2025/12/25 21:53:09  -gwaslab format meta info:\n",
      "2025/12/25 21:53:09   - format_name  : gwaslab\n",
      "2025/12/25 21:53:09   - format_source  : https://cloufield.github.io/gwaslab/\n",
      "2025/12/25 21:53:09   - format_version  : 20231220_v4\n",
      "2025/12/25 21:53:09  -Output path: ./mysumstats.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Output columns: SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P,DIRECTION,N\n",
      "2025/12/25 21:53:09  -Writing sumstats to: ./mysumstats.gwaslab.tsv.gz...\n",
      "2025/12/25 21:53:09  -Saving log file to: ./mysumstats.gwaslab.log\n",
      "2025/12/25 21:53:09 Finished outputting successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.to_format(\"./mysumstats\",fmt=\"gwaslab\",xymt_number=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ea7fe203",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNPID\tCHR\tPOS\tEA\tNEA\tSTATUS\tEAF\tBETA\tSE\tP\tDIRECTION\tN\n",
      "1:2005486_C_T\t1\t2005486\tC\tT\t1960099\t0.9863\t-0.0969\t0.0471\t3.9820e-02\t+---\t191764\n",
      "1:2247939_AAGG_A\t1\t2247939\tAAGG\tA\t1960399\t0.9966\t0.0330\t0.1249\t7.9190e-01\t++--\t191764\n",
      "1:3741853_G_A\t1\t3741853\tG\tA\t1960099\t0.8849\t-0.0375\t0.0142\t8.2820e-03\t----\t191764\n",
      "1:5017526_G_A\t1\t5017526\tG\tA\t1960099\t0.9822\t0.0126\t0.0373\t7.3620e-01\t+-++\t191764\n",
      "1:5843475_C_T\t1\t5843475\tC\tT\t1960099\t0.9857\t-0.0011\t0.0433\t9.8010e-01\t--++\t191764\n",
      "1:9405103_T_C\t1\t9405103\tC\tT\t1960099\t0.0021\t-0.0729\t0.1516\t6.3050e-01\t+---\t191764\n",
      "1:9443411_G_A\t1\t9443411\tG\tA\t1960099\t0.9916\t0.0362\t0.0532\t4.9690e-01\t+-++\t191764\n",
      "1:12866348_G_C\t1\t12866348\tG\tC\t1960099\t0.9728\t-0.0352\t0.0431\t4.1450e-01\t---+\t191764\n",
      "1:14466316_A_G\t1\t14466316\tG\tA\t1960099\t0.6942\t-0.0042\t0.0096\t6.6360e-01\t--+-\t191764\n",
      "\n",
      "gzip: stdout: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "!zcat mysumstats.gwaslab.tsv.gz | head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9e5ff639",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X:121023171_A_ACTT\t23\t121023171\tACTT\tA\t1960399\t0.5117\t0.0096\t0.0071\t1.7390e-01\t++++\t191764\n",
      "X:134838698_C_CTA\t23\t134838698\tC\tCTA\t1960399\t0.9355\t0.0060\t0.0145\t6.8080e-01\t++-+\t191764\n",
      "X:135939006_G_T\t23\t135939006\tG\tT\t1960099\t0.2068\t-0.0037\t0.0085\t6.6300e-01\t-+-+\t191764\n",
      "X:136020644_C_T\t23\t136020644\tC\tT\t1960099\t0.8756\t0.0103\t0.0106\t3.3370e-01\t+++-\t191764\n",
      "X:138148816_C_T\t23\t138148816\tC\tT\t1960099\t0.7842\t0.0089\t0.0088\t3.1330e-01\t++++\t191764\n",
      "X:139318378_A_G\t23\t139318378\tG\tA\t1960099\t0.5252\t-0.0132\t0.0071\t6.1670e-02\t+---\t191764\n",
      "X:144540038_G_T\t23\t144540038\tG\tT\t1960099\t0.9866\t0.0379\t0.0363\t2.9580e-01\t++++\t191764\n",
      "X:145299627_C_T\t23\t145299627\tC\tT\t1960099\t0.9984\t-0.0663\t0.1671\t6.9140e-01\t---+\t191764\n",
      "X:146441317_G_A\t23\t146441317\tG\tA\t1960099\t0.7345\t0.0037\t0.0078\t6.3510e-01\t++-+\t191764\n",
      "X:152025052_A_G\t23\t152025052\tG\tA\t1960099\t0.2417\t0.0041\t0.0082\t6.2220e-01\t-++-\t191764\n"
     ]
    }
   ],
   "source": [
    "!zcat mysumstats.gwaslab.tsv.gz | tail"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68af95ae",
   "metadata": {},
   "source": [
    "## output each chromosome to a single file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d26a2ccc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/25 21:53:09 Start to convert the output sumstats in:  gwaslab  format\n",
      "2025/12/25 21:53:09  -Formatting statistics ...\n",
      "2025/12/25 21:53:09  -Float statistics formats:\n",
      "2025/12/25 21:53:09   - Columns       : ['EAF', 'BETA', 'SE', 'P']\n",
      "2025/12/25 21:53:09   - Output formats: ['{:.4g}', '{:.4f}', '{:.4f}', '{:.4e}']\n",
      "2025/12/25 21:53:09  -Start outputting sumstats in gwaslab format...\n",
      "2025/12/25 21:53:09  -gwaslab format will be loaded...\n",
      "2025/12/25 21:53:09  -gwaslab format meta info:\n",
      "2025/12/25 21:53:09   - format_name  : gwaslab\n",
      "2025/12/25 21:53:09   - format_source  : https://cloufield.github.io/gwaslab/\n",
      "2025/12/25 21:53:09   - format_version  : 20231220_v4\n",
      "2025/12/25 21:53:09  -Output path: ./mysumstats.@.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Output columns: SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P,DIRECTION,N\n",
      "2025/12/25 21:53:09  -Writing sumstats to: ./mysumstats.@.gwaslab.tsv.gz...\n",
      "2025/12/25 21:53:09   -@ detected: writing each chromosome to a single file...\n",
      "2025/12/25 21:53:09   -Chromosomes:[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]...\n",
      "2025/12/25 21:53:09  -Saving log file to: ./mysumstats.@.gwaslab.log\n",
      "2025/12/25 21:53:09 Finished outputting successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.to_format(\"./mysumstats.@\",fmt=\"gwaslab\",xymt_number=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2244e9ad",
   "metadata": {},
   "source": [
    "## load sumstats for each chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e62ca850",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/25 21:53:09 GWASLab v4.0.0 https://cloufield.github.io/gwaslab/\n",
      "2025/12/25 21:53:09 (C) 2022-2025, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2025/12/25 21:53:09 Python version: 3.12.0 | packaged by conda-forge | (main, Oct  3 2023, 08:43:22) [GCC 12.3.0]\n",
      "2025/12/25 21:53:09 Start to load format from formatbook....\n",
      "2025/12/25 21:53:09  -gwaslab format meta info:\n",
      "2025/12/25 21:53:09   - format_name  : gwaslab\n",
      "2025/12/25 21:53:09   - format_source  : https://cloufield.github.io/gwaslab/\n",
      "2025/12/25 21:53:09   - format_version  : 20231220_v4\n",
      "2025/12/25 21:53:09  -Detected @ in path: load sumstats by each chromosome...\n",
      "2025/12/25 21:53:09  -Chromosomes detected: 1,10,11,12,13,14,15,16,17,18,19,2,20,21,22,23,3,4,5,6,7,8,9\n",
      "2025/12/25 21:53:09 Start to initialize gl.Sumstats from files with pattern :./mysumstats.@.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.1.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.10.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.11.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.12.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.13.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.14.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.15.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.16.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.17.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.18.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.19.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.2.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.20.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.21.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.22.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.23.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.3.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.4.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.5.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.6.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.7.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.8.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Loading:./mysumstats.9.gwaslab.tsv.gz\n",
      "2025/12/25 21:53:09  -Merging sumstats for chromosomes: 1,10,11,12,13,14,15,16,17,18,19,2,20,21,22,23,3,4,5,6,7,8,9\n",
      "2025/12/25 21:53:09  -Reading columns          : POS,N,CHR,DIRECTION,BETA,EA,P,EAF,SNPID,STATUS,NEA,SE\n",
      "2025/12/25 21:53:09  -Renaming columns to      : POS,N,CHR,DIRECTION,BETA,EA,P,EAF,SNPID,STATUS,NEA,SE\n",
      "2025/12/25 21:53:09  -Current Dataframe shape : 1000  x  12\n",
      "2025/12/25 21:53:09  -Initiating a status column: STATUS ...\n",
      "2025/12/25 21:53:09 #WARNING! Version of genomic coordinates is unknown...\n",
      "2025/12/25 21:53:09 Start to reorder the columns ...(v4.0.0)\n",
      "2025/12/25 21:53:09  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P,DIRECTION,N\n",
      "2025/12/25 21:53:09 Finished reordering the columns.\n",
      "2025/12/25 21:53:09  -Trying to convert datatype for CHR: string -> Int64...Success\n",
      "2025/12/25 21:53:09  -Column  : SNPID  CHR   POS   EA       NEA      STATUS EAF     BETA    SE      P       DIRECTION N    \n",
      "2025/12/25 21:53:09  -DType   : object Int64 int64 category category int64  float64 float64 float64 float64 object    int64\n",
      "2025/12/25 21:53:09  -Verified: T      T     T     T        T        T      T       T       T       T       T         T    \n",
      "2025/12/25 21:53:09  -Current Dataframe memory usage: 0.08 MB\n",
      "2025/12/25 21:53:09 Finished loading data successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(\"./mysumstats.@.gwaslab.tsv.gz\",\n",
    "             fmt=\"gwaslab\", verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab477af2",
   "metadata": {},
   "source": [
    "## Check available formats"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5081e7af",
   "metadata": {},
   "source": [
    "List the formats that GWASLab supports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ff45fe2c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/25 21:53:09 Available formats: auto,auto_0,auto_1,auto_2,auto_neaf,auto_ref,bolt_lmm,ccgwas,cojo,fastgwa,genomicsem,gwascatalog,gwascatalog_hm,gwaslab,ldsc,mesusie,metal,mrmega,mtag,pgscatalog,pgscatalog_hm,pheweb,plink,plink2,plink2_firth,plink2_linear,plink2_logistic,plink_assoc,plink_bim,plink_dosage,plink_fam,plink_fisher,plink_linear,plink_logistic,plink_psam,plink_pvar,popcorn,regenie,regenie_gene,saige,ssf,template,vcf\n"
     ]
    }
   ],
   "source": [
    "formats = gl.list_formats()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38d598c4",
   "metadata": {},
   "source": [
    "Check the contents of the specified format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8009e7f4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/25 21:53:09 Available formats:2025/12/25 21:53:09 meta_data2025/12/25 21:53:09 format_dict2025/12/25 21:53:09 \n",
      "2025/12/25 21:53:09 {'format_name': 'ssf', 'format_source': 'https://www.biorxiv.org/content/10.1101/2022.07.15.500230v1.full', 'format_cite_name': 'GWAS-SSF v0.1', 'format_separator': '\\t', 'format_na': '#NA', 'format_comment': None, 'format_col_order': ['chromosome', 'base_pair_location', 'effect_allele', 'other_allele', 'beta', 'odds_ratio', 'hazard_ratio', 'standard_error', 'effect_allele_frequency', 'p_value', 'neg_log_10_p_value', 'ci_upper', 'ci_lower', 'rsid', 'variant_id', 'info', 'ref_allele', 'n'], 'format_version': 20230328}2025/12/25 21:53:09 {'variant_id': 'SNPID', 'rsid': 'rsID', 'chromosome': 'CHR', 'base_pair_location': 'POS', 'other_allele': 'NEA', 'effect_allele': 'EA', 'effect_allele_frequency': 'EAF', 'n': 'N', 'beta': 'BETA', 'standard_error': 'SE', 'p_value': 'P', 'neg_log_10_p_value': 'MLOG10P', 'info': 'INFO', 'odds_ratio': 'OR', 'hazard_ratio': 'HR', 'ci_lower': 'OR_95L', 'ci_upper': 'OR_95U'}"
     ]
    }
   ],
   "source": [
    "ssf_format_dict = gl.check_format(\"ssf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7391280",
   "metadata": {},
   "source": [
    "## Formatting and saving"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a565395",
   "metadata": {},
   "source": [
    "### get ready for submission to gwas catalog (GWAS-ssf format)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c33d4a82",
   "metadata": {},
   "source": [
    "- `fmt`: specify the output format\n",
    "- `ssfmeta`: if True, output the meta file\n",
    "- `md5sum`: if True, create a file with the md5sum of the output sumstats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "69c2668f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/25 21:53:09 Start to convert the output sumstats in:  ssf  format\n",
      "2025/12/25 21:53:09  -Formatting statistics ...\n",
      "2025/12/25 21:53:09  -Float statistics formats:\n",
      "2025/12/25 21:53:09   - Columns       : ['EAF', 'BETA', 'SE', 'P']\n",
      "2025/12/25 21:53:09   - Output formats: ['{:.4g}', '{:.4f}', '{:.4f}', '{:.4e}']\n",
      "2025/12/25 21:53:09  -Replacing SNPID separator from \":\" to \"_\"...\n",
      "2025/12/25 21:53:09  -Start outputting sumstats in ssf format...\n",
      "2025/12/25 21:53:09  -ssf format will be loaded...\n",
      "2025/12/25 21:53:09  -ssf format meta info:\n",
      "2025/12/25 21:53:09   - format_name  : ssf\n",
      "2025/12/25 21:53:09   - format_source  : https://www.biorxiv.org/content/10.1101/2022.07.15.500230v1.full\n",
      "2025/12/25 21:53:09   - format_cite_name  : GWAS-SSF v0.1\n",
      "2025/12/25 21:53:09   - format_separator  : \\t\n",
      "2025/12/25 21:53:09   - format_na  : #NA\n",
      "2025/12/25 21:53:09   - format_col_order  : chromosome,base_pair_location,effect_allele,other_allele,beta,odds_ratio,hazard_ratio,standard_error,effect_allele_frequency,p_value,neg_log_10_p_value,ci_upper,ci_lower,rsid,variant_id,info,ref_allele,n\n",
      "2025/12/25 21:53:09   - format_version  :  20230328\n",
      "2025/12/25 21:53:09  -gwaslab to ssf format dictionary:\n",
      "2025/12/25 21:53:09   - gwaslab keys: SNPID,rsID,CHR,POS,NEA,EA,EAF,N,BETA,SE,P,MLOG10P,INFO,OR,HR,OR_95L,OR_95U\n",
      "2025/12/25 21:53:09   - ssf values: variant_id,rsid,chromosome,base_pair_location,other_allele,effect_allele,effect_allele_frequency,n,beta,standard_error,p_value,neg_log_10_p_value,info,odds_ratio,hazard_ratio,ci_lower,ci_upper\n",
      "2025/12/25 21:53:09  -Output path: ./mysumstats.ssf.tsv.gz\n",
      "2025/12/25 21:53:09  -Output columns: chromosome,base_pair_location,effect_allele,other_allele,beta,standard_error,effect_allele_frequency,p_value,variant_id,n\n",
      "2025/12/25 21:53:09  -Writing sumstats to: ./mysumstats.ssf.tsv.gz...\n",
      "2025/12/25 21:53:09  -md5sum hashing for the file: ./mysumstats.ssf.tsv.gz\n",
      "2025/12/25 21:53:09  -md5sum path: ./mysumstats.ssf.tsv.gz.md5sum\n",
      "2025/12/25 21:53:09  -md5sum: e8e9b4cf01d7b0166d0cd9a208a2808f\n",
      "2025/12/25 21:53:09  -Exporting SSF-style meta data to ./mysumstats.ssf.tsv.gz.ssf.tsv-meta.yaml\n",
      "2025/12/25 21:53:09  -Saving log file to: ./mysumstats.ssf.log\n",
      "2025/12/25 21:53:09 Finished outputting successfully!\n",
      "2025/12/25 21:53:09 Validating SSF format output...\n",
      "2025/12/25 21:53:09 Using built-in SSF validator (gwas-ssf CLI not available)...\n",
      "2025/12/25 21:53:09 Validating file extension...\n",
      "2025/12/25 21:53:09 ✓ File extension OK\n",
      "2025/12/25 21:53:09 Validating column order...\n",
      "2025/12/25 21:53:09 ✓ Column order OK\n",
      "2025/12/25 21:53:09 Validating chromosomes...\n",
      "2025/12/25 21:53:09 ✓ Chromosomes OK (All autosomes exist. Optional chromosomes ['24', '25'] do not exist.)\n",
      "2025/12/25 21:53:09 Validating minimum row count...\n",
      "2025/12/25 21:53:09 #WARNING! ✗ SSF validation failed: The file has fewer than the minimum rows required: 1000 < 100000\n",
      "2025/12/25 21:53:09 #WARNING!   - The file has fewer than the minimum rows required: 1000 < 100000\n"
     ]
    }
   ],
   "source": [
    "mysumstats.to_format(\"./mysumstats\", fmt=\"ssf\", ssfmeta=True, md5sum=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d276a228",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosome\tbase_pair_location\teffect_allele\tother_allele\tbeta\tstandard_error\teffect_allele_frequency\tp_value\tvariant_id\tn\n",
      "1\t2005486\tC\tT\t-0.0969\t0.0471\t0.9863\t3.9820e-02\t1_2005486_C_T\t191764\n",
      "1\t2247939\tAAGG\tA\t0.0330\t0.1249\t0.9966\t7.9190e-01\t1_2247939_AAGG_A\t191764\n",
      "1\t3741853\tG\tA\t-0.0375\t0.0142\t0.8849\t8.2820e-03\t1_3741853_G_A\t191764\n",
      "1\t5017526\tG\tA\t0.0126\t0.0373\t0.9822\t7.3620e-01\t1_5017526_G_A\t191764\n",
      "1\t5843475\tC\tT\t-0.0011\t0.0433\t0.9857\t9.8010e-01\t1_5843475_C_T\t191764\n",
      "1\t9405103\tC\tT\t-0.0729\t0.1516\t0.0021\t6.3050e-01\t1_9405103_T_C\t191764\n",
      "1\t9443411\tG\tA\t0.0362\t0.0532\t0.9916\t4.9690e-01\t1_9443411_G_A\t191764\n",
      "1\t12866348\tG\tC\t-0.0352\t0.0431\t0.9728\t4.1450e-01\t1_12866348_G_C\t191764\n",
      "1\t14466316\tG\tA\t-0.0042\t0.0096\t0.6942\t6.6360e-01\t1_14466316_A_G\t191764\n"
     ]
    }
   ],
   "source": [
    "!zcat mysumstats.ssf.tsv.gz | head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "46f76d7c-d17c-40bb-99cc-29de58552f2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "e8e9b4cf01d7b0166d0cd9a208a2808f\n"
     ]
    }
   ],
   "source": [
    "!head mysumstats.ssf.tsv.gz.md5sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4b5668e6-e368-42b1-9763-2ca1488c6f1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coordinate_system: 1-based\n",
      "data_file_md5sum: e8e9b4cf01d7b0166d0cd9a208a2808f\n",
      "data_file_name: ./mysumstats.ssf.tsv.gz\n",
      "date_last_modified: 2025-12-25-21:53:09\n",
      "file_type: GWAS-SSF v0.1\n",
      "genome_assembly: Unknown\n",
      "genotyping_technology: Unknown\n",
      "gwas_id: Unknown\n",
      "gwaslab:\n",
      "  basic_check:\n"
     ]
    }
   ],
   "source": [
    "!head ./mysumstats.ssf.tsv.gz.ssf.tsv-meta.yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45077731",
   "metadata": {},
   "source": [
    "### ldsc default format"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac68ed6f",
   "metadata": {},
   "source": [
    "- `hapmap3`: if True, only output hapmap3 SNPs\n",
    "- `exclude_hla`: if True, exclude variants in HLA region from output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e8204a6e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/25 21:53:10 Start to convert the output sumstats in:  ldsc  format\n",
      "2025/12/25 21:53:10 Start to exclude variants in HLA regions ...(v4.0.0)\n",
      "2025/12/25 21:53:10  -Current Dataframe shape : 1000 x 12 ; Memory usage: 0.08 MB\n",
      "2025/12/25 21:53:10  -Excluded 3 variants in HLA region (chr6: 25000000-34000000 )...\n",
      "2025/12/25 21:53:10  -Filtered out variants: 3\n",
      "2025/12/25 21:53:10  -Current Dataframe shape : 997 x 12 ; Memory usage: 0.09 MB\n",
      "2025/12/25 21:53:10 Finished filtering variants.\n",
      "2025/12/25 21:53:10  -Genomic coordinates are based on GRCh37/hg19...\n",
      "2025/12/25 21:53:10 Start to extract HapMap3 SNPs ...(v4.0.0)\n",
      "2025/12/25 21:53:10  -Current Dataframe shape : 997 x 12 ; Memory usage: 0.09 MB\n",
      "2025/12/25 21:53:10  -Loading Hapmap3 variants from built-in datasets...\n",
      "2025/12/25 21:53:11  -Since rsID not in sumstats, CHR:POS( build 19) will be used for matching...\n",
      "2025/12/25 21:53:12  -Checking if alleles are same...\n",
      "2025/12/25 21:53:12  -Variants with macthed alleles: 81\n",
      "2025/12/25 21:53:12  -Raw input contains 81 Hapmap3 variants based on CHR:POS...\n",
      "2025/12/25 21:53:12  -Current Dataframe shape : 997 x 13 ; Memory usage: 0.10 MB\n",
      "2025/12/25 21:53:12 Finished extracting HapMap3 SNPs.\n",
      "2025/12/25 21:53:12  -Extract 81 variants in Hapmap3 datasets for build 19.\n",
      "2025/12/25 21:53:12  -Formatting statistics ...\n",
      "2025/12/25 21:53:12  -Float statistics formats:\n",
      "2025/12/25 21:53:12   - Columns       : ['EAF', 'BETA', 'SE', 'P']\n",
      "2025/12/25 21:53:12   - Output formats: ['{:.4g}', '{:.4f}', '{:.4f}', '{:.4e}']\n",
      "2025/12/25 21:53:12  -Start outputting sumstats in ldsc format...\n",
      "2025/12/25 21:53:12  -ldsc format will be loaded...\n",
      "2025/12/25 21:53:12  -ldsc format meta info:\n",
      "2025/12/25 21:53:12   - format_name  : ldsc\n",
      "2025/12/25 21:53:12   - format_source  : https://github.com/bulik/ldsc/wiki/Summary-Statistics-File-Format\n",
      "2025/12/25 21:53:12   - format_source2  : https://github.com/bulik/ldsc/blob/master/munge_sumstats.py\n",
      "2025/12/25 21:53:12   - format_version  :  20150306\n",
      "2025/12/25 21:53:12  -gwaslab to ldsc format dictionary:\n",
      "2025/12/25 21:53:12   - gwaslab keys: rsID,NEA,EA,EAF,N,BETA,P,Z,INFO,OR,CHR,POS\n",
      "2025/12/25 21:53:12   - ldsc values: SNP,A2,A1,Frq,N,Beta,P,Z,INFO,OR,CHR,POS\n",
      "2025/12/25 21:53:12  -Output path: ./mysumstats.hapmap3.noMHC.ldsc.tsv.gz\n",
      "2025/12/25 21:53:12  -Output columns: CHR,POS,A1,A2,Frq,Beta,P,N,SNP\n",
      "2025/12/25 21:53:12  -Writing sumstats to: ./mysumstats.hapmap3.noMHC.ldsc.tsv.gz...\n",
      "2025/12/25 21:53:12  -Saving log file to: ./mysumstats.hapmap3.noMHC.ldsc.log\n",
      "2025/12/25 21:53:12 Finished outputting successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.to_format(\"./mysumstats\",fmt=\"ldsc\",hapmap3=True,exclude_hla=True,build=\"19\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "31c947fd-8499-4aea-8358-938ce267f2b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHR\tPOS\tA1\tA2\tFrq\tBeta\tP\tN\tSNP\n",
      "1\t14900419\tG\tA\t0.3952\t0.0144\t1.3750e-01\t191764\trs6703840\n",
      "1\t19593199\tC\tT\t0.1323\t-0.0127\t3.2570e-01\t191764\trs7527253\n",
      "1\t35282297\tG\tA\t0.5434\t0.0041\t6.4190e-01\t191764\trs1407135\n",
      "1\t66001402\tC\tT\t0.2103\t-0.0148\t1.7720e-01\t191764\trs1171261\n",
      "1\t83510491\tG\tA\t0.0025\t0.0378\t6.9800e-01\t191764\trs2022427\n",
      "1\t166110693\tC\tT\t0.8627\t0.0286\t2.5250e-02\t191764\trs4656480\n",
      "1\t175886511\tG\tA\t0.1828\t-0.0141\t2.2480e-01\t191764\trs6656281\n",
      "1\t181612041\tC\tT\t0.9603\t0.0135\t5.5050e-01\t191764\trs199955\n",
      "1\t196329362\tC\tT\t0.0301\t0.0300\t2.5060e-01\t191764\trs11801881\n"
     ]
    }
   ],
   "source": [
    "!zcat ./mysumstats.hapmap3.noMHC.ldsc.tsv.gz | head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20f2c297",
   "metadata": {},
   "source": [
    "### vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6cdcf20",
   "metadata": {},
   "source": [
    "- `bgzip` : if True, bgzip the output vcf/bed\n",
    "- `tabix` : if True, index the bgzipped file with tabix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7ddc3164",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/25 21:53:12 Start to convert the output sumstats in:  vcf  format\n",
      "2025/12/25 21:53:12  -Formatting statistics ...\n",
      "2025/12/25 21:53:12  -Float statistics formats:\n",
      "2025/12/25 21:53:12   - Columns       : ['EAF', 'BETA', 'SE', 'P']\n",
      "2025/12/25 21:53:12   - Output formats: ['{:.4g}', '{:.4f}', '{:.4f}', '{:.4e}']\n",
      "2025/12/25 21:53:12  -Start outputting sumstats in vcf format...\n",
      "2025/12/25 21:53:12  -vcf format will be loaded...\n",
      "2025/12/25 21:53:12  -vcf format meta info:\n",
      "2025/12/25 21:53:12   - format_name  : vcf\n",
      "2025/12/25 21:53:12   - format_source  : https://github.com/MRCIEU/gwas-vcf-specification/tree/1.0.0\n",
      "2025/12/25 21:53:12   - format_version  :  20220923\n",
      "2025/12/25 21:53:12   - format_citation  : Lyon, M.S., Andrews, S.J., Elsworth, B. et al. The variant call format provides efficient and robust storage of GWAS summary statistics. Genome Biol 22, 32 (2021). https://doi.org/10.1186/s13059-020-02248-0\n",
      "2025/12/25 21:53:12   - format_fixed  : #CHROM,POS,ID,REF,ALT,QUAL,FILTER,INFO,FORMAT\n",
      "2025/12/25 21:53:12   - format_format  : ID,SS,ES,SE,LP,SI,EZ\n",
      "2025/12/25 21:53:12  -gwaslab to vcf format dictionary:\n",
      "2025/12/25 21:53:12   - gwaslab keys: rsID,CHR,POS,NEA,EA,N,EAF,BETA,SE,MLOG10P,INFO,Z\n",
      "2025/12/25 21:53:12   - vcf values: ID,#CHROM,POS,REF,ALT,SS,AF,ES,SE,LP,SI,EZ\n",
      "2025/12/25 21:53:12  -Creating VCF file header...\n",
      "2025/12/25 21:53:12   -VCF header contig build:19\n",
      "2025/12/25 21:53:12   -ID:Study_1\n",
      "2025/12/25 21:53:12   -StudyType:Unknown\n",
      "2025/12/25 21:53:12   -TotalVariants:1000\n",
      "2025/12/25 21:53:12   -HarmonisedVariants:0\n",
      "2025/12/25 21:53:12   -VariantsNotHarmonised:1000\n",
      "2025/12/25 21:53:12   -SwitchedAlleles:0\n",
      "2025/12/25 21:53:12  -Writing sumstats to: ./mysumstats.vcf...\n",
      "2025/12/25 21:53:12  -bgzip compressing : ./mysumstats.vcf.gz...\n",
      "2025/12/25 21:53:12  -tabix indexing : : ./mysumstats.vcf.gz.tbi...\n",
      "2025/12/25 21:53:12  -Saving log file to: ./mysumstats.vcf.log\n",
      "2025/12/25 21:53:12 Finished outputting successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.to_format(\"./mysumstats\",fmt=\"vcf\",bgzip=True,tabix=True,build=\"19\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cbfa9a8",
   "metadata": {},
   "source": [
    "### parquet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9ef3c5fa-021a-424e-88d1-a6f8e66c89e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p mysumstats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "847d2b32",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/25 21:53:12 Start to convert the output sumstats in:  gwaslab  format\n",
      "2025/12/25 21:53:12  -Start outputting sumstats in gwaslab format...\n",
      "2025/12/25 21:53:12  -gwaslab format will be loaded...\n",
      "2025/12/25 21:53:12  -gwaslab format meta info:\n",
      "2025/12/25 21:53:12   - format_name  : gwaslab\n",
      "2025/12/25 21:53:12   - format_source  : https://cloufield.github.io/gwaslab/\n",
      "2025/12/25 21:53:12   - format_version  : 20231220_v4\n",
      "2025/12/25 21:53:12  -Output path: ./mysumstats.gwaslab.parquet\n",
      "2025/12/25 21:53:12  -Output columns: SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P,DIRECTION,N\n",
      "2025/12/25 21:53:12  -Writing sumstats to: ./mysumstats.gwaslab.parquet...\n",
      "2025/12/25 21:53:12   -Partitioned parquet will be written to directory: ./mysumstats.gwaslab\n",
      "2025/12/25 21:53:13  -Saving log file to: ./mysumstats.gwaslab.log\n",
      "2025/12/25 21:53:13 Finished outputting successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.to_format(\"./mysumstats\",\n",
    "                     fmt=\"gwaslab\",\n",
    "                     tab_fmt=\"parquet\",\n",
    "                     to_tabular_kwargs={\"partition_cols\":[\"CHR\"]})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a360ba8f",
   "metadata": {},
   "source": [
    "## For annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0192b1ae",
   "metadata": {},
   "source": [
    "### convert to bed format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "acc2d8cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/25 21:53:13 Start to convert the output sumstats in:  bed  format\n",
      "2025/12/25 21:53:13  -Formatting statistics ...\n",
      "2025/12/25 21:53:13  -Float statistics formats:\n",
      "2025/12/25 21:53:13   - Columns       : ['EAF', 'BETA', 'SE', 'P']\n",
      "2025/12/25 21:53:13   - Output formats: ['{:.4g}', '{:.4f}', '{:.4f}', '{:.4e}']\n",
      "2025/12/25 21:53:13  -Start outputting sumstats in bed format...\n",
      "2025/12/25 21:53:13  -Number of SNPs : 920\n",
      "2025/12/25 21:53:13  -Number of Insertions : 52\n",
      "2025/12/25 21:53:13  -Number of Deletions : 28\n",
      "2025/12/25 21:53:13  -formatting to 0-based bed-like file...\n",
      "2025/12/25 21:53:13  -format description: https://genome.ucsc.edu/FAQ/FAQformat.html#format1\n",
      "2025/12/25 21:53:13  -Adjusting positions in format-specific manner..\n",
      "2025/12/25 21:53:13  -Output columns: CHR,START,END,NEA/EA,STRAND,SNPID\n",
      "2025/12/25 21:53:13  -Writing sumstats to: ./mysumstats.bed...\n",
      "2025/12/25 21:53:13  -Saving log file to: ./mysumstats.bed.log\n",
      "2025/12/25 21:53:13 Finished outputting successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.to_format(\"./mysumstats\",fmt=\"bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4234b747-de03-4ae6-b3f9-0377df0782f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\t2005485\t2005486\tT/C\t+\t1:2005486_C_T\n",
      "1\t2247939\t2247939\t-/AGG\t+\t1:2247939_AAGG_A\n",
      "1\t3741852\t3741853\tA/G\t+\t1:3741853_G_A\n",
      "1\t5017525\t5017526\tA/G\t+\t1:5017526_G_A\n",
      "1\t5843474\t5843475\tT/C\t+\t1:5843475_C_T\n",
      "1\t9405102\t9405103\tT/C\t+\t1:9405103_T_C\n",
      "1\t9443410\t9443411\tA/G\t+\t1:9443411_G_A\n",
      "1\t12866347\t12866348\tC/G\t+\t1:12866348_G_C\n",
      "1\t14466315\t14466316\tA/G\t+\t1:14466316_A_G\n",
      "1\t14900418\t14900419\tA/G\t+\t1:14900419_A_G\n"
     ]
    }
   ],
   "source": [
    "!cat mysumstats.bed | head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "027f04d4",
   "metadata": {},
   "source": [
    "### convert to vep default format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "10d6e0bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/25 21:53:13 Start to convert the output sumstats in:  vep  format\n",
      "2025/12/25 21:53:13  -Formatting statistics ...\n",
      "2025/12/25 21:53:13  -Float statistics formats:\n",
      "2025/12/25 21:53:13   - Columns       : ['EAF', 'BETA', 'SE', 'P']\n",
      "2025/12/25 21:53:13   - Output formats: ['{:.4g}', '{:.4f}', '{:.4f}', '{:.4e}']\n",
      "2025/12/25 21:53:13  -Start outputting sumstats in vep format...\n",
      "2025/12/25 21:53:13  -Number of SNPs : 920\n",
      "2025/12/25 21:53:13  -Number of Insertions : 52\n",
      "2025/12/25 21:53:13  -Number of Deletions : 28\n",
      "2025/12/25 21:53:13  -formatting to 1-based bed-like file (for vep)...\n",
      "2025/12/25 21:53:13  -format description: http://asia.ensembl.org/info/docs/tools/vep/vep_formats.html\n",
      "2025/12/25 21:53:13  -Adjusting positions in format-specific manner..\n",
      "2025/12/25 21:53:13  -Output columns: CHR,START,END,NEA/EA,STRAND,SNPID\n",
      "2025/12/25 21:53:13  -Writing sumstats to: ./mysumstats.vep...\n",
      "2025/12/25 21:53:13  -Saving log file to: ./mysumstats.vep.log\n",
      "2025/12/25 21:53:13 Finished outputting successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.to_format(\"./mysumstats\",fmt=\"vep\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0237b9da-e592-4a5b-a8fd-bc06cce43a70",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\t2005486\t2005486\tT/C\t+\t1:2005486_C_T\n",
      "1\t2247940\t2247939\t-/AGG\t+\t1:2247939_AAGG_A\n",
      "1\t3741853\t3741853\tA/G\t+\t1:3741853_G_A\n",
      "1\t5017526\t5017526\tA/G\t+\t1:5017526_G_A\n",
      "1\t5843475\t5843475\tT/C\t+\t1:5843475_C_T\n",
      "1\t9405103\t9405103\tT/C\t+\t1:9405103_T_C\n",
      "1\t9443411\t9443411\tA/G\t+\t1:9443411_G_A\n",
      "1\t12866348\t12866348\tC/G\t+\t1:12866348_G_C\n",
      "1\t14466316\t14466316\tA/G\t+\t1:14466316_A_G\n",
      "1\t14900419\t14900419\tA/G\t+\t1:14900419_A_G\n"
     ]
    }
   ],
   "source": [
    "!cat mysumstats.vep | head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5aa3c522",
   "metadata": {},
   "source": [
    "### convert to annovar default input format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8c96349d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/25 21:53:13 Start to convert the output sumstats in:  annovar  format\n",
      "2025/12/25 21:53:13  -Formatting statistics ...\n",
      "2025/12/25 21:53:13  -Float statistics formats:\n",
      "2025/12/25 21:53:13   - Columns       : ['EAF', 'BETA', 'SE', 'P']\n",
      "2025/12/25 21:53:13   - Output formats: ['{:.4g}', '{:.4f}', '{:.4f}', '{:.4e}']\n",
      "2025/12/25 21:53:13  -Start outputting sumstats in annovar format...\n",
      "2025/12/25 21:53:13  -Number of SNPs : 920\n",
      "2025/12/25 21:53:13  -Number of Insertions : 52\n",
      "2025/12/25 21:53:13  -Number of Deletions : 28\n",
      "2025/12/25 21:53:13  -formatting to 1-based bed-like file...\n",
      "2025/12/25 21:53:13  -format description: https://annovar.openbioinformatics.org/en/latest/user-guide/input/\n",
      "2025/12/25 21:53:13  -Adjusting positions in format-specific manner..\n",
      "2025/12/25 21:53:13  -Output columns: CHR,START,END,NEA_out,EA_out,SNPID\n",
      "2025/12/25 21:53:13  -Writing sumstats to: ./mysumstats.annovar...\n",
      "2025/12/25 21:53:13  -Saving log file to: ./mysumstats.annovar.log\n",
      "2025/12/25 21:53:13 Finished outputting successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.to_format(\"./mysumstats\",fmt=\"annovar\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e7406a88-f6a9-45a9-8426-08d943d2eaa8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\t2005486\t2005486\tT\tC\t1:2005486_C_T\n",
      "1\t2247939\t2247939\t-\tAGG\t1:2247939_AAGG_A\n",
      "1\t3741853\t3741853\tA\tG\t1:3741853_G_A\n",
      "1\t5017526\t5017526\tA\tG\t1:5017526_G_A\n",
      "1\t5843475\t5843475\tT\tC\t1:5843475_C_T\n",
      "1\t9405103\t9405103\tT\tC\t1:9405103_T_C\n",
      "1\t9443411\t9443411\tA\tG\t1:9443411_G_A\n",
      "1\t12866348\t12866348\tC\tG\t1:12866348_G_C\n",
      "1\t14466316\t14466316\tA\tG\t1:14466316_A_G\n",
      "1\t14900419\t14900419\tA\tG\t1:14900419_A_G\n"
     ]
    }
   ],
   "source": [
    "!cat mysumstats.annovar | head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1349a798-8bac-4cbf-add4-4b78fac194bc",
   "metadata": {},
   "source": [
    "## Filter and then output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ea44a53e-6e1d-4f86-8c1e-5c1c8b77c7d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/25 21:53:13 Start to filter variants by condition... ...(v4.0.0)\n",
      "2025/12/25 21:53:13  -Current Dataframe shape : 1000 x 12 ; Memory usage: 0.08 MB\n",
      "2025/12/25 21:53:13  -Expression: EAF >0.05 and EAF < 0.95\n",
      "2025/12/25 21:53:13  -Filtered out variants: 483\n",
      "2025/12/25 21:53:13  -Current Dataframe shape : 517 x 12 ; Memory usage: 0.05 MB\n",
      "2025/12/25 21:53:13 Finished filtering variants.\n",
      "2025/12/25 21:53:13 Start to convert the output sumstats in:  ssf  format\n",
      "2025/12/25 21:53:13  -Formatting statistics ...\n",
      "2025/12/25 21:53:13  -Float statistics formats:\n",
      "2025/12/25 21:53:13   - Columns       : ['EAF', 'BETA', 'SE', 'P']\n",
      "2025/12/25 21:53:13   - Output formats: ['{:.4g}', '{:.4f}', '{:.4f}', '{:.4e}']\n",
      "2025/12/25 21:53:13  -Replacing SNPID separator from \":\" to \"_\"...\n",
      "2025/12/25 21:53:13  -Start outputting sumstats in ssf format...\n",
      "2025/12/25 21:53:13  -ssf format will be loaded...\n",
      "2025/12/25 21:53:13  -ssf format meta info:\n",
      "2025/12/25 21:53:13   - format_name  : ssf\n",
      "2025/12/25 21:53:13   - format_source  : https://www.biorxiv.org/content/10.1101/2022.07.15.500230v1.full\n",
      "2025/12/25 21:53:13   - format_cite_name  : GWAS-SSF v0.1\n",
      "2025/12/25 21:53:13   - format_separator  : \\t\n",
      "2025/12/25 21:53:13   - format_na  : #NA\n",
      "2025/12/25 21:53:13   - format_col_order  : chromosome,base_pair_location,effect_allele,other_allele,beta,odds_ratio,hazard_ratio,standard_error,effect_allele_frequency,p_value,neg_log_10_p_value,ci_upper,ci_lower,rsid,variant_id,info,ref_allele,n\n",
      "2025/12/25 21:53:13   - format_version  :  20230328\n",
      "2025/12/25 21:53:13  -gwaslab to ssf format dictionary:\n",
      "2025/12/25 21:53:13   - gwaslab keys: SNPID,rsID,CHR,POS,NEA,EA,EAF,N,BETA,SE,P,MLOG10P,INFO,OR,HR,OR_95L,OR_95U\n",
      "2025/12/25 21:53:13   - ssf values: variant_id,rsid,chromosome,base_pair_location,other_allele,effect_allele,effect_allele_frequency,n,beta,standard_error,p_value,neg_log_10_p_value,info,odds_ratio,hazard_ratio,ci_lower,ci_upper\n",
      "2025/12/25 21:53:13  -Output path: ./mysumstats_maf005.ssf.tsv.gz\n",
      "2025/12/25 21:53:13  -Output columns: chromosome,base_pair_location,effect_allele,other_allele,beta,standard_error,effect_allele_frequency,p_value,variant_id,n\n",
      "2025/12/25 21:53:13  -Writing sumstats to: ./mysumstats_maf005.ssf.tsv.gz...\n",
      "2025/12/25 21:53:13  -md5sum hashing for the file: ./mysumstats_maf005.ssf.tsv.gz\n",
      "2025/12/25 21:53:13  -md5sum path: ./mysumstats_maf005.ssf.tsv.gz.md5sum\n",
      "2025/12/25 21:53:13  -md5sum: 0370f9f0dcb5360b8d1f13460fba43c0\n",
      "2025/12/25 21:53:13  -Exporting SSF-style meta data to ./mysumstats_maf005.ssf.tsv.gz.ssf.tsv-meta.yaml\n",
      "2025/12/25 21:53:13  -Saving log file to: ./mysumstats_maf005.ssf.log\n",
      "2025/12/25 21:53:13 Finished outputting successfully!\n",
      "2025/12/25 21:53:13 Validating SSF format output...\n",
      "2025/12/25 21:53:13 Using built-in SSF validator (gwas-ssf CLI not available)...\n",
      "2025/12/25 21:53:13 Validating file extension...\n",
      "2025/12/25 21:53:13 ✓ File extension OK\n",
      "2025/12/25 21:53:13 Validating column order...\n",
      "2025/12/25 21:53:13 ✓ Column order OK\n",
      "2025/12/25 21:53:13 Validating chromosomes...\n",
      "2025/12/25 21:53:13 ✓ Chromosomes OK (All autosomes exist. Optional chromosomes ['24', '25'] do not exist.)\n",
      "2025/12/25 21:53:13 Validating minimum row count...\n",
      "2025/12/25 21:53:13 #WARNING! ✗ SSF validation failed: The file has fewer than the minimum rows required: 517 < 100000\n",
      "2025/12/25 21:53:13 #WARNING!   - The file has fewer than the minimum rows required: 517 < 100000\n"
     ]
    }
   ],
   "source": [
    "mysumstats.filter_value(\"EAF >0.05 and EAF < 0.95\").to_format(\"./mysumstats_maf005\", fmt=\"ssf\", ssfmeta=True, md5sum=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "512px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
