{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "498f762a",
   "metadata": {},
   "source": [
    "# Finemapping by calling susieR"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e749209",
   "metadata": {},
   "source": [
    "## Load gwaslab and sample data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "894c82d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0,\"/home/yunye/work/gwaslab/src\")\n",
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "934a9a6e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024/12/23 16:46:23 GWASLab v3.5.4 https://cloufield.github.io/gwaslab/\n",
      "2024/12/23 16:46:23 (C) 2022-2024, Yunye He, Kamatani Lab, MIT License, gwaslab@gmail.com\n"
     ]
    }
   ],
   "source": [
    "gl.show_version()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b8ad1679",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024/12/23 16:46:23 GWASLab v3.5.4 https://cloufield.github.io/gwaslab/\n",
      "2024/12/23 16:46:23 (C) 2022-2024, Yunye He, Kamatani Lab, MIT License, gwaslab@gmail.com\n",
      "2024/12/23 16:46:23 Start to initialize gl.Sumstats from file :./0_sample_data/t2d_bbj.txt.gz\n",
      "2024/12/23 16:46:24  -Reading columns          : BETA,POS,CHR,P,ALT,SE,REF,SNP\n",
      "2024/12/23 16:46:24  -Renaming columns to      : BETA,POS,CHR,P,EA,SE,NEA,SNPID\n",
      "2024/12/23 16:46:24  -Current Dataframe shape : 1000000  x  8\n",
      "2024/12/23 16:46:24  -Initiating a status column: STATUS ...\n",
      "2024/12/23 16:46:24  #WARNING! Version of genomic coordinates is unknown...\n",
      "2024/12/23 16:46:25 Start to reorder the columns...v3.5.4\n",
      "2024/12/23 16:46:25  -Current Dataframe shape : 1000000 x 10 ; Memory usage: 82.88 MB\n",
      "2024/12/23 16:46:25  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,BETA,SE,P,N,STATUS\n",
      "2024/12/23 16:46:25 Finished reordering the columns.\n",
      "2024/12/23 16:46:25  -Column  : SNPID  CHR    POS   EA       NEA      BETA    SE      P       N     STATUS  \n",
      "2024/12/23 16:46:25  -DType   : object string int64 category category float64 float64 float64 int64 category\n",
      "2024/12/23 16:46:25  -Verified: T      F      T     T        T        T       T       T       T     T       \n",
      "2024/12/23 16:46:25  #WARNING! Columns with possibly incompatible dtypes: CHR\n",
      "2024/12/23 16:46:25  -Current Dataframe memory usage: 82.88 MB\n",
      "2024/12/23 16:46:25 Finished loading data successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(\"./0_sample_data/t2d_bbj.txt.gz\",\n",
    "             snpid=\"SNP\",\n",
    "             chrom=\"CHR\",\n",
    "             ea=\"ALT\",nea=\"REF\",\n",
    "             pos=\"POS\",\n",
    "             beta=\"BETA\",\n",
    "             se=\"SE\",\n",
    "             n=150000,\n",
    "             p=\"P\",nrows=1000000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "12e0c512",
   "metadata": {},
   "outputs": [],
   "source": [
    "mysumstats.fix_id(fixsep=True,verbose=False)\n",
    "mysumstats.basic_check(verbose=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3043b360",
   "metadata": {},
   "source": [
    "## run finemapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d161c22",
   "metadata": {},
   "source": [
    "### calculate ld matrix for each locus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c2f2d1f9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024/12/23 16:46:39 Start to calculate LD matrix...v3.5.4\n",
      "2024/12/23 16:46:39  -Current Dataframe shape : 1000000 x 10 ; Memory usage: 86.00 MB\n",
      "2024/12/23 16:46:39  -Loci were not provided. All significant loci will be automatically extracted...\n",
      "2024/12/23 16:46:39 Start to extract lead variants...v3.5.4\n",
      "2024/12/23 16:46:39  -Current Dataframe shape : 1000000 x 10 ; Memory usage: 86.00 MB\n",
      "2024/12/23 16:46:39  -Processing 1000000 variants...\n",
      "2024/12/23 16:46:39  -Significance threshold : 5e-08\n",
      "2024/12/23 16:46:39  -Sliding window size: 1000  kb\n",
      "2024/12/23 16:46:40  -Using P for extracting lead variants...\n",
      "2024/12/23 16:46:40  -Found 543 significant variants in total...\n",
      "2024/12/23 16:46:40  -Identified 4 lead variants!\n",
      "2024/12/23 16:46:40 Finished extracting lead variants.\n",
      "2024/12/23 16:46:40  -plink1.9 path: /home/yunye/tools/bin/plink\n",
      "2024/12/23 16:46:40  -plink2 path: /home/yunye/tools/bin/plink2\n",
      "2024/12/23 16:46:40  -Dropping duplicated SNPIDs...\n",
      "2024/12/23 16:46:40  -Locus #1---------------------------------------------------------------\n",
      "2024/12/23 16:46:40  -Processing locus with lead variant 1:22068326:A:G at CHR 1 POS 22068326 ...\n",
      "2024/12/23 16:46:40  -Processing VCF : /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz...\n",
      "2024/12/23 16:46:40  -PLINK version: PLINK v2.00a5.9LM AVX2 AMD (12 Dec 2023)\n",
      "2024/12/23 16:46:40   -Processing VCF for CHR 1...\n",
      "2024/12/23 16:46:40   -Plink bfile for CHR 1 exists: /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.1. Skipping...\n",
      "2024/12/23 16:46:44    -Variants in ref file: 6471908\n",
      "2024/12/23 16:46:44    -Variants in locus (1:22068326:A:G): 8867\n",
      "2024/12/23 16:46:47    -Variants with perfect matched alleles:3611\n",
      "2024/12/23 16:46:47    -Variants with flipped alleles:5253\n",
      "2024/12/23 16:46:47    -Total Variants matched:8864\n",
      "2024/12/23 16:46:47    -Flipping BETA for variants with flipped alleles...\n",
      "2024/12/23 16:46:47  -Exporting SNP list of 8864  to: ./Study_1_1:22068326:A:G_1000.snplist.raw...\n",
      "2024/12/23 16:46:47  -Exporting locus sumstats to: ./Study_1_1:22068326:A:G_1000.sumstats.gz...\n",
      "2024/12/23 16:46:47  -Exported columns: ['SNPID', 'CHR', 'POS', 'EA_bim', 'NEA_bim', 'BETA', 'SE', 'N']...\n",
      "2024/12/23 16:46:48  -Start to calculate LD r matrix...\n",
      "2024/12/23 16:46:48  -PLINK version: PLINK v1.90b7.2 64-bit (11 Dec 2023)\n",
      "2024/12/23 16:47:24  -Finished calculating LD r for locus with lead variant 1:22068326:A:G at CHR 1 POS 22068326...\n",
      "2024/12/23 16:47:24  -Sumstats SNPID order and LD matrix SNPID order are matched.\n",
      "2024/12/23 16:47:24  -Locus #2---------------------------------------------------------------\n",
      "2024/12/23 16:47:24  -Processing locus with lead variant 1:51103268:T:C at CHR 1 POS 51103268 ...\n",
      "2024/12/23 16:47:24  -Processing VCF : /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz...\n",
      "2024/12/23 16:47:24  -PLINK version: PLINK v2.00a5.9LM AVX2 AMD (12 Dec 2023)\n",
      "2024/12/23 16:47:24   -Processing VCF for CHR 1...\n",
      "2024/12/23 16:47:24   -Plink bfile for CHR 1 exists: /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.1. Skipping...\n",
      "2024/12/23 16:47:28    -Variants in ref file: 6471908\n",
      "2024/12/23 16:47:28    -Variants in locus (1:51103268:T:C): 4545\n",
      "2024/12/23 16:47:31    -Variants with perfect matched alleles:2139\n",
      "2024/12/23 16:47:31    -Variants with flipped alleles:2406\n",
      "2024/12/23 16:47:31    -Total Variants matched:4545\n",
      "2024/12/23 16:47:31    -Flipping BETA for variants with flipped alleles...\n",
      "2024/12/23 16:47:31  -Exporting SNP list of 4545  to: ./Study_1_1:51103268:T:C_1000.snplist.raw...\n",
      "2024/12/23 16:47:31  -Exporting locus sumstats to: ./Study_1_1:51103268:T:C_1000.sumstats.gz...\n",
      "2024/12/23 16:47:31  -Exported columns: ['SNPID', 'CHR', 'POS', 'EA_bim', 'NEA_bim', 'BETA', 'SE', 'N']...\n",
      "2024/12/23 16:47:31  -Start to calculate LD r matrix...\n",
      "2024/12/23 16:47:31  -PLINK version: PLINK v1.90b7.2 64-bit (11 Dec 2023)\n",
      "2024/12/23 16:47:40  -Finished calculating LD r for locus with lead variant 1:51103268:T:C at CHR 1 POS 51103268...\n",
      "2024/12/23 16:47:40  -Sumstats SNPID order and LD matrix SNPID order are matched.\n",
      "2024/12/23 16:47:40  -Locus #3---------------------------------------------------------------\n",
      "2024/12/23 16:47:40  -Processing locus with lead variant 1:154309595:TA:T at CHR 1 POS 154309595 ...\n",
      "2024/12/23 16:47:40  -Processing VCF : /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz...\n",
      "2024/12/23 16:47:40  -PLINK version: PLINK v2.00a5.9LM AVX2 AMD (12 Dec 2023)\n",
      "2024/12/23 16:47:40   -Processing VCF for CHR 1...\n",
      "2024/12/23 16:47:40   -Plink bfile for CHR 1 exists: /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.1. Skipping...\n",
      "2024/12/23 16:47:44    -Variants in ref file: 6471908\n",
      "2024/12/23 16:47:44    -Variants in locus (1:154309595:TA:T): 5737\n",
      "2024/12/23 16:47:47    -Variants with perfect matched alleles:2420\n",
      "2024/12/23 16:47:47    -Variants with flipped alleles:3312\n",
      "2024/12/23 16:47:47    -Total Variants matched:5732\n",
      "2024/12/23 16:47:47    -Flipping BETA for variants with flipped alleles...\n",
      "2024/12/23 16:47:47  -Exporting SNP list of 5732  to: ./Study_1_1:154309595:TA:T_1000.snplist.raw...\n",
      "2024/12/23 16:47:47  -Exporting locus sumstats to: ./Study_1_1:154309595:TA:T_1000.sumstats.gz...\n",
      "2024/12/23 16:47:47  -Exported columns: ['SNPID', 'CHR', 'POS', 'EA_bim', 'NEA_bim', 'BETA', 'SE', 'N']...\n",
      "2024/12/23 16:47:48  -Start to calculate LD r matrix...\n",
      "2024/12/23 16:47:48  -PLINK version: PLINK v1.90b7.2 64-bit (11 Dec 2023)\n",
      "2024/12/23 16:48:03  -Finished calculating LD r for locus with lead variant 1:154309595:TA:T at CHR 1 POS 154309595...\n",
      "2024/12/23 16:48:03  -Sumstats SNPID order and LD matrix SNPID order are matched.\n",
      "2024/12/23 16:48:03  -Locus #4---------------------------------------------------------------\n",
      "2024/12/23 16:48:03  -Processing locus with lead variant 2:640986:CACAT:C at CHR 2 POS 640986 ...\n",
      "2024/12/23 16:48:03  -Processing VCF : /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz...\n",
      "2024/12/23 16:48:03  -PLINK version: PLINK v2.00a5.9LM AVX2 AMD (12 Dec 2023)\n",
      "2024/12/23 16:48:03   -Processing VCF for CHR 2...\n",
      "2024/12/23 16:48:03   -Plink bfile for CHR 2 exists: /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.2. Skipping...\n",
      "2024/12/23 16:48:07    -Variants in ref file: 7085912\n",
      "2024/12/23 16:48:07    -Variants in locus (2:640986:CACAT:C): 9584\n",
      "2024/12/23 16:48:11    -Variants with perfect matched alleles:4160\n",
      "2024/12/23 16:48:11    -Variants with flipped alleles:5424\n",
      "2024/12/23 16:48:11    -Total Variants matched:9584\n",
      "2024/12/23 16:48:11    -Flipping BETA for variants with flipped alleles...\n",
      "2024/12/23 16:48:11  -Exporting SNP list of 9584  to: ./Study_1_2:640986:CACAT:C_1000.snplist.raw...\n",
      "2024/12/23 16:48:11  -Exporting locus sumstats to: ./Study_1_2:640986:CACAT:C_1000.sumstats.gz...\n",
      "2024/12/23 16:48:11  -Exported columns: ['SNPID', 'CHR', 'POS', 'EA_bim', 'NEA_bim', 'BETA', 'SE', 'N']...\n",
      "2024/12/23 16:48:11  -Start to calculate LD r matrix...\n",
      "2024/12/23 16:48:11  -PLINK version: PLINK v1.90b7.2 64-bit (11 Dec 2023)\n",
      "2024/12/23 16:48:52  -Finished calculating LD r for locus with lead variant 2:640986:CACAT:C at CHR 2 POS 640986...\n",
      "2024/12/23 16:48:52  -Sumstats SNPID order and LD matrix SNPID order are matched.\n",
      "2024/12/23 16:48:52  -File list is saved to: ./Study_1_4loci_1000kb.filelist\n",
      "2024/12/23 16:48:52  -Finished LD matrix calculation.\n",
      "2024/12/23 16:48:52 Finished calculating LD matrix.\n"
     ]
    }
   ],
   "source": [
    "mysumstats.calculate_ld_matrix( plink=\"/home/yunye/tools/bin/plink\", \n",
    "                                plink2=\"/home/yunye/tools/bin/plink2\",\n",
    "                                vcf = gl.get_path(\"1kg_eas_hg19\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5e5a6d50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>SNPID_LIST</th>\n",
       "      <th>LD_R_MATRIX</th>\n",
       "      <th>LOCUS_SUMSTATS</th>\n",
       "      <th>STUDY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:22068326:A:G</td>\n",
       "      <td>./Study_1_1:22068326:A:G_1000.snplist.raw</td>\n",
       "      <td>./Study_1_1:22068326:A:G_1000.ld.gz</td>\n",
       "      <td>./Study_1_1:22068326:A:G_1000.sumstats.gz</td>\n",
       "      <td>Study_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:51103268:T:C</td>\n",
       "      <td>./Study_1_1:51103268:T:C_1000.snplist.raw</td>\n",
       "      <td>./Study_1_1:51103268:T:C_1000.ld.gz</td>\n",
       "      <td>./Study_1_1:51103268:T:C_1000.sumstats.gz</td>\n",
       "      <td>Study_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:154309595:TA:T</td>\n",
       "      <td>./Study_1_1:154309595:TA:T_1000.snplist.raw</td>\n",
       "      <td>./Study_1_1:154309595:TA:T_1000.ld.gz</td>\n",
       "      <td>./Study_1_1:154309595:TA:T_1000.sumstats.gz</td>\n",
       "      <td>Study_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2:640986:CACAT:C</td>\n",
       "      <td>./Study_1_2:640986:CACAT:C_1000.snplist.raw</td>\n",
       "      <td>./Study_1_2:640986:CACAT:C_1000.ld.gz</td>\n",
       "      <td>./Study_1_2:640986:CACAT:C_1000.sumstats.gz</td>\n",
       "      <td>Study_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              SNPID                                   SNPID_LIST  \\\n",
       "0    1:22068326:A:G    ./Study_1_1:22068326:A:G_1000.snplist.raw   \n",
       "1    1:51103268:T:C    ./Study_1_1:51103268:T:C_1000.snplist.raw   \n",
       "2  1:154309595:TA:T  ./Study_1_1:154309595:TA:T_1000.snplist.raw   \n",
       "3  2:640986:CACAT:C  ./Study_1_2:640986:CACAT:C_1000.snplist.raw   \n",
       "\n",
       "                             LD_R_MATRIX  \\\n",
       "0    ./Study_1_1:22068326:A:G_1000.ld.gz   \n",
       "1    ./Study_1_1:51103268:T:C_1000.ld.gz   \n",
       "2  ./Study_1_1:154309595:TA:T_1000.ld.gz   \n",
       "3  ./Study_1_2:640986:CACAT:C_1000.ld.gz   \n",
       "\n",
       "                                LOCUS_SUMSTATS    STUDY  \n",
       "0    ./Study_1_1:22068326:A:G_1000.sumstats.gz  Study_1  \n",
       "1    ./Study_1_1:51103268:T:C_1000.sumstats.gz  Study_1  \n",
       "2  ./Study_1_1:154309595:TA:T_1000.sumstats.gz  Study_1  \n",
       "3  ./Study_1_2:640986:CACAT:C_1000.sumstats.gz  Study_1  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.finemapping[\"file\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "14652f29",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>STATUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:725932:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>725932</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59700</td>\n",
       "      <td>150000</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:725933:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>725933</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59730</td>\n",
       "      <td>150000</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:737801:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>737801</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0490</td>\n",
       "      <td>0.1231</td>\n",
       "      <td>0.69080</td>\n",
       "      <td>150000</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:749963:T:TAA</td>\n",
       "      <td>1</td>\n",
       "      <td>749963</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0213</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.28460</td>\n",
       "      <td>150000</td>\n",
       "      <td>9960399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:751343:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>751343</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>150000</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999995</th>\n",
       "      <td>2:6347639:C:A</td>\n",
       "      <td>2</td>\n",
       "      <td>6347639</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>-0.0159</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>0.15100</td>\n",
       "      <td>150000</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999996</th>\n",
       "      <td>2:6347694:G:C</td>\n",
       "      <td>2</td>\n",
       "      <td>6347694</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0151</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>0.17250</td>\n",
       "      <td>150000</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999997</th>\n",
       "      <td>2:6348478:G:A</td>\n",
       "      <td>2</td>\n",
       "      <td>6348478</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-0.0152</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>0.17160</td>\n",
       "      <td>150000</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999998</th>\n",
       "      <td>2:6348490:G:C</td>\n",
       "      <td>2</td>\n",
       "      <td>6348490</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.3180</td>\n",
       "      <td>0.2032</td>\n",
       "      <td>0.11750</td>\n",
       "      <td>150000</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999999</th>\n",
       "      <td>2:6348754:A:C</td>\n",
       "      <td>2</td>\n",
       "      <td>6348754</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0297</td>\n",
       "      <td>0.0126</td>\n",
       "      <td>0.01846</td>\n",
       "      <td>150000</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000000 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 SNPID  CHR      POS   EA NEA    BETA      SE        P  \\\n",
       "0         1:725932:G:A    1   725932    G   A -0.0737  0.1394  0.59700   \n",
       "1         1:725933:A:G    1   725933    G   A  0.0737  0.1394  0.59730   \n",
       "2         1:737801:T:C    1   737801    C   T  0.0490  0.1231  0.69080   \n",
       "3       1:749963:T:TAA    1   749963  TAA   T  0.0213  0.0199  0.28460   \n",
       "4         1:751343:T:A    1   751343    T   A  0.0172  0.0156  0.27050   \n",
       "...                ...  ...      ...  ...  ..     ...     ...      ...   \n",
       "999995   2:6347639:C:A    2  6347639    C   A -0.0159  0.0111  0.15100   \n",
       "999996   2:6347694:G:C    2  6347694    G   C -0.0151  0.0111  0.17250   \n",
       "999997   2:6348478:G:A    2  6348478    G   A -0.0152  0.0111  0.17160   \n",
       "999998   2:6348490:G:C    2  6348490    G   C -0.3180  0.2032  0.11750   \n",
       "999999   2:6348754:A:C    2  6348754    C   A  0.0297  0.0126  0.01846   \n",
       "\n",
       "             N   STATUS  \n",
       "0       150000  9960099  \n",
       "1       150000  9960099  \n",
       "2       150000  9960099  \n",
       "3       150000  9960399  \n",
       "4       150000  9960099  \n",
       "...        ...      ...  \n",
       "999995  150000  9960099  \n",
       "999996  150000  9960099  \n",
       "999997  150000  9960099  \n",
       "999998  150000  9960099  \n",
       "999999  150000  9960099  \n",
       "\n",
       "[1000000 rows x 10 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df169c99",
   "metadata": {},
   "source": [
    "### run susie rss by calling SuSieR in R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0b773e60",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024/12/23 16:48:52 Start to run finemapping using SuSieR from command line...v3.5.4\n",
      "2024/12/23 16:48:52  -R version: R scripting front-end version 4.1.3 (2022-03-10)\n",
      "2024/12/23 16:48:52  -SuSieR version: [1] ‘0.12.35’\n",
      "2024/12/23 16:48:52  -Running for: 1:22068326:A:G - Study_1\n",
      "2024/12/23 16:48:52   -Locus sumstats:./Study_1_1:22068326:A:G_1000.sumstats.gz\n",
      "2024/12/23 16:48:52   -LD r matrix:./Study_1_1:22068326:A:G_1000.ld.gz\n",
      "2024/12/23 16:48:52   -output_prefix:./Study_1_1:22068326:A:G_1000\n",
      "2024/12/23 16:48:52   -SuSieR script: susie_rss(bhat = sumstats$BETA,shat = sumstats$SE,, n = n, R = R, max_iter = 100000, min_abs_corr=0.1, refine = TRUE, L = 10)\n",
      "2024/12/23 16:50:41   -Running SuSieR from command line...\n",
      "2024/12/23 16:50:41   -SuSieR result summary to: ./Study_1_1:22068326:A:G_1000.pipcs\n",
      "2024/12/23 16:50:42  -Running for: 1:51103268:T:C - Study_1\n",
      "2024/12/23 16:50:42   -Locus sumstats:./Study_1_1:51103268:T:C_1000.sumstats.gz\n",
      "2024/12/23 16:50:42   -LD r matrix:./Study_1_1:51103268:T:C_1000.ld.gz\n",
      "2024/12/23 16:50:42   -output_prefix:./Study_1_1:51103268:T:C_1000\n",
      "2024/12/23 16:50:42   -SuSieR script: susie_rss(bhat = sumstats$BETA,shat = sumstats$SE,, n = n, R = R, max_iter = 100000, min_abs_corr=0.1, refine = TRUE, L = 10)\n",
      "2024/12/23 16:51:59   -Running SuSieR from command line...\n",
      "2024/12/23 16:51:59   -SuSieR result summary to: ./Study_1_1:51103268:T:C_1000.pipcs\n",
      "2024/12/23 16:51:59  -Running for: 1:154309595:TA:T - Study_1\n",
      "2024/12/23 16:51:59   -Locus sumstats:./Study_1_1:154309595:TA:T_1000.sumstats.gz\n",
      "2024/12/23 16:51:59   -LD r matrix:./Study_1_1:154309595:TA:T_1000.ld.gz\n",
      "2024/12/23 16:51:59   -output_prefix:./Study_1_1:154309595:TA:T_1000\n",
      "2024/12/23 16:51:59   -SuSieR script: susie_rss(bhat = sumstats$BETA,shat = sumstats$SE,, n = n, R = R, max_iter = 100000, min_abs_corr=0.1, refine = TRUE, L = 10)\n",
      "2024/12/23 16:52:23   -Running SuSieR from command line...\n",
      "2024/12/23 16:52:23   -SuSieR result summary to: ./Study_1_1:154309595:TA:T_1000.pipcs\n",
      "2024/12/23 16:52:23  -Running for: 2:640986:CACAT:C - Study_1\n",
      "2024/12/23 16:52:23   -Locus sumstats:./Study_1_2:640986:CACAT:C_1000.sumstats.gz\n",
      "2024/12/23 16:52:23   -LD r matrix:./Study_1_2:640986:CACAT:C_1000.ld.gz\n",
      "2024/12/23 16:52:23   -output_prefix:./Study_1_2:640986:CACAT:C_1000\n",
      "2024/12/23 16:52:23   -SuSieR script: susie_rss(bhat = sumstats$BETA,shat = sumstats$SE,, n = n, R = R, max_iter = 100000, min_abs_corr=0.1, refine = TRUE, L = 10)\n",
      "2024/12/23 16:54:33   -Running SuSieR from command line...\n",
      "2024/12/23 16:54:33   -SuSieR result summary to: ./Study_1_2:640986:CACAT:C_1000.pipcs\n",
      "2024/12/23 16:54:33 Finished running finemapping using SuSieR from command line.\n"
     ]
    }
   ],
   "source": [
    "mysumstats.run_susie_rss(r=\"/home/yunye/anaconda3/envs/r41/bin/Rscript\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ba11db33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N_SNP</th>\n",
       "      <th>PIP</th>\n",
       "      <th>CREDIBLE_SET_INDEX</th>\n",
       "      <th>SNPID</th>\n",
       "      <th>LOCUS</th>\n",
       "      <th>STUDY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3656</td>\n",
       "      <td>0.132195</td>\n",
       "      <td>1</td>\n",
       "      <td>1:22068326:A:G</td>\n",
       "      <td>1:22068326:A:G</td>\n",
       "      <td>Study_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3645</td>\n",
       "      <td>0.115444</td>\n",
       "      <td>1</td>\n",
       "      <td>1:22065128:G:C</td>\n",
       "      <td>1:22068326:A:G</td>\n",
       "      <td>Study_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3658</td>\n",
       "      <td>0.094082</td>\n",
       "      <td>1</td>\n",
       "      <td>1:22069036:CA:C</td>\n",
       "      <td>1:22068326:A:G</td>\n",
       "      <td>Study_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3636</td>\n",
       "      <td>0.088931</td>\n",
       "      <td>1</td>\n",
       "      <td>1:22061876:T:TA</td>\n",
       "      <td>1:22068326:A:G</td>\n",
       "      <td>Study_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3596</td>\n",
       "      <td>0.080825</td>\n",
       "      <td>1</td>\n",
       "      <td>1:22046611:G:C</td>\n",
       "      <td>1:22068326:A:G</td>\n",
       "      <td>Study_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28720</th>\n",
       "      <td>3252</td>\n",
       "      <td>0.000593</td>\n",
       "      <td>-1</td>\n",
       "      <td>2:715893:C:T</td>\n",
       "      <td>2:640986:CACAT:C</td>\n",
       "      <td>Study_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28721</th>\n",
       "      <td>9389</td>\n",
       "      <td>0.000593</td>\n",
       "      <td>-1</td>\n",
       "      <td>2:1615003:C:T</td>\n",
       "      <td>2:640986:CACAT:C</td>\n",
       "      <td>Study_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28722</th>\n",
       "      <td>7887</td>\n",
       "      <td>0.000593</td>\n",
       "      <td>-1</td>\n",
       "      <td>2:1416855:C:T</td>\n",
       "      <td>2:640986:CACAT:C</td>\n",
       "      <td>Study_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28723</th>\n",
       "      <td>8095</td>\n",
       "      <td>0.000593</td>\n",
       "      <td>-1</td>\n",
       "      <td>2:1441305:C:T</td>\n",
       "      <td>2:640986:CACAT:C</td>\n",
       "      <td>Study_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28724</th>\n",
       "      <td>1765</td>\n",
       "      <td>0.000593</td>\n",
       "      <td>-1</td>\n",
       "      <td>2:469552:C:T</td>\n",
       "      <td>2:640986:CACAT:C</td>\n",
       "      <td>Study_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28725 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       N_SNP       PIP  CREDIBLE_SET_INDEX            SNPID             LOCUS  \\\n",
       "0       3656  0.132195                   1   1:22068326:A:G    1:22068326:A:G   \n",
       "1       3645  0.115444                   1   1:22065128:G:C    1:22068326:A:G   \n",
       "2       3658  0.094082                   1  1:22069036:CA:C    1:22068326:A:G   \n",
       "3       3636  0.088931                   1  1:22061876:T:TA    1:22068326:A:G   \n",
       "4       3596  0.080825                   1   1:22046611:G:C    1:22068326:A:G   \n",
       "...      ...       ...                 ...              ...               ...   \n",
       "28720   3252  0.000593                  -1     2:715893:C:T  2:640986:CACAT:C   \n",
       "28721   9389  0.000593                  -1    2:1615003:C:T  2:640986:CACAT:C   \n",
       "28722   7887  0.000593                  -1    2:1416855:C:T  2:640986:CACAT:C   \n",
       "28723   8095  0.000593                  -1    2:1441305:C:T  2:640986:CACAT:C   \n",
       "28724   1765  0.000593                  -1     2:469552:C:T  2:640986:CACAT:C   \n",
       "\n",
       "         STUDY  \n",
       "0      Study_1  \n",
       "1      Study_1  \n",
       "2      Study_1  \n",
       "3      Study_1  \n",
       "4      Study_1  \n",
       "...        ...  \n",
       "28720  Study_1  \n",
       "28721  Study_1  \n",
       "28722  Study_1  \n",
       "28723  Study_1  \n",
       "28724  Study_1  \n",
       "\n",
       "[28725 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.pipcs.to_csv(\"mysumstats.pipcs\",index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e44e30d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "mysumstats.pipcs.to_csv(\"mysumstats.pipcs\",index=None)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
