{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "da50a8cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/yunye/anaconda3/lib/python3.9/site-packages/pandas/core/computation/expressions.py:21: UserWarning: Pandas requires version '2.8.4' or newer of 'numexpr' (version '2.7.3' currently installed).\n",
      "  from pandas.core.computation.check import NUMEXPR_INSTALLED\n",
      "/home/yunye/anaconda3/lib/python3.9/site-packages/pandas/core/arrays/masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.2' currently installed).\n",
      "  from pandas.core import (\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import sys\n",
    "sys.path.insert(0,\"/home/yunye/work/gwaslab/src\")\n",
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "271bed64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-01-30 11:39:04--  https://koges.leelabsg.org/download/KoGES_BMI\n",
      "Resolving koges.leelabsg.org (koges.leelabsg.org)... 35.192.148.108, 216.239.36.101, 216.239.38.101\n",
      "Connecting to koges.leelabsg.org (koges.leelabsg.org)|35.192.148.108|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 158394405 (151M) [text/tab-separated-values]\n",
      "Saving to: ‘koges_bmi.tsv.gz’\n",
      "\n",
      "koges_bmi.tsv.gz    100%[===================>] 151.06M  16.9MB/s    in 10s     \n",
      "\n",
      "2024-01-30 11:39:15 (14.9 MB/s) - ‘koges_bmi.tsv.gz’ saved [158394405/158394405]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget -O \"koges_bmi.tsv.gz\" https://koges.leelabsg.org/download/KoGES_BMI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "aa9ccc8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "mysumstats = gl.Sumstats(\"./koges_bmi.tsv.gz\", \n",
    "                             chrom=\"chrom\",\n",
    "                             pos=\"pos\",\n",
    "                             ea=\"alt\",\n",
    "                             nea=\"ref\",\n",
    "                             eaf=\"af\",\n",
    "                             beta=\"beta\",\n",
    "                             se=\"sebeta\",\n",
    "                             p=\"pval\",nrows=1000000,\n",
    "                            verbose=False) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ba61a118",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>STATUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>714646</td>\n",
       "      <td>T</td>\n",
       "      <td>TTTGAAAC</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.0093</td>\n",
       "      <td>0.0170</td>\n",
       "      <td>0.5900</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>725256</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.027</td>\n",
       "      <td>0.0076</td>\n",
       "      <td>0.0170</td>\n",
       "      <td>0.6500</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>726805</td>\n",
       "      <td>GA</td>\n",
       "      <td>G</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>0.0180</td>\n",
       "      <td>0.6600</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>726807</td>\n",
       "      <td>T</td>\n",
       "      <td>TTG</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>0.0180</td>\n",
       "      <td>0.6600</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>726812</td>\n",
       "      <td>AG</td>\n",
       "      <td>A</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>0.0180</td>\n",
       "      <td>0.6600</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999995</th>\n",
       "      <td>2</td>\n",
       "      <td>142257738</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.020</td>\n",
       "      <td>-0.0030</td>\n",
       "      <td>0.0180</td>\n",
       "      <td>0.8700</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999996</th>\n",
       "      <td>2</td>\n",
       "      <td>142257971</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.540</td>\n",
       "      <td>-0.0057</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.2700</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999997</th>\n",
       "      <td>2</td>\n",
       "      <td>142258001</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.370</td>\n",
       "      <td>0.0140</td>\n",
       "      <td>0.0053</td>\n",
       "      <td>0.0068</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999998</th>\n",
       "      <td>2</td>\n",
       "      <td>142258020</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.0093</td>\n",
       "      <td>0.0230</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999999</th>\n",
       "      <td>2</td>\n",
       "      <td>142258089</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.530</td>\n",
       "      <td>-0.0110</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.0330</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000000 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CHR        POS  EA       NEA    EAF    BETA      SE       P   STATUS\n",
       "0        1     714646   T  TTTGAAAC  0.026  0.0093  0.0170  0.5900  9999999\n",
       "1        1     725256   T         C  0.027  0.0076  0.0170  0.6500  9999999\n",
       "2        1     726805  GA         G  0.026  0.0078  0.0180  0.6600  9999999\n",
       "3        1     726807   T       TTG  0.026  0.0078  0.0180  0.6600  9999999\n",
       "4        1     726812  AG         A  0.026  0.0078  0.0180  0.6600  9999999\n",
       "...     ..        ...  ..       ...    ...     ...     ...     ...      ...\n",
       "999995   2  142257738   A         G  0.020 -0.0030  0.0180  0.8700  9999999\n",
       "999996   2  142257971   G         T  0.540 -0.0057  0.0051  0.2700  9999999\n",
       "999997   2  142258001   A         G  0.370  0.0140  0.0053  0.0068  9999999\n",
       "999998   2  142258020   A         G  0.082  0.0210  0.0093  0.0230  9999999\n",
       "999999   2  142258089   G         A  0.530 -0.0110  0.0051  0.0330  9999999\n",
       "\n",
       "[1000000 rows x 9 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d5dd6550",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 30 12:04:50 2024 Start to check IDs...v3.4.37\n",
      "Tue Jan 30 12:04:50 2024  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 73.81 MB\n",
      "Tue Jan 30 12:04:50 2024 Finished checking IDs successfully!\n",
      "Tue Jan 30 12:04:50 2024 Start to fix chromosome notation...v3.4.37\n",
      "Tue Jan 30 12:04:50 2024  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 73.81 MB\n",
      "Tue Jan 30 12:04:50 2024  -Checking CHR data type...\n",
      "Tue Jan 30 12:04:52 2024  -Variants with standardized chromosome notation: 1000000\n",
      "Tue Jan 30 12:04:54 2024  -All CHR are already fixed...\n",
      "Tue Jan 30 12:04:57 2024 Finished fixing chromosome notation successfully!\n",
      "Tue Jan 30 12:04:57 2024 Start to fix basepair positions...v3.4.37\n",
      "Tue Jan 30 12:04:57 2024  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 74.77 MB\n",
      "Tue Jan 30 12:04:57 2024  -Converting to Int64 data type ...\n",
      "Tue Jan 30 12:04:58 2024  -Position bound:(0 , 250,000,000)\n",
      "Tue Jan 30 12:05:00 2024  -Removed outliers: 0\n",
      "Tue Jan 30 12:05:00 2024  -Converted all position to datatype Int64.\n",
      "Tue Jan 30 12:05:00 2024 Finished fixing basepair position successfully!\n",
      "Tue Jan 30 12:05:00 2024 Start to fix alleles...v3.4.37\n",
      "Tue Jan 30 12:05:00 2024  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 83.35 MB\n",
      "Tue Jan 30 12:05:00 2024  -Converted all bases to string datatype and UPPERCASE.\n",
      "Tue Jan 30 12:05:01 2024  -Variants with bad EA  : 464\n",
      "Tue Jan 30 12:05:01 2024  -Variants with bad NEA : 0\n",
      "Tue Jan 30 12:05:01 2024  -Variants with NA for EA or NEA: 0\n",
      "Tue Jan 30 12:05:01 2024  -Variants with same EA and NEA: 0\n",
      "Tue Jan 30 12:05:01 2024  -A look at the non-ATCG EA: {'<CN0>', '<INS:ME:ALU>'} ...\n",
      "Tue Jan 30 12:05:01 2024  -Detected 464 variants with alleles that contain bases other than A/C/T/G .\n",
      "Tue Jan 30 12:05:05 2024 Finished fixing allele successfully!\n",
      "Tue Jan 30 12:05:05 2024 Start sanity check for statistics...v3.4.37\n",
      "Tue Jan 30 12:05:05 2024  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 83.54 MB\n",
      "Tue Jan 30 12:05:05 2024  -Comparison tolerance for floats: 1e-07\n",
      "Tue Jan 30 12:05:05 2024  -Checking if  -1e-07 <EAF< 1.0000001  ...\n",
      "Tue Jan 30 12:05:05 2024  -Removed 0 variants with bad EAF.\n",
      "Tue Jan 30 12:05:05 2024  -Checking if  -1e-07 < P < 1.0000001  ...\n",
      "Tue Jan 30 12:05:05 2024  -Removed 0 variants with bad P.\n",
      "Tue Jan 30 12:05:05 2024  -Checking if  -100.0000001 <BETA< 100.0000001  ...\n",
      "Tue Jan 30 12:05:05 2024  -Removed 0 variants with bad BETA.\n",
      "Tue Jan 30 12:05:05 2024  -Checking if  -1e-07 <SE< inf  ...\n",
      "Tue Jan 30 12:05:05 2024  -Removed 0 variants with bad SE.\n",
      "Tue Jan 30 12:05:05 2024  -Checking STATUS and converting STATUS to categories....\n",
      "Tue Jan 30 12:05:06 2024  -Removed 0 variants with bad statistics in total.\n",
      "Tue Jan 30 12:05:06 2024  -Data types for each column:\n",
      "Tue Jan 30 12:05:06 2024  -Column  : CHR   POS   EA       NEA      EAF     BETA    SE      P       STATUS  \n",
      "Tue Jan 30 12:05:06 2024  -DType   : Int64 Int64 category category float64 float64 float64 float64 category\n",
      "Tue Jan 30 12:05:06 2024  -Verified: T     T     T        T        T       T       T       T       T       \n",
      "Tue Jan 30 12:05:06 2024 Finished sanity check successfully!\n",
      "Tue Jan 30 12:05:06 2024 Start to check data consistency across columns...v3.4.37\n",
      "Tue Jan 30 12:05:06 2024  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 83.54 MB\n",
      "Tue Jan 30 12:05:06 2024  -Tolerance: 0.001 (Relative) and 0.001 (Absolute)\n",
      "Tue Jan 30 12:05:06 2024  -SNPID/rsID not available...SKipping\n",
      "Tue Jan 30 12:05:06 2024 Finished checking data consistency across columns.\n",
      "Tue Jan 30 12:05:06 2024 Start to normalize variants...v3.4.37\n",
      "Tue Jan 30 12:05:06 2024  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 83.54 MB\n",
      "Tue Jan 30 12:05:07 2024  -Not normalized allele IDs:[656] [1089] [1592] [2366] [3221] ... \n",
      "Tue Jan 30 12:05:07 2024  -Not normalized allele:['TTGTGTG' 'TTGTG']['TTG' 'CTG']['TAC' 'TACAC']['CAGAGAGAG' 'CAGAGAG']['TGT' 'CGT']... \n",
      "Tue Jan 30 12:05:07 2024  -Modified 3787 variants according to parsimony and left alignment principal.\n",
      "Tue Jan 30 12:05:07 2024 Finished normalizing variants successfully!\n",
      "Tue Jan 30 12:05:07 2024 Start to sort the genome coordinates...v3.4.37\n",
      "Tue Jan 30 12:05:07 2024  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 83.54 MB\n",
      "Tue Jan 30 12:05:07 2024  -Sorting genome coordinates...\n",
      "Tue Jan 30 12:05:07 2024 Finished sorting genome coordinates successfully!\n",
      "Tue Jan 30 12:05:07 2024 Start to reorder the columns...v3.4.37\n",
      "Tue Jan 30 12:05:07 2024  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 75.91 MB\n",
      "Tue Jan 30 12:05:07 2024  -Reordering columns to    : CHR,POS,EA,NEA,EAF,BETA,SE,P,STATUS\n",
      "Tue Jan 30 12:05:07 2024 Finished sorting columns successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.basic_check(verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c159267f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 30 12:05:07 2024 Start to check IDs...v3.4.37\n",
      "Tue Jan 30 12:05:07 2024  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 75.91 MB\n",
      "Tue Jan 30 12:05:10 2024  -Filling 0 SNPID using CHR:POS...\n",
      "Tue Jan 30 12:05:10 2024  -Filling 1000000 SNPID using CHR:POS:NEA:EA...\n",
      "Tue Jan 30 12:05:12 2024  -Fixed 1000000 variants ID...\n",
      "Tue Jan 30 12:05:12 2024 Finished checking IDs successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats.fix_id(fixid=True,forcefixid=True,overwrite=True,verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "887b35ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>SNPID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>714646</td>\n",
       "      <td>T</td>\n",
       "      <td>TTTGAAAC</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.0093</td>\n",
       "      <td>0.0170</td>\n",
       "      <td>0.5900</td>\n",
       "      <td>9960399</td>\n",
       "      <td>1:714646:TTTGAAAC:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>725256</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.027</td>\n",
       "      <td>0.0076</td>\n",
       "      <td>0.0170</td>\n",
       "      <td>0.6500</td>\n",
       "      <td>9960099</td>\n",
       "      <td>1:725256:C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>726805</td>\n",
       "      <td>GA</td>\n",
       "      <td>G</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>0.0180</td>\n",
       "      <td>0.6600</td>\n",
       "      <td>9960399</td>\n",
       "      <td>1:726805:G:GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>726807</td>\n",
       "      <td>T</td>\n",
       "      <td>TTG</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>0.0180</td>\n",
       "      <td>0.6600</td>\n",
       "      <td>9960399</td>\n",
       "      <td>1:726807:TTG:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>726812</td>\n",
       "      <td>AG</td>\n",
       "      <td>A</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>0.0180</td>\n",
       "      <td>0.6600</td>\n",
       "      <td>9960399</td>\n",
       "      <td>1:726812:A:AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999995</th>\n",
       "      <td>2</td>\n",
       "      <td>142257738</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.020</td>\n",
       "      <td>-0.0030</td>\n",
       "      <td>0.0180</td>\n",
       "      <td>0.8700</td>\n",
       "      <td>9960099</td>\n",
       "      <td>2:142257738:G:A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999996</th>\n",
       "      <td>2</td>\n",
       "      <td>142257971</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.540</td>\n",
       "      <td>-0.0057</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.2700</td>\n",
       "      <td>9960099</td>\n",
       "      <td>2:142257971:T:G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999997</th>\n",
       "      <td>2</td>\n",
       "      <td>142258001</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.370</td>\n",
       "      <td>0.0140</td>\n",
       "      <td>0.0053</td>\n",
       "      <td>0.0068</td>\n",
       "      <td>9960099</td>\n",
       "      <td>2:142258001:G:A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999998</th>\n",
       "      <td>2</td>\n",
       "      <td>142258020</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.0093</td>\n",
       "      <td>0.0230</td>\n",
       "      <td>9960099</td>\n",
       "      <td>2:142258020:G:A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999999</th>\n",
       "      <td>2</td>\n",
       "      <td>142258089</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.530</td>\n",
       "      <td>-0.0110</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.0330</td>\n",
       "      <td>9960099</td>\n",
       "      <td>2:142258089:A:G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000000 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHR        POS  EA       NEA    EAF    BETA      SE       P   STATUS  \\\n",
       "0         1     714646   T  TTTGAAAC  0.026  0.0093  0.0170  0.5900  9960399   \n",
       "1         1     725256   T         C  0.027  0.0076  0.0170  0.6500  9960099   \n",
       "2         1     726805  GA         G  0.026  0.0078  0.0180  0.6600  9960399   \n",
       "3         1     726807   T       TTG  0.026  0.0078  0.0180  0.6600  9960399   \n",
       "4         1     726812  AG         A  0.026  0.0078  0.0180  0.6600  9960399   \n",
       "...     ...        ...  ..       ...    ...     ...     ...     ...      ...   \n",
       "999995    2  142257738   A         G  0.020 -0.0030  0.0180  0.8700  9960099   \n",
       "999996    2  142257971   G         T  0.540 -0.0057  0.0051  0.2700  9960099   \n",
       "999997    2  142258001   A         G  0.370  0.0140  0.0053  0.0068  9960099   \n",
       "999998    2  142258020   A         G  0.082  0.0210  0.0093  0.0230  9960099   \n",
       "999999    2  142258089   G         A  0.530 -0.0110  0.0051  0.0330  9960099   \n",
       "\n",
       "                      SNPID  \n",
       "0       1:714646:TTTGAAAC:T  \n",
       "1              1:725256:C:T  \n",
       "2             1:726805:G:GA  \n",
       "3            1:726807:TTG:T  \n",
       "4             1:726812:A:AG  \n",
       "...                     ...  \n",
       "999995      2:142257738:G:A  \n",
       "999996      2:142257971:T:G  \n",
       "999997      2:142258001:G:A  \n",
       "999998      2:142258020:G:A  \n",
       "999999      2:142258089:A:G  \n",
       "\n",
       "[1000000 rows x 10 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5ca780bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "dirty_sumstats = gl.Sumstats(\"../t2d_bbj.txt.gz\", \n",
    "                             snpid=\"SNP\",\n",
    "                             chrom=\"CHR\",\n",
    "                             pos=\"POS\",\n",
    "                             ea=\"REF\",\n",
    "                             nea=\"ALT\",\n",
    "                             eaf=\"Frq\",\n",
    "                             beta=\"BETA\",\n",
    "                             se=\"SE\",\n",
    "                             p=\"P\",\n",
    "                             nrows=1000000,\n",
    "                             verbose=False) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "af4e0665",
   "metadata": {},
   "outputs": [],
   "source": [
    "dirty_sumstats.basic_check(verbose=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d02ba9d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 30 12:05:34 2024 Dropping old IDs:['SNPID']\n",
      "Tue Jan 30 12:05:34 2024 After merging by CHR and POS:613216\n",
      "Tue Jan 30 12:05:34 2024 Matched variants:612148\n",
      "Tue Jan 30 12:05:36 2024  -Current Dataframe shape : 612148 x 17 ; Memory usage: 107.64 MB\n",
      "Tue Jan 30 12:05:36 2024 Start to flip allele-specific stats for SNPs with status xxxxx[35]x: alt->ea , ref->nea ...v3.4.37\n",
      "Tue Jan 30 12:05:36 2024  -Flipping 282001 variants...\n",
      "Tue Jan 30 12:05:36 2024  -Swapping column: NEA <=> EA...\n",
      "Tue Jan 30 12:05:37 2024  -Flipping column: BETA = - BETA...\n",
      "Tue Jan 30 12:05:37 2024  -Flipping column: EAF = 1 - EAF...\n",
      "Tue Jan 30 12:05:37 2024  -Changed the status for flipped variants : xxxxx[35]x -> xxxxx[12]x\n",
      "Tue Jan 30 12:05:38 2024 Finished converting successfully!\n",
      "Tue Jan 30 12:05:38 2024  -Renaming sumstats2 columns by adding suffix _2\n",
      "Tue Jan 30 12:05:38 2024 Start to reorder the columns...\n",
      "Tue Jan 30 12:05:38 2024  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,EAF_1,BETA_1,SE_1,P_1,STATUS_1,EAF_2,BETA_2,SE_2,P_2,STATUS_2\n",
      "Tue Jan 30 12:05:38 2024 Finished sorting columns successfully!\n"
     ]
    }
   ],
   "source": [
    "pair = gl.SumstatsPair(mysumstats, dirty_sumstats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ad62414e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF_1</th>\n",
       "      <th>BETA_1</th>\n",
       "      <th>SE_1</th>\n",
       "      <th>P_1</th>\n",
       "      <th>STATUS_1</th>\n",
       "      <th>EAF_2</th>\n",
       "      <th>BETA_2</th>\n",
       "      <th>SE_2</th>\n",
       "      <th>P_2</th>\n",
       "      <th>STATUS_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:751343:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>751343</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.16</td>\n",
       "      <td>-0.00160</td>\n",
       "      <td>0.0074</td>\n",
       "      <td>0.82</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.1407</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:751488:G:GA</td>\n",
       "      <td>1</td>\n",
       "      <td>751488</td>\n",
       "      <td>GA</td>\n",
       "      <td>G</td>\n",
       "      <td>0.16</td>\n",
       "      <td>-0.00390</td>\n",
       "      <td>0.0077</td>\n",
       "      <td>0.61</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.1470</td>\n",
       "      <td>0.0187</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0.28300</td>\n",
       "      <td>9960319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:751756:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>751756</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.16</td>\n",
       "      <td>-0.00150</td>\n",
       "      <td>0.0074</td>\n",
       "      <td>0.84</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.1400</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>9960019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:752566:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>752566</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.00091</td>\n",
       "      <td>0.0073</td>\n",
       "      <td>0.90</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.8422</td>\n",
       "      <td>-0.0155</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.23500</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:752894:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>752894</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.00130</td>\n",
       "      <td>0.0074</td>\n",
       "      <td>0.86</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.8560</td>\n",
       "      <td>-0.0169</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0.25250</td>\n",
       "      <td>9960019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613211</th>\n",
       "      <td>2:6347478:T:C</td>\n",
       "      <td>2</td>\n",
       "      <td>6347478</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.22</td>\n",
       "      <td>-0.00270</td>\n",
       "      <td>0.0063</td>\n",
       "      <td>0.67</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.2093</td>\n",
       "      <td>-0.0149</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>0.17790</td>\n",
       "      <td>9960019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613212</th>\n",
       "      <td>2:6347639:C:A</td>\n",
       "      <td>2</td>\n",
       "      <td>6347639</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.22</td>\n",
       "      <td>-0.00280</td>\n",
       "      <td>0.0063</td>\n",
       "      <td>0.66</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.2100</td>\n",
       "      <td>-0.0159</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>0.15100</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613213</th>\n",
       "      <td>2:6347694:G:C</td>\n",
       "      <td>2</td>\n",
       "      <td>6347694</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.22</td>\n",
       "      <td>-0.00270</td>\n",
       "      <td>0.0063</td>\n",
       "      <td>0.67</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.2093</td>\n",
       "      <td>-0.0151</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>0.17250</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613214</th>\n",
       "      <td>2:6348478:G:A</td>\n",
       "      <td>2</td>\n",
       "      <td>6348478</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.22</td>\n",
       "      <td>-0.00280</td>\n",
       "      <td>0.0063</td>\n",
       "      <td>0.65</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.2140</td>\n",
       "      <td>-0.0152</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>0.17160</td>\n",
       "      <td>9960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613215</th>\n",
       "      <td>2:6348754:A:C</td>\n",
       "      <td>2</td>\n",
       "      <td>6348754</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.14</td>\n",
       "      <td>-0.00350</td>\n",
       "      <td>0.0073</td>\n",
       "      <td>0.64</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.1572</td>\n",
       "      <td>-0.0297</td>\n",
       "      <td>0.0126</td>\n",
       "      <td>0.01846</td>\n",
       "      <td>9960019</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>612148 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                SNPID  CHR      POS  EA NEA  EAF_1   BETA_1    SE_1   P_1  \\\n",
       "0        1:751343:T:A    1   751343   A   T   0.16 -0.00160  0.0074  0.82   \n",
       "1       1:751488:G:GA    1   751488  GA   G   0.16 -0.00390  0.0077  0.61   \n",
       "2        1:751756:T:C    1   751756   C   T   0.16 -0.00150  0.0074  0.84   \n",
       "3        1:752566:G:A    1   752566   A   G   0.83  0.00091  0.0073  0.90   \n",
       "4        1:752894:T:C    1   752894   C   T   0.84  0.00130  0.0074  0.86   \n",
       "...               ...  ...      ...  ..  ..    ...      ...     ...   ...   \n",
       "613211  2:6347478:T:C    2  6347478   C   T   0.22 -0.00270  0.0063  0.67   \n",
       "613212  2:6347639:C:A    2  6347639   A   C   0.22 -0.00280  0.0063  0.66   \n",
       "613213  2:6347694:G:C    2  6347694   C   G   0.22 -0.00270  0.0063  0.67   \n",
       "613214  2:6348478:G:A    2  6348478   A   G   0.22 -0.00280  0.0063  0.65   \n",
       "613215  2:6348754:A:C    2  6348754   C   A   0.14 -0.00350  0.0073  0.64   \n",
       "\n",
       "       STATUS_1   EAF_2  BETA_2    SE_2      P_2 STATUS_2  \n",
       "0       9960099  0.1407  0.0172  0.0156  0.27050  9960099  \n",
       "1       9960399  0.1470  0.0187  0.0174  0.28300  9960319  \n",
       "2       9960099  0.1400  0.0172  0.0156  0.27050  9960019  \n",
       "3       9960099  0.8422 -0.0155  0.0131  0.23500  9960099  \n",
       "4       9960099  0.8560 -0.0169  0.0147  0.25250  9960019  \n",
       "...         ...     ...     ...     ...      ...      ...  \n",
       "613211  9960099  0.2093 -0.0149  0.0111  0.17790  9960019  \n",
       "613212  9960099  0.2100 -0.0159  0.0111  0.15100  9960099  \n",
       "613213  9960099  0.2093 -0.0151  0.0111  0.17250  9960099  \n",
       "613214  9960099  0.2140 -0.0152  0.0111  0.17160  9960099  \n",
       "613215  9960099  0.1572 -0.0297  0.0126  0.01846  9960019  \n",
       "\n",
       "[612148 rows x 15 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pair.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a656347d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 30 12:05:38 2024 Start to extract lead variants...\n",
      "Tue Jan 30 12:05:38 2024  -Processing 612148 variants...\n",
      "Tue Jan 30 12:05:38 2024  -Significance threshold : 5e-08\n",
      "Tue Jan 30 12:05:38 2024  -Sliding window size: 1000  kb\n",
      "Tue Jan 30 12:05:38 2024  -Using P_1 for extracting lead variants...\n",
      "Tue Jan 30 12:05:38 2024  -Found 289 significant variants in total...\n",
      "Tue Jan 30 12:05:39 2024  -Identified 2 lead variants!\n",
      "Tue Jan 30 12:05:39 2024 Finished extracting lead variants successfully!\n",
      "Tue Jan 30 12:05:39 2024  -Dropping duplicated SNPIDs...\n",
      "Tue Jan 30 12:05:39 2024  -Processing locus with lead variant 1:177894591:T:G at CHR 1 POS 177894591 ...\n",
      "Tue Jan 30 12:05:39 2024  -Processing VCF : /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz...\n",
      "Tue Jan 30 12:05:39 2024    -PLINK version: PLINK v2.00a5.9LM AVX2 AMD (12 Dec 2023)\n",
      "Tue Jan 30 12:05:39 2024   -Processing VCF for CHR 1...\n",
      "Tue Jan 30 12:05:39 2024   -Plink bfile for CHR 1 exists. Skipping...\n",
      "Tue Jan 30 12:05:44 2024    -#variants in ref file: 6471908\n",
      "Tue Jan 30 12:05:44 2024    -#variants in locus (1:177894591:T:G): 4678\n",
      "Tue Jan 30 12:05:50 2024    -#Variants with matched alleles:4675\n",
      "Tue Jan 30 12:05:50 2024    -#Variants with flipped alleles:4675\n",
      "Tue Jan 30 12:05:50 2024  -Start to calculate LD r matrix...\n",
      "Tue Jan 30 12:05:50 2024    -PLINK version: PLINK v1.90b7.2 64-bit (11 Dec 2023)\n",
      "Tue Jan 30 12:06:03 2024  -Finished calculating LD r for locus with lead variant 1:177894591:T:G at CHR 1 POS 177894591...\n",
      "Tue Jan 30 12:06:03 2024  -Sumstats SNPID order and LD matrix SNPID order are matched.\n",
      "Tue Jan 30 12:06:03 2024  -Processing locus with lead variant 2:632723:C:T at CHR 2 POS 632723 ...\n",
      "Tue Jan 30 12:06:03 2024  -Processing VCF : /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz...\n",
      "Tue Jan 30 12:06:03 2024    -PLINK version: PLINK v2.00a5.9LM AVX2 AMD (12 Dec 2023)\n",
      "Tue Jan 30 12:06:03 2024   -Processing VCF for CHR 2...\n",
      "Tue Jan 30 12:06:03 2024   -Plink bfile for CHR 2 exists. Skipping...\n",
      "Tue Jan 30 12:06:09 2024    -#variants in ref file: 7085910\n",
      "Tue Jan 30 12:06:09 2024    -#variants in locus (2:632723:C:T): 6644\n",
      "Tue Jan 30 12:06:15 2024    -#Variants with matched alleles:6644\n",
      "Tue Jan 30 12:06:15 2024    -#Variants with flipped alleles:6644\n",
      "Tue Jan 30 12:06:15 2024  -Start to calculate LD r matrix...\n",
      "Tue Jan 30 12:06:15 2024    -PLINK version: PLINK v1.90b7.2 64-bit (11 Dec 2023)\n",
      "Tue Jan 30 12:06:40 2024  -Finished calculating LD r for locus with lead variant 2:632723:C:T at CHR 2 POS 632723...\n",
      "Tue Jan 30 12:06:40 2024  -Sumstats SNPID order and LD matrix SNPID order are matched.\n",
      "Tue Jan 30 12:06:40 2024  -File list is saved to: ./Study_1_Study_1_2loci_1000kb.filelist\n",
      "Tue Jan 30 12:06:40 2024  -Finished LD matrix calculation.\n"
     ]
    }
   ],
   "source": [
    "pair.to_coloc(vcf = \"/home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz\",n_cores=1, memory=10000)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d421ca2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Jan 30 12:06:40 2024  Start to run coloc.susie from command line:\n",
      "Tue Jan 30 12:06:40 2024  -Phenotype types: quant and cc\n",
      "Tue Jan 30 12:06:40 2024  -Ns: 50000 and 150000\n",
      "Tue Jan 30 12:06:40 2024  -R version: Rscript (R) version 4.2.3 (2023-03-15)\n",
      "Tue Jan 30 12:06:40 2024  -Running for: 1:177894591:T:G - Study_1_Study_1\n",
      "Tue Jan 30 12:06:40 2024   -Locus sumstats:./Study_1_Study_1_1:177894591:T:G_1000.sumstats.gz\n",
      "Tue Jan 30 12:06:40 2024   -LD r matrix:./Study_1_Study_1_1:177894591:T:G_1000.ld.gz\n",
      "Tue Jan 30 12:06:40 2024   -output_prefix:./Study_1_Study_1_1:177894591:T:G_1000\n",
      "Tue Jan 30 12:06:40 2024   -coloc script: coloc.susie(S1,S2)\n",
      "Tue Jan 30 12:07:14 2024  Running coloc.SuSieR from command line...\n",
      "Tue Jan 30 12:07:14 2024   -SuSieR result for ./Study_1_Study_1_1:177894591:T:G_1000 is empty. Please check parameters.\n",
      "Tue Jan 30 12:07:14 2024   -SuSieR result summary to: ./Study_1_Study_1_1:177894591:T:G_1000.pipcs\n",
      "Tue Jan 30 12:07:14 2024  -Running for: 2:632723:C:T - Study_1_Study_1\n",
      "Tue Jan 30 12:07:14 2024   -Locus sumstats:./Study_1_Study_1_2:632723:C:T_1000.sumstats.gz\n",
      "Tue Jan 30 12:07:14 2024   -LD r matrix:./Study_1_Study_1_2:632723:C:T_1000.ld.gz\n",
      "Tue Jan 30 12:07:14 2024   -output_prefix:./Study_1_Study_1_2:632723:C:T_1000\n",
      "Tue Jan 30 12:07:14 2024   -coloc script: coloc.susie(S1,S2)\n",
      "Tue Jan 30 12:08:18 2024  Running coloc.SuSieR from command line...\n",
      "Tue Jan 30 12:08:18 2024   -SuSieR result summary to: ./Study_1_Study_1_2:632723:C:T_1000.pipcs\n",
      "Tue Jan 30 12:08:18 2024 Finished finemapping using SuSieR.\n"
     ]
    }
   ],
   "source": [
    "pair.run_coloc_susie(r=\"/home/yunye/anaconda3/envs/r42/bin/Rscript\",types=(\"quant\",\"cc\"),d1_args=\",sdY=1\", ns=(50000,150000))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d874253b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
