{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e015760b",
   "metadata": {},
   "source": [
    "# Workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3e7a1f9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46661590-88b9-439e-be24-a6eb7614a529",
   "metadata": {},
   "source": [
    "## Load test data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0b49d3af-5ba4-4400-8f36-6cfe005fdcfe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Feb  2 19:36:24 2024 GWASLab v3.4.38 https://cloufield.github.io/gwaslab/\n",
      "Fri Feb  2 19:36:24 2024 (C) 2022-2024, Yunye He, Kamatani Lab, MIT License, gwaslab@gmail.com\n",
      "Fri Feb  2 19:36:24 2024 Start to load format from formatbook....\n",
      "Fri Feb  2 19:36:24 2024  -gwaslab format meta info:\n",
      "Fri Feb  2 19:36:24 2024   - format_name  : gwaslab\n",
      "Fri Feb  2 19:36:24 2024   - format_source  : https://cloufield.github.io/gwaslab/\n",
      "Fri Feb  2 19:36:24 2024   - format_version  : 20231220_v4\n",
      "Fri Feb  2 19:36:24 2024 Start to initialize gl.Sumstats from file :/home/yunye/work/gwaslab/examples/toy_data/to_harmonize.tsv\n",
      "Fri Feb  2 19:36:24 2024  -Reading columns          : CHR,NEA,SE,P,DIRECTION,BETA,SNPID,NOTE,N,EAF,EA,POS\n",
      "Fri Feb  2 19:36:24 2024  -Renaming columns to      : CHR,NEA,SE,P,DIRECTION,BETA,SNPID,NOTE,N,EAF,EA,POS\n",
      "Fri Feb  2 19:36:24 2024  -Current Dataframe shape : 15  x  12\n",
      "Fri Feb  2 19:36:24 2024  -Initiating a status column: STATUS ...\n",
      "Fri Feb  2 19:36:24 2024  #WARNING! Version of genomic coordinates is unknown...\n",
      "Fri Feb  2 19:36:24 2024 Start to reorder the columns...v3.4.38\n",
      "Fri Feb  2 19:36:24 2024  -Current Dataframe shape : 15 x 13 ; Memory usage: 19.94 MB\n",
      "Fri Feb  2 19:36:24 2024  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,EAF,BETA,SE,P,N,DIRECTION,STATUS,NOTE\n",
      "Fri Feb  2 19:36:24 2024 Finished reordering the columns.\n",
      "Fri Feb  2 19:36:24 2024  -Column  : SNPID  CHR    POS   EA       NEA      EAF     BETA    SE      P       N     DIRECTION STATUS   NOTE  \n",
      "Fri Feb  2 19:36:24 2024  -DType   : object string int64 category category float64 float64 float64 float64 int64 object    category object\n",
      "Fri Feb  2 19:36:24 2024  -Verified: T      F      T     T        T        T       T       T       T       T     T         T        NA    \n",
      "Fri Feb  2 19:36:24 2024  #WARNING! Columns with possibly incompatable dtypes: CHR\n",
      "Fri Feb  2 19:36:24 2024  -Current Dataframe memory usage: 19.94 MB\n",
      "Fri Feb  2 19:36:24 2024 Finished loading data successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(\"/home/yunye/work/gwaslab/examples/toy_data/to_harmonize.tsv\",fmt=\"gwaslab\",other=[\"NOTE\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32fe1ac1",
   "metadata": {},
   "source": [
    "A list of variants that needs to be harmonized. Issues are described in NOTE column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2e7b6fbe-c71a-4816-a531-ffca02fefc15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>DIRECTION</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>NOTE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:1066403:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>1066403</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.5276</td>\n",
       "      <td>0.0043</td>\n",
       "      <td>0.0109</td>\n",
       "      <td>0.68910</td>\n",
       "      <td>191764</td>\n",
       "      <td>++-+</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Clean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:1163266:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1163266</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.8355</td>\n",
       "      <td>0.0122</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>0.30810</td>\n",
       "      <td>191764</td>\n",
       "      <td>-+++</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Flip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:1213224:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1213224</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.3345</td>\n",
       "      <td>0.0071</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>0.45280</td>\n",
       "      <td>191764</td>\n",
       "      <td>+++-</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Flip + Palindromic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:3066761:A:T</td>\n",
       "      <td>1</td>\n",
       "      <td>3066761</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1245</td>\n",
       "      <td>-0.0066</td>\n",
       "      <td>0.0157</td>\n",
       "      <td>0.67420</td>\n",
       "      <td>191764</td>\n",
       "      <td>-+++</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Palindromic + No flip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:3997271:T:TTTTA</td>\n",
       "      <td>1</td>\n",
       "      <td>3997271</td>\n",
       "      <td>TTTTA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.3840</td>\n",
       "      <td>0.0188</td>\n",
       "      <td>0.0135</td>\n",
       "      <td>0.16300</td>\n",
       "      <td>191764</td>\n",
       "      <td>++++</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Indel + Both on genome + No flip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:4007705:TC:T</td>\n",
       "      <td>1</td>\n",
       "      <td>4007705</td>\n",
       "      <td>TC</td>\n",
       "      <td>T</td>\n",
       "      <td>0.4368</td>\n",
       "      <td>0.0306</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>0.01421</td>\n",
       "      <td>191764</td>\n",
       "      <td>++++</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Indel  + Both on genome + Flip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1:5164281:A:AAGAG</td>\n",
       "      <td>1</td>\n",
       "      <td>5164281</td>\n",
       "      <td>AAGAG</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9871</td>\n",
       "      <td>0.0167</td>\n",
       "      <td>0.0454</td>\n",
       "      <td>0.71200</td>\n",
       "      <td>191764</td>\n",
       "      <td>-+++</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Indel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1:12799025:C:CCT</td>\n",
       "      <td>1</td>\n",
       "      <td>12799025</td>\n",
       "      <td>C</td>\n",
       "      <td>CCT</td>\n",
       "      <td>0.1730</td>\n",
       "      <td>-0.0186</td>\n",
       "      <td>0.0118</td>\n",
       "      <td>0.11500</td>\n",
       "      <td>191764</td>\n",
       "      <td>---+</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Indel + Flip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1:19323015:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>19323015</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4389</td>\n",
       "      <td>-0.0165</td>\n",
       "      <td>0.0088</td>\n",
       "      <td>0.05949</td>\n",
       "      <td>191764</td>\n",
       "      <td>----</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Palindromic + Flip + Indistinguishable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1:22841855:A:T</td>\n",
       "      <td>1</td>\n",
       "      <td>22841855</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4527</td>\n",
       "      <td>0.0080</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>0.39980</td>\n",
       "      <td>191764</td>\n",
       "      <td>-++-</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Palindromic + Indistinguishable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3:183629306:T:TTCTC</td>\n",
       "      <td>3</td>\n",
       "      <td>183629306</td>\n",
       "      <td>TTCTC</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0082</td>\n",
       "      <td>-0.0083</td>\n",
       "      <td>0.0555</td>\n",
       "      <td>0.88060</td>\n",
       "      <td>191764</td>\n",
       "      <td>+--+</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Indel  + Both on genome + No information</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>4:99731866:G:C</td>\n",
       "      <td>4</td>\n",
       "      <td>99731866</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.3883</td>\n",
       "      <td>0.0052</td>\n",
       "      <td>0.0093</td>\n",
       "      <td>0.57360</td>\n",
       "      <td>191764</td>\n",
       "      <td>+-++</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Palindromic+ Flip + Different strand (REF G, T...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>8:89935201:C:G</td>\n",
       "      <td>8</td>\n",
       "      <td>89935201</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.3533</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>0.0094</td>\n",
       "      <td>0.27560</td>\n",
       "      <td>191764</td>\n",
       "      <td>++++</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Palindromic + Different strand (REF A; TOMMO G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>X:4206466:G:C</td>\n",
       "      <td>X</td>\n",
       "      <td>4206466</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.8727</td>\n",
       "      <td>-0.0055</td>\n",
       "      <td>0.0104</td>\n",
       "      <td>0.59840</td>\n",
       "      <td>191764</td>\n",
       "      <td>-+--</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Palindromic + Flip + No information</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>X:5053229:A:T</td>\n",
       "      <td>X</td>\n",
       "      <td>5053229</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9747</td>\n",
       "      <td>0.0218</td>\n",
       "      <td>0.0225</td>\n",
       "      <td>0.33180</td>\n",
       "      <td>191764</td>\n",
       "      <td>++++</td>\n",
       "      <td>9999999</td>\n",
       "      <td>Palindromic + No information</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  SNPID CHR        POS     EA  NEA     EAF    BETA      SE  \\\n",
       "0         1:1066403:T:C   1    1066403      C    T  0.5276  0.0043  0.0109   \n",
       "1         1:1163266:G:A   1    1163266      G    A  0.8355  0.0122  0.0120   \n",
       "2         1:1213224:T:A   1    1213224      T    A  0.3345  0.0071  0.0095   \n",
       "3         1:3066761:A:T   1    3066761      T    A  0.1245 -0.0066  0.0157   \n",
       "4     1:3997271:T:TTTTA   1    3997271  TTTTA    T  0.3840  0.0188  0.0135   \n",
       "5        1:4007705:TC:T   1    4007705     TC    T  0.4368  0.0306  0.0125   \n",
       "6     1:5164281:A:AAGAG   1    5164281  AAGAG    A  0.9871  0.0167  0.0454   \n",
       "7      1:12799025:C:CCT   1   12799025      C  CCT  0.1730 -0.0186  0.0118   \n",
       "8        1:19323015:T:A   1   19323015      T    A  0.4389 -0.0165  0.0088   \n",
       "9        1:22841855:A:T   1   22841855      T    A  0.4527  0.0080  0.0095   \n",
       "10  3:183629306:T:TTCTC   3  183629306  TTCTC    T  0.0082 -0.0083  0.0555   \n",
       "11       4:99731866:G:C   4   99731866      G    C  0.3883  0.0052  0.0093   \n",
       "12       8:89935201:C:G   8   89935201      G    C  0.3533  0.0103  0.0094   \n",
       "13        X:4206466:G:C   X    4206466      G    C  0.8727 -0.0055  0.0104   \n",
       "14        X:5053229:A:T   X    5053229      T    A  0.9747  0.0218  0.0225   \n",
       "\n",
       "          P       N DIRECTION   STATUS  \\\n",
       "0   0.68910  191764      ++-+  9999999   \n",
       "1   0.30810  191764      -+++  9999999   \n",
       "2   0.45280  191764      +++-  9999999   \n",
       "3   0.67420  191764      -+++  9999999   \n",
       "4   0.16300  191764      ++++  9999999   \n",
       "5   0.01421  191764      ++++  9999999   \n",
       "6   0.71200  191764      -+++  9999999   \n",
       "7   0.11500  191764      ---+  9999999   \n",
       "8   0.05949  191764      ----  9999999   \n",
       "9   0.39980  191764      -++-  9999999   \n",
       "10  0.88060  191764      +--+  9999999   \n",
       "11  0.57360  191764      +-++  9999999   \n",
       "12  0.27560  191764      ++++  9999999   \n",
       "13  0.59840  191764      -+--  9999999   \n",
       "14  0.33180  191764      ++++  9999999   \n",
       "\n",
       "                                                 NOTE  \n",
       "0                                               Clean  \n",
       "1                                                Flip  \n",
       "2                                  Flip + Palindromic  \n",
       "3                               Palindromic + No flip  \n",
       "4                    Indel + Both on genome + No flip  \n",
       "5                      Indel  + Both on genome + Flip  \n",
       "6                                              Indel   \n",
       "7                                        Indel + Flip  \n",
       "8              Palindromic + Flip + Indistinguishable  \n",
       "9                     Palindromic + Indistinguishable  \n",
       "10           Indel  + Both on genome + No information  \n",
       "11  Palindromic+ Flip + Different strand (REF G, T...  \n",
       "12  Palindromic + Different strand (REF A; TOMMO G...  \n",
       "13                Palindromic + Flip + No information  \n",
       "14                       Palindromic + No information  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "615e6907-62ad-4f3c-bf78-8b75a946168c",
   "metadata": {},
   "source": [
    "## Perform harmonization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce20b6c7-e927-462d-8215-ea464a57b399",
   "metadata": {},
   "source": [
    "- ref_seq : reference genome fasta file for allele alignment\n",
    "- ref_infer : vcf file with allele frequency information for inferring strand and comparing allele frequency \n",
    "- ref_alt_freq : field in INFO of vcf file for alternative allele frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f2551025-eeaf-4c59-a428-7def9d07afdf",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Feb  2 19:36:24 2024 Start to check SNPID/rsID...v3.4.38\n",
      "Fri Feb  2 19:36:24 2024  -Current Dataframe shape : 15 x 13 ; Memory usage: 19.94 MB\n",
      "Fri Feb  2 19:36:24 2024  -Checking SNPID data type...\n",
      "Fri Feb  2 19:36:24 2024  -Converting SNPID to pd.string data type...\n",
      "Fri Feb  2 19:36:24 2024  -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\n",
      "Fri Feb  2 19:36:25 2024 Finished checking SNPID/rsID.\n",
      "Fri Feb  2 19:36:25 2024 Start to fix chromosome notation (CHR)...v3.4.38\n",
      "Fri Feb  2 19:36:25 2024  -Current Dataframe shape : 15 x 13 ; Memory usage: 19.94 MB\n",
      "Fri Feb  2 19:36:25 2024  -Checking CHR data type...\n",
      "Fri Feb  2 19:36:25 2024  -Variants with standardized chromosome notation: 13\n",
      "Fri Feb  2 19:36:25 2024  -Variants with fixable chromosome notations: 2\n",
      "Fri Feb  2 19:36:25 2024  -No unrecognized chromosome notations...\n",
      "Fri Feb  2 19:36:25 2024  -Identifying non-autosomal chromosomes : X, Y, and MT ...\n",
      "Fri Feb  2 19:36:25 2024  -Identified  2  variants on sex chromosomes...\n",
      "Fri Feb  2 19:36:25 2024  -Standardizing sex chromosome notations: X to 23...\n",
      "Fri Feb  2 19:36:27 2024 Finished fixing chromosome notation (CHR).\n",
      "Fri Feb  2 19:36:27 2024 Start to fix basepair positions (POS)...v3.4.38\n",
      "Fri Feb  2 19:36:27 2024  -Current Dataframe shape : 15 x 13 ; Memory usage: 19.94 MB\n",
      "Fri Feb  2 19:36:27 2024  -Converting to Int64 data type ...\n",
      "Fri Feb  2 19:36:28 2024  -Position bound:(0 , 250,000,000)\n",
      "Fri Feb  2 19:36:28 2024  -Removed outliers: 0\n",
      "Fri Feb  2 19:36:28 2024 Finished fixing basepair positions (POS).\n",
      "Fri Feb  2 19:36:28 2024 Start to fix alleles (EA and NEA)...v3.4.38\n",
      "Fri Feb  2 19:36:28 2024  -Current Dataframe shape : 15 x 13 ; Memory usage: 19.94 MB\n",
      "Fri Feb  2 19:36:28 2024  -Converted all bases to string datatype and UPPERCASE.\n",
      "Fri Feb  2 19:36:28 2024  -Variants with bad EA  : 0\n",
      "Fri Feb  2 19:36:28 2024  -Variants with bad NEA : 0\n",
      "Fri Feb  2 19:36:28 2024  -Variants with NA for EA or NEA: 0\n",
      "Fri Feb  2 19:36:28 2024  -Variants with same EA and NEA: 0\n",
      "Fri Feb  2 19:36:28 2024  -Detected 0 variants with alleles that contain bases other than A/C/T/G .\n",
      "Fri Feb  2 19:36:30 2024 Finished fixing alleles (EA and NEA).\n",
      "Fri Feb  2 19:36:30 2024 Start to perform sanity check for statistics...v3.4.38\n",
      "Fri Feb  2 19:36:30 2024  -Current Dataframe shape : 15 x 13 ; Memory usage: 19.94 MB\n",
      "Fri Feb  2 19:36:30 2024  -Comparison tolerance for floats: 1e-07\n",
      "Fri Feb  2 19:36:30 2024  -Checking if 0 <= N <= 2147483647 ...\n",
      "Fri Feb  2 19:36:30 2024  -Removed 0 variants with bad/na N.\n",
      "Fri Feb  2 19:36:30 2024  -Checking if -1e-07 < EAF < 1.0000001 ...\n",
      "Fri Feb  2 19:36:30 2024  -Removed 0 variants with bad/na EAF.\n",
      "Fri Feb  2 19:36:30 2024  -Checking if -1e-07 < P < 1.0000001 ...\n",
      "Fri Feb  2 19:36:30 2024  -Removed 0 variants with bad/na P.\n",
      "Fri Feb  2 19:36:30 2024  -Checking if -100.0000001 < BETA < 100.0000001 ...\n",
      "Fri Feb  2 19:36:30 2024  -Removed 0 variants with bad/na BETA.\n",
      "Fri Feb  2 19:36:30 2024  -Checking if -1e-07 < SE < inf ...\n",
      "Fri Feb  2 19:36:30 2024  -Removed 0 variants with bad/na SE.\n",
      "Fri Feb  2 19:36:30 2024  -Checking STATUS and converting STATUS to categories....\n",
      "Fri Feb  2 19:36:30 2024  -Removed 0 variants with bad statistics in total.\n",
      "Fri Feb  2 19:36:30 2024  -Data types for each column:\n",
      "Fri Feb  2 19:36:30 2024  -Column  : SNPID  CHR   POS   EA       NEA      EAF     BETA    SE      P       N     DIRECTION STATUS   NOTE  \n",
      "Fri Feb  2 19:36:30 2024  -DType   : string Int64 Int64 category category float32 float64 float64 float64 Int64 object    category object\n",
      "Fri Feb  2 19:36:30 2024  -Verified: T      T     T     T        T        T       T       T       T       T     T         T        NA    \n",
      "Fri Feb  2 19:36:30 2024 Finished sanity check for statistics.\n",
      "Fri Feb  2 19:36:30 2024 Start to normalize indels...v3.4.38\n",
      "Fri Feb  2 19:36:30 2024  -Current Dataframe shape : 15 x 13 ; Memory usage: 19.94 MB\n",
      "Fri Feb  2 19:36:30 2024  -No available variants to normalize..\n",
      "Fri Feb  2 19:36:30 2024 Finished normalizing variants successfully!\n",
      "Fri Feb  2 19:36:30 2024 Start to reorder the columns...v3.4.38\n",
      "Fri Feb  2 19:36:30 2024  -Current Dataframe shape : 15 x 13 ; Memory usage: 19.94 MB\n",
      "Fri Feb  2 19:36:30 2024  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,EAF,BETA,SE,P,N,DIRECTION,STATUS,NOTE\n",
      "Fri Feb  2 19:36:30 2024 Finished reordering the columns.\n",
      "Fri Feb  2 19:36:31 2024 Start to check if NEA is aligned with reference sequence...v3.4.38\n",
      "Fri Feb  2 19:36:31 2024  -Current Dataframe shape : 15 x 13 ; Memory usage: 19.94 MB\n",
      "Fri Feb  2 19:36:31 2024  -Reference genome FASTA file: /home/yunye/.gwaslab/hg19.fa\n",
      "Fri Feb  2 19:36:31 2024  -Checking records: 1  2  3  4  5  6  7  X  8  9  10  11  12  13  14  15  16  17  18  20  Y  19  22  21  M  \n",
      "Fri Feb  2 19:36:47 2024  -Variants allele on given reference sequence :  6\n",
      "Fri Feb  2 19:36:47 2024  -Variants flipped :  6\n",
      "Fri Feb  2 19:36:47 2024   -Raw Matching rate :  80.00%\n",
      "Fri Feb  2 19:36:47 2024  -Variants inferred reverse_complement :  0\n",
      "Fri Feb  2 19:36:47 2024  -Variants inferred reverse_complement_flipped :  0\n",
      "Fri Feb  2 19:36:47 2024  -Both allele on genome + unable to distinguish :  3\n",
      "Fri Feb  2 19:36:47 2024  -Variants not on given reference sequence :  0\n",
      "Fri Feb  2 19:36:47 2024 Finished checking if NEA is aligned with reference sequence.\n",
      "Fri Feb  2 19:36:47 2024 Start to adjust statistics based on STATUS code...v3.4.38\n",
      "Fri Feb  2 19:36:47 2024  -Current Dataframe shape : 15 x 13 ; Memory usage: 0.00 MB\n",
      "Fri Feb  2 19:36:47 2024 Start to flip allele-specific stats for SNPs with status xxxxx[35]x: ALT->EA , REF->NEA ...v3.4.38\n",
      "Fri Feb  2 19:36:47 2024  -Flipping 6 variants...\n",
      "Fri Feb  2 19:36:47 2024  -Swapping column: NEA <=> EA...\n",
      "Fri Feb  2 19:36:47 2024  -Flipping column: BETA = - BETA...\n",
      "Fri Feb  2 19:36:47 2024  -Flipping column: DIRECTION +-?0 <=> -+?0 ...\n",
      "Fri Feb  2 19:36:47 2024  -Flipping column: EAF = 1 - EAF...\n",
      "Fri Feb  2 19:36:47 2024  -Changed the status for flipped variants : xxxxx[35]x -> xxxxx[12]x\n",
      "Fri Feb  2 19:36:47 2024 Finished adjusting.\n",
      "Fri Feb  2 19:36:47 2024 Start to infer strand for palindromic SNPs/align indistinguishable indels...v3.4.38\n",
      "Fri Feb  2 19:36:47 2024  -Current Dataframe shape : 15 x 13 ; Memory usage: 0.00 MB\n",
      "Fri Feb  2 19:36:47 2024  -Number of threads/cores to use: 1\n",
      "Fri Feb  2 19:36:47 2024  -Reference VCF: /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz\n",
      "Fri Feb  2 19:36:47 2024  -Checking prefix for chromosomes in vcf files...\n",
      "Fri Feb  2 19:36:47 2024  -No prefix for chromosomes in the VCF files.\n",
      "Fri Feb  2 19:36:47 2024  -Field for alternative allele frequency in VCF INFO: AF\n",
      "Fri Feb  2 19:36:47 2024  -Identified  8  palindromic SNPs...\n",
      "Fri Feb  2 19:36:47 2024  -After filtering by MAF< 0.4 , 6 palindromic SNPs with unknown strand will be inferred...\n",
      "Fri Feb  2 19:36:47 2024   -Non-palindromic :  2\n",
      "Fri Feb  2 19:36:47 2024   -Palindromic SNPs on + strand:  2\n",
      "Fri Feb  2 19:36:47 2024   -Palindromic SNPs on - strand and needed to be flipped: 2\n",
      "Fri Feb  2 19:36:47 2024   -Palindromic SNPs with MAF not available to infer :  2\n",
      "Fri Feb  2 19:36:47 2024   -Palindromic SNPs with no macthes or no information :  1\n",
      "Fri Feb  2 19:36:47 2024  -Identified  3  indistinguishable Indels...\n",
      "Fri Feb  2 19:36:47 2024  -Indistinguishable indels will be inferred from reference vcf ref and alt...\n",
      "Fri Feb  2 19:36:47 2024  -DAF tolerance: 0.2\n",
      "Fri Feb  2 19:36:47 2024   -Indels ea/nea match reference :  1\n",
      "Fri Feb  2 19:36:47 2024   -Indels ea/nea need to be flipped :  1\n",
      "Fri Feb  2 19:36:47 2024   -Indels with no macthes or no information :  1\n",
      "Fri Feb  2 19:36:47 2024 Finished inferring strand for palindromic SNPs/align indistinguishable indels.\n",
      "Fri Feb  2 19:36:47 2024 Start to adjust statistics based on STATUS code...v3.4.38\n",
      "Fri Feb  2 19:36:47 2024  -Current Dataframe shape : 15 x 13 ; Memory usage: 0.00 MB\n",
      "Fri Feb  2 19:36:47 2024 Start to flip allele-specific stats for standardized indels with status xxxx[123][67][6]: ALT->EA , REF->NEA...v3.4.38\n",
      "Fri Feb  2 19:36:47 2024  -Flipping 1 variants...\n",
      "Fri Feb  2 19:36:47 2024  -Swapping column: NEA <=> EA...\n",
      "Fri Feb  2 19:36:47 2024  -Flipping column: BETA = - BETA...\n",
      "Fri Feb  2 19:36:47 2024  -Flipping column: DIRECTION +-?0 <=> -+?0 ...\n",
      "Fri Feb  2 19:36:47 2024  -Flipping column: EAF = 1 - EAF...\n",
      "Fri Feb  2 19:36:47 2024  -Changed the status for flipped variants xxxx[123][67]6 -> xxxx[123][67]4\n",
      "Fri Feb  2 19:36:47 2024 Start to flip allele-specific stats for palindromic SNPs with status xxxxx[12]5: (-)strand <=> (+)strand...v3.4.38\n",
      "Fri Feb  2 19:36:47 2024  -Flipping 2 variants...\n",
      "Fri Feb  2 19:36:47 2024  -Flipping column: BETA = - BETA...\n",
      "Fri Feb  2 19:36:47 2024  -Flipping column: DIRECTION +-?0 <=> -+?0 ...\n",
      "Fri Feb  2 19:36:47 2024  -Flipping column: EAF = 1 - EAF...\n",
      "Fri Feb  2 19:36:47 2024  -Changed the status for flipped variants:  xxxxx[012]5: ->  xxxxx[012]2\n",
      "Fri Feb  2 19:36:47 2024 Finished adjusting.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Feb  2 19:36:48 2024 Start to sort the genome coordinates...v3.4.38\n",
      "Fri Feb  2 19:36:48 2024  -Current Dataframe shape : 15 x 13 ; Memory usage: 0.00 MB\n",
      "Fri Feb  2 19:36:48 2024 Finished sorting coordinates.\n",
      "Fri Feb  2 19:36:48 2024 Start to reorder the columns...v3.4.38\n",
      "Fri Feb  2 19:36:48 2024  -Current Dataframe shape : 15 x 13 ; Memory usage: 0.00 MB\n",
      "Fri Feb  2 19:36:48 2024  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,EAF,BETA,SE,P,N,DIRECTION,STATUS,NOTE\n",
      "Fri Feb  2 19:36:48 2024 Finished reordering the columns.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x7f6c69f86550>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.harmonize(   basic_check=True,\n",
    "                        n_cores=1,\n",
    "                        ref_seq=gl.get_path(\"ucsc_genome_hg19\"),\n",
    "                        ref_infer=gl.get_path(\"1kg_eas_hg19\"),\n",
    "                        ref_alt_freq=\"AF\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6e64d22",
   "metadata": {},
   "source": [
    "All variants were checked and harmonized based on the reference datasets. The manipulations are reflected in STATUS column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "37d7cf7d-624a-4f4b-a25d-4f26ca0213ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>DIRECTION</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>NOTE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:1066403:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>1066403</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.5276</td>\n",
       "      <td>0.0043</td>\n",
       "      <td>0.0109</td>\n",
       "      <td>0.68910</td>\n",
       "      <td>191764</td>\n",
       "      <td>++-+</td>\n",
       "      <td>9960000</td>\n",
       "      <td>Clean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:1163266:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1163266</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.1645</td>\n",
       "      <td>-0.0122</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>0.30810</td>\n",
       "      <td>191764</td>\n",
       "      <td>+---</td>\n",
       "      <td>9960010</td>\n",
       "      <td>Flip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:1213224:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>1213224</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.6655</td>\n",
       "      <td>-0.0071</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>0.45280</td>\n",
       "      <td>191764</td>\n",
       "      <td>---+</td>\n",
       "      <td>9960011</td>\n",
       "      <td>Flip + Palindromic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:3066761:A:T</td>\n",
       "      <td>1</td>\n",
       "      <td>3066761</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1245</td>\n",
       "      <td>-0.0066</td>\n",
       "      <td>0.0157</td>\n",
       "      <td>0.67420</td>\n",
       "      <td>191764</td>\n",
       "      <td>-+++</td>\n",
       "      <td>9960001</td>\n",
       "      <td>Palindromic + No flip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:3997271:T:TTTTA</td>\n",
       "      <td>1</td>\n",
       "      <td>3997271</td>\n",
       "      <td>TTTTA</td>\n",
       "      <td>T</td>\n",
       "      <td>0.3840</td>\n",
       "      <td>0.0188</td>\n",
       "      <td>0.0135</td>\n",
       "      <td>0.16300</td>\n",
       "      <td>191764</td>\n",
       "      <td>++++</td>\n",
       "      <td>9960363</td>\n",
       "      <td>Indel + Both on genome + No flip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:4007705:TC:T</td>\n",
       "      <td>1</td>\n",
       "      <td>4007705</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>0.5632</td>\n",
       "      <td>-0.0306</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>0.01421</td>\n",
       "      <td>191764</td>\n",
       "      <td>----</td>\n",
       "      <td>9960364</td>\n",
       "      <td>Indel  + Both on genome + Flip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1:5164281:A:AAGAG</td>\n",
       "      <td>1</td>\n",
       "      <td>5164281</td>\n",
       "      <td>AAGAG</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9871</td>\n",
       "      <td>0.0167</td>\n",
       "      <td>0.0454</td>\n",
       "      <td>0.71200</td>\n",
       "      <td>191764</td>\n",
       "      <td>-+++</td>\n",
       "      <td>9960309</td>\n",
       "      <td>Indel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1:12799025:C:CCT</td>\n",
       "      <td>1</td>\n",
       "      <td>12799025</td>\n",
       "      <td>CCT</td>\n",
       "      <td>C</td>\n",
       "      <td>0.8270</td>\n",
       "      <td>0.0186</td>\n",
       "      <td>0.0118</td>\n",
       "      <td>0.11500</td>\n",
       "      <td>191764</td>\n",
       "      <td>+++-</td>\n",
       "      <td>9960319</td>\n",
       "      <td>Indel + Flip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1:19323015:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>19323015</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.5611</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.0088</td>\n",
       "      <td>0.05949</td>\n",
       "      <td>191764</td>\n",
       "      <td>++++</td>\n",
       "      <td>9960017</td>\n",
       "      <td>Palindromic + Flip + Indistinguishable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1:22841855:A:T</td>\n",
       "      <td>1</td>\n",
       "      <td>22841855</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4527</td>\n",
       "      <td>0.0080</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>0.39980</td>\n",
       "      <td>191764</td>\n",
       "      <td>-++-</td>\n",
       "      <td>9960007</td>\n",
       "      <td>Palindromic + Indistinguishable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3:183629306:T:TTCTC</td>\n",
       "      <td>3</td>\n",
       "      <td>183629306</td>\n",
       "      <td>TTCTC</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0082</td>\n",
       "      <td>-0.0083</td>\n",
       "      <td>0.0555</td>\n",
       "      <td>0.88060</td>\n",
       "      <td>191764</td>\n",
       "      <td>+--+</td>\n",
       "      <td>9960368</td>\n",
       "      <td>Indel  + Both on genome + No information</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>4:99731866:G:C</td>\n",
       "      <td>4</td>\n",
       "      <td>99731866</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.3883</td>\n",
       "      <td>0.0052</td>\n",
       "      <td>0.0093</td>\n",
       "      <td>0.57360</td>\n",
       "      <td>191764</td>\n",
       "      <td>+-++</td>\n",
       "      <td>9960012</td>\n",
       "      <td>Palindromic+ Flip + Different strand (REF G, T...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>8:89935201:C:G</td>\n",
       "      <td>8</td>\n",
       "      <td>89935201</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.6467</td>\n",
       "      <td>-0.0103</td>\n",
       "      <td>0.0094</td>\n",
       "      <td>0.27560</td>\n",
       "      <td>191764</td>\n",
       "      <td>----</td>\n",
       "      <td>9960002</td>\n",
       "      <td>Palindromic + Different strand (REF A; TOMMO G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>X:4206466:G:C</td>\n",
       "      <td>23</td>\n",
       "      <td>4206466</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.1273</td>\n",
       "      <td>0.0055</td>\n",
       "      <td>0.0104</td>\n",
       "      <td>0.59840</td>\n",
       "      <td>191764</td>\n",
       "      <td>+-++</td>\n",
       "      <td>9960018</td>\n",
       "      <td>Palindromic + Flip + No information</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>X:5053229:A:T</td>\n",
       "      <td>23</td>\n",
       "      <td>5053229</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9747</td>\n",
       "      <td>0.0218</td>\n",
       "      <td>0.0225</td>\n",
       "      <td>0.33180</td>\n",
       "      <td>191764</td>\n",
       "      <td>++++</td>\n",
       "      <td>9960008</td>\n",
       "      <td>Palindromic + No information</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  SNPID  CHR        POS     EA NEA     EAF    BETA      SE  \\\n",
       "0         1:1066403:T:C    1    1066403      C   T  0.5276  0.0043  0.0109   \n",
       "1         1:1163266:G:A    1    1163266      A   G  0.1645 -0.0122  0.0120   \n",
       "2         1:1213224:T:A    1    1213224      A   T  0.6655 -0.0071  0.0095   \n",
       "3         1:3066761:A:T    1    3066761      T   A  0.1245 -0.0066  0.0157   \n",
       "4     1:3997271:T:TTTTA    1    3997271  TTTTA   T  0.3840  0.0188  0.0135   \n",
       "5        1:4007705:TC:T    1    4007705      T  TC  0.5632 -0.0306  0.0125   \n",
       "6     1:5164281:A:AAGAG    1    5164281  AAGAG   A  0.9871  0.0167  0.0454   \n",
       "7      1:12799025:C:CCT    1   12799025    CCT   C  0.8270  0.0186  0.0118   \n",
       "8        1:19323015:T:A    1   19323015      A   T  0.5611  0.0165  0.0088   \n",
       "9        1:22841855:A:T    1   22841855      T   A  0.4527  0.0080  0.0095   \n",
       "10  3:183629306:T:TTCTC    3  183629306  TTCTC   T  0.0082 -0.0083  0.0555   \n",
       "11       4:99731866:G:C    4   99731866      C   G  0.3883  0.0052  0.0093   \n",
       "12       8:89935201:C:G    8   89935201      G   C  0.6467 -0.0103  0.0094   \n",
       "13        X:4206466:G:C   23    4206466      C   G  0.1273  0.0055  0.0104   \n",
       "14        X:5053229:A:T   23    5053229      T   A  0.9747  0.0218  0.0225   \n",
       "\n",
       "          P       N DIRECTION   STATUS  \\\n",
       "0   0.68910  191764      ++-+  9960000   \n",
       "1   0.30810  191764      +---  9960010   \n",
       "2   0.45280  191764      ---+  9960011   \n",
       "3   0.67420  191764      -+++  9960001   \n",
       "4   0.16300  191764      ++++  9960363   \n",
       "5   0.01421  191764      ----  9960364   \n",
       "6   0.71200  191764      -+++  9960309   \n",
       "7   0.11500  191764      +++-  9960319   \n",
       "8   0.05949  191764      ++++  9960017   \n",
       "9   0.39980  191764      -++-  9960007   \n",
       "10  0.88060  191764      +--+  9960368   \n",
       "11  0.57360  191764      +-++  9960012   \n",
       "12  0.27560  191764      ----  9960002   \n",
       "13  0.59840  191764      +-++  9960018   \n",
       "14  0.33180  191764      ++++  9960008   \n",
       "\n",
       "                                                 NOTE  \n",
       "0                                               Clean  \n",
       "1                                                Flip  \n",
       "2                                  Flip + Palindromic  \n",
       "3                               Palindromic + No flip  \n",
       "4                    Indel + Both on genome + No flip  \n",
       "5                      Indel  + Both on genome + Flip  \n",
       "6                                              Indel   \n",
       "7                                        Indel + Flip  \n",
       "8              Palindromic + Flip + Indistinguishable  \n",
       "9                     Palindromic + Indistinguishable  \n",
       "10           Indel  + Both on genome + No information  \n",
       "11  Palindromic+ Flip + Different strand (REF G, T...  \n",
       "12  Palindromic + Different strand (REF A; TOMMO G...  \n",
       "13                Palindromic + Flip + No information  \n",
       "14                       Palindromic + No information  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd817741-2f1e-4758-bc90-bf3b22bbe106",
   "metadata": {},
   "source": [
    "## Check summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d4d0708c-5031-4184-850a-4803168051cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Values</th>\n",
       "      <th>Percentage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Category</th>\n",
       "      <th>Items</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">META</th>\n",
       "      <th>Row_num</th>\n",
       "      <td>15</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Column_num</th>\n",
       "      <td>13</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Column_names</th>\n",
       "      <td>SNPID,CHR,POS,EA,NEA,EAF,BETA,SE,P,N,DIRECTION...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Last_checked_time</th>\n",
       "      <td>Fri Feb  2 19:36:48 2024</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MISSING</th>\n",
       "      <th>Missing_total</th>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">MAF</th>\n",
       "      <th>Common</th>\n",
       "      <td>12</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Low_frequency</th>\n",
       "      <td>2</td>\n",
       "      <td>13.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rare</th>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">P</th>\n",
       "      <th>Minimum</th>\n",
       "      <td>0.01421</td>\n",
       "      <td>0.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Significant</th>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Suggestive</th>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"15\" valign=\"top\">STATUS</th>\n",
       "      <th>9960000</th>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960001</th>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960002</th>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960007</th>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960008</th>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960010</th>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960011</th>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960012</th>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960017</th>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960018</th>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960309</th>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960319</th>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960363</th>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960364</th>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960368</th>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                       Values  \\\n",
       "Category Items                                                                  \n",
       "META     Row_num                                                           15   \n",
       "         Column_num                                                        13   \n",
       "         Column_names       SNPID,CHR,POS,EA,NEA,EAF,BETA,SE,P,N,DIRECTION...   \n",
       "         Last_checked_time                           Fri Feb  2 19:36:48 2024   \n",
       "MISSING  Missing_total                                                      0   \n",
       "MAF      Common                                                            12   \n",
       "         Low_frequency                                                      2   \n",
       "         Rare                                                               1   \n",
       "P        Minimum                                                      0.01421   \n",
       "         Significant                                                        0   \n",
       "         Suggestive                                                         0   \n",
       "STATUS   9960000                                                            1   \n",
       "         9960001                                                            1   \n",
       "         9960002                                                            1   \n",
       "         9960007                                                            1   \n",
       "         9960008                                                            1   \n",
       "         9960010                                                            1   \n",
       "         9960011                                                            1   \n",
       "         9960012                                                            1   \n",
       "         9960017                                                            1   \n",
       "         9960018                                                            1   \n",
       "         9960309                                                            1   \n",
       "         9960319                                                            1   \n",
       "         9960363                                                            1   \n",
       "         9960364                                                            1   \n",
       "         9960368                                                            1   \n",
       "\n",
       "                            Percentage  \n",
       "Category Items                          \n",
       "META     Row_num                  <NA>  \n",
       "         Column_num               <NA>  \n",
       "         Column_names             <NA>  \n",
       "         Last_checked_time        <NA>  \n",
       "MISSING  Missing_total             0.0  \n",
       "MAF      Common                   80.0  \n",
       "         Low_frequency           13.33  \n",
       "         Rare                     6.67  \n",
       "P        Minimum                  0.09  \n",
       "         Significant               0.0  \n",
       "         Suggestive                0.0  \n",
       "STATUS   9960000                  6.67  \n",
       "         9960001                  6.67  \n",
       "         9960002                  6.67  \n",
       "         9960007                  6.67  \n",
       "         9960008                  6.67  \n",
       "         9960010                  6.67  \n",
       "         9960011                  6.67  \n",
       "         9960012                  6.67  \n",
       "         9960017                  6.67  \n",
       "         9960018                  6.67  \n",
       "         9960309                  6.67  \n",
       "         9960319                  6.67  \n",
       "         9960363                  6.67  \n",
       "         9960364                  6.67  \n",
       "         9960368                  6.67  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bd41743e-ff13-4f52-b46b-5946d4da1087",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome_Build</th>\n",
       "      <th>rsID&amp;SNPID</th>\n",
       "      <th>CHR&amp;POS</th>\n",
       "      <th>Stadardize&amp;Normalize</th>\n",
       "      <th>Align</th>\n",
       "      <th>Panlidromic_SNP&amp;Indel</th>\n",
       "      <th>Count</th>\n",
       "      <th>Percentage(%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9960000</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Match: NEA=REF</td>\n",
       "      <td>Not_palindromic_SNPs</td>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960001</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Match: NEA=REF</td>\n",
       "      <td>Palindromic+strand</td>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960002</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Match: NEA=REF</td>\n",
       "      <td>Palindromic-strand_fixed</td>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960007</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Match: NEA=REF</td>\n",
       "      <td>Indistinguishable</td>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960008</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Match: NEA=REF</td>\n",
       "      <td>No_matching_or_no_info</td>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960010</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Flipped_fixed</td>\n",
       "      <td>Not_palindromic_SNPs</td>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960011</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Flipped_fixed</td>\n",
       "      <td>Palindromic+strand</td>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960012</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Flipped_fixed</td>\n",
       "      <td>Palindromic-strand_fixed</td>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960017</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Flipped_fixed</td>\n",
       "      <td>Indistinguishable</td>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960018</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Flipped_fixed</td>\n",
       "      <td>No_matching_or_no_info</td>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960309</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized &amp; normalized indel</td>\n",
       "      <td>Match: NEA=REF</td>\n",
       "      <td>Unchecked</td>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960319</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized &amp; normalized indel</td>\n",
       "      <td>Flipped_fixed</td>\n",
       "      <td>Unchecked</td>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960363</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized &amp; normalized indel</td>\n",
       "      <td>Both_alleles_on_ref+indistinguishable</td>\n",
       "      <td>Indel_match</td>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960364</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized &amp; normalized indel</td>\n",
       "      <td>Both_alleles_on_ref+indistinguishable</td>\n",
       "      <td>Indel_flipped_fixed</td>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960368</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized &amp; normalized indel</td>\n",
       "      <td>Both_alleles_on_ref+indistinguishable</td>\n",
       "      <td>No_matching_or_no_info</td>\n",
       "      <td>1</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Genome_Build                  rsID&SNPID                CHR&POS  \\\n",
       "9960000    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960001    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960002    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960007    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960008    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960010    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960011    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960012    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960017    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960018    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960309    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960319    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960363    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960364    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960368    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "\n",
       "                    Stadardize&Normalize  \\\n",
       "9960000                 standardized SNP   \n",
       "9960001                 standardized SNP   \n",
       "9960002                 standardized SNP   \n",
       "9960007                 standardized SNP   \n",
       "9960008                 standardized SNP   \n",
       "9960010                 standardized SNP   \n",
       "9960011                 standardized SNP   \n",
       "9960012                 standardized SNP   \n",
       "9960017                 standardized SNP   \n",
       "9960018                 standardized SNP   \n",
       "9960309  standardized & normalized indel   \n",
       "9960319  standardized & normalized indel   \n",
       "9960363  standardized & normalized indel   \n",
       "9960364  standardized & normalized indel   \n",
       "9960368  standardized & normalized indel   \n",
       "\n",
       "                                         Align     Panlidromic_SNP&Indel  \\\n",
       "9960000                         Match: NEA=REF      Not_palindromic_SNPs   \n",
       "9960001                         Match: NEA=REF        Palindromic+strand   \n",
       "9960002                         Match: NEA=REF  Palindromic-strand_fixed   \n",
       "9960007                         Match: NEA=REF         Indistinguishable   \n",
       "9960008                         Match: NEA=REF    No_matching_or_no_info   \n",
       "9960010                          Flipped_fixed      Not_palindromic_SNPs   \n",
       "9960011                          Flipped_fixed        Palindromic+strand   \n",
       "9960012                          Flipped_fixed  Palindromic-strand_fixed   \n",
       "9960017                          Flipped_fixed         Indistinguishable   \n",
       "9960018                          Flipped_fixed    No_matching_or_no_info   \n",
       "9960309                         Match: NEA=REF                 Unchecked   \n",
       "9960319                          Flipped_fixed                 Unchecked   \n",
       "9960363  Both_alleles_on_ref+indistinguishable               Indel_match   \n",
       "9960364  Both_alleles_on_ref+indistinguishable       Indel_flipped_fixed   \n",
       "9960368  Both_alleles_on_ref+indistinguishable    No_matching_or_no_info   \n",
       "\n",
       "        Count Percentage(%)  \n",
       "9960000     1          6.67  \n",
       "9960001     1          6.67  \n",
       "9960002     1          6.67  \n",
       "9960007     1          6.67  \n",
       "9960008     1          6.67  \n",
       "9960010     1          6.67  \n",
       "9960011     1          6.67  \n",
       "9960012     1          6.67  \n",
       "9960017     1          6.67  \n",
       "9960018     1          6.67  \n",
       "9960309     1          6.67  \n",
       "9960319     1          6.67  \n",
       "9960363     1          6.67  \n",
       "9960364     1          6.67  \n",
       "9960368     1          6.67  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.lookup_status()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b93e292",
   "metadata": {},
   "source": [
    "## VCF : NCBI Sequence Identifiers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d57803bc",
   "metadata": {},
   "source": [
    "For some reference file the chromosome notation is not in the form of 1, or chr1, instead, they are using NCBI Sequence Identifiers (for example: NC_000001.11)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c57e719",
   "metadata": {},
   "source": [
    "See https://www.ncbi.nlm.nih.gov/genbank/sequenceids/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b316a1f2",
   "metadata": {},
   "source": [
    "For example: GCF_000001405.25.vcf.gz\n",
    "\n",
    "NC_000001.10    10059   rs1570391745    C       G       .       .       RS=1570391745;dbSNPBuildID=154;SSR=0;PSEUDOGENEINFO=DDX11L1:100287102;VC=SNV;R5;GNO;FREQ=KOREAN:0.9997,0.0003425|dbGaP_PopFreq:1,0\n",
    "\n",
    "NC_000001.10    10060   rs1639544146    C       CT      .       .       RS=1639544146;SSR=0;PSEUDOGENEINFO=DDX11L1:100287102;VC=INDEL;R5;GNO;FREQ=dbGaP_PopFreq:1,0\n",
    "\n",
    "NC_000001.10    10060   rs1639544159    CT      C       .       .       RS=1639544159;SSR=0;PSEUDOGENEINFO=DDX11L1:100287102;VC=DEL;R5;GNO;FREQ=dbGaP_PopFreq:1,0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e6f4d30",
   "metadata": {},
   "source": [
    " In such case, gwaslab provide built-in conversion table. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a1098cc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: 'NC_000001.10',\n",
       " 2: 'NC_000002.11',\n",
       " 3: 'NC_000003.11',\n",
       " 4: 'NC_000004.11',\n",
       " 5: 'NC_000005.9',\n",
       " 6: 'NC_000006.11',\n",
       " 7: 'NC_000007.13',\n",
       " 8: 'NC_000008.10',\n",
       " 9: 'NC_000009.11',\n",
       " 10: 'NC_000010.10',\n",
       " 11: 'NC_000011.9',\n",
       " 12: 'NC_000012.11',\n",
       " 13: 'NC_000013.10',\n",
       " 14: 'NC_000014.8',\n",
       " 15: 'NC_000015.9',\n",
       " 16: 'NC_000016.9',\n",
       " 17: 'NC_000017.10',\n",
       " 18: 'NC_000018.9',\n",
       " 19: 'NC_000019.9',\n",
       " 20: 'NC_000020.10',\n",
       " 21: 'NC_000021.8',\n",
       " 22: 'NC_000022.10',\n",
       " 23: 'NC_000023.10',\n",
       " 24: 'NC_000024.9',\n",
       " 25: 'NC_012920.1'}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gl.get_number_to_NC(build=\"19\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7757b8bd",
   "metadata": {},
   "source": [
    "Specify it in assignrsid_args and inferstrand_args for the all-in-one function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2534554d-e504-4899-a57e-53da71a17d77",
   "metadata": {},
   "outputs": [],
   "source": [
    "mysumstats.harmonize(   basic_check=False,\n",
    "                        n_cores=3,\n",
    "                        ref_infer=\"/home/yunye/mydata/d_disk/dbsnp/GCF_000001405.25.vcf.gz\",\n",
    "                        inferstrand_args= {\"chr_dict\" : gl.get_number_to_NC(build=\"19\")})"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
