{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lead/ novel variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load sample data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use only first 1000000 variants as example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "mysumstats = gl.Sumstats(\"t2d_bbj.txt.gz\",\n",
    "             snpid=\"SNP\",\n",
    "             chrom=\"CHR\",\n",
    "             pos=\"POS\",\n",
    "             ea=\"ALT\",\n",
    "             nea=\"REF\",            \n",
    "             beta=\"BETA\",\n",
    "             se=\"SE\",\n",
    "             p=\"P\", \n",
    "             build=\"19\",\n",
    "             verbose=False, \n",
    "             nrows=1000000)\n",
    "mysumstats.basic_check(verbose=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get lead variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GWASLab will use MLOG10P first by default. If MLOG10P is not avaiable, it will look for P column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sat Feb  3 01:27:18 2024 Start to extract lead variants...v3.4.38\n",
      "Sat Feb  3 01:27:18 2024  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 75.89 MB\n",
      "Sat Feb  3 01:27:18 2024  -Processing 1000000 variants...\n",
      "Sat Feb  3 01:27:18 2024  -Significance threshold : 5e-08\n",
      "Sat Feb  3 01:27:18 2024  -Sliding window size: 500  kb\n",
      "Sat Feb  3 01:27:19 2024  -Using P for extracting lead variants...\n",
      "Sat Feb  3 01:27:19 2024  -Found 543 significant variants in total...\n",
      "Sat Feb  3 01:27:19 2024  -Identified 4 lead variants!\n",
      "Sat Feb  3 01:27:19 2024 Finished extracting lead variants.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>STATUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96739</th>\n",
       "      <td>1:22068326_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>22068326</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0621</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>1.629000e-09</td>\n",
       "      <td>1960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213860</th>\n",
       "      <td>1:51103268_T_C</td>\n",
       "      <td>1</td>\n",
       "      <td>51103268</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0802</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>2.519000e-11</td>\n",
       "      <td>1960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534095</th>\n",
       "      <td>1:154309595_TA_T</td>\n",
       "      <td>1</td>\n",
       "      <td>154309595</td>\n",
       "      <td>TA</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0915</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>3.289000e-08</td>\n",
       "      <td>1960399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>969974</th>\n",
       "      <td>2:640986_CACAT_C</td>\n",
       "      <td>2</td>\n",
       "      <td>640986</td>\n",
       "      <td>C</td>\n",
       "      <td>CACAT</td>\n",
       "      <td>-0.0946</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>2.665000e-10</td>\n",
       "      <td>1960399</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   SNPID  CHR        POS  EA    NEA    BETA      SE  \\\n",
       "96739     1:22068326_A_G    1   22068326   G      A  0.0621  0.0103   \n",
       "213860    1:51103268_T_C    1   51103268   C      T -0.0802  0.0120   \n",
       "534095  1:154309595_TA_T    1  154309595  TA      T -0.0915  0.0166   \n",
       "969974  2:640986_CACAT_C    2     640986   C  CACAT -0.0946  0.0150   \n",
       "\n",
       "                   P   STATUS  \n",
       "96739   1.629000e-09  1960099  \n",
       "213860  2.519000e-11  1960099  \n",
       "534095  3.289000e-08  1960399  \n",
       "969974  2.665000e-10  1960399  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.get_lead()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get lead variants with gene name annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sat Feb  3 01:27:19 2024 Start to extract lead variants...v3.4.38\n",
      "Sat Feb  3 01:27:19 2024  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 75.89 MB\n",
      "Sat Feb  3 01:27:19 2024  -Processing 1000000 variants...\n",
      "Sat Feb  3 01:27:19 2024  -Significance threshold : 5e-08\n",
      "Sat Feb  3 01:27:19 2024  -Sliding window size: 500  kb\n",
      "Sat Feb  3 01:27:19 2024  -Using P for extracting lead variants...\n",
      "Sat Feb  3 01:27:19 2024  -Found 543 significant variants in total...\n",
      "Sat Feb  3 01:27:19 2024  -Identified 4 lead variants!\n",
      "Sat Feb  3 01:27:19 2024  -Annotating variants using references:ensembl\n",
      "Sat Feb  3 01:27:19 2024  -Annotating variants using references based on genome build:19\n",
      "Sat Feb  3 01:27:19 2024 Start to annotate variants with nearest gene name(s)...\n",
      "Sat Feb  3 01:27:19 2024  -Assigning Gene name using ensembl_hg19_gtf for protein coding genes\n",
      "Sat Feb  3 01:27:19 2024 Finished annotating variants with nearest gene name(s) successfully!\n",
      "Sat Feb  3 01:27:19 2024 Finished extracting lead variants.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>LOCATION</th>\n",
       "      <th>GENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96739</th>\n",
       "      <td>1:22068326_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>22068326</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0621</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>1.629000e-09</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0</td>\n",
       "      <td>USP48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213860</th>\n",
       "      <td>1:51103268_T_C</td>\n",
       "      <td>1</td>\n",
       "      <td>51103268</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0802</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>2.519000e-11</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0</td>\n",
       "      <td>FAF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534095</th>\n",
       "      <td>1:154309595_TA_T</td>\n",
       "      <td>1</td>\n",
       "      <td>154309595</td>\n",
       "      <td>TA</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0915</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>3.289000e-08</td>\n",
       "      <td>1960399</td>\n",
       "      <td>0</td>\n",
       "      <td>ATP8B2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>969974</th>\n",
       "      <td>2:640986_CACAT_C</td>\n",
       "      <td>2</td>\n",
       "      <td>640986</td>\n",
       "      <td>C</td>\n",
       "      <td>CACAT</td>\n",
       "      <td>-0.0946</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>2.665000e-10</td>\n",
       "      <td>1960399</td>\n",
       "      <td>26349</td>\n",
       "      <td>TMEM18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   SNPID  CHR        POS  EA    NEA    BETA      SE  \\\n",
       "96739     1:22068326_A_G    1   22068326   G      A  0.0621  0.0103   \n",
       "213860    1:51103268_T_C    1   51103268   C      T -0.0802  0.0120   \n",
       "534095  1:154309595_TA_T    1  154309595  TA      T -0.0915  0.0166   \n",
       "969974  2:640986_CACAT_C    2     640986   C  CACAT -0.0946  0.0150   \n",
       "\n",
       "                   P   STATUS  LOCATION    GENE  \n",
       "96739   1.629000e-09  1960099         0   USP48  \n",
       "213860  2.519000e-11  1960099         0    FAF1  \n",
       "534095  3.289000e-08  1960399         0  ATP8B2  \n",
       "969974  2.665000e-10  1960399     26349  TMEM18  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.get_lead(anno=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Different window sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sat Feb  3 01:27:19 2024 Start to extract lead variants...v3.4.38\n",
      "Sat Feb  3 01:27:19 2024  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 75.89 MB\n",
      "Sat Feb  3 01:27:19 2024  -Processing 1000000 variants...\n",
      "Sat Feb  3 01:27:19 2024  -Significance threshold : 5e-08\n",
      "Sat Feb  3 01:27:19 2024  -Sliding window size: 1000  kb\n",
      "Sat Feb  3 01:27:19 2024  -Using P for extracting lead variants...\n",
      "Sat Feb  3 01:27:19 2024  -Found 543 significant variants in total...\n",
      "Sat Feb  3 01:27:19 2024  -Identified 4 lead variants!\n",
      "Sat Feb  3 01:27:19 2024 Finished extracting lead variants.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>STATUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96739</th>\n",
       "      <td>1:22068326_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>22068326</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0621</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>1.629000e-09</td>\n",
       "      <td>1960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213860</th>\n",
       "      <td>1:51103268_T_C</td>\n",
       "      <td>1</td>\n",
       "      <td>51103268</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0802</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>2.519000e-11</td>\n",
       "      <td>1960099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534095</th>\n",
       "      <td>1:154309595_TA_T</td>\n",
       "      <td>1</td>\n",
       "      <td>154309595</td>\n",
       "      <td>TA</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0915</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>3.289000e-08</td>\n",
       "      <td>1960399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>969974</th>\n",
       "      <td>2:640986_CACAT_C</td>\n",
       "      <td>2</td>\n",
       "      <td>640986</td>\n",
       "      <td>C</td>\n",
       "      <td>CACAT</td>\n",
       "      <td>-0.0946</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>2.665000e-10</td>\n",
       "      <td>1960399</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   SNPID  CHR        POS  EA    NEA    BETA      SE  \\\n",
       "96739     1:22068326_A_G    1   22068326   G      A  0.0621  0.0103   \n",
       "213860    1:51103268_T_C    1   51103268   C      T -0.0802  0.0120   \n",
       "534095  1:154309595_TA_T    1  154309595  TA      T -0.0915  0.0166   \n",
       "969974  2:640986_CACAT_C    2     640986   C  CACAT -0.0946  0.0150   \n",
       "\n",
       "                   P   STATUS  \n",
       "96739   1.629000e-09  1960099  \n",
       "213860  2.519000e-11  1960099  \n",
       "534095  3.289000e-08  1960399  \n",
       "969974  2.665000e-10  1960399  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.get_lead(windowsizekb=1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Different thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sat Feb  3 01:27:19 2024 Start to extract lead variants...v3.4.38\n",
      "Sat Feb  3 01:27:19 2024  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 75.89 MB\n",
      "Sat Feb  3 01:27:19 2024  -Processing 1000000 variants...\n",
      "Sat Feb  3 01:27:19 2024  -Significance threshold : 1e-10\n",
      "Sat Feb  3 01:27:19 2024  -Sliding window size: 500  kb\n",
      "Sat Feb  3 01:27:20 2024  -Using P for extracting lead variants...\n",
      "Sat Feb  3 01:27:20 2024  -Found 1 significant variants in total...\n",
      "Sat Feb  3 01:27:20 2024  -Identified 1 lead variants!\n",
      "Sat Feb  3 01:27:20 2024 Finished extracting lead variants.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>STATUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>213860</th>\n",
       "      <td>1:51103268_T_C</td>\n",
       "      <td>1</td>\n",
       "      <td>51103268</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0802</td>\n",
       "      <td>0.012</td>\n",
       "      <td>2.519000e-11</td>\n",
       "      <td>1960099</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 SNPID  CHR       POS EA NEA    BETA     SE             P  \\\n",
       "213860  1:51103268_T_C    1  51103268  C   T -0.0802  0.012  2.519000e-11   \n",
       "\n",
       "         STATUS  \n",
       "213860  1960099  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.get_lead(sig_level=1e-10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check if novel against a tabular file"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "known_loci.txt file looks like:\n",
    "\n",
    "CHR POS\n",
    "1 154309595\n",
    "1 51103268"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sat Feb  3 01:22:31 2024 Start to check if lead variants are known...v3.4.38\n",
      "Sat Feb  3 01:22:31 2024  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 75.89 MB\n",
      "Sat Feb  3 01:22:31 2024 Start to extract lead variants...v3.4.38\n",
      "Sat Feb  3 01:22:31 2024  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 75.89 MB\n",
      "Sat Feb  3 01:22:31 2024  -Processing 1000000 variants...\n",
      "Sat Feb  3 01:22:31 2024  -Significance threshold : 5e-08\n",
      "Sat Feb  3 01:22:31 2024  -Sliding window size: 500  kb\n",
      "Sat Feb  3 01:22:31 2024  -Using P for extracting lead variants...\n",
      "Sat Feb  3 01:22:31 2024  -Found 543 significant variants in total...\n",
      "Sat Feb  3 01:22:31 2024  -Identified 4 lead variants!\n",
      "Sat Feb  3 01:22:31 2024 Finished extracting lead variants.\n",
      "Sat Feb  3 01:22:31 2024  -Lead variants in known loci: 2\n",
      "Sat Feb  3 01:22:31 2024  -Checking the minimum distance between identified lead variants and provided known variants...\n",
      "Sat Feb  3 01:22:31 2024  -Identified  2  known vairants in current sumstats...\n",
      "Sat Feb  3 01:22:31 2024  -Identified  2  novel vairants in current sumstats...\n",
      "Sat Feb  3 01:22:31 2024 Finished checking if lead variants are known.\n"
     ]
    }
   ],
   "source": [
    "novel = mysumstats.get_novel(known=\"./toy_data/known_loci.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>DISTANCE_TO_KNOWN</th>\n",
       "      <th>KNOWN_ID</th>\n",
       "      <th>NOVEL</th>\n",
       "      <th>LOCATION_OF_KNOWN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:22068326_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>22068326</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0621</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>1.629000e-09</td>\n",
       "      <td>1960099</td>\n",
       "      <td>29034942.0</td>\n",
       "      <td>1:51103268</td>\n",
       "      <td>True</td>\n",
       "      <td>Upstream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:51103268_T_C</td>\n",
       "      <td>1</td>\n",
       "      <td>51103268</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0802</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>2.519000e-11</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1:51103268</td>\n",
       "      <td>False</td>\n",
       "      <td>Same</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:154309595_TA_T</td>\n",
       "      <td>1</td>\n",
       "      <td>154309595</td>\n",
       "      <td>TA</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0915</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>3.289000e-08</td>\n",
       "      <td>1960399</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1:154309595</td>\n",
       "      <td>False</td>\n",
       "      <td>Same</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2:640986_CACAT_C</td>\n",
       "      <td>2</td>\n",
       "      <td>640986</td>\n",
       "      <td>C</td>\n",
       "      <td>CACAT</td>\n",
       "      <td>-0.0946</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>2.665000e-10</td>\n",
       "      <td>1960399</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>True</td>\n",
       "      <td>NoneOnThisChr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              SNPID  CHR        POS  EA    NEA    BETA      SE             P  \\\n",
       "0    1:22068326_A_G    1   22068326   G      A  0.0621  0.0103  1.629000e-09   \n",
       "1    1:51103268_T_C    1   51103268   C      T -0.0802  0.0120  2.519000e-11   \n",
       "2  1:154309595_TA_T    1  154309595  TA      T -0.0915  0.0166  3.289000e-08   \n",
       "3  2:640986_CACAT_C    2     640986   C  CACAT -0.0946  0.0150  2.665000e-10   \n",
       "\n",
       "    STATUS  DISTANCE_TO_KNOWN     KNOWN_ID  NOVEL LOCATION_OF_KNOWN  \n",
       "0  1960099         29034942.0   1:51103268   True          Upstream  \n",
       "1  1960099                0.0   1:51103268  False              Same  \n",
       "2  1960399                0.0  1:154309595  False              Same  \n",
       "3  1960399                NaN         <NA>   True     NoneOnThisChr  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ccheck against GWAS Catalog using EFO ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sat Feb  3 01:22:31 2024 Start to check if lead variants are known...v3.4.38\n",
      "Sat Feb  3 01:22:31 2024  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 75.89 MB\n",
      "Sat Feb  3 01:22:31 2024 Start to extract lead variants...v3.4.38\n",
      "Sat Feb  3 01:22:31 2024  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 75.89 MB\n",
      "Sat Feb  3 01:22:31 2024  -Processing 1000000 variants...\n",
      "Sat Feb  3 01:22:31 2024  -Significance threshold : 5e-08\n",
      "Sat Feb  3 01:22:31 2024  -Sliding window size: 500  kb\n",
      "Sat Feb  3 01:22:31 2024  -Using P for extracting lead variants...\n",
      "Sat Feb  3 01:22:31 2024  -Found 543 significant variants in total...\n",
      "Sat Feb  3 01:22:31 2024  -Identified 4 lead variants!\n",
      "Sat Feb  3 01:22:31 2024 Finished extracting lead variants.\n",
      "Sat Feb  3 01:22:32 2024 Start to retrieve data using EFO: MONDO_0005148...\n",
      "Sat Feb  3 01:22:32 2024 Start to retrieve data from GWASCatalog...\n",
      "Sat Feb  3 01:22:32 2024  -Please make sure your sumstats is based on GRCh38...\n",
      "Sat Feb  3 01:22:32 2024  -Requesting (GET) trait information through the GWASCatalog API...\n",
      "Sat Feb  3 01:22:32 2024  -EFO trait api: https://www.ebi.ac.uk/gwas/rest/api/efoTraits/MONDO_0005148\n",
      "Sat Feb  3 01:22:33 2024  -Status code: 200\n",
      "Sat Feb  3 01:22:33 2024  -Trait Name: type 2 diabetes mellitus\n",
      "Sat Feb  3 01:22:33 2024  -Trait URL: http://purl.obolibrary.org/obo/MONDO_0005148\n",
      "Sat Feb  3 01:22:33 2024  -Requesting (GET) GWAS associations through the GWASCatalog API...\n",
      "Sat Feb  3 01:22:33 2024  -associationsByTraitSummary API: https://www.ebi.ac.uk/gwas/rest/api/efoTraits/MONDO_0005148/associations?projection=associationByEfoTrait\n",
      "Sat Feb  3 01:22:33 2024  -Note: this step might take a while...\n",
      "Sat Feb  3 01:25:44 2024  -Status code 200 OK: Retrieved data from GWASCatalog successffully ...\n",
      "Sat Feb  3 01:25:44 2024  -Loading json ...\n",
      "Sat Feb  3 01:25:45 2024  -Parsing json ...\n",
      "Sat Feb  3 01:25:45 2024  -Number of reported associations for MONDO_0005148 in GWASCatalog: 5450\n",
      "Sat Feb  3 01:25:45 2024  -Loading retrieved data into gwaslab Sumstats object ...\n",
      "Sat Feb  3 01:25:50 2024 Finished retrieving data from GWASCatalog...\n",
      "Sat Feb  3 01:25:50 2024  -Retrieved 4176 associations from GWAS catalog.\n",
      "Sat Feb  3 01:25:50 2024  -Lead variants in known loci: 4176\n",
      "Sat Feb  3 01:25:50 2024  -Checking the minimum distance between identified lead variants and provided known variants...\n",
      "Sat Feb  3 01:25:50 2024  -Identified  4  known vairants in current sumstats...\n",
      "Sat Feb  3 01:25:50 2024  -Identified  0  novel vairants in current sumstats...\n",
      "Sat Feb  3 01:25:50 2024 Finished checking if lead variants are known.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>DISTANCE_TO_KNOWN</th>\n",
       "      <th>KNOWN_ID</th>\n",
       "      <th>KNOWN_PUBMED_ID</th>\n",
       "      <th>KNOWN_AUTHOR</th>\n",
       "      <th>NOVEL</th>\n",
       "      <th>LOCATION_OF_KNOWN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:22068326_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>22068326</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0621</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>1.629000e-09</td>\n",
       "      <td>1960099</td>\n",
       "      <td>-279528</td>\n",
       "      <td>rs34759301</td>\n",
       "      <td>34594039</td>\n",
       "      <td>Sakaue S</td>\n",
       "      <td>False</td>\n",
       "      <td>Downstream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:51103268_T_C</td>\n",
       "      <td>1</td>\n",
       "      <td>51103268</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0802</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>2.519000e-11</td>\n",
       "      <td>1960099</td>\n",
       "      <td>-62054</td>\n",
       "      <td>rs12088739</td>\n",
       "      <td>30054458</td>\n",
       "      <td>Xue A</td>\n",
       "      <td>False</td>\n",
       "      <td>Downstream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:154309595_TA_T</td>\n",
       "      <td>1</td>\n",
       "      <td>154309595</td>\n",
       "      <td>TA</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0915</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>3.289000e-08</td>\n",
       "      <td>1960399</td>\n",
       "      <td>12189</td>\n",
       "      <td>rs1194606</td>\n",
       "      <td>32541925</td>\n",
       "      <td>Vujkovic M</td>\n",
       "      <td>False</td>\n",
       "      <td>Upstream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2:640986_CACAT_C</td>\n",
       "      <td>2</td>\n",
       "      <td>640986</td>\n",
       "      <td>C</td>\n",
       "      <td>CACAT</td>\n",
       "      <td>-0.0946</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>2.665000e-10</td>\n",
       "      <td>1960399</td>\n",
       "      <td>1</td>\n",
       "      <td>rs72156956</td>\n",
       "      <td>30718926</td>\n",
       "      <td>Suzuki K</td>\n",
       "      <td>False</td>\n",
       "      <td>Upstream</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              SNPID  CHR        POS  EA    NEA    BETA      SE             P  \\\n",
       "0    1:22068326_A_G    1   22068326   G      A  0.0621  0.0103  1.629000e-09   \n",
       "1    1:51103268_T_C    1   51103268   C      T -0.0802  0.0120  2.519000e-11   \n",
       "2  1:154309595_TA_T    1  154309595  TA      T -0.0915  0.0166  3.289000e-08   \n",
       "3  2:640986_CACAT_C    2     640986   C  CACAT -0.0946  0.0150  2.665000e-10   \n",
       "\n",
       "    STATUS  DISTANCE_TO_KNOWN    KNOWN_ID KNOWN_PUBMED_ID KNOWN_AUTHOR  NOVEL  \\\n",
       "0  1960099            -279528  rs34759301        34594039     Sakaue S  False   \n",
       "1  1960099             -62054  rs12088739        30054458        Xue A  False   \n",
       "2  1960399              12189   rs1194606        32541925   Vujkovic M  False   \n",
       "3  1960399                  1  rs72156956        30718926     Suzuki K  False   \n",
       "\n",
       "  LOCATION_OF_KNOWN  \n",
       "0        Downstream  \n",
       "1        Downstream  \n",
       "2          Upstream  \n",
       "3          Upstream  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# EFO ID can be found on gwas catalog\n",
    "mysumstats.get_novel(efo=\"MONDO_0005148\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
