
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="HE Yunye">
      
      
      
        <link rel="prev" href="../StatusCode/">
      
      
        <link rel="next" href="../QC%26Filtering/">
      
      
      <link rel="icon" href="../assets/gwaslab.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.6">
    
    
      
        <title>Standardization and Normalization - GWASLab</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.50c56a3b.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue" data-md-color-accent="blue">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#standardization-and-normalization" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="GWASLab" class="md-header__button md-logo" aria-label="GWASLab" data-md-component="logo">
      
  <img src="../assets/gwaslab.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            GWASLab
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Standardization and Normalization
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/Cloufield/gwaslab/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href=".." class="md-tabs__link">
          
  
  Home

        </a>
      </li>
    
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../tutorial_3.4/" class="md-tabs__link">
        
  
    
  
  Tutorial for gwaslab

      </a>
    </li>
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../SumstatsObject/" class="md-tabs__link">
          
  
  Sumstats manipulation

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Visualization/" class="md-tabs__link">
          
  
  Visualization

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../ExtractLead/" class="md-tabs__link">
          
  
  Utilities

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Download/" class="md-tabs__link">
          
  
  References

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../visualization_regional/" class="md-tabs__link">
          
  
  Examples

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="GWASLab" class="md-nav__button md-logo" aria-label="GWASLab" data-md-component="logo">
      
  <img src="../assets/gwaslab.png" alt="logo">

    </a>
    GWASLab
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Cloufield/gwaslab/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
    
    
    
      
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" >
        
          
          <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Home
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            Home
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../UpdateLogs/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Update logs
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../KnownIssues/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Known Issues
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../tutorial_3.4/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Tutorial for gwaslab
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
      
      
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Sumstats manipulation
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Sumstats manipulation
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../SumstatsObject/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Sumstats Object
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../StatusCode/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Status Code
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Standardization and Normalization
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Standardization and Normalization
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1-methods-summary" class="md-nav__link">
    <span class="md-ellipsis">
      1. Methods Summary
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-ids-snpid-and-rsid" class="md-nav__link">
    <span class="md-ellipsis">
      2. IDs - SNPID and rsID
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3-chr-chromosomes" class="md-nav__link">
    <span class="md-ellipsis">
      3. CHR - Chromosomes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-pos-base-pair-positions" class="md-nav__link">
    <span class="md-ellipsis">
      4. POS - Base-pair positions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#5-allele" class="md-nav__link">
    <span class="md-ellipsis">
      5. Allele
    </span>
  </a>
  
    <nav class="md-nav" aria-label="5. Allele">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#51-allele-notation-standardization" class="md-nav__link">
    <span class="md-ellipsis">
      5.1 Allele notation Standardization
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#52-variant-normalization" class="md-nav__link">
    <span class="md-ellipsis">
      5.2 Variant Normalization
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#6-genome-coordinate-sorting" class="md-nav__link">
    <span class="md-ellipsis">
      6. Genome coordinate sorting
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#7-column-sorting" class="md-nav__link">
    <span class="md-ellipsis">
      7. Column sorting
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#8-all-in-one-basic_check" class="md-nav__link">
    <span class="md-ellipsis">
      8. All-in-one .basic_check()
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../QC%26Filtering/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    QC and Filtering
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Harmonization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Harmonization
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../AssignrsID/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assigning rsID
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../AssignCHRPOS/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assign CHR/POS using rsID
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../LiftOver/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Liftover
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Conversion/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Data conversion
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
      
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Visualization
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Visualization
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Visualization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Manhattan and QQ plot
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../RegionalPlot/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Regional Plot
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../BrisbanePlot/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Brisbane Plot
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../MiamiPlot/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Miami Plot
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../TrumpetPlot/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Trumpet Plot (testing)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../EffectSize/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Effect Size Comparison
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../AlleleFrequency/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Allele Frequency Comparison
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../GeneticCorrelation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Genetic correlation heatmap
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Gallery/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Gallery
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
      
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Utilities
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Utilities
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ExtractLead/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Extract Lead Variants
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ExtractNovel/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Extract Novel Variants
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Format/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Format and Save
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Pickle/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Save Sumstats Object
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../LoadLDSC/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Load LDSC log
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../HeritabilityConversion/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Convert Heritability
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../PerSNPh2/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Per SNP Heritability and F statistics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../InferBuild/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Infer Genome Build
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
      
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="">
            
  
  <span class="md-ellipsis">
    References
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            References
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Download/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Download
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Reference/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Reference data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../CommonData/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Common data
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
      
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Examples
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            Examples
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../visualization_regional/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Regional plot
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../visualization_miami2/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Miami plot
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1-methods-summary" class="md-nav__link">
    <span class="md-ellipsis">
      1. Methods Summary
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-ids-snpid-and-rsid" class="md-nav__link">
    <span class="md-ellipsis">
      2. IDs - SNPID and rsID
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3-chr-chromosomes" class="md-nav__link">
    <span class="md-ellipsis">
      3. CHR - Chromosomes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-pos-base-pair-positions" class="md-nav__link">
    <span class="md-ellipsis">
      4. POS - Base-pair positions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#5-allele" class="md-nav__link">
    <span class="md-ellipsis">
      5. Allele
    </span>
  </a>
  
    <nav class="md-nav" aria-label="5. Allele">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#51-allele-notation-standardization" class="md-nav__link">
    <span class="md-ellipsis">
      5.1 Allele notation Standardization
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#52-variant-normalization" class="md-nav__link">
    <span class="md-ellipsis">
      5.2 Variant Normalization
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#6-genome-coordinate-sorting" class="md-nav__link">
    <span class="md-ellipsis">
      6. Genome coordinate sorting
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#7-column-sorting" class="md-nav__link">
    <span class="md-ellipsis">
      7. Column sorting
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#8-all-in-one-basic_check" class="md-nav__link">
    <span class="md-ellipsis">
      8. All-in-one .basic_check()
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="standardization-and-normalization">Standardization and Normalization</h1>
<p>After loading raw sumstats into gwaslab.Sumstats Object, the first thing we probably want to do is to standardize the variant-related notations, normalize the indels and check if there are any unexpected errors in the statistics. When checking is finished, the status code will be automatically changed.</p>
<h2 id="1-methods-summary">1. Methods Summary</h2>
<table>
<thead>
<tr>
<th>Sumstats Methods</th>
<th>Options</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>.fix_id()</code></td>
<td><code>fixchrpos=False</code>, <br/><code>fixid=False</code>, <br/><code>fixsep=False</code>,<br/><code>overwrite=False</code>,<br/><code>forcefixid=False</code></td>
<td>check and  fix rsID or SNPID(chr:pos:ref:alt), or use snpid to fix CHR and POS</td>
</tr>
<tr>
<td><code>.fix_chr()</code></td>
<td><code>remove=False</code>, <code>x=("X",23)</code>, <code>y=("Y",24)</code>, <code>mt=("MT",25)</code>, <code>chrom_list = gl.get_chr_list()</code></td>
<td>standardize chromsome notation</td>
</tr>
<tr>
<td><code>.fix_pos()</code></td>
<td><code>remove=False</code> , <code>limit=250000000</code></td>
<td>standardize basepair posituion notation and filter out bad values</td>
</tr>
<tr>
<td><code>.fix_allele()</code></td>
<td><code>remove=False</code></td>
<td>standardize base notations to ATCG</td>
</tr>
<tr>
<td><code>.normalize_allele()</code></td>
<td><code>n_cores=1</code></td>
<td>normalize indels (only support ATA:AA -&gt; AT:A but not -:T)</td>
</tr>
<tr>
<td><code>.sort_coordinate()</code></td>
<td></td>
<td>sort the variant coordinates</td>
</tr>
<tr>
<td><code>.sort_column()</code></td>
<td><code>order</code></td>
<td>sort the column order to GWASLab default</td>
</tr>
<tr>
<td><code>.basic_check()</code></td>
<td></td>
<td>all-in-one function to perform the default pipeline</td>
</tr>
</tbody>
</table>
<h2 id="2-ids-snpid-and-rsid">2. IDs - SNPID and rsID</h2>
<p>GWASLab requires at least one ID column for sumstats, either in the form of SNPID or rsID, (or both). GWASLab will automatically check if SNPID is mixed in rsID.</p>
<ul>
<li><code>SNPID</code> : it could be user-provided IDs, or in <code>CHR:POS:REF:ALT</code> format, delimiters can be <code>":"</code>,<code>"_"</code> or <code>"-"</code>.</li>
<li><code>rsID</code>: dbSNP rsIDs</li>
</ul>
<p>GWASLab will check if the IDs are valid SNPID or rsID.  It can also extract CHR and POS information from the <code>CHR:POS:REF:ALT</code> formatted IDs using <code>.fix_id()</code> method.</p>
<p>SNPID will be fixed by <code>CHR:POS:NEA:EA</code>  only when the variants are already aligned with the reference genome. Otherwise, a temporary SNPID in the format of <code>CHR:POS</code> will be assigned.</p>
<p><code>.fix_id()</code>: check or fix SNPID and rsID.   </p>
<table>
<thead>
<tr>
<th><code>.fix_id()</code> options</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>fixchrpos</code></td>
<td><code>Boolean</code></td>
<td>If True, extract CHR and POS from SNPID (CHR:POS:NEA:EA) to fill CHR and POS columns (default:<code>False</code>)</td>
</tr>
<tr>
<td><code>fixid</code></td>
<td><code>Boolean</code></td>
<td>If True, use CHR/POS/NEA/EA to reconstruct the SNPID. For variant that are not aligned with reference genome. Only CHR/POS will be used. (default:<code>False</code>)</td>
</tr>
<tr>
<td><code>forcefixid</code></td>
<td><code>Boolean</code></td>
<td>If True, use CHR/POS/NEA/EA to reconstruct the SNPID without checking if the variant is aligned. (default:<code>False</code>)</td>
</tr>
<tr>
<td><code>fixsep</code></td>
<td><code>Boolean</code></td>
<td>If True, fix SNPID delimiter (For example: 1:123_A_C to 1:123:A:C) (default:<code>False</code>)</td>
</tr>
<tr>
<td><code>overwrite</code></td>
<td><code>Boolean</code></td>
<td>If True, overwrite existing data. (default:<code>False</code>)</td>
</tr>
</tbody>
</table>
<div class="admonition example">
<p class="admonition-title">Example</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="n">sumstats</span><span class="o">.</span><span class="n">fix_id</span><span class="p">(</span><span class="n">fixchrpos</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                <span class="n">fixid</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
</span><span id="__span-0-3"><a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>                <span class="n">fixsep</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
</span><span id="__span-0-4"><a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>                <span class="n">forcefixid</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
</span><span id="__span-0-5"><a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>                <span class="n">overwrite</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</span></code></pre></div>
</div>
<h2 id="3-chr-chromosomes">3. CHR - Chromosomes</h2>
<div class="language-text highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>.fix_chr()
</span></code></pre></div>
<p>CHR will be standardized. </p>
<ul>
<li>Prefixes like "chr" will be removed.</li>
<li>String datatype will be converted to integers.</li>
<li>Sex chromosomes will be converted to integers.</li>
<li>Variants with unrecognized chromosome notations will be removed. (notations not in <code>chrom_list</code>)</li>
<li>Variants with CHR&lt;=0 will be removed. </li>
</ul>
<p>Options:</p>
<table>
<thead>
<tr>
<th><code>.fix_chr()</code> options</th>
<th>DataType</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>x</code></td>
<td><code>tuple</code></td>
<td>how to convert X chromosomes. For examplem <code>x = ("X",23)</code></td>
<td><code>("X",23)</code></td>
</tr>
<tr>
<td><code>y</code></td>
<td><code>tuple</code></td>
<td>how to convert Y chromosomes.</td>
<td><code>("Y",24)</code></td>
</tr>
<tr>
<td><code>mt</code></td>
<td><code>tuple</code></td>
<td>how to convert Mitochondrial DNA.</td>
<td><code>("MT",25)</code></td>
</tr>
<tr>
<td><code>chrom_list</code></td>
<td><code>list</code></td>
<td>vaild chromosome notations for filtering. Datatype should be <code>string</code></td>
<td><code>gl.get_chr_list()</code></td>
</tr>
</tbody>
</table>
<p>(by default) For human chromosomes, CHR will be converted to integers for autosomes, <code>23</code> for <code>X</code>, <code>24</code> for <code>Y</code>, and <code>25</code> for <code>MT</code>. </p>
<p>For other species, you can change it like:
- <code>x=("X",26)</code> 
- <code>y=("Y",27)</code><br />
- <code>mt=("MT",28)</code>
- <code>chrom_list=gl.get_chr_list(n=28)</code></p>
<div class="admonition example">
<p class="admonition-title">Example</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a><span class="n">sumstats</span><span class="o">.</span><span class="n">fix_chr</span><span class="p">()</span>
</span></code></pre></div>
</div>
<p>!!! tip gl.get_chr_list()
    Get a chromosome list for n autosomes plus 'X', 'Y', 'M', 'MT. (`string`` data type)</p>
<div class="language-text highlight"><pre><span></span><code>```python
gl.get_chr_list()
[&#39;1&#39;,&#39;2&#39;,&#39;3&#39;,&#39;4&#39;,&#39;5&#39;,&#39;6&#39;,&#39;7&#39;,&#39;8&#39;,&#39;9&#39;,&#39;10&#39;,&#39;11&#39;,&#39;12&#39;,&#39;13&#39;,&#39;14&#39;,&#39;15&#39;,&#39;16&#39;,&#39;17&#39;,&#39;18&#39;,&#39;19&#39;,&#39;20&#39;,&#39;21&#39;,&#39;22&#39;,&#39;23&#39;,&#39;24&#39;,&#39;25&#39;,&#39;X&#39;,&#39;Y&#39;,&#39;M&#39;,&#39;MT&#39;]

gl.get_chr_list(n=10)
[&#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;, &#39;6&#39;, &#39;7&#39;, &#39;8&#39;, &#39;9&#39;, &#39;10&#39;, &#39;X&#39;, &#39;Y&#39;, &#39;M&#39;, &#39;MT&#39;]
```
</code></pre></div>
<h2 id="4-pos-base-pair-positions">4. POS - Base-pair positions</h2>
<div class="language-text highlight"><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>.fix_pos()
</span></code></pre></div>
<p>Check and fix values in POS.</p>
<ul>
<li>Values in POS must be positive integer numbers. </li>
<li>Basepair position will be force converted to integers. </li>
<li>Invalid POS values will be converted to NA. </li>
<li>POS outliers will be removed.</li>
</ul>
<table>
<thead>
<tr>
<th><code>.fix_pos()</code> options</th>
<th>DataType</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>limit</code></td>
<td><code>integer</code></td>
<td>After conversion, GWASLab will also perform a sanity check for POS (if is in the range of 1 to 250,000,000). (The longest chromosome for human, namely chromosome 1, is less than 250,000,000bp long)</td>
<td><code>250000000</code></td>
</tr>
</tbody>
</table>
<div class="admonition example">
<p class="admonition-title">Example</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a><span class="n">sumstats</span><span class="o">.</span><span class="n">fix_pos</span><span class="p">()</span>
</span></code></pre></div>
</div>
<h2 id="5-allele">5. Allele</h2>
<h3 id="51-allele-notation-standardization">5.1 Allele notation Standardization</h3>
<p><code>.fix_allele()</code></p>
<ul>
<li>Currently, GWASLab only supports processing SNPs and INDELs. </li>
<li>All alleles will be checked if containing letters other than <code>A</code>,<code>T</code>,<code>C</code>,<code>G</code>.</li>
<li>Copy number variants (CNV) like <code>&lt;CN0&gt;</code> won't be recognized.</li>
<li>Lower cases will be converted to UPPERCASES.</li>
</ul>
<div class="admonition example">
<p class="admonition-title">Example</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-6-1"><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a><span class="n">sumstats</span><span class="o">.</span><span class="n">fix_allele</span><span class="p">()</span>
</span></code></pre></div>
</div>
<h3 id="52-variant-normalization">5.2 Variant Normalization</h3>
<p><code>.normalize_allele()</code></p>
<p>Alleles will be normalized according to the left alignment and parsimony principle. For example, chr1:123456:ATG:AT will be normalized to chr1:123457:TG:T.</p>
<table>
<thead>
<tr>
<th><code>.normalize_allele()</code> options</th>
<th>DataType</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>n_cores</code></td>
<td><code>integer</code></td>
<td>threads to use for normalization.</td>
<td><code>1</code></td>
</tr>
</tbody>
</table>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Currently, the normalization is implemented without checking reference, which means it can not normalize variants like chr1:123456:G:- if the missing allele information needs to be obtained from a reference genome.  </p>
</div>
<div class="admonition quote">
<p class="admonition-title">Variant Normalization</p>
<p>For details on variant normalization, please check: <a href="https://genome.sph.umich.edu/wiki/Variant_Normalization">https://genome.sph.umich.edu/wiki/Variant_Normalization</a></p>
</div>
<div class="admonition example">
<p class="admonition-title">Example</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-7-1"><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a><span class="n">sumstats</span><span class="o">.</span><span class="n">normalize_allele</span><span class="p">(</span><span class="n">n_cores</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</span></code></pre></div>
<p>Before:</p>
<p><img width="345" alt="image" src="https://user-images.githubusercontent.com/40289485/212257048-1a5517a4-dd4d-4210-9e19-1dcf7747b7f5.png"></p>
<p>After:</p>
<p><img width="345" alt="image" src="https://user-images.githubusercontent.com/40289485/212256576-7808a1ec-5cf2-42ec-819e-a6f1c9c200bb.png"></p>
</div>
<h2 id="6-genome-coordinate-sorting">6. Genome coordinate sorting</h2>
<p>Sort genomic coordinates. Make sure CHR and POS are fixed beforehand.</p>
<div class="admonition example">
<p class="admonition-title">Example</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-8-1"><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a><span class="n">sumstats</span><span class="o">.</span><span class="n">sort_coordinate</span><span class="p">()</span>
</span></code></pre></div>
</div>
<h2 id="7-column-sorting">7. Column sorting</h2>
<div class="language-text highlight"><pre><span></span><code><span id="__span-9-1"><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>.sort_column()
</span></code></pre></div>
<table>
<thead>
<tr>
<th><code>.sort_column()</code> options</th>
<th>DataType</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>order</code></td>
<td><code>list</code></td>
<td>sort the columns</td>
<td>The default column order is <code>"SNPID","rsID", "CHR", "POS", "EA", "NEA", "EAF", "MAF", "BETA", "SE","BETA_95L","BETA_95U", "Z", "CHISQ", "P", "MLOG10P", "OR", "OR_95L", "OR_95U","HR", "HR_95L", "HR_95U","INFO", "N","N_CASE","N_CONTROL","DIRECTION","I2","P_HET","DOF","SNPR2","STATUS"</code> and other additional columns.</td>
</tr>
</tbody>
</table>
<div class="admonition example">
<p class="admonition-title">Example</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-10-1"><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a><span class="n">sumstats</span><span class="o">.</span><span class="n">sort_column</span><span class="p">()</span>
</span></code></pre></div>
</div>
<h2 id="8-all-in-one-basic_check">8. All-in-one <code>.basic_check()</code></h2>
<p><div class="language-text highlight"><pre><span></span><code><span id="__span-11-1"><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>.basic_check()
</span></code></pre></div>
Fix data without dropping any variant.</p>
<div class="admonition note">
<p class="admonition-title"><code>basic_check()</code> pipeline</p>
<ul>
<li>fix_id()</li>
<li>remove_dup()</li>
<li>fix_chr()</li>
<li>fix_pos()</li>
<li>fix_allele()</li>
<li>normalize_allele()</li>
<li>check_sanity()  </li>
<li>sort_coordinate()</li>
<li>sort_column()</li>
</ul>
</div>
<p>For <code>remove_dup()</code> and <code>check_sanity()</code>, please check <a href="https://cloufield.github.io/gwaslab/QC%26Filtering/">QC and filtering</a></p>
<table>
<thead>
<tr>
<th><code>.basic_check()</code> options</th>
<th>DataType</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>remove</code></td>
<td><code>boolean</code></td>
<td>if True, remove duplicate and multi-allelic variants</td>
<td><code>False</code></td>
</tr>
<tr>
<td><code>removedup_args</code></td>
<td><code>dict</code></td>
<td>options for remove_dup()</td>
<td><code>{}</code></td>
</tr>
<tr>
<td><code>fixid_args</code></td>
<td><code>dict</code></td>
<td>options for <code>.fix_id()</code></td>
<td><code>{}</code></td>
</tr>
<tr>
<td><code>fixchr_agrs</code></td>
<td><code>dict</code></td>
<td>options for <code>.fix_chr()</code></td>
<td><code>{}</code></td>
</tr>
<tr>
<td><code>fixpos_args</code></td>
<td><code>dict</code></td>
<td>options for <code>.fix_pos()</code></td>
<td><code>{}</code></td>
</tr>
<tr>
<td><code>fixallele_args</code></td>
<td><code>dict</code></td>
<td>options for <code>.fix_allele()</code></td>
<td><code>{}</code></td>
</tr>
<tr>
<td><code>sanitycheckstats_args</code></td>
<td><code>dict</code></td>
<td>options for <code>.check_sanity()</code></td>
<td><code>{}</code></td>
</tr>
<tr>
<td><code>normalizeallele_args</code></td>
<td><code>dict</code></td>
<td>options for <code>.normalize_allele()</code></td>
<td><code>{}</code></td>
</tr>
<tr>
<td><code>verbose</code></td>
<td><code>boolean</code></td>
<td>if True, print log</td>
<td><code>True</code></td>
</tr>
</tbody>
</table>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      GWASLab is licensed under the MIT license
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs"], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.e1c3ead8.min.js"></script>
      
        <script src="../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>