<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    
    <meta name="author" content="HE Yunye">
    
    <link rel="shortcut icon" href="../img/favicon.ico">

    
    <title>Standardization&normalization - GWASLab</title>
    

    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.12.0/css/all.css">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.12.0/css/v4-shims.css">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/npm/hack-font@3.3.0/build/web/hack.min.css">
    <link href='//rsms.me/inter/inter.css' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,700italic,400,300,600,700&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href="../css/bootstrap-custom.min.css" rel="stylesheet">
    <link href="../css/base.min.css" rel="stylesheet">
    <link href="../css/cinder.min.css" rel="stylesheet">

    
        
        <link rel="stylesheet" href="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/styles/github.min.css">
        
    
    <link href="../stylesheets/extra.css" rel="stylesheet">

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js"></script>
            <script src="https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js"></script>
        <![endif]-->

    

     
</head>

<body>

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->

            
              <a class="navbar-brand" href="..">GWASLab</a>
            
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="..">Home</a>
                    </li>
                
                
                
                    <li >
                        <a href="../Tutorial/">Tutorial</a>
                    </li>
                
                
                
                    <li class="dropdown active">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Sumstats manipulation <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../SumstatsObject/">SumstatsObject</a>
</li>

                        
                            
<li >
    <a href="../StatusCode/">StatusCode</a>
</li>

                        
                            
<li class="active">
    <a href="./">Standardization&normalization</a>
</li>

                        
                            
<li >
    <a href="../QC%26Filtering/">QC&Filtering</a>
</li>

                        
                            
<li >
    <a href="../Harmonization/">Harmonization</a>
</li>

                        
                            
<li >
    <a href="../LiftOver/">Liftover</a>
</li>

                        
                            
<li >
    <a href="../Conversion/">Data conversion</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Visualization <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../Visualization/">Manhattan & QQ plot</a>
</li>

                        
                            
<li >
    <a href="../RegionalPlot/">Regional Plot</a>
</li>

                        
                            
<li >
    <a href="../BrisbanePlot/">Brisbane Plot</a>
</li>

                        
                            
<li >
    <a href="../MiamiPlot/">Miami Plot</a>
</li>

                        
                            
<li >
    <a href="../EffectSize/">Scatter: effect size comparison</a>
</li>

                        
                            
<li >
    <a href="../AlleleFrequency/">Scatter: allele frequency comparison</a>
</li>

                        
                            
<li >
    <a href="../GeneticCorrelation/">Heatmap: genetic correlation</a>
</li>

                        
                            
<li >
    <a href="../ForestPlot/">Forest Plot</a>
</li>

                        
                            
<li >
    <a href="../Gallery/">Gallery</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Utilities <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../ExtractLead/">Extract Lead Variants</a>
</li>

                        
                            
<li >
    <a href="../ExtractNovel/">Extract Novel Variants</a>
</li>

                        
                            
<li >
    <a href="../Format/">Format & Save</a>
</li>

                        
                            
<li >
    <a href="../LoadLDSC/">Load LDSC log</a>
</li>

                        
                            
<li >
    <a href="../HeritabilityConversion/">Convert Heritability</a>
</li>

                        
                            
<li >
    <a href="../InferBuild/">Infer Genome Build</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">References <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../Reference/">Reference data</a>
</li>

                        
                            
<li >
    <a href="../CommonData/">Common data</a>
</li>

                        
                            
<li >
    <a href="../UpdateLogs/">Update logs</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Examples <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Standardization and QC example</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../standardization_workflow/">Standardization Workflow</a>
</li>

        
            
<li >
    <a href="../quality_control_and_filtering/">QC &amp; Filtering</a>
</li>

        
    </ul>
  </li>

                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Harmonization example</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../harmonization_workflow/">Workflow</a>
</li>

        
            
<li >
    <a href="../harmonization_liftover/">Liftover</a>
</li>

        
    </ul>
  </li>

                        
                            
<li >
    <a href="../format_load_save/">Formatting and saving</a>
</li>

                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Utility example</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../utility_data_conversion/">Data conversion</a>
</li>

        
            
<li >
    <a href="../utility_get_novel/">Get novel variants</a>
</li>

        
    </ul>
  </li>

                        
                            
  <li class="dropdown-submenu">
    <a tabindex="-1" href="">Visualization example</a>
    <ul class="dropdown-menu">
        
            
<li >
    <a href="../visualization_mqq/">Manhattan and Q-Q plot</a>
</li>

        
            
<li >
    <a href="../visualization_miami/">Miami plot</a>
</li>

        
            
<li >
    <a href="../visualization_regional/">Regional plot</a>
</li>

        
            
<li >
    <a href="../visualization_brisbane/">Brisbane plot</a>
</li>

        
    </ul>
  </li>

                        
                        </ul>
                    </li>
                
                
                </ul>

            <ul class="nav navbar-nav navbar-right">
                    <li >
                        <a rel="prev" href="../StatusCode/">
                            <i class="fas fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="next" href="../QC%26Filtering/">
                            Next <i class="fas fa-arrow-right"></i>
                        </a>
                    </li>
                    <li>
                        <a href="https://github.com/Cloufield/gwaslab/"><i class="fab fa-github"></i> GitHub</a>
                    </li>
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
        <li class="first-level active"><a href="#standardization-and-normalization">Standardization and normalization</a></li>
            <li class="second-level"><a href="#methods-summary">Methods Summary</a></li>
                
            <li class="second-level"><a href="#1-ids">1. IDs</a></li>
                
            <li class="second-level"><a href="#2-chr">2. CHR</a></li>
                
            <li class="second-level"><a href="#3-pos">3. POS</a></li>
                
            <li class="second-level"><a href="#4-allele">4. Allele</a></li>
                
                <li class="third-level"><a href="#41-allele-notation-standardization">4.1 Allele notation Standardization</a></li>
                <li class="third-level"><a href="#42-variant-normalization">4.2 Variant Normalization</a></li>
            <li class="second-level"><a href="#5-coordinate-sorting">5. Coordinate sorting</a></li>
                
            <li class="second-level"><a href="#6-column-sorting">6. Column sorting</a></li>
                
            <li class="second-level"><a href="#example">Example</a></li>
                
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<h1 id="standardization-and-normalization">Standardization and normalization</h1>
<pre><code class="language-python">import gwaslab as gl
sumstats = gl.Sumstats(...)
</code></pre>
<p>After loading raw sumstats into gwaslab Sumstats Object, the first thing we probably want to do is to standardize the variant-related notations and check if there are any unexpected errors in the statistics. When checking is finished, the status code will be automatically changed.</p>
<h2 id="methods-summary">Methods Summary</h2>
<table>
<thead>
<tr>
<th>Sumstats Methods</th>
<th>Options</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>.fix_ID()</code></td>
<td><code>fixchrpos=False</code>, <br/><code>fixid=False</code>, <br/><code>fixsep=False</code>,<br/><code>overwrite=False</code>,<br/><code>forcefixid=False</code></td>
<td>check and  fix rsID or SNPID(chr:pos:ref:alt), or use snpid to fix CHR and POS</td>
</tr>
<tr>
<td><code>.fix_CHR()</code></td>
<td><code>remove=False</code></td>
<td>standardize chromsome notation</td>
</tr>
<tr>
<td><code>.fix_POS()</code></td>
<td><code>remove=False</code></td>
<td>standardize basepair posituion notation and filter out bad values</td>
</tr>
<tr>
<td><code>.fix_allele()</code></td>
<td><code>remove=False</code></td>
<td>standardize base notation to ATCG</td>
</tr>
<tr>
<td><code>.normalize_allele()</code></td>
<td><code>n_cores=1</code></td>
<td>normalize indels (only support ATA:AA -&gt; AT:A but not -:T)</td>
</tr>
<tr>
<td><code>.sort_coordinate()</code></td>
<td></td>
<td>sort the variant coordinates</td>
</tr>
<tr>
<td><code>.sort_column()</code></td>
<td></td>
<td>sort the column order to gwaslab default</td>
</tr>
<tr>
<td><code>.basic_check()</code></td>
<td></td>
<td>all in one function</td>
</tr>
</tbody>
</table>
<h2 id="1-ids">1. IDs</h2>
<p>Gwaslab requires at least one ID columns for sumstats, either in the form of SNPID or rsID, (or both). Gwaslab will automatically check if SNPID is mixed in rsID.</p>
<p><code>SNPID</code> : it could be user provided IDs, or in <code>CHR:POS:REF:ALT</code> format, delimiter can be":","_" or "-"</p>
<p><code>rsID</code>: dbSNP rsIDs</p>
<p>gwaslab checks if the IDs you provided is valid SNPID or rsID.  It can also extract CHR and POS information from the CHR:POS:REF:ALT formatted IDs using <code>.fix_ID()</code> method.</p>
<p>SNPID will be fixed by <code>CHR:POS:NEA:EA</code>  only when the variants is already aligned with reference genome. Otherwise, a temporary SNPID in the format of <code>CHR:POS</code> will be given.</p>
<p><code>.fix_ID()</code> : check or fix SNPID and rsID.   </p>
<pre><code class="language-python">sumstats.fixID(fixchrpos=False,
               fixid=False,
               fixsep=False,
               overwrite=False)
</code></pre>
<p><code>.fix_ID()</code> options:</p>
<ul>
<li><code>fixchrpos</code> : <code>Boolean</code>, extract CHR and POS from SNPID (CHR:POS:NEA:EA) to fill CHR and POS columns (deaful:<code>False</code>)</li>
<li><code>fixid</code> : <code>Boolean</code> ,  use CHR/POS/NEA/EA to reconstruct the SNPID. For variant that are not aligned with reference genome. Only CHR/POS will be used. (deaful:<code>False</code>)</li>
<li><code>forcefixid</code>: <code>Boolean</code> ,  use CHR/POS/NEA/EA to reconstruct the SNPID without checking if the variant is aligned. (deaful:<code>False</code>)</li>
<li><code>fixsep</code> : <code>Boolean</code>, fix SNPID delimiter (For example: 1:123_A_C to 1:123:A:C) (deaful:<code>False</code>)</li>
<li><code>overwrite</code> : <code>Boolean</code>, whether overwrite existing data. (deaful:<code>False</code>)</li>
</ul>
<h2 id="2-chr">2. CHR</h2>
<p><code>.fix_CHR()</code></p>
<p>CHR will be standardized to integers <code>1-22,23,24,25</code>
- autosomes: 1-22
- <code>x="X"</code> : 23
- <code>x="Y"</code> : 24
- <code>x="MT"</code> : 25</p>
<p>Leading "chr" or leading 0s will be stripped. Before conversion, CHR will be converted to uppercase.</p>
<pre><code class="language-python">sumstats.fix_CHR(remove=False,x=&quot;X&quot;,y=&quot;Y&quot;,mt=&quot;MT&quot;)
</code></pre>
<h2 id="3-pos">3. POS</h2>
<p><code>.fix_POS()</code></p>
<p>Values in POS must be positive integer numbers.</p>
<p>Basepair position will be force converted to integers.</p>
<p>Invalid pos will be converted to NA. </p>
<p><code>limit=250000000</code> : After conversion, gwaslab will also sanity check if POS is in the range of 1 to 250,000,000. (the longest chromosome, CHR1, is around 250,000,000bp long)  </p>
<pre><code class="language-python">sumstats.fix_POS(remove=False, limit=250000000)
</code></pre>
<h2 id="4-allele">4. Allele</h2>
<h3 id="41-allele-notation-standardization">4.1 Allele notation Standardization</h3>
<p><code>.fix_allele()</code></p>
<p>Currently, gwaslab only support processing SNPs and INDELs.</p>
<p>All alleles will be checked if containing letters other than <code>A</code>,<code>T</code>,<code>C</code>,<code>G</code>.</p>
<p>Copy number variant (CNV) like <code>&lt;CN0&gt;</code> won't be recognized.</p>
<p>Lower cases will converted to UPPERCASES.</p>
<pre><code class="language-python">sumstats.fix_allele(remove=False)
</code></pre>
<h3 id="42-variant-normalization">4.2 Variant Normalization</h3>
<p><code>.normalize_allele()</code></p>
<p>Alleles will be normalized accroding to left alignment and parsimony principal. (For details: <a href="https://genome.sph.umich.edu/wiki/Variant_Normalization">https://genome.sph.umich.edu/wiki/Variant_Normalization</a> )</p>
<p>For example, chr1:123456:ATG:AT will be normalized to chr1:123455:TG:T.</p>
<p>Note: Currently, the normalizeation is implemented without checking reference, which means it can not normalize variants like chr1:123456:G:- if the missing information need to be obtained from a reference genome. 
<code>n_cores</code> : threads to use for normalization.</p>
<pre><code class="language-python">sumstats.normalize_allele(n_cores=1)
</code></pre>
<h2 id="5-coordinate-sorting">5. Coordinate sorting</h2>
<p>Sort genomic coordinates， 1-25 (23:X, 24:Y ,25:MT)</p>
<pre><code class="language-python">sumstats.sort_coordinate()
</code></pre>
<h2 id="6-column-sorting">6. Column sorting</h2>
<p>The default column order is "SNPID","rsID", "CHR", "POS", "EA", "NEA", "EAF", "BETA", "SE", "Z",  "CHISQ", "P", "MLOG10P", "OR", "OR_95L", "OR_95U", "INFO", "N","DIRECTION","STATUS" and other additional columns.</p>
<p><code>order</code>:<code>list</code> , column order.</p>
<pre><code class="language-python">sumstats.sort_column(order = [
        &quot;SNPID&quot;,&quot;rsID&quot;, &quot;CHR&quot;, &quot;POS&quot;, &quot;EA&quot;, &quot;NEA&quot;, &quot;EAF&quot;, &quot;BETA&quot;, &quot;SE&quot;, &quot;Z&quot;,
        &quot;CHISQ&quot;, &quot;P&quot;, &quot;MLOG10P&quot;, &quot;OR&quot;, &quot;OR_SE&quot;, &quot;OR_95L&quot;, &quot;OR_95U&quot;, &quot;INFO&quot;, &quot;N&quot;,&quot;DIRECTION&quot;,&quot;STATUS&quot;
           ])
</code></pre>
<h2 id="example">Example</h2>
<p>Please check <a href="https://github.com/Cloufield/gwaslab/blob/main/examples/standardization_workflow.ipynb">https://github.com/Cloufield/gwaslab/blob/main/examples/standardization_workflow.ipynb</a></p></div>
        
        
    </div>

    
      <footer class="col-md-12 text-center">
          
          
            <hr>
            <p>
            <small>GWASLab is licensed under the MIT license</small><br>
            
            <small>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</small>
            </p>
          

          
          
      </footer>
    
    <script src="//ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="../js/bootstrap-3.0.3.min.js"></script>

    
    <script src="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/highlight.min.js"></script>
        
    <script>hljs.initHighlightingOnLoad();</script>
    

    <script>var base_url = ".."</script>
    
    <script src="../js/base.js"></script>

    <div class="modal" id="mkdocs_keyboard_modal" tabindex="-1" role="dialog" aria-labelledby="keyboardModalLabel" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
            </div>
            <div class="modal-body">
              <table class="table">
                <thead>
                  <tr>
                    <th style="width: 20%;">Keys</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td class="help shortcut"><kbd>?</kbd></td>
                    <td>Open this help</td>
                  </tr>
                  <tr>
                    <td class="next shortcut"><kbd>n</kbd></td>
                    <td>Next page</td>
                  </tr>
                  <tr>
                    <td class="prev shortcut"><kbd>p</kbd></td>
                    <td>Previous page</td>
                  </tr>
                  <tr>
                    <td class="search shortcut"><kbd>s</kbd></td>
                    <td>Search</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>
    </body>

</html>
