{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lead and novel variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/26 11:36:20 GWASLab v4.0.0 https://cloufield.github.io/gwaslab/\n",
      "2025/12/26 11:36:20 (C) 2022-2025, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2025/12/26 11:36:20 Python version: 3.12.0 | packaged by conda-forge | (main, Oct  3 2023, 08:43:22) [GCC 12.3.0]\n"
     ]
    }
   ],
   "source": [
    "gl.show_version()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load sample data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use only first 1000000 variants as example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:725932_G_A</td>\n",
       "      <td>1</td>\n",
       "      <td>725932</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>-0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:725933_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>725933</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:737801_T_C</td>\n",
       "      <td>1</td>\n",
       "      <td>737801</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.0490</td>\n",
       "      <td>0.1231</td>\n",
       "      <td>0.69080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:749963_T_TAA</td>\n",
       "      <td>1</td>\n",
       "      <td>749963</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>1960399</td>\n",
       "      <td>0.0213</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.28460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:751343_T_A</td>\n",
       "      <td>1</td>\n",
       "      <td>751343</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999995</th>\n",
       "      <td>2:6347639_C_A</td>\n",
       "      <td>2</td>\n",
       "      <td>6347639</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>-0.0159</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>0.15100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999996</th>\n",
       "      <td>2:6347694_G_C</td>\n",
       "      <td>2</td>\n",
       "      <td>6347694</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1960099</td>\n",
       "      <td>-0.0151</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>0.17250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999997</th>\n",
       "      <td>2:6348478_G_A</td>\n",
       "      <td>2</td>\n",
       "      <td>6348478</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>-0.0152</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>0.17160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999998</th>\n",
       "      <td>2:6348490_G_C</td>\n",
       "      <td>2</td>\n",
       "      <td>6348490</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1960099</td>\n",
       "      <td>-0.3180</td>\n",
       "      <td>0.2032</td>\n",
       "      <td>0.11750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999999</th>\n",
       "      <td>2:6348754_A_C</td>\n",
       "      <td>2</td>\n",
       "      <td>6348754</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.0297</td>\n",
       "      <td>0.0126</td>\n",
       "      <td>0.01846</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000000 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x75413507c4a0>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(\"../0_sample_data/t2d_bbj.txt.gz\",\n",
    "             snpid=\"SNP\",\n",
    "             chrom=\"CHR\",\n",
    "             pos=\"POS\",\n",
    "             ea=\"ALT\",\n",
    "             nea=\"REF\",            \n",
    "             beta=\"BETA\",\n",
    "             se=\"SE\",\n",
    "             p=\"P\", \n",
    "             build=\"19\",\n",
    "             verbose=False, \n",
    "             nrows=1000000)\n",
    "mysumstats.basic_check(verbose=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get lead variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GWASLab will use MLOG10P first by default. If MLOG10P is not avaiable, it will look for P column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/26 11:36:24  -Genomic coordinates are based on GRCh37/hg19...\n",
      "2025/12/26 11:36:24 Start to extract lead variants ...(v4.0.0)\n",
      "2025/12/26 11:36:24  -Processing 1000000 variants...\n",
      "2025/12/26 11:36:24  -Significance threshold : 5e-08\n",
      "2025/12/26 11:36:24  -Sliding window size: 500  kb\n",
      "2025/12/26 11:36:24  -Using P for extracting lead variants...\n",
      "2025/12/26 11:36:24  -Found 543 significant variants in total...\n",
      "2025/12/26 11:36:24  -Identified 4 lead variants!\n",
      "2025/12/26 11:36:24 Finished extracting lead variants.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96739</th>\n",
       "      <td>1:22068326_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>22068326</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.0621</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>1.629000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213860</th>\n",
       "      <td>1:51103268_T_C</td>\n",
       "      <td>1</td>\n",
       "      <td>51103268</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>-0.0802</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>2.519000e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534095</th>\n",
       "      <td>1:154309595_TA_T</td>\n",
       "      <td>1</td>\n",
       "      <td>154309595</td>\n",
       "      <td>TA</td>\n",
       "      <td>T</td>\n",
       "      <td>1960399</td>\n",
       "      <td>-0.0915</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>3.289000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>969974</th>\n",
       "      <td>2:640986_CACAT_C</td>\n",
       "      <td>2</td>\n",
       "      <td>640986</td>\n",
       "      <td>C</td>\n",
       "      <td>CACAT</td>\n",
       "      <td>1960399</td>\n",
       "      <td>-0.0946</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>2.665000e-10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   SNPID  CHR        POS  EA    NEA   STATUS    BETA      SE  \\\n",
       "96739     1:22068326_A_G    1   22068326   G      A  1960099  0.0621  0.0103   \n",
       "213860    1:51103268_T_C    1   51103268   C      T  1960099 -0.0802  0.0120   \n",
       "534095  1:154309595_TA_T    1  154309595  TA      T  1960399 -0.0915  0.0166   \n",
       "969974  2:640986_CACAT_C    2     640986   C  CACAT  1960399 -0.0946  0.0150   \n",
       "\n",
       "                   P  \n",
       "96739   1.629000e-09  \n",
       "213860  2.519000e-11  \n",
       "534095  3.289000e-08  \n",
       "969974  2.665000e-10  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.get_lead()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get lead variants with gene name annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/26 11:36:24  -Genomic coordinates are based on GRCh37/hg19...\n",
      "2025/12/26 11:36:24 Start to extract lead variants ...(v4.0.0)\n",
      "2025/12/26 11:36:24  -Processing 1000000 variants...\n",
      "2025/12/26 11:36:24  -Significance threshold : 5e-08\n",
      "2025/12/26 11:36:24  -Sliding window size: 500  kb\n",
      "2025/12/26 11:36:24  -Using P for extracting lead variants...\n",
      "2025/12/26 11:36:24  -Found 543 significant variants in total...\n",
      "2025/12/26 11:36:24  -Identified 4 lead variants!\n",
      "2025/12/26 11:36:24  -Annotating variants using references:ensembl\n",
      "2025/12/26 11:36:24  -Annotating variants using references based on genome build:19\n",
      "2025/12/26 11:36:24  -Genomic coordinates are based on GRCh37/hg19...\n",
      "2025/12/26 11:36:24 Start to annotate variants with nearest gene name(s) ...(v4.0.0)\n",
      "2025/12/26 11:36:24  -Current Dataframe shape : 4 x 10 ; Memory usage: 0.64 MB\n",
      "2025/12/26 11:36:24  -Genomic coordinates are based on GRCh37/hg19...\n",
      "2025/12/26 11:36:24  -Assigning Gene name using ensembl_hg19_gtf for protein coding genes\n",
      "2025/12/26 11:36:24  -Loading and indexing GTF file...\n",
      "2025/12/26 11:36:26  -Processing variants by chromosome...\n",
      "2025/12/26 11:36:26    -Processed 3 variants on chromosome 1\n",
      "2025/12/26 11:36:26    -Processed 1 variants on chromosome 2\n",
      "2025/12/26 11:36:26 Finished annotating variants with nearest gene name(s) successfully!.\n",
      "2025/12/26 11:36:26 Finished extracting lead variants.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>LOCATION</th>\n",
       "      <th>GENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96739</th>\n",
       "      <td>1:22068326_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>22068326</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.0621</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>1.629000e-09</td>\n",
       "      <td>0</td>\n",
       "      <td>USP48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213860</th>\n",
       "      <td>1:51103268_T_C</td>\n",
       "      <td>1</td>\n",
       "      <td>51103268</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>-0.0802</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>2.519000e-11</td>\n",
       "      <td>0</td>\n",
       "      <td>FAF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534095</th>\n",
       "      <td>1:154309595_TA_T</td>\n",
       "      <td>1</td>\n",
       "      <td>154309595</td>\n",
       "      <td>TA</td>\n",
       "      <td>T</td>\n",
       "      <td>1960399</td>\n",
       "      <td>-0.0915</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>3.289000e-08</td>\n",
       "      <td>0</td>\n",
       "      <td>ATP8B2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>969974</th>\n",
       "      <td>2:640986_CACAT_C</td>\n",
       "      <td>2</td>\n",
       "      <td>640986</td>\n",
       "      <td>C</td>\n",
       "      <td>CACAT</td>\n",
       "      <td>1960399</td>\n",
       "      <td>-0.0946</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>2.665000e-10</td>\n",
       "      <td>-26349</td>\n",
       "      <td>TMEM18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   SNPID  CHR        POS  EA    NEA   STATUS    BETA      SE  \\\n",
       "96739     1:22068326_A_G    1   22068326   G      A  1960099  0.0621  0.0103   \n",
       "213860    1:51103268_T_C    1   51103268   C      T  1960099 -0.0802  0.0120   \n",
       "534095  1:154309595_TA_T    1  154309595  TA      T  1960399 -0.0915  0.0166   \n",
       "969974  2:640986_CACAT_C    2     640986   C  CACAT  1960399 -0.0946  0.0150   \n",
       "\n",
       "                   P  LOCATION    GENE  \n",
       "96739   1.629000e-09         0   USP48  \n",
       "213860  2.519000e-11         0    FAF1  \n",
       "534095  3.289000e-08         0  ATP8B2  \n",
       "969974  2.665000e-10    -26349  TMEM18  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.get_lead(anno=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Different window sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/26 11:36:28  -Genomic coordinates are based on GRCh37/hg19...\n",
      "2025/12/26 11:36:28 Start to extract lead variants ...(v4.0.0)\n",
      "2025/12/26 11:36:28  -Current Dataframe shape : 1000000 x 9 ; Memory usage: 60.72 MB\n",
      "2025/12/26 11:36:28  -Processing 1000000 variants...\n",
      "2025/12/26 11:36:28  -Significance threshold : 5e-08\n",
      "2025/12/26 11:36:28  -Sliding window size: 1000  kb\n",
      "2025/12/26 11:36:29  -Using P for extracting lead variants...\n",
      "2025/12/26 11:36:29  -Found 543 significant variants in total...\n",
      "2025/12/26 11:36:29  -Identified 4 lead variants!\n",
      "2025/12/26 11:36:29 Finished extracting lead variants.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96739</th>\n",
       "      <td>1:22068326_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>22068326</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.0621</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>1.629000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213860</th>\n",
       "      <td>1:51103268_T_C</td>\n",
       "      <td>1</td>\n",
       "      <td>51103268</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>-0.0802</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>2.519000e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534095</th>\n",
       "      <td>1:154309595_TA_T</td>\n",
       "      <td>1</td>\n",
       "      <td>154309595</td>\n",
       "      <td>TA</td>\n",
       "      <td>T</td>\n",
       "      <td>1960399</td>\n",
       "      <td>-0.0915</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>3.289000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>969974</th>\n",
       "      <td>2:640986_CACAT_C</td>\n",
       "      <td>2</td>\n",
       "      <td>640986</td>\n",
       "      <td>C</td>\n",
       "      <td>CACAT</td>\n",
       "      <td>1960399</td>\n",
       "      <td>-0.0946</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>2.665000e-10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   SNPID  CHR        POS  EA    NEA   STATUS    BETA      SE  \\\n",
       "96739     1:22068326_A_G    1   22068326   G      A  1960099  0.0621  0.0103   \n",
       "213860    1:51103268_T_C    1   51103268   C      T  1960099 -0.0802  0.0120   \n",
       "534095  1:154309595_TA_T    1  154309595  TA      T  1960399 -0.0915  0.0166   \n",
       "969974  2:640986_CACAT_C    2     640986   C  CACAT  1960399 -0.0946  0.0150   \n",
       "\n",
       "                   P  \n",
       "96739   1.629000e-09  \n",
       "213860  2.519000e-11  \n",
       "534095  3.289000e-08  \n",
       "969974  2.665000e-10  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.get_lead(windowsizekb=1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Different thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/26 11:36:30  -Genomic coordinates are based on GRCh37/hg19...\n",
      "2025/12/26 11:36:30 Start to extract lead variants ...(v4.0.0)\n",
      "2025/12/26 11:36:30  -Processing 1000000 variants...\n",
      "2025/12/26 11:36:30  -Significance threshold : 1e-10\n",
      "2025/12/26 11:36:30  -Sliding window size: 500  kb\n",
      "2025/12/26 11:36:30  -Using P for extracting lead variants...\n",
      "2025/12/26 11:36:30  -Found 1 significant variants in total...\n",
      "2025/12/26 11:36:30  -Identified 1 lead variants!\n",
      "2025/12/26 11:36:30 Finished extracting lead variants.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>213860</th>\n",
       "      <td>1:51103268_T_C</td>\n",
       "      <td>1</td>\n",
       "      <td>51103268</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>-0.0802</td>\n",
       "      <td>0.012</td>\n",
       "      <td>2.519000e-11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 SNPID  CHR       POS EA NEA   STATUS    BETA     SE  \\\n",
       "213860  1:51103268_T_C    1  51103268  C   T  1960099 -0.0802  0.012   \n",
       "\n",
       "                   P  \n",
       "213860  2.519000e-11  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.get_lead(sig_level=1e-10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check if novel against a tabular file"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "known_loci.txt file looks like:\n",
    "\n",
    "CHR POS\n",
    "1 154309595\n",
    "1 51103268"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/26 11:36:36  -Genomic coordinates are based on GRCh37/hg19...\n",
      "2025/12/26 11:36:36 Start to check if lead variants are known ...(v4.0.0)\n",
      "2025/12/26 11:36:36  -Genomic coordinates are based on GRCh37/hg19...\n",
      "2025/12/26 11:36:36 Start to extract lead variants ...(v4.0.0)\n",
      "2025/12/26 11:36:36  -Processing 1000000 variants...\n",
      "2025/12/26 11:36:36  -Significance threshold : 5e-08\n",
      "2025/12/26 11:36:36  -Sliding window size: 500  kb\n",
      "2025/12/26 11:36:36  -Using P for extracting lead variants...\n",
      "2025/12/26 11:36:36  -Found 543 significant variants in total...\n",
      "2025/12/26 11:36:36  -Identified 4 lead variants!\n",
      "2025/12/26 11:36:36 Finished extracting lead variants.\n",
      "2025/12/26 11:36:36  -Lead variants in known loci: 2\n",
      "2025/12/26 11:36:36  -Checking the minimum distance between identified lead variants and provided known variants...\n",
      "2025/12/26 11:36:36  -Identified  2  known vairants in current sumstats...\n",
      "2025/12/26 11:36:36  -Identified  2  novel vairants in current sumstats...\n",
      "2025/12/26 11:36:36 Finished checking if lead variants are known.\n"
     ]
    }
   ],
   "source": [
    "novel = mysumstats.get_novel(known=\"../0_sample_data//toy_data/known_loci.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>DISTANCE_TO_KNOWN</th>\n",
       "      <th>KNOWN_ID</th>\n",
       "      <th>NOVEL</th>\n",
       "      <th>LOCATION_OF_KNOWN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:22068326_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>22068326</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.0621</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>1.629000e-09</td>\n",
       "      <td>29034942.0</td>\n",
       "      <td>1:51103268</td>\n",
       "      <td>True</td>\n",
       "      <td>Upstream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:51103268_T_C</td>\n",
       "      <td>1</td>\n",
       "      <td>51103268</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>-0.0802</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>2.519000e-11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1:51103268</td>\n",
       "      <td>False</td>\n",
       "      <td>Same</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:154309595_TA_T</td>\n",
       "      <td>1</td>\n",
       "      <td>154309595</td>\n",
       "      <td>TA</td>\n",
       "      <td>T</td>\n",
       "      <td>1960399</td>\n",
       "      <td>-0.0915</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>3.289000e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1:154309595</td>\n",
       "      <td>False</td>\n",
       "      <td>Same</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2:640986_CACAT_C</td>\n",
       "      <td>2</td>\n",
       "      <td>640986</td>\n",
       "      <td>C</td>\n",
       "      <td>CACAT</td>\n",
       "      <td>1960399</td>\n",
       "      <td>-0.0946</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>2.665000e-10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>True</td>\n",
       "      <td>NoneOnThisChr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              SNPID  CHR        POS  EA    NEA   STATUS    BETA      SE  \\\n",
       "0    1:22068326_A_G    1   22068326   G      A  1960099  0.0621  0.0103   \n",
       "1    1:51103268_T_C    1   51103268   C      T  1960099 -0.0802  0.0120   \n",
       "2  1:154309595_TA_T    1  154309595  TA      T  1960399 -0.0915  0.0166   \n",
       "3  2:640986_CACAT_C    2     640986   C  CACAT  1960399 -0.0946  0.0150   \n",
       "\n",
       "              P  DISTANCE_TO_KNOWN     KNOWN_ID  NOVEL LOCATION_OF_KNOWN  \n",
       "0  1.629000e-09         29034942.0   1:51103268   True          Upstream  \n",
       "1  2.519000e-11                0.0   1:51103268  False              Same  \n",
       "2  3.289000e-08                0.0  1:154309595  False              Same  \n",
       "3  2.665000e-10                NaN         <NA>   True     NoneOnThisChr  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ccheck against GWAS Catalog using EFO ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/26 11:37:18 Start to perform liftover ...(v4.0.0)\n",
      "2025/12/26 11:37:18  -Using built-in chain file: /home/yunye/anaconda3/envs/py312/lib/python3.12/site-packages/gwaslab/data/chains/hg19ToHg38.over.chain.gz\n",
      "2025/12/26 11:37:18  -Converting variants with status code xxx0xxx: 1,000,000\n",
      "2025/12/26 11:37:18  -Target build: 38\n",
      "2025/12/26 11:37:18  -Input positions are 1-based\n",
      "2025/12/26 11:37:18  -Output positions will be 1-based\n",
      "2025/12/26 11:37:19  -Chromosome mismatches detected: 52 variants (treated as unmapped)\n",
      "2025/12/26 11:37:19  -Examples of chromosome mismatches:\n",
      "2025/12/26 11:37:19    SNPID=1:13127868_C_A | CHR=1 | POS=13127868 | CHR_LIFT=1_KI270766v1_alt | POS_LIFT=5939 | STATUS=1960099\n",
      "2025/12/26 11:37:19    SNPID=1:146705000_G_A | CHR=1 | POS=146705000 | CHR_LIFT=15 | POS_LIFT=39894340 | STATUS=1960099\n",
      "2025/12/26 11:37:19    SNPID=1:147758522_T_C | CHR=1 | POS=147758522 | CHR_LIFT=8 | POS_LIFT=38661844 | STATUS=1960099\n",
      "2025/12/26 11:37:19    SNPID=1:223725692_T_C | CHR=1 | POS=223725692 | CHR_LIFT=9 | POS_LIFT=136033409 | STATUS=1960099\n",
      "2025/12/26 11:37:19    SNPID=1:223725695_T_G | CHR=1 | POS=223725695 | CHR_LIFT=9 | POS_LIFT=136033412 | STATUS=1960099\n",
      "2025/12/26 11:37:19  -Mapped: 999381 variants\n",
      "2025/12/26 11:37:19  -Unmapped: 619 variants\n",
      "2025/12/26 11:37:19  -Examples of unmapped variants:\n",
      "2025/12/26 11:37:19    SNPID=1:1581586_T_C | CHR=1 | POS=1581586 | STATUS=1960099\n",
      "2025/12/26 11:37:19    SNPID=1:1584117_G_C | CHR=1 | POS=1584117 | STATUS=1960099\n",
      "2025/12/26 11:37:19    SNPID=1:1584218_T_C | CHR=1 | POS=1584218 | STATUS=1960099\n",
      "2025/12/26 11:37:19    SNPID=1:1585257_A_G | CHR=1 | POS=1585257 | STATUS=1960099\n",
      "2025/12/26 11:37:19    SNPID=1:1585313_CACAA_C | CHR=1 | POS=1585313 | STATUS=1960399\n",
      "2025/12/26 11:37:20  -Removed 619 unmapped variants\n",
      "2025/12/26 11:37:20 Start to fix chromosome notation (CHR) ...(v4.0.0)\n",
      "2025/12/26 11:37:20  -Current Dataframe shape : 999381 x 9 ; Memory usage: 67.36 MB\n",
      "2025/12/26 11:37:20  -Checking CHR data type...\n",
      "2025/12/26 11:37:20  -Variants with standardized chromosome notation: 999381\n",
      "2025/12/26 11:37:20  -All CHR are already fixed...\n",
      "2025/12/26 11:37:20 Finished fixing chromosome notation (CHR).\n",
      "2025/12/26 11:37:20 Start to fix basepair positions (POS) ...(v4.0.0)\n",
      "2025/12/26 11:37:20  -Trying to convert datatype for POS: Int64 -> Int64...\n",
      "2025/12/26 11:37:20  -Position bound:(0 , 250,000,000)\n",
      "2025/12/26 11:37:20  -Removed variants outliers: 0\n",
      "2025/12/26 11:37:20  -Removed variants with bad positions: 0\n",
      "2025/12/26 11:37:20 Finished fixing basepair positions (POS).\n",
      "2025/12/26 11:37:20 Finished liftover.\n"
     ]
    }
   ],
   "source": [
    "mysumstats.liftover(to_build=\"38\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/12/26 11:37:25  -Genomic coordinates are based on GRCh38/hg38...\n",
      "2025/12/26 11:37:25 Start to check if lead variants are known ...(v4.0.0)\n",
      "2025/12/26 11:37:25  -Genomic coordinates are based on GRCh38/hg38...\n",
      "2025/12/26 11:37:25 Start to extract lead variants ...(v4.0.0)\n",
      "2025/12/26 11:37:25  -Processing 999381 variants...\n",
      "2025/12/26 11:37:25  -Significance threshold : 5e-08\n",
      "2025/12/26 11:37:25  -Sliding window size: 500  kb\n",
      "2025/12/26 11:37:25  -Using P for extracting lead variants...\n",
      "2025/12/26 11:37:25  -Found 543 significant variants in total...\n",
      "2025/12/26 11:37:25  -Identified 4 lead variants!\n",
      "2025/12/26 11:37:25 Finished extracting lead variants.\n",
      "2025/12/26 11:37:25  -Genomic coordinates are based on GRCh38/hg38...\n",
      "2025/12/26 11:37:25  -Genomic coordinates are based on GRCh38/hg38...\n",
      "2025/12/26 11:37:25  -Sumstats build matches target build\n",
      "2025/12/26 11:37:25 Start to retrieve data using EFO: MONDO_0005148...\n",
      "2025/12/26 11:37:25  -Querying GWAS Catalog API v2 for trait: MONDO_0005148...\n",
      "2025/12/26 11:37:25  -MONDO ID detected, looking up EFO equivalent...\n",
      "2025/12/26 11:37:25 Requesting: https://www.ebi.ac.uk/gwas/rest/api/v2/efoTraits/MONDO_0005148\n",
      "2025/12/26 11:37:27 #WARNING! Error: Received status code 404\n",
      "2025/12/26 11:37:27 #WARNING! URL: https://www.ebi.ac.uk/gwas/rest/api/v2/efoTraits/MONDO_0005148\n",
      "2025/12/26 11:37:27 #WARNING! Response: {\"timestamp\":\"2025-12-26T02:37:28.946+00:00\",\"status\":404,\"error\":\"Not Found\",\"path\":\"/gwas/rest/api/v2/efoTraits/MONDO_0005148\"}\n",
      "2025/12/26 11:37:27  -Warning: Could not retrieve trait info for MONDO ID. Trying directly...\n",
      "2025/12/26 11:37:27 Requesting: https://www.ebi.ac.uk/gwas/rest/api/v2/associations\n",
      "2025/12/26 11:37:27 Parameters: {'efo_trait': 'MONDO_0005148', 'sort': 'p_value', 'direction': 'asc', 'size': 200, 'page': 0}\n",
      "2025/12/26 11:37:27 Status code: 200\n",
      "2025/12/26 11:37:27 #WARNING! No more data or an error occurred. Stopping.\n",
      "2025/12/26 11:37:27 Retrieved 0 total items\n",
      "2025/12/26 11:37:27  -No results from API v2, falling back to old API v1 for MONDO ID...\n",
      "2025/12/26 11:37:27 Start to retrieve data from GWASCatalog ...(v4.0.0)\n",
      "2025/12/26 11:37:27 searching cache in : ./\n",
      "2025/12/26 11:37:27  -Cache not found for False... Downloading from GWASCatalog...\n",
      "2025/12/26 11:37:27  -Requesting (GET) trait information through the GWASCatalog API...\n",
      "2025/12/26 11:37:27  -EFO trait api: https://www.ebi.ac.uk/gwas/rest/api/efoTraits/MONDO_0005148\n",
      "2025/12/26 11:37:28  -Status code: 200\n",
      "2025/12/26 11:37:28  -Trait Name: type 2 diabetes mellitus\n",
      "2025/12/26 11:37:28  -Trait URL: http://purl.obolibrary.org/obo/MONDO_0005148\n",
      "2025/12/26 11:37:28  -Requesting (GET) GWAS associations through the GWASCatalog API...\n",
      "2025/12/26 11:37:28  -associationsByTraitSummary API: https://www.ebi.ac.uk/gwas/rest/api/efoTraits/MONDO_0005148/associations?projection=associationByEfoTrait\n",
      "2025/12/26 11:37:28  -Note: this step might take a while...\n",
      "2025/12/26 11:42:13  -Status code 200 OK: Retrieved data from GWASCatalog successffully ...\n",
      "2025/12/26 11:42:13  -Loading json ...\n",
      "2025/12/26 11:42:14  -Saving json to: ./GWASCatalog_MONDO_0005148_associationsByTraitSummary_text_20251226.json ...\n",
      "2025/12/26 11:42:20  -Parsing json ...\n",
      "2025/12/26 11:42:20  -Number of reported associations for MONDO_0005148 in GWASCatalog: 8794\n",
      "2025/12/26 11:42:21  -Loading retrieved data into gwaslab Sumstats object ...\n",
      "2025/12/26 11:42:21 Finished retrieving data from GWASCatalog...\n",
      "2025/12/26 11:42:21 Finished retrieving data from GWASCatalog.\n",
      "2025/12/26 11:42:21  -Retrieved 6927 variants using old API v1\n",
      "2025/12/26 11:42:21  -Removed 2512 duplicate variants (kept 4415 unique)\n",
      "2025/12/26 11:42:21  -Processed 4415 unique variants from GWAS Catalog\n",
      "2025/12/26 11:42:21  -Retrieved 4415 associations from GWAS catalog.\n",
      "2025/12/26 11:42:21  -Lead variants in known loci: 4415\n",
      "2025/12/26 11:42:21  -Checking the minimum distance between identified lead variants and provided known variants...\n",
      "2025/12/26 11:42:21  -Identified  4  known vairants in current sumstats...\n",
      "2025/12/26 11:42:21  -Identified  0  novel vairants in current sumstats...\n",
      "2025/12/26 11:42:21 Finished checking if lead variants are known.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>DISTANCE_TO_KNOWN</th>\n",
       "      <th>KNOWN_ID</th>\n",
       "      <th>KNOWN_PUBMED_ID</th>\n",
       "      <th>KNOWN_AUTHOR</th>\n",
       "      <th>NOVEL</th>\n",
       "      <th>LOCATION_OF_KNOWN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:22068326_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>21741833</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>3860099</td>\n",
       "      <td>0.0621</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>1.629000e-09</td>\n",
       "      <td>0</td>\n",
       "      <td>rs1825307</td>\n",
       "      <td>30718926</td>\n",
       "      <td>Suzuki K</td>\n",
       "      <td>False</td>\n",
       "      <td>Same</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:51103268_T_C</td>\n",
       "      <td>1</td>\n",
       "      <td>50637596</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>3860099</td>\n",
       "      <td>-0.0802</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>2.519000e-11</td>\n",
       "      <td>0</td>\n",
       "      <td>rs12031188</td>\n",
       "      <td>30718926</td>\n",
       "      <td>Suzuki K</td>\n",
       "      <td>False</td>\n",
       "      <td>Same</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:154309595_TA_T</td>\n",
       "      <td>1</td>\n",
       "      <td>154337119</td>\n",
       "      <td>TA</td>\n",
       "      <td>T</td>\n",
       "      <td>3860399</td>\n",
       "      <td>-0.0915</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>3.289000e-08</td>\n",
       "      <td>1</td>\n",
       "      <td>rs68062313</td>\n",
       "      <td>30718926</td>\n",
       "      <td>Suzuki K</td>\n",
       "      <td>False</td>\n",
       "      <td>Upstream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2:640986_CACAT_C</td>\n",
       "      <td>2</td>\n",
       "      <td>640986</td>\n",
       "      <td>C</td>\n",
       "      <td>CACAT</td>\n",
       "      <td>3860399</td>\n",
       "      <td>-0.0946</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>2.665000e-10</td>\n",
       "      <td>-1931</td>\n",
       "      <td>rs7564708</td>\n",
       "      <td>34594039</td>\n",
       "      <td>Sakaue S</td>\n",
       "      <td>False</td>\n",
       "      <td>Downstream</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              SNPID  CHR        POS  EA    NEA   STATUS    BETA      SE  \\\n",
       "0    1:22068326_A_G    1   21741833   G      A  3860099  0.0621  0.0103   \n",
       "1    1:51103268_T_C    1   50637596   C      T  3860099 -0.0802  0.0120   \n",
       "2  1:154309595_TA_T    1  154337119  TA      T  3860399 -0.0915  0.0166   \n",
       "3  2:640986_CACAT_C    2     640986   C  CACAT  3860399 -0.0946  0.0150   \n",
       "\n",
       "              P  DISTANCE_TO_KNOWN    KNOWN_ID KNOWN_PUBMED_ID KNOWN_AUTHOR  \\\n",
       "0  1.629000e-09                  0   rs1825307        30718926     Suzuki K   \n",
       "1  2.519000e-11                  0  rs12031188        30718926     Suzuki K   \n",
       "2  3.289000e-08                  1  rs68062313        30718926     Suzuki K   \n",
       "3  2.665000e-10              -1931   rs7564708        34594039     Sakaue S   \n",
       "\n",
       "   NOVEL LOCATION_OF_KNOWN  \n",
       "0  False              Same  \n",
       "1  False              Same  \n",
       "2  False          Upstream  \n",
       "3  False        Downstream  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# EFO ID can be found on gwas catalog\n",
    "mysumstats.get_novel(efo=\"MONDO_0005148\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
