{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b992c2a2",
   "metadata": {},
   "source": [
    "# Clumping by calling PLINK"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4d232022",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1576b950",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/05/11 19:15:52 GWASLab v3.6.3 https://cloufield.github.io/gwaslab/\n",
      "2025/05/11 19:15:52 (C) 2022-2025, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2025/05/11 19:15:52 Python version: 3.12.0 | packaged by conda-forge | (main, Oct  3 2023, 08:43:22) [GCC 12.3.0]\n"
     ]
    }
   ],
   "source": [
    "gl.show_version()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e377891c-214e-4e8d-8e76-83c3244f3618",
   "metadata": {},
   "source": [
    "## Load sample data and perform QC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cdc90019",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/05/11 19:13:45 GWASLab v3.6.2 https://cloufield.github.io/gwaslab/\n",
      "2025/05/11 19:13:45 (C) 2022-2025, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2025/05/11 19:13:45 Python version: 3.12.0 | packaged by conda-forge | (main, Oct  3 2023, 08:43:22) [GCC 12.3.0]\n",
      "2025/05/11 19:13:45 Start to initialize gl.Sumstats from file :../0_sample_data/t2d_bbj.txt.gz\n",
      "2025/05/11 19:13:46  -Reading columns          : SE,SNP,P,CHR,POS\n",
      "2025/05/11 19:13:46  -Renaming columns to      : SE,SNPID,P,CHR,POS\n",
      "2025/05/11 19:13:46  -Current Dataframe shape : 1000000  x  5\n",
      "2025/05/11 19:13:46  -Initiating a status column: STATUS ...\n",
      "2025/05/11 19:13:46  #WARNING! Version of genomic coordinates is unknown...\n",
      "2025/05/11 19:13:47 Start to reorder the columns...v3.6.2\n",
      "2025/05/11 19:13:47  -Current Dataframe shape : 1000000 x 6 ; Memory usage: 63.43 MB\n",
      "2025/05/11 19:13:47  -Reordering columns to    : SNPID,CHR,POS,SE,P,STATUS\n",
      "2025/05/11 19:13:47 Finished reordering the columns.\n",
      "2025/05/11 19:13:47  -Trying to convert datatype for CHR: string -> Int64...Int64\n",
      "2025/05/11 19:13:47  -Column  : SNPID  CHR   POS   SE      P       STATUS  \n",
      "2025/05/11 19:13:47  -DType   : object Int64 int64 float64 float64 category\n",
      "2025/05/11 19:13:47  -Verified: T      T     T     T       T       T       \n",
      "2025/05/11 19:13:47  -Current Dataframe memory usage: 64.38 MB\n",
      "2025/05/11 19:13:47 Finished loading data successfully!\n",
      "2025/05/11 19:13:58  #WARNING! Necessary columns for .fix_allele() were not detected:EA,NEA\n",
      "2025/05/11 19:14:00  #WARNING! Necessary columns for .normalize() were not detected:EA,NEA\n",
      "2025/05/11 19:14:01 Start to check SNPID/rsID...v3.6.2\n",
      "2025/05/11 19:14:01  -Current Dataframe shape : 1000000 x 6 ; Memory usage: 65.33 MB\n",
      "2025/05/11 19:14:01  -Checking SNPID data type...\n",
      "2025/05/11 19:14:01  -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\n",
      "2025/05/11 19:14:04  -Replacing [_-] in SNPID with \":\" ...\n",
      "2025/05/11 19:14:05 Finished checking SNPID/rsID.\n"
     ]
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(\"../0_sample_data/t2d_bbj.txt.gz\",\n",
    "             snpid=\"SNP\",\n",
    "             chrom=\"CHR\",\n",
    "             pos=\"POS\",\n",
    "             se=\"SE\",\n",
    "             p=\"P\",\n",
    "             nrows=1000000  # Load only the first 1 million lines for demonstration purposes\n",
    "             )\n",
    "mysumstats.basic_check(verbose=False)\n",
    "mysumstats.fix_id(fixsep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6ecf141b-f800-4136-980d-264622a780f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>STATUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:725932:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>725932</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59700</td>\n",
       "      <td>9960999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:725933:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>725933</td>\n",
       "      <td>0.1394</td>\n",
       "      <td>0.59730</td>\n",
       "      <td>9960999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:737801:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>737801</td>\n",
       "      <td>0.1231</td>\n",
       "      <td>0.69080</td>\n",
       "      <td>9960999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:749963:T:TAA</td>\n",
       "      <td>1</td>\n",
       "      <td>749963</td>\n",
       "      <td>0.0199</td>\n",
       "      <td>0.28460</td>\n",
       "      <td>9960999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:751343:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>751343</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.27050</td>\n",
       "      <td>9960999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999995</th>\n",
       "      <td>2:6347639:C:A</td>\n",
       "      <td>2</td>\n",
       "      <td>6347639</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>0.15100</td>\n",
       "      <td>9960999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999996</th>\n",
       "      <td>2:6347694:G:C</td>\n",
       "      <td>2</td>\n",
       "      <td>6347694</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>0.17250</td>\n",
       "      <td>9960999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999997</th>\n",
       "      <td>2:6348478:G:A</td>\n",
       "      <td>2</td>\n",
       "      <td>6348478</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>0.17160</td>\n",
       "      <td>9960999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999998</th>\n",
       "      <td>2:6348490:G:C</td>\n",
       "      <td>2</td>\n",
       "      <td>6348490</td>\n",
       "      <td>0.2032</td>\n",
       "      <td>0.11750</td>\n",
       "      <td>9960999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999999</th>\n",
       "      <td>2:6348754:A:C</td>\n",
       "      <td>2</td>\n",
       "      <td>6348754</td>\n",
       "      <td>0.0126</td>\n",
       "      <td>0.01846</td>\n",
       "      <td>9960999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000000 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 SNPID  CHR      POS      SE        P   STATUS\n",
       "0         1:725932:G:A    1   725932  0.1394  0.59700  9960999\n",
       "1         1:725933:A:G    1   725933  0.1394  0.59730  9960999\n",
       "2         1:737801:T:C    1   737801  0.1231  0.69080  9960999\n",
       "3       1:749963:T:TAA    1   749963  0.0199  0.28460  9960999\n",
       "4         1:751343:T:A    1   751343  0.0156  0.27050  9960999\n",
       "...                ...  ...      ...     ...      ...      ...\n",
       "999995   2:6347639:C:A    2  6347639  0.0111  0.15100  9960999\n",
       "999996   2:6347694:G:C    2  6347694  0.0111  0.17250  9960999\n",
       "999997   2:6348478:G:A    2  6348478  0.0111  0.17160  9960999\n",
       "999998   2:6348490:G:C    2  6348490  0.2032  0.11750  9960999\n",
       "999999   2:6348754:A:C    2  6348754  0.0126  0.01846  9960999\n",
       "\n",
       "[1000000 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6698f10e",
   "metadata": {},
   "source": [
    "## Run clumping by calling plink2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "297f52c6",
   "metadata": {},
   "source": [
    "GWASLab will extract the chromosomes with lead variants and prepare input files for running clumping by calling PLINK2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dca9179",
   "metadata": {},
   "source": [
    "- VCF: reference LD VCF file (it will be converted to bfile)\n",
    "- bfile: PLINK bfile prefix\n",
    "- pfile: PLINK2 pfile prefix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "506d7c7c",
   "metadata": {},
   "source": [
    "Default parameters:\n",
    "    \n",
    "- clump_p1=5e-8\n",
    "- clump_p2=5e-8\n",
    "- clump_r2=0.01\n",
    "- clump_kb=250"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b5c0702-9e71-48ff-a6e5-5a556c57c742",
   "metadata": {},
   "source": [
    "Note:\n",
    "- plink2 need to be avaialble in your environment, or you can specify the path for PLINK2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3842cd76",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/05/11 19:14:05 Start to perfrom clumping...v3.6.2\n",
      "2025/05/11 19:14:05  -Current Dataframe shape : 1000000 x 6 ; Memory usage: 65.33 MB\n",
      "2025/05/11 19:14:05 Start to perform clumping...\n",
      "2025/05/11 19:14:05  -Clumping parameters for PLINK2:\n",
      "2025/05/11 19:14:05   -clump_p1 : 5e-08...\n",
      "2025/05/11 19:14:05   -clump_p2 : 1e-06...\n",
      "2025/05/11 19:14:05   -clump_kb : 250...\n",
      "2025/05/11 19:14:05   -clump_r2 : 0.1...\n",
      "2025/05/11 19:14:05  -Clumping will be performed using P\n",
      "2025/05/11 19:14:05  -Significant variants on CHR:  [1, 2]\n",
      "2025/05/11 19:14:05  -Processing VCF : /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz...\n",
      "2025/05/11 19:14:05  -PLINK version: PLINK v2.00a5.9LM AVX2 AMD (12 Dec 2023)\n",
      "2025/05/11 19:14:05   -Processing VCF for CHR 1...\n",
      "2025/05/11 19:14:05   -Plink bfile for CHR 1 exists: /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.1. Skipping...\n",
      "2025/05/11 19:14:05   -Processing VCF for CHR 2...\n",
      "2025/05/11 19:14:05   -Plink bfile for CHR 2 exists: /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.2. Skipping...\n",
      "2025/05/11 19:14:05  -Processing sumstats for CHR 1...\n",
      "2025/05/11 19:14:12  -Variants in reference file: 6471908...\n",
      "2025/05/11 19:14:12  -Variants in sumstats: 624...\n",
      "2025/05/11 19:14:12  -Variants available in both reference and sumstats: 624...\n",
      "2025/05/11 19:14:12  -Processing sumstats for CHR 2...\n",
      "2025/05/11 19:14:20  -Variants in reference file: 7085910...\n",
      "2025/05/11 19:14:20  -Variants in sumstats: 233...\n",
      "2025/05/11 19:14:20  -Variants available in both reference and sumstats: 233...\n",
      "2025/05/11 19:14:20  -Performing clumping for CHR 1...\n",
      "2025/05/11 19:14:20  -PLINK version: PLINK v2.00a5.9LM AVX2 AMD (12 Dec 2023)\n",
      "2025/05/11 19:14:21  -Saved results for CHR 1 to : clumping_plink2.1.clumps\n",
      "2025/05/11 19:14:21  -Performing clumping for CHR 2...\n",
      "2025/05/11 19:14:21  -PLINK version: PLINK v2.00a5.9LM AVX2 AMD (12 Dec 2023)\n",
      "2025/05/11 19:14:23  -Saved results for CHR 2 to : clumping_plink2.2.clumps\n",
      "2025/05/11 19:14:23 Finished clumping.\n",
      "2025/05/11 19:14:23 Finished clumping.\n"
     ]
    }
   ],
   "source": [
    "mysumstats.clump(   plink2=\"plink2\", # default\n",
    "                    clump_r2=0.1,\n",
    "                    clump_p2=1e-6,\n",
    "                    vcf=gl.get_path(\"1kg_eas_hg19\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f47adcc-9236-492d-9613-39f14bfe64b5",
   "metadata": {},
   "source": [
    "if using multiple bfile or pfile, @ can be used to indicate each chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "41a9ef70-0666-4e9e-875d-faad2b297392",
   "metadata": {},
   "outputs": [],
   "source": [
    "#mysumstats.clump(   plink2=\"plink2\", # default\n",
    "#                    clump_r2=0.1,\n",
    "#                    clump_p2=1e-6,\n",
    "#                    bfile=\"/home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.@\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "027da179",
   "metadata": {},
   "source": [
    "### clump results are stored in mysumstats.clumps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "29886935",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>STATUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96739</th>\n",
       "      <td>1:22068326:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>22068326</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>1.629000e-09</td>\n",
       "      <td>9960999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213860</th>\n",
       "      <td>1:51103268:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>51103268</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>2.519000e-11</td>\n",
       "      <td>9960999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214500</th>\n",
       "      <td>1:51401146:CT:C</td>\n",
       "      <td>1</td>\n",
       "      <td>51401146</td>\n",
       "      <td>0.0145</td>\n",
       "      <td>6.090000e-10</td>\n",
       "      <td>9960999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534095</th>\n",
       "      <td>1:154309595:TA:T</td>\n",
       "      <td>1</td>\n",
       "      <td>154309595</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>3.289000e-08</td>\n",
       "      <td>9960999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>969974</th>\n",
       "      <td>2:640986:CACAT:C</td>\n",
       "      <td>2</td>\n",
       "      <td>640986</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>2.665000e-10</td>\n",
       "      <td>9960999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   SNPID  CHR        POS      SE             P   STATUS\n",
       "96739     1:22068326:A:G    1   22068326  0.0103  1.629000e-09  9960999\n",
       "213860    1:51103268:T:C    1   51103268  0.0120  2.519000e-11  9960999\n",
       "214500   1:51401146:CT:C    1   51401146  0.0145  6.090000e-10  9960999\n",
       "534095  1:154309595:TA:T    1  154309595  0.0166  3.289000e-08  9960999\n",
       "969974  2:640986:CACAT:C    2     640986  0.0150  2.665000e-10  9960999"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.clumps[\"clumps\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1c4f6cdf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>SNPID</th>\n",
       "      <th>P</th>\n",
       "      <th>TOTAL</th>\n",
       "      <th>NONSIG</th>\n",
       "      <th>S0.05</th>\n",
       "      <th>S0.01</th>\n",
       "      <th>S0.001</th>\n",
       "      <th>S0.0001</th>\n",
       "      <th>SP2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>22068326</td>\n",
       "      <td>1:22068326:A:G</td>\n",
       "      <td>1.629000e-09</td>\n",
       "      <td>89</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>89</td>\n",
       "      <td>1:21982702:A:G,1:21983254:C:T,1:21983611:C:T,1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>51103268</td>\n",
       "      <td>1:51103268:T:C</td>\n",
       "      <td>2.519000e-11</td>\n",
       "      <td>347</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>347</td>\n",
       "      <td>1:50854197:G:C,1:50854654:A:G,1:50855083:A:G,1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>51401146</td>\n",
       "      <td>1:51401146:CT:C</td>\n",
       "      <td>6.090000e-10</td>\n",
       "      <td>134</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>134</td>\n",
       "      <td>1:51353720:C:G,1:51355823:A:G,1:51356091:G:T,1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>154309595</td>\n",
       "      <td>1:154309595:TA:T</td>\n",
       "      <td>3.289000e-08</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>1:154264194:G:A,1:154321382:C:T,1:154345344:G:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>640986</td>\n",
       "      <td>2:640986:CACAT:C</td>\n",
       "      <td>2.665000e-10</td>\n",
       "      <td>232</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>232</td>\n",
       "      <td>2:601905:T:G,2:605176:T:A,2:609177:A:G,2:61060...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHR        POS             SNPID             P  TOTAL  NONSIG  S0.05  \\\n",
       "2    1   22068326    1:22068326:A:G  1.629000e-09     89       0      0   \n",
       "0    1   51103268    1:51103268:T:C  2.519000e-11    347       0      0   \n",
       "1    1   51401146   1:51401146:CT:C  6.090000e-10    134       0      0   \n",
       "3    1  154309595  1:154309595:TA:T  3.289000e-08      5       0      0   \n",
       "4    2     640986  2:640986:CACAT:C  2.665000e-10    232       0      0   \n",
       "\n",
       "   S0.01  S0.001  S0.0001                                                SP2  \n",
       "2      0       0       89  1:21982702:A:G,1:21983254:C:T,1:21983611:C:T,1...  \n",
       "0      0       0      347  1:50854197:G:C,1:50854654:A:G,1:50855083:A:G,1...  \n",
       "1      0       0      134  1:51353720:C:G,1:51355823:A:G,1:51356091:G:T,1...  \n",
       "3      0       0        5  1:154264194:G:A,1:154321382:C:T,1:154345344:G:...  \n",
       "4      0       0      232  2:601905:T:G,2:605176:T:A,2:609177:A:G,2:61060...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.clumps[\"clumps_raw\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "591daf58",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.00a5.9LM AVX2 AMD (12 Dec 2023)       www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to clumping_plink2.1.log.\n",
      "Options in effect:\n",
      "  --bfile /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.1\n",
      "  --chr 1\n",
      "  --clump _gwaslab_tmp.1.SNPIDP\n",
      "  --clump-id-field SNPID\n",
      "  --clump-kb 250\n",
      "  --clump-p-field P\n",
      "  --clump-p1 5e-08\n",
      "  --clump-p2 1e-06\n",
      "  --clump-r2 0.1\n",
      "  --out clumping_plink2.1\n",
      "  --threads 1\n",
      "\n",
      "Start time: Sun May 11 19:14:20 2025\n",
      "31093 MiB RAM detected, ~10647 available; reserving 10583 MiB for main\n",
      "workspace.\n",
      "Using 1 compute thread.\n",
      "504 samples (0 females, 0 males, 504 ambiguous; 504 founders) loaded from\n",
      "/home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.1.fam.\n",
      "6471908 variants loaded from\n",
      "/home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.1.bim.\n",
      "Note: No phenotype data present.\n",
      "\n",
      "--clump: 0/312 index candidates processed.\n",
      "--clump: 4 clumps formed from 312 index candidates.  \n",
      "Results written to clumping_plink2.1.clumps .\n",
      "End time: Sun May 11 19:14:21 2025\n",
      "\n",
      "PLINK v2.00a5.9LM AVX2 AMD (12 Dec 2023)       www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to clumping_plink2.2.log.\n",
      "Options in effect:\n",
      "  --bfile /home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.2\n",
      "  --chr 2\n",
      "  --clump _gwaslab_tmp.2.SNPIDP\n",
      "  --clump-id-field SNPID\n",
      "  --clump-kb 250\n",
      "  --clump-p-field P\n",
      "  --clump-p1 5e-08\n",
      "  --clump-p2 1e-06\n",
      "  --clump-r2 0.1\n",
      "  --out clumping_plink2.2\n",
      "  --threads 1\n",
      "\n",
      "Start time: Sun May 11 19:14:22 2025\n",
      "31093 MiB RAM detected, ~10619 available; reserving 10555 MiB for main\n",
      "workspace.\n",
      "Using 1 compute thread.\n",
      "504 samples (0 females, 0 males, 504 ambiguous; 504 founders) loaded from\n",
      "/home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.2.fam.\n",
      "7085910 variants loaded from\n",
      "/home/yunye/.gwaslab/EAS.ALL.split_norm_af.1kgp3v5.hg19.2.bim.\n",
      "Note: No phenotype data present.\n",
      "\n",
      "--clump: 0/231 index candidates processed.\n",
      "--clump: 1 clump formed from 231 index candidates.  \n",
      "Results written to clumping_plink2.2.clumps .\n",
      "End time: Sun May 11 19:14:23 2025\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(mysumstats.clumps[\"plink_log\"])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
